{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\liyoa\\anaconda3\\envs\\sbna\\Lib\\site-packages\\paramiko\\transport.py:219: CryptographyDeprecationWarning: Blowfish has been deprecated and will be removed in a future release\n",
      "  \"class\": algorithms.Blowfish,\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from collections import defaultdict\n",
    "import paramiko # for ssh\n",
    "import traceback #  for error handling\n",
    "import ast\n",
    "import requests\n",
    "import re\n",
    "from Bio.SeqUtils import seq1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Name</th>\n",
       "      <th>Entrez GeneId</th>\n",
       "      <th>Genome Location</th>\n",
       "      <th>Tier</th>\n",
       "      <th>Hallmark</th>\n",
       "      <th>Chr Band</th>\n",
       "      <th>Somatic</th>\n",
       "      <th>Germline</th>\n",
       "      <th>Tumour Types(Somatic)</th>\n",
       "      <th>Tumour Types(Germline)</th>\n",
       "      <th>Cancer Syndrome</th>\n",
       "      <th>Tissue Type</th>\n",
       "      <th>Molecular Genetics</th>\n",
       "      <th>Role in Cancer</th>\n",
       "      <th>Mutation Types</th>\n",
       "      <th>Translocation Partner</th>\n",
       "      <th>Other Germline Mut</th>\n",
       "      <th>Other Syndrome</th>\n",
       "      <th>Synonyms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>29974.0</td>\n",
       "      <td>10:50799421-50885675</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>melanoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACF,ACF64,ACF65,APOBEC1CF,ASP,CCDS73133.1,ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>abl-interactor 1</td>\n",
       "      <td>10006.0</td>\n",
       "      <td>10:26746593-26860935</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>10p12.1</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>TSG, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>KMT2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABI-1,CCDS7150.1,E3B1,ENSG00000136754.17,NM_00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>v-abl Abelson murine leukemia viral oncogene h...</td>\n",
       "      <td>25.0</td>\n",
       "      <td>9:130713946-130885683</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>9q34.12</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CML, ALL, T-ALL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T, Mis</td>\n",
       "      <td>BCR, ETV6, NUP214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL,CCDS35165.1,ENSG00000097007.17,JTK7,NM_007...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABL2</td>\n",
       "      <td>c-abl oncogene 2, non-receptor tyrosine kinase</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1:179099327-179229601</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1q25.2</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>ETV6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABLL,ARG,CCDS30947.1,ENSG00000143322.19,NM_007...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACKR3</td>\n",
       "      <td>atypical chemokine receptor 3</td>\n",
       "      <td>57007.0</td>\n",
       "      <td>2:236569641-236582358</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2q37.3</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lipoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>HMGA2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCDS2516.1,CMKOR1,CXCR7,ENSG00000144476.5,GPR1...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene Symbol                                               Name  \\\n",
       "0        A1CF                     APOBEC1 complementation factor   \n",
       "1        ABI1                                   abl-interactor 1   \n",
       "2        ABL1  v-abl Abelson murine leukemia viral oncogene h...   \n",
       "3        ABL2     c-abl oncogene 2, non-receptor tyrosine kinase   \n",
       "4       ACKR3                      atypical chemokine receptor 3   \n",
       "\n",
       "   Entrez GeneId        Genome Location  Tier Hallmark  Chr Band Somatic  \\\n",
       "0        29974.0   10:50799421-50885675     2      NaN  10q11.23     yes   \n",
       "1        10006.0   10:26746593-26860935     1      Yes   10p12.1     yes   \n",
       "2           25.0  9:130713946-130885683     1      Yes   9q34.12     yes   \n",
       "3           27.0  1:179099327-179229601     1      NaN    1q25.2     yes   \n",
       "4        57007.0  2:236569641-236582358     1      Yes    2q37.3     yes   \n",
       "\n",
       "  Germline Tumour Types(Somatic) Tumour Types(Germline) Cancer Syndrome  \\\n",
       "0      NaN              melanoma                    NaN             NaN   \n",
       "1      NaN                   AML                    NaN             NaN   \n",
       "2      NaN       CML, ALL, T-ALL                    NaN             NaN   \n",
       "3      NaN                   AML                    NaN             NaN   \n",
       "4      NaN                lipoma                    NaN             NaN   \n",
       "\n",
       "  Tissue Type Molecular Genetics    Role in Cancer Mutation Types  \\\n",
       "0           E                NaN          oncogene            Mis   \n",
       "1           L                Dom       TSG, fusion              T   \n",
       "2           L                Dom  oncogene, fusion         T, Mis   \n",
       "3           L                Dom  oncogene, fusion              T   \n",
       "4           M                Dom  oncogene, fusion              T   \n",
       "\n",
       "  Translocation Partner Other Germline Mut Other Syndrome  \\\n",
       "0                   NaN                NaN            NaN   \n",
       "1                 KMT2A                NaN            NaN   \n",
       "2     BCR, ETV6, NUP214                NaN            NaN   \n",
       "3                  ETV6                NaN            NaN   \n",
       "4                 HMGA2                NaN            NaN   \n",
       "\n",
       "                                            Synonyms  \n",
       "0  ACF,ACF64,ACF65,APOBEC1CF,ASP,CCDS73133.1,ENSG...  \n",
       "1  ABI-1,CCDS7150.1,E3B1,ENSG00000136754.17,NM_00...  \n",
       "2  ABL,CCDS35165.1,ENSG00000097007.17,JTK7,NM_007...  \n",
       "3  ABLL,ARG,CCDS30947.1,ENSG00000143322.19,NM_007...  \n",
       "4  CCDS2516.1,CMKOR1,CXCR7,ENSG00000144476.5,GPR1...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cgc_data = pd.read_csv('Census_allTue Apr  9 04_58_38 2024.csv')\n",
    "cgc_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# response = requests.get(\"https://gtexportal.org/api/v2/expression/geneExpression/gene-expression?gencodeId=ENSG00000065613.9,ENSG00000203782.5\")\n",
    "# response"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pdb_ids(gene_symbol, entrez_gene_id, synonyms):\n",
    "    api_url = f\"http://mygene.info/v3/query?q={gene_symbol}&species=human&fields=entrezgene,ensembl,pdb\"\n",
    "    response = requests.get(api_url)\n",
    "    # check response\n",
    "    if response.status_code != 200:\n",
    "        return response.status_code, None\n",
    "    response = response.json()\n",
    "    for hit in response['hits']:\n",
    "        if ( str(hit['_id']) == str(int(entrez_gene_id)) ) or ( synonyms and 'ensembl' in hit and hit['ensembl']['gene'] in synonyms ): # (sometimes entrez_gene_id might not match but ensembl gene does\n",
    "            if 'pdb' in hit:\n",
    "                return 200, hit['pdb'] # status_code, pdb_ids\n",
    "            else: # no structures found \n",
    "                continue\n",
    "    return 200, None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to get pdb ids for ACSL3 with status code 200\n",
      "Failed to get pdb ids for ACSL6 with status code 200\n",
      "Failed to get pdb ids for AFF3 with status code 200\n",
      "Failed to get pdb ids for AKAP9 with status code 200\n",
      "Failed to get pdb ids for ARHGEF10 with status code 200\n",
      "Failed to get pdb ids for ARHGEF10L with status code 200\n",
      "Failed to get pdb ids for ASPM with status code 200\n",
      "Failed to get pdb ids for ASXL2 with status code 200\n",
      "Failed to get pdb ids for ATF1 with status code 200\n",
      "Failed to get pdb ids for ATP2B3 with status code 200\n",
      "Failed to get pdb ids for BCL11B with status code 200\n",
      "Failed to get pdb ids for BCL2L12 with status code 200\n",
      "Failed to get pdb ids for BCL7A with status code 200\n",
      "Failed to get pdb ids for BMP5 with status code 200\n",
      "Failed to get pdb ids for BTG1 with status code 200\n",
      "Failed to get pdb ids for CARS with error list indices must be integers or slices, not str\n",
      "Failed to get pdb ids for CBFA2T3 with status code 200\n",
      "Failed to get pdb ids for CCDC6 with status code 200\n",
      "Failed to get pdb ids for CCNB1IP1 with status code 200\n",
      "Failed to get pdb ids for CCR4 with error list indices must be integers or slices, not str\n",
      "Failed to get pdb ids for CDH10 with status code 200\n",
      "Failed to get pdb ids for CDH11 with status code 200\n",
      "Failed to get pdb ids for CEP89 with status code 200\n",
      "Failed to get pdb ids for CHD2 with error list indices must be integers or slices, not str\n",
      "Failed to get pdb ids for CHIC2 with status code 200\n",
      "Failed to get pdb ids for CHST11 with status code 200\n",
      "Failed to get pdb ids for CIITA with status code 200\n",
      "Failed to get pdb ids for CLTCL1 with status code 200\n",
      "Failed to get pdb ids for CNBD1 with status code 200\n",
      "Failed to get pdb ids for CNBP with status code 200\n",
      "Failed to get pdb ids for CNTRL with status code 200\n",
      "Failed to get pdb ids for CREB3L1 with status code 200\n",
      "Failed to get pdb ids for CREB3L2 with status code 200\n",
      "Failed to get pdb ids for CRTC3 with status code 200\n",
      "Failed to get pdb ids for CSMD3 with status code 200\n",
      "Failed to get pdb ids for CTNND2 with status code 200\n",
      "Failed to get pdb ids for CUX1 with status code 200\n",
      "Failed to get pdb ids for CYSLTR2 with status code 200\n",
      "Failed to get pdb ids for DCAF12L2 with status code 200\n",
      "Failed to get pdb ids for DDIT3 with status code 200\n",
      "Failed to get pdb ids for ECT2L with status code 200\n",
      "Failed to get pdb ids for ELF4 with status code 200\n",
      "Failed to get pdb ids for ELN with status code 200\n",
      "Failed to get pdb ids for ERC1 with status code 200\n",
      "Failed to get pdb ids for ETNK1 with status code 200\n",
      "Failed to get pdb ids for FAM131B with status code 200\n",
      "Failed to get pdb ids for FAM135B with status code 200\n",
      "Failed to get pdb ids for FAM47C with status code 200\n",
      "Failed to get pdb ids for FAT1 with status code 200\n",
      "Failed to get pdb ids for FAT3 with status code 200\n",
      "Failed to get pdb ids for FBLN2 with status code 200\n",
      "Failed to get pdb ids for FCRL4 with status code 200\n",
      "Failed to get pdb ids for FGFR1OP with error list indices must be integers or slices, not str\n",
      "Failed to get pdb ids for FKBP9 with status code 200\n",
      "Failed to get pdb ids for FOXR1 with status code 200\n",
      "Failed to get pdb ids for GOLGA5 with status code 200\n",
      "Failed to get pdb ids for GPC5 with status code 200\n",
      "Failed to get pdb ids for HLF with status code 200\n",
      "Failed to get pdb ids for HMGA2 with status code 200\n",
      "Failed to get pdb ids for HMGN2P46 with error cannot convert float NaN to integer\n",
      "Failed to get pdb ids for HOXA11 with status code 200\n",
      "Failed to get pdb ids for HOXA9 with status code 200\n",
      "Failed to get pdb ids for HOXC11 with status code 200\n",
      "Failed to get pdb ids for HOXC13 with status code 200\n",
      "Failed to get pdb ids for HOXD11 with status code 200\n",
      "Failed to get pdb ids for HOXD13 with status code 200\n",
      "Failed to get pdb ids for IGH with error argument of type 'float' is not iterable\n",
      "Failed to get pdb ids for IGK with status code 200\n",
      "Failed to get pdb ids for IGL with error argument of type 'float' is not iterable\n",
      "Failed to get pdb ids for IKZF3 with status code 200\n",
      "Failed to get pdb ids for IRS4 with status code 200\n",
      "Failed to get pdb ids for ISX with status code 200\n",
      "Failed to get pdb ids for JAZF1 with status code 200\n",
      "Failed to get pdb ids for KCNJ5 with status code 200\n",
      "Failed to get pdb ids for KDSR with status code 200\n",
      "Failed to get pdb ids for KIAA1549 with status code 200\n",
      "Failed to get pdb ids for KLF6 with status code 200\n",
      "Failed to get pdb ids for KNSTRN with status code 200\n",
      "Failed to get pdb ids for KTN1 with status code 200\n",
      "Failed to get pdb ids for LEF1 with status code 200\n",
      "Failed to get pdb ids for LEPROTL1 with status code 200\n",
      "Failed to get pdb ids for LHFPL6 with status code 200\n",
      "Failed to get pdb ids for LMO1 with status code 200\n",
      "Failed to get pdb ids for LPP with error list indices must be integers or slices, not str\n",
      "Failed to get pdb ids for LRIG3 with status code 200\n",
      "Failed to get pdb ids for LRP1B with status code 200\n",
      "Failed to get pdb ids for LYL1 with status code 200\n",
      "Failed to get pdb ids for LZTR1 with status code 200\n",
      "Failed to get pdb ids for MACC1 with status code 200\n",
      "Failed to get pdb ids for MAF with error list indices must be integers or slices, not str\n",
      "Failed to get pdb ids for MAFB with status code 200\n",
      "Failed to get pdb ids for MALAT1 with error cannot convert float NaN to integer\n",
      "Failed to get pdb ids for MAML2 with status code 200\n",
      "Failed to get pdb ids for MAP3K13 with status code 200\n",
      "Failed to get pdb ids for MDS2 with error cannot convert float NaN to integer\n",
      "Failed to get pdb ids for MED12 with status code 200\n",
      "Failed to get pdb ids for MLLT11 with status code 200\n",
      "Failed to get pdb ids for MLLT6 with status code 200\n",
      "Failed to get pdb ids for MN1 with status code 200\n",
      "Failed to get pdb ids for MNX1 with status code 200\n",
      "Failed to get pdb ids for MUC4 with status code 200\n",
      "Failed to get pdb ids for MUC6 with status code 200\n",
      "Failed to get pdb ids for MYB with status code 200\n",
      "Failed to get pdb ids for MYCL with status code 200\n",
      "Failed to get pdb ids for MYH11 with status code 200\n",
      "Failed to get pdb ids for MYOD1 with status code 200\n",
      "Failed to get pdb ids for NAB2 with status code 200\n",
      "Failed to get pdb ids for NFIB with status code 200\n",
      "Failed to get pdb ids for NFKBIE with status code 200\n",
      "Failed to get pdb ids for NIN with status code 200\n",
      "Failed to get pdb ids for NKX2-1 with status code 200\n",
      "Failed to get pdb ids for NR4A3 with status code 200\n",
      "Failed to get pdb ids for NUTM2B with status code 200\n",
      "Failed to get pdb ids for NUTM2D with status code 200\n",
      "Failed to get pdb ids for OLIG2 with status code 200\n",
      "Failed to get pdb ids for P2RY8 with error list indices must be integers or slices, not str\n",
      "Failed to get pdb ids for PAX7 with error list indices must be integers or slices, not str\n",
      "Failed to get pdb ids for PDE4DIP with status code 200\n",
      "Failed to get pdb ids for PER1 with status code 200\n",
      "Failed to get pdb ids for PICALM with status code 200\n",
      "Failed to get pdb ids for PIK3CB with status code 200\n",
      "Failed to get pdb ids for PLAG1 with status code 200\n",
      "Failed to get pdb ids for PPFIBP1 with status code 200\n",
      "Failed to get pdb ids for PPM1D with status code 200\n",
      "Failed to get pdb ids for PPP6C with status code 200\n",
      "Failed to get pdb ids for PRCC with error list indices must be integers or slices, not str\n",
      "Failed to get pdb ids for PRF1 with status code 200\n",
      "Failed to get pdb ids for PRKD1 with status code 200\n",
      "Failed to get pdb ids for PRPF40B with status code 200\n",
      "Failed to get pdb ids for PRRX1 with status code 200\n",
      "Failed to get pdb ids for PWWP2A with status code 200\n",
      "Failed to get pdb ids for REL with status code 200\n",
      "Failed to get pdb ids for RGPD3 with status code 200\n",
      "Failed to get pdb ids for RHOH with status code 200\n",
      "Failed to get pdb ids for RNF213 with status code 200\n",
      "Failed to get pdb ids for RSPO2 with status code 200\n",
      "Failed to get pdb ids for RSPO3 with status code 200\n",
      "Failed to get pdb ids for SETBP1 with status code 200\n",
      "Failed to get pdb ids for SETD1B with status code 200\n",
      "Failed to get pdb ids for SFRP4 with status code 200\n",
      "Failed to get pdb ids for SH2B3 with status code 200\n",
      "Failed to get pdb ids for SH3GL1 with status code 200\n",
      "Failed to get pdb ids for SHTN1 with status code 200\n",
      "Failed to get pdb ids for SIX2 with status code 200\n",
      "Failed to get pdb ids for SLC34A2 with status code 200\n",
      "Failed to get pdb ids for SLC45A3 with status code 200\n",
      "Failed to get pdb ids for SNX29 with status code 200\n",
      "Failed to get pdb ids for SOCS1 with status code 200\n",
      "Failed to get pdb ids for SOX21 with status code 200\n",
      "Failed to get pdb ids for SPECC1 with status code 200\n",
      "Failed to get pdb ids for SRGAP3 with status code 200\n",
      "Failed to get pdb ids for SS18L1 with status code 200\n",
      "Failed to get pdb ids for SSX1 with status code 200\n",
      "Failed to get pdb ids for SSX4 with status code 200\n",
      "Failed to get pdb ids for STRN with status code 200\n",
      "Failed to get pdb ids for TAL2 with status code 200\n",
      "Failed to get pdb ids for TFG with status code 200\n",
      "Failed to get pdb ids for TFPT with status code 200\n",
      "Failed to get pdb ids for THRAP3 with status code 200\n",
      "Failed to get pdb ids for TLX1 with error list indices must be integers or slices, not str\n",
      "Failed to get pdb ids for TLX3 with status code 200\n",
      "Failed to get pdb ids for TMEM127 with status code 200\n",
      "Failed to get pdb ids for TPM4 with status code 200\n",
      "Failed to get pdb ids for TRA with status code 200\n",
      "Failed to get pdb ids for TRAF7 with status code 200\n",
      "Failed to get pdb ids for TRB with error argument of type 'float' is not iterable\n",
      "Failed to get pdb ids for TRD with error argument of type 'float' is not iterable\n",
      "Failed to get pdb ids for TRIM27 with status code 200\n",
      "Failed to get pdb ids for TRIP11 with status code 200\n",
      "Failed to get pdb ids for USP44 with status code 200\n",
      "Failed to get pdb ids for USP6 with status code 200\n",
      "Failed to get pdb ids for VTI1A with status code 200\n",
      "Failed to get pdb ids for WDCP with status code 200\n",
      "Failed to get pdb ids for ZMYM2 with status code 200\n",
      "Failed to get pdb ids for ZMYM3 with status code 200\n",
      "Failed to get pdb ids for ZNF331 with status code 200\n",
      "Failed to get pdb ids for ZNF384 with status code 200\n",
      "Failed to get pdb ids for ZNF429 with status code 200\n",
      "Failed to get pdb ids for ZNF479 with status code 200\n",
      "Failed to get pdb ids for ZNF521 with status code 200\n",
      "Failed to get pdb ids for ZNRF3 with status code 200\n",
      "Failed to get pdb ids for ZRSR2 with status code 200\n"
     ]
    }
   ],
   "source": [
    "# add pdb structures to cgc_data\n",
    "pdb_structures = []\n",
    "count = 1\n",
    "for _, row in cgc_data.iterrows():\n",
    "    # print curent count, replace previous line    \n",
    "    gene_symbol = row['Gene Symbol']\n",
    "    entrez_gene_id = row['Entrez GeneId']\n",
    "    synonyms = row['Synonyms']\n",
    "    print(f\"Processing {count} of {len(cgc_data)} ({gene_symbol})\", end='\\r')\n",
    "    count += 1\n",
    "\n",
    "    try:\n",
    "        status, pdb = get_pdb_ids(gene_symbol, entrez_gene_id, synonyms)\n",
    "    except Exception as e:\n",
    "        print(f\"Failed to get pdb ids for {gene_symbol} with error {e}\")\n",
    "        pdb_structures.append(None)\n",
    "        continue\n",
    "\n",
    "    if not pdb:\n",
    "        print(f\"Failed to get pdb ids for {gene_symbol} with status code {status}\")\n",
    "        pdb_structures.append(None)\n",
    "    else:\n",
    "        if type(pdb) == list:\n",
    "            pdb_structures.append(pdb)\n",
    "        else:\n",
    "            pdb_structures.append([pdb])\n",
    "\n",
    "cgc_data['PDB Structures'] = pdb_structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Name</th>\n",
       "      <th>Entrez GeneId</th>\n",
       "      <th>Genome Location</th>\n",
       "      <th>Tier</th>\n",
       "      <th>Hallmark</th>\n",
       "      <th>Chr Band</th>\n",
       "      <th>Somatic</th>\n",
       "      <th>Germline</th>\n",
       "      <th>Tumour Types(Somatic)</th>\n",
       "      <th>...</th>\n",
       "      <th>Tissue Type</th>\n",
       "      <th>Molecular Genetics</th>\n",
       "      <th>Role in Cancer</th>\n",
       "      <th>Mutation Types</th>\n",
       "      <th>Translocation Partner</th>\n",
       "      <th>Other Germline Mut</th>\n",
       "      <th>Other Syndrome</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>PDB Structures</th>\n",
       "      <th>n_structures</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>29974.0</td>\n",
       "      <td>10:50799421-50885675</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>melanoma</td>\n",
       "      <td>...</td>\n",
       "      <td>E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACF,ACF64,ACF65,APOBEC1CF,ASP,CCDS73133.1,ENSG...</td>\n",
       "      <td>[2CPD]</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>abl-interactor 1</td>\n",
       "      <td>10006.0</td>\n",
       "      <td>10:26746593-26860935</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>10p12.1</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AML</td>\n",
       "      <td>...</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>TSG, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>KMT2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABI-1,CCDS7150.1,E3B1,ENSG00000136754.17,NM_00...</td>\n",
       "      <td>[7LXE]</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>v-abl Abelson murine leukemia viral oncogene h...</td>\n",
       "      <td>25.0</td>\n",
       "      <td>9:130713946-130885683</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>9q34.12</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CML, ALL, T-ALL</td>\n",
       "      <td>...</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T, Mis</td>\n",
       "      <td>BCR, ETV6, NUP214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL,CCDS35165.1,ENSG00000097007.17,JTK7,NM_007...</td>\n",
       "      <td>[1AB2, 1AWO, 1BBZ, 1JU5, 1OPL, 1ZZP, 2ABL, 2E2...</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene Symbol                                               Name  \\\n",
       "0        A1CF                     APOBEC1 complementation factor   \n",
       "1        ABI1                                   abl-interactor 1   \n",
       "2        ABL1  v-abl Abelson murine leukemia viral oncogene h...   \n",
       "\n",
       "   Entrez GeneId        Genome Location  Tier Hallmark  Chr Band Somatic  \\\n",
       "0        29974.0   10:50799421-50885675     2      NaN  10q11.23     yes   \n",
       "1        10006.0   10:26746593-26860935     1      Yes   10p12.1     yes   \n",
       "2           25.0  9:130713946-130885683     1      Yes   9q34.12     yes   \n",
       "\n",
       "  Germline Tumour Types(Somatic)  ... Tissue Type Molecular Genetics  \\\n",
       "0      NaN              melanoma  ...           E                NaN   \n",
       "1      NaN                   AML  ...           L                Dom   \n",
       "2      NaN       CML, ALL, T-ALL  ...           L                Dom   \n",
       "\n",
       "     Role in Cancer Mutation Types Translocation Partner Other Germline Mut  \\\n",
       "0          oncogene            Mis                   NaN                NaN   \n",
       "1       TSG, fusion              T                 KMT2A                NaN   \n",
       "2  oncogene, fusion         T, Mis     BCR, ETV6, NUP214                NaN   \n",
       "\n",
       "  Other Syndrome                                           Synonyms  \\\n",
       "0            NaN  ACF,ACF64,ACF65,APOBEC1CF,ASP,CCDS73133.1,ENSG...   \n",
       "1            NaN  ABI-1,CCDS7150.1,E3B1,ENSG00000136754.17,NM_00...   \n",
       "2            NaN  ABL,CCDS35165.1,ENSG00000097007.17,JTK7,NM_007...   \n",
       "\n",
       "                                      PDB Structures n_structures  \n",
       "0                                             [2CPD]            4  \n",
       "1                                             [7LXE]            4  \n",
       "2  [1AB2, 1AWO, 1BBZ, 1JU5, 1OPL, 1ZZP, 2ABL, 2E2...           81  \n",
       "\n",
       "[3 rows x 22 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cgc_data.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dwl_pdb_file(pdb_id):\n",
    "    base_url = 'https://files.rcsb.org/download/'\n",
    "    pdb_url = base_url + pdb_id + '.pdb'\n",
    "    response = requests.get(pdb_url)\n",
    "    # Check if the request was successful\n",
    "    if response.status_code == 200:\n",
    "        # Save the PDB file to pdb_files folder\n",
    "        with open(f'pdb_files/{pdb_id}.pdb', 'wb') as f:\n",
    "            f.write(response.content)\n",
    "    else:\n",
    "        print(f\"Failed to download PDB file {pdb_id}. Status code:\", response.status_code)\n",
    "    return\n",
    "\n",
    "def get_resolution(pdb_id):\n",
    "    try:\n",
    "        with open(f'pdb_files/{pdb_id}.pdb', 'r') as f:\n",
    "            pdb_text = f.read()\n",
    "\n",
    "            # find line REMARK   2 RESOLUTION.\n",
    "            resolution = None\n",
    "            for line in pdb_text.split('\\n'):\n",
    "                if line.startswith('REMARK   2 RESOLUTION.'):\n",
    "                    resolution = line.split(\"REMARK   2 RESOLUTION.\")[1].strip()\n",
    "                    break\n",
    "        if \"ANGSTROM\" in resolution:\n",
    "            resolution = resolution.replace(\" ANGSTROMS.\", \"\").strip()\n",
    "        else:\n",
    "            # print(f\"Resolution not found for {pdb_id}\")\n",
    "            resolution = None\n",
    "    except Exception as e:\n",
    "        print(f\"Failed to get resolution for {pdb_id} with error {e}\")\n",
    "        resolution = None\n",
    "    return resolution\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Name</th>\n",
       "      <th>Entrez GeneId</th>\n",
       "      <th>Genome Location</th>\n",
       "      <th>Tier</th>\n",
       "      <th>Hallmark</th>\n",
       "      <th>Chr Band</th>\n",
       "      <th>Somatic</th>\n",
       "      <th>Germline</th>\n",
       "      <th>Tumour Types(Somatic)</th>\n",
       "      <th>...</th>\n",
       "      <th>Tissue Type</th>\n",
       "      <th>Molecular Genetics</th>\n",
       "      <th>Role in Cancer</th>\n",
       "      <th>Mutation Types</th>\n",
       "      <th>Translocation Partner</th>\n",
       "      <th>Other Germline Mut</th>\n",
       "      <th>Other Syndrome</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>PDB Structures</th>\n",
       "      <th>n_structures</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>29974.0</td>\n",
       "      <td>10:50799421-50885675</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>melanoma</td>\n",
       "      <td>...</td>\n",
       "      <td>E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACF,ACF64,ACF65,APOBEC1CF,ASP,CCDS73133.1,ENSG...</td>\n",
       "      <td>[2CPD]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>abl-interactor 1</td>\n",
       "      <td>10006.0</td>\n",
       "      <td>10:26746593-26860935</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>10p12.1</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AML</td>\n",
       "      <td>...</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>TSG, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>KMT2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABI-1,CCDS7150.1,E3B1,ENSG00000136754.17,NM_00...</td>\n",
       "      <td>[7LXE]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>v-abl Abelson murine leukemia viral oncogene h...</td>\n",
       "      <td>25.0</td>\n",
       "      <td>9:130713946-130885683</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>9q34.12</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CML, ALL, T-ALL</td>\n",
       "      <td>...</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T, Mis</td>\n",
       "      <td>BCR, ETV6, NUP214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL,CCDS35165.1,ENSG00000097007.17,JTK7,NM_007...</td>\n",
       "      <td>[1AB2, 1AWO, 1BBZ, 1JU5, 1OPL, 1ZZP, 2ABL, 2E2...</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene Symbol                                               Name  \\\n",
       "0        A1CF                     APOBEC1 complementation factor   \n",
       "1        ABI1                                   abl-interactor 1   \n",
       "2        ABL1  v-abl Abelson murine leukemia viral oncogene h...   \n",
       "\n",
       "   Entrez GeneId        Genome Location  Tier Hallmark  Chr Band Somatic  \\\n",
       "0        29974.0   10:50799421-50885675     2      NaN  10q11.23     yes   \n",
       "1        10006.0   10:26746593-26860935     1      Yes   10p12.1     yes   \n",
       "2           25.0  9:130713946-130885683     1      Yes   9q34.12     yes   \n",
       "\n",
       "  Germline Tumour Types(Somatic)  ... Tissue Type Molecular Genetics  \\\n",
       "0      NaN              melanoma  ...           E                NaN   \n",
       "1      NaN                   AML  ...           L                Dom   \n",
       "2      NaN       CML, ALL, T-ALL  ...           L                Dom   \n",
       "\n",
       "     Role in Cancer Mutation Types Translocation Partner Other Germline Mut  \\\n",
       "0          oncogene            Mis                   NaN                NaN   \n",
       "1       TSG, fusion              T                 KMT2A                NaN   \n",
       "2  oncogene, fusion         T, Mis     BCR, ETV6, NUP214                NaN   \n",
       "\n",
       "  Other Syndrome                                           Synonyms  \\\n",
       "0            NaN  ACF,ACF64,ACF65,APOBEC1CF,ASP,CCDS73133.1,ENSG...   \n",
       "1            NaN  ABI-1,CCDS7150.1,E3B1,ENSG00000136754.17,NM_00...   \n",
       "2            NaN  ABL,CCDS35165.1,ENSG00000097007.17,JTK7,NM_007...   \n",
       "\n",
       "                                      PDB Structures n_structures  \n",
       "0                                             [2CPD]            1  \n",
       "1                                             [7LXE]            1  \n",
       "2  [1AB2, 1AWO, 1BBZ, 1JU5, 1OPL, 1ZZP, 2ABL, 2E2...           81  \n",
       "\n",
       "[3 rows x 22 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count number of elements in each pdb_structure\n",
    "lens = []\n",
    "for i in pdb_structures:\n",
    "    # if value is list\n",
    "    if isinstance(i, list):\n",
    "        lens.append(len(i))\n",
    "    elif i is not None:\n",
    "        lens.append(1)\n",
    "    else:\n",
    "        lens.append(0)\n",
    "\n",
    "cgc_data['n_structures'] = lens\n",
    "cgc_data.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "cgc_data.to_csv('Census_all_with_pdb.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp_genes = [\"KRAS\", \"EGFR\", \"TP53\", \"HRAS\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to download PDB file 7Q9U. Status code: 404 remaining))\n",
      "Failed to get resolution for 7Q9U with error [Errno 2] No such file or directory: 'pdb_files/7Q9U.pdb'\n",
      "Downloading PDB file 8FG4 for HRAS (0 structures remaining)))\r"
     ]
    }
   ],
   "source": [
    "# download pdb files\n",
    "\n",
    "resolutions = defaultdict(dict)\n",
    "\n",
    "# count number of structures for all gene\n",
    "n_total = cgc_data[cgc_data['Gene Symbol'].isin(tmp_genes)][\"n_structures\"].sum()\n",
    "\n",
    "for gene in tmp_genes:\n",
    "    pdb_structures = cgc_data[cgc_data['Gene Symbol'] == gene]['PDB Structures'].values[0]\n",
    "    if pdb_structures:\n",
    "        for pdb_id in pdb_structures:\n",
    "            n_total -= 1\n",
    "            print(f\"Downloading PDB file {pdb_id} for {gene} ({n_total} structures remaining)\", end='\\r')\n",
    "            # check if the file already exists\n",
    "            if not os.path.exists(f'pdb_files/{pdb_id}.pdb'):\n",
    "                dwl_pdb_file(pdb_id)\n",
    "            resolution = get_resolution(pdb_id)\n",
    "            resolutions[gene][pdb_id] = resolution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Top 30 structures for Gene KRAS:\n",
      "PDB ID: 6P0Z, Resolution: 1.01\n",
      "PDB ID: 8ONV, Resolution: 1.01\n",
      "PDB ID: 8AZZ, Resolution: 1.02\n",
      "PDB ID: 4QL3, Resolution: 1.04\n",
      "PDB ID: 8AZX, Resolution: 1.04\n",
      "PDB ID: 8B00, Resolution: 1.04\n",
      "PDB ID: 8AZV, Resolution: 1.05\n",
      "PDB ID: 8AZY, Resolution: 1.09\n",
      "PDB ID: 8B78, Resolution: 1.11\n",
      "PDB ID: 8AFB, Resolution: 1.12\n",
      "PDB ID: 4TQA, Resolution: 1.13\n",
      "PDB ID: 4LDJ, Resolution: 1.15\n",
      "PDB ID: 6TAN, Resolution: 1.16\n",
      "PDB ID: 7O70, Resolution: 1.18\n",
      "PDB ID: 8EDY, Resolution: 1.18\n",
      "PDB ID: 8EER, Resolution: 1.18\n",
      "PDB ID: 4Q03, Resolution: 1.20\n",
      "PDB ID: 7R0N, Resolution: 1.20\n",
      "PDB ID: 8EBZ, Resolution: 1.20\n",
      "PDB ID: 8TXH, Resolution: 1.20\n",
      "PDB ID: 4OBE, Resolution: 1.24\n",
      "PDB ID: 6QUW, Resolution: 1.24\n",
      "PDB ID: 5XCO, Resolution: 1.25\n",
      "PDB ID: 7RT1, Resolution: 1.27\n",
      "PDB ID: 7T47, Resolution: 1.27\n",
      "PDB ID: 4LUC, Resolution: 1.29\n",
      "PDB ID: 4Q01, Resolution: 1.29\n",
      "PDB ID: 6B0V, Resolution: 1.29\n",
      "PDB ID: 7RT5, Resolution: 1.29\n",
      "PDB ID: 7RPZ, Resolution: 1.30\n",
      "\n",
      "Top 30 structures for Gene EGFR:\n",
      "PDB ID: 8A27, Resolution: 1.07\n",
      "PDB ID: 8A2D, Resolution: 1.11\n",
      "PDB ID: 5UG9, Resolution: 1.33\n",
      "PDB ID: 5HG8, Resolution: 1.42\n",
      "PDB ID: 8A2A, Resolution: 1.43\n",
      "PDB ID: 5UG8, Resolution: 1.46\n",
      "PDB ID: 3POZ, Resolution: 1.50\n",
      "PDB ID: 6TFV, Resolution: 1.50\n",
      "PDB ID: 6TG0, Resolution: 1.50\n",
      "PDB ID: 3VRP, Resolution: 1.52\n",
      "PDB ID: 5HG5, Resolution: 1.52\n",
      "PDB ID: 5CNO, Resolution: 1.55\n",
      "PDB ID: 5UGC, Resolution: 1.58\n",
      "PDB ID: 3G5Y, Resolution: 1.59\n",
      "PDB ID: 5U8L, Resolution: 1.60\n",
      "PDB ID: 6TG1, Resolution: 1.60\n",
      "PDB ID: 7SI1, Resolution: 1.60\n",
      "PDB ID: 8A2B, Resolution: 1.69\n",
      "PDB ID: 3W33, Resolution: 1.70\n",
      "PDB ID: 6TFY, Resolution: 1.70\n",
      "PDB ID: 4I22, Resolution: 1.71\n",
      "PDB ID: 6WXN, Resolution: 1.76\n",
      "PDB ID: 6V66, Resolution: 1.79\n",
      "PDB ID: 3P0Y, Resolution: 1.80\n",
      "PDB ID: 3W32, Resolution: 1.80\n",
      "PDB ID: 4I24, Resolution: 1.80\n",
      "PDB ID: 5GNK, Resolution: 1.80\n",
      "PDB ID: 6TFZ, Resolution: 1.80\n",
      "PDB ID: 5UGA, Resolution: 1.82\n",
      "PDB ID: 7JXQ, Resolution: 1.83\n",
      "\n",
      "Top 30 structures for Gene TP53:\n",
      "PDB ID: 3D06, Resolution: 1.20\n",
      "PDB ID: 5MHC, Resolution: 1.20\n",
      "PDB ID: 6GGC, Resolution: 1.24\n",
      "PDB ID: 6SHZ, Resolution: 1.24\n",
      "PDB ID: 4MZI, Resolution: 1.25\n",
      "PDB ID: 6GGE, Resolution: 1.25\n",
      "PDB ID: 3LW1, Resolution: 1.28\n",
      "PDB ID: 5O1E, Resolution: 1.30\n",
      "PDB ID: 6RL3, Resolution: 1.30\n",
      "PDB ID: 8E7A, Resolution: 1.30\n",
      "PDB ID: 5O1C, Resolution: 1.32\n",
      "PDB ID: 5O1H, Resolution: 1.32\n",
      "PDB ID: 6GGB, Resolution: 1.32\n",
      "PDB ID: 6GGF, Resolution: 1.32\n",
      "PDB ID: 7B4N, Resolution: 1.32\n",
      "PDB ID: 3ZME, Resolution: 1.35\n",
      "PDB ID: 5AOK, Resolution: 1.35\n",
      "PDB ID: 5G4N, Resolution: 1.35\n",
      "PDB ID: 5O1G, Resolution: 1.35\n",
      "PDB ID: 6V4F, Resolution: 1.35\n",
      "PDB ID: 5AB9, Resolution: 1.36\n",
      "PDB ID: 5O1D, Resolution: 1.36\n",
      "PDB ID: 8A92, Resolution: 1.37\n",
      "PDB ID: 5G4M, Resolution: 1.38\n",
      "PDB ID: 5O1F, Resolution: 1.38\n",
      "PDB ID: 7B4H, Resolution: 1.39\n",
      "PDB ID: 1YC5, Resolution: 1.40\n",
      "PDB ID: 3D08, Resolution: 1.40\n",
      "PDB ID: 5A7B, Resolution: 1.40\n",
      "PDB ID: 5O1I, Resolution: 1.40\n",
      "\n",
      "Top 30 structures for Gene HRAS:\n",
      "PDB ID: 2CE2, Resolution: 1.00\n",
      "PDB ID: 2EVW, Resolution: 1.05\n",
      "PDB ID: 2CLD, Resolution: 1.22\n",
      "PDB ID: 2CL6, Resolution: 1.24\n",
      "PDB ID: 2CL7, Resolution: 1.25\n",
      "PDB ID: 5WDQ, Resolution: 1.25\n",
      "PDB ID: 1CTQ, Resolution: 1.26\n",
      "PDB ID: 2CLC, Resolution: 1.30\n",
      "PDB ID: 3K8Y, Resolution: 1.30\n",
      "PDB ID: 3OIW, Resolution: 1.30\n",
      "PDB ID: 3TGP, Resolution: 1.31\n",
      "PDB ID: 3OIU, Resolution: 1.32\n",
      "PDB ID: 4DLR, Resolution: 1.32\n",
      "PDB ID: 8BWG, Resolution: 1.32\n",
      "PDB ID: 2RGB, Resolution: 1.35\n",
      "PDB ID: 5P21, Resolution: 1.35\n",
      "PDB ID: 5WDP, Resolution: 1.35\n",
      "PDB ID: 8CNJ, Resolution: 1.35\n",
      "PDB ID: 3I3S, Resolution: 1.36\n",
      "PDB ID: 2RGE, Resolution: 1.40\n",
      "PDB ID: 3RS0, Resolution: 1.40\n",
      "PDB ID: 5E95, Resolution: 1.40\n",
      "PDB ID: 3L8Z, Resolution: 1.44\n",
      "PDB ID: 2RGG, Resolution: 1.45\n",
      "PDB ID: 8CNN, Resolution: 1.48\n",
      "PDB ID: 2QUZ, Resolution: 1.49\n",
      "PDB ID: 8FG3, Resolution: 1.49\n",
      "PDB ID: 1ZW6, Resolution: 1.50\n",
      "PDB ID: 6MQT, Resolution: 1.50\n",
      "PDB ID: 821P, Resolution: 1.50\n"
     ]
    }
   ],
   "source": [
    "def sort_pdb_ids_by_resolution(gene_id):\n",
    "    pdb_resolutions = resolutions[gene_id]\n",
    "    sorted_pdb_ids = sorted(pdb_resolutions, key=lambda pdb_id: float('inf') if pdb_resolutions[pdb_id] is None else float(pdb_resolutions[pdb_id]))\n",
    "    sorted_resolutions = [(pdb_id, pdb_resolutions[pdb_id]) for pdb_id in sorted_pdb_ids]\n",
    "    return sorted_resolutions\n",
    "\n",
    "top_structures = dict()\n",
    "n = 30\n",
    "for gene_id in resolutions:\n",
    "    sorted_pdb_resolutions = sort_pdb_ids_by_resolution(gene_id)\n",
    "    print(f\"\\nTop {n} structures for Gene {gene_id}:\")\n",
    "    for pdb_id, resolution in sorted_pdb_resolutions[:n]:\n",
    "        print(f\"PDB ID: {pdb_id}, Resolution: {resolution}\")\n",
    "    top_structures[gene_id] = [i[0] for i in sorted_pdb_resolutions[:n]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save top_structures to a file\n",
    "with open(f'top_{n}_structures.txt', 'w') as f:\n",
    "\tf.write(str(top_structures))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'KRAS': ['6P0Z',\n",
       "  '8ONV',\n",
       "  '8AZZ',\n",
       "  '4QL3',\n",
       "  '8AZX',\n",
       "  '8B00',\n",
       "  '8AZV',\n",
       "  '8AZY',\n",
       "  '8B78',\n",
       "  '8AFB',\n",
       "  '4TQA',\n",
       "  '4LDJ',\n",
       "  '6TAN',\n",
       "  '7O70',\n",
       "  '8EDY',\n",
       "  '8EER',\n",
       "  '4Q03',\n",
       "  '7R0N',\n",
       "  '8EBZ',\n",
       "  '8TXH',\n",
       "  '4OBE',\n",
       "  '6QUW',\n",
       "  '5XCO',\n",
       "  '7RT1',\n",
       "  '7T47',\n",
       "  '4LUC',\n",
       "  '4Q01',\n",
       "  '6B0V',\n",
       "  '7RT5',\n",
       "  '7RPZ'],\n",
       " 'EGFR': ['8A27',\n",
       "  '8A2D',\n",
       "  '5UG9',\n",
       "  '5HG8',\n",
       "  '8A2A',\n",
       "  '5UG8',\n",
       "  '3POZ',\n",
       "  '6TFV',\n",
       "  '6TG0',\n",
       "  '3VRP',\n",
       "  '5HG5',\n",
       "  '5CNO',\n",
       "  '5UGC',\n",
       "  '3G5Y',\n",
       "  '5U8L',\n",
       "  '6TG1',\n",
       "  '7SI1',\n",
       "  '8A2B',\n",
       "  '3W33',\n",
       "  '6TFY',\n",
       "  '4I22',\n",
       "  '6WXN',\n",
       "  '6V66',\n",
       "  '3P0Y',\n",
       "  '3W32',\n",
       "  '4I24',\n",
       "  '5GNK',\n",
       "  '6TFZ',\n",
       "  '5UGA',\n",
       "  '7JXQ'],\n",
       " 'TP53': ['3D06',\n",
       "  '5MHC',\n",
       "  '6GGC',\n",
       "  '6SHZ',\n",
       "  '4MZI',\n",
       "  '6GGE',\n",
       "  '3LW1',\n",
       "  '5O1E',\n",
       "  '6RL3',\n",
       "  '8E7A',\n",
       "  '5O1C',\n",
       "  '5O1H',\n",
       "  '6GGB',\n",
       "  '6GGF',\n",
       "  '7B4N',\n",
       "  '3ZME',\n",
       "  '5AOK',\n",
       "  '5G4N',\n",
       "  '5O1G',\n",
       "  '6V4F',\n",
       "  '5AB9',\n",
       "  '5O1D',\n",
       "  '8A92',\n",
       "  '5G4M',\n",
       "  '5O1F',\n",
       "  '7B4H',\n",
       "  '1YC5',\n",
       "  '3D08',\n",
       "  '5A7B',\n",
       "  '5O1I'],\n",
       " 'HRAS': ['2CE2',\n",
       "  '2EVW',\n",
       "  '2CLD',\n",
       "  '2CL6',\n",
       "  '2CL7',\n",
       "  '5WDQ',\n",
       "  '1CTQ',\n",
       "  '2CLC',\n",
       "  '3K8Y',\n",
       "  '3OIW',\n",
       "  '3TGP',\n",
       "  '3OIU',\n",
       "  '4DLR',\n",
       "  '8BWG',\n",
       "  '2RGB',\n",
       "  '5P21',\n",
       "  '5WDP',\n",
       "  '8CNJ',\n",
       "  '3I3S',\n",
       "  '2RGE',\n",
       "  '3RS0',\n",
       "  '5E95',\n",
       "  '3L8Z',\n",
       "  '2RGG',\n",
       "  '8CNN',\n",
       "  '2QUZ',\n",
       "  '8FG3',\n",
       "  '1ZW6',\n",
       "  '6MQT',\n",
       "  '821P']}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read top_structures from file\n",
    "with open(f'top_30_structures.txt', 'r') as f:\n",
    "\ttop_structures = ast.literal_eval(f.read())\n",
    "top_structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'chain': 'A', 'uniprot': 'Q93009', 'start': 53, 'end': 200},\n",
       " {'chain': 'A', 'uniprot': 'P04637', 'start': 201, 'end': 209},\n",
       " {'chain': 'B', 'uniprot': 'Q93009', 'start': 53, 'end': 200},\n",
       " {'chain': 'B', 'uniprot': 'P04637', 'start': 201, 'end': 209}]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from stat import S_ISDIR\n",
    "\n",
    "def download_directory(sftp, remote_dir, local_dir):\n",
    "    for item in sftp.listdir_attr(remote_dir):\n",
    "        remote_path = remote_dir + '/' + item.filename\n",
    "        local_path = os.path.join(local_dir, item.filename)\n",
    "        if S_ISDIR(item.st_mode):\n",
    "            os.makedirs(local_path, exist_ok=True)\n",
    "            download_directory(sftp, remote_path, local_path)\n",
    "        else:\n",
    "            sftp.get(remote_path, local_path)\n",
    "\n",
    "# get all chains, uniprot id, and start and end in pdb file\n",
    "def get_dbref_data(pdb_id):\n",
    "    with open(f'pdb_files/{pdb_id}.pdb', 'r') as f:\n",
    "        pdb_text = f.read()\n",
    "        data = []\n",
    "        for line in pdb_text.split('\\n'):\n",
    "            # contents of a DBREF line:  https://www.wwpdb.org/documentation/file-format-content/format33/sect3.html\n",
    "            # only reading relevant contents    \n",
    "            if line.startswith('DBREF') and line[26:32].strip()==\"UNP\": # if database is UNP \n",
    "                data.append({\n",
    "                    'chain': line[12],\n",
    "                    'uniprot': line[33:41].strip(),\n",
    "                    'start': int(line[14:18]),\n",
    "                    'end': int(line[20:24])\n",
    "                })  \n",
    "    return data\n",
    "\n",
    "get_dbref_data('2F1X')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'chain_id': 'A', 'seq_begin': 2, 'seq_end': 149, 'database': 'UNP', 'db_accession': 'Q07817'}\n"
     ]
    }
   ],
   "source": [
    "def parse_dbref_line(line):\n",
    "    # contents of a DBREF line:  https://www.wwpdb.org/documentation/file-format-content/format33/sect3.html\n",
    "    # only reading relevant contents\n",
    "    chain_id = line[12]               # Chain identifier\n",
    "    seq_begin = int(line[14:18])      # Initial sequence number of the PDB sequence segment\n",
    "    seq_end = int(line[20:24])        # Ending sequence number of the PDB sequence segment\n",
    "    database = line[26:32].strip()    # Sequence database name\n",
    "    db_accession = line[33:41].strip()# Sequence database accession code\n",
    "    \n",
    "    return {\n",
    "        \"chain_id\": chain_id,\n",
    "        \"seq_begin\": seq_begin,\n",
    "        \"seq_end\": seq_end,\n",
    "        \"database\": database,\n",
    "        \"db_accession\": db_accession,\n",
    "    }\n",
    "\n",
    "# Example usage:\n",
    "line = \"DBREF  6LHD A    2   149L UNP    Q07817   B2CL1_HUMAN      2    201\"\n",
    "parsed_line = parse_dbref_line(line)\n",
    "print(parsed_line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_sbna_and_download(client, sftp, pdb_id, base_path, chains=None):\n",
    "    # get available chains\n",
    "    if not chains: # chain not specified\n",
    "        try:\n",
    "            chains = get_dbref_data(pdb_id)\n",
    "            chains = [chain['chain'] for chain in chains]\n",
    "        except Exception as e:\n",
    "            print(f\"Failed to get chains for {pdb_id} with error {e}\")\n",
    "            return\n",
    "    \n",
    "    sftp.mkdir(f'{base_path}/{pdb_id}')\n",
    "    for chain in chains: \n",
    "        local_dir = f\"sbna_results/{pdb_id}/{chain}\"\n",
    "        if os.path.exists(f\"{local_dir}/{pdb_id}_monomer/FinalSum\"): # skip if analysis already done and downloaded\n",
    "            continue\n",
    "        try:\n",
    "            print(f\"\\nProcessing {pdb_id} with chain {chain}...\")\n",
    "\n",
    "            # remove the directory if already exists and create new one\n",
    "            client.exec_command(f\"rm -r {base_path}/{pdb_id}/{pdb_id}_monomer\")\n",
    "            client.exec_command(f\"rm -r {base_path}/{pdb_id}/{pdb_id}_multimer\") \n",
    "            \n",
    "            # upload pdb file to the remote directory\n",
    "            sftp.put(f\"pdb_files/{pdb_id}.pdb\", f'{base_path}/{pdb_id}/{pdb_id}.pdb') \n",
    "            \n",
    "            # copy the sbna.sh file to the subdirectory\n",
    "            client.exec_command(f\"cp {base_path}/sbna.sh {base_path}/{pdb_id}/sbna.sh\") \n",
    "\n",
    "            # run the sbna.sh script\n",
    "            _,stdout,stderr=client.exec_command(f'cd {base_path}/{pdb_id}; sh sbna.sh {pdb_id}.pdb \"Chain {chain}\"') \n",
    "            print(stdout.read().decode())\n",
    "            \n",
    "            # check if Final_sum file exists\n",
    "            print(stderr.read().decode('utf-8'))\n",
    "            _,stdout,stderr=client.exec_command(f\"ls {base_path}/{pdb_id}/{pdb_id}_monomer/FinalSum\")\n",
    "            # print error message if no output\n",
    "\n",
    "            # download the sbna results\n",
    "            if not stdout.read().decode():\n",
    "                print(f\"Final_sum file not found for {pdb_id}\")\n",
    "                # client.exec_command(f\"rm -rf {base_path}/{pdb_id}\") \n",
    "                continue\n",
    "            \n",
    "            # download the pdb_id directory\n",
    "            os.makedirs(local_dir, exist_ok=True) # create the directory if it doesn't exist\n",
    "            download_directory(sftp, f\"{base_path}/{pdb_id}\", local_dir)\n",
    "            print(f\"SBNA results for {pdb_id} succesfully downloaded\")\n",
    "        \n",
    "        except Exception:\n",
    "            # print full error message\n",
    "            print(f\"Failed to process {pdb_id} with error:\")\n",
    "            print(traceback.print_exc())\n",
    "\n",
    "            # remove directory\n",
    "            client.exec_command(f\"rm -r {base_path}/{pdb_id}/{pdb_id}_monomer\")\n",
    "            client.exec_command(f\"rm -r {base_path}/{pdb_id}/{pdb_id}_multimer\")\n",
    "            continue\n",
    "    \n",
    "    client.exec_command(f\"rm -rf {base_path}/{pdb_id}\") # remove the pdb_id remote directory\n",
    "    return\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "host = \"m3-dtn.massive.org.au\"\n",
    "username = \"yliy0004\"\n",
    "with open(\"pw.txt\", \"r\") as f:\n",
    "    password = f.read()\n",
    "base_path = 'ym65_scratch/yliy0004/NetworkAnalysis'\n",
    "\n",
    "client = paramiko.client.SSHClient()\n",
    "client.set_missing_host_key_policy(paramiko.AutoAddPolicy())\n",
    "client.connect(host, username=username, password=password)\n",
    "\n",
    "# check if R module is loaded (uses .bashrc - not the best solution because STRUDEL cannot be used, make sure to correct this if using STRUDEL)\n",
    "stdin,stdout,stderr=client.exec_command('R --version') \n",
    "print(stdout.read().decode())\n",
    "print(stderr.read().decode())\n",
    "\n",
    "sftp_client=client.open_sftp()\n",
    "\n",
    "# ==================== SBNA ====================\n",
    "# ## multiple structures by gene\n",
    "# for gene_id in top_structures:\n",
    "#     for pdb_id in top_structures[gene_id]:\n",
    "\n",
    "# ## or certain genes\n",
    "# # for gene in ['HRAS']:\n",
    "# #     count = 0\n",
    "#     # tmp = ast.literal_eval(cgc_data.loc[cgc_data['Gene Symbol'] == gene, 'PDB Structures'].values[0])\n",
    "#     # for pdb_id in tmp:\n",
    "#     # for pdb_id in top_structures[gene]:\n",
    "#         try:\n",
    "#             resolution = get_resolution(pdb_id)\n",
    "#             if resolution and float(resolution) <= 3.0:\n",
    "#                 calculate_sbna_and_download(client, sftp_client, pdb_id, base_path)\n",
    "#         except Exception as e:\n",
    "#             print(f\"Failed to process {pdb_id} with error:\")\n",
    "#             print(traceback.print_exc())\n",
    "#             continue\n",
    "\n",
    "## or simply a few pdbs\n",
    "for pdb_id in [\"1D5R\", \"1JM7\", \"1T29\", \"4FMQ\", \"4NIF\"]:\n",
    "    try:\n",
    "        calculate_sbna_and_download(client, sftp_client, pdb_id, base_path)\n",
    "    except Exception as e:\n",
    "        print(f\"Failed to process {pdb_id} with error:\")\n",
    "        print(traceback.print_exc())\n",
    "        continue\n",
    "\n",
    "## or one pdb\n",
    "# calculate_sbna_and_download(client, sftp_client, \"1TSR\", base_path, chains=['C'])\n",
    "\n",
    "# close the connection\n",
    "client.close()\n",
    "sftp_client.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "client.close()\n",
    "sftp_client.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# if you want to add more results to final_data\n",
    "(after having done more sbna iterations)\n",
    "* reads the pdb structures related to each gene in `cgc_data`\n",
    "* read the current `final_data.csv`\n",
    "* appends the new sbna_results to `final_data`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "cgc_data = pd.read_csv('Census_all_with_pdb.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Name</th>\n",
       "      <th>Entrez GeneId</th>\n",
       "      <th>Genome Location</th>\n",
       "      <th>Tier</th>\n",
       "      <th>Hallmark</th>\n",
       "      <th>Chr Band</th>\n",
       "      <th>Somatic</th>\n",
       "      <th>Germline</th>\n",
       "      <th>Tumour Types(Somatic)</th>\n",
       "      <th>...</th>\n",
       "      <th>Tissue Type</th>\n",
       "      <th>Molecular Genetics</th>\n",
       "      <th>Role in Cancer</th>\n",
       "      <th>Mutation Types</th>\n",
       "      <th>Translocation Partner</th>\n",
       "      <th>Other Germline Mut</th>\n",
       "      <th>Other Syndrome</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>PDB Structures</th>\n",
       "      <th>n_structures</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>29974.0</td>\n",
       "      <td>10:50799421-50885675</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>melanoma</td>\n",
       "      <td>...</td>\n",
       "      <td>E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACF,ACF64,ACF65,APOBEC1CF,ASP,CCDS73133.1,ENSG...</td>\n",
       "      <td>['2CPD']</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>abl-interactor 1</td>\n",
       "      <td>10006.0</td>\n",
       "      <td>10:26746593-26860935</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>10p12.1</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AML</td>\n",
       "      <td>...</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>TSG, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>KMT2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABI-1,CCDS7150.1,E3B1,ENSG00000136754.17,NM_00...</td>\n",
       "      <td>['7LXE']</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>v-abl Abelson murine leukemia viral oncogene h...</td>\n",
       "      <td>25.0</td>\n",
       "      <td>9:130713946-130885683</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>9q34.12</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CML, ALL, T-ALL</td>\n",
       "      <td>...</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T, Mis</td>\n",
       "      <td>BCR, ETV6, NUP214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL,CCDS35165.1,ENSG00000097007.17,JTK7,NM_007...</td>\n",
       "      <td>['1AB2', '1AWO', '1BBZ', '1JU5', '1OPL', '1ZZP...</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABL2</td>\n",
       "      <td>c-abl oncogene 2, non-receptor tyrosine kinase</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1:179099327-179229601</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1q25.2</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AML</td>\n",
       "      <td>...</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>ETV6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABLL,ARG,CCDS30947.1,ENSG00000143322.19,NM_007...</td>\n",
       "      <td>['2ECD', '2KK1', '2XYN', '3GVU', '3HMI', '3ULR...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACKR3</td>\n",
       "      <td>atypical chemokine receptor 3</td>\n",
       "      <td>57007.0</td>\n",
       "      <td>2:236569641-236582358</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2q37.3</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lipoma</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>HMGA2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCDS2516.1,CMKOR1,CXCR7,ENSG00000144476.5,GPR1...</td>\n",
       "      <td>['6K3F', '7SK3', '7SK4', '7SK5', '7SK6', '7SK7...</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>738</th>\n",
       "      <td>ZNF429</td>\n",
       "      <td>zinc finger protein 429</td>\n",
       "      <td>353088.0</td>\n",
       "      <td>19:21505564-21538078</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19p12</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GBM</td>\n",
       "      <td>...</td>\n",
       "      <td>O</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCDS42537.1,ENSG00000197013.9,NM_001001415.2,N...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>739</th>\n",
       "      <td>ZNF479</td>\n",
       "      <td>zinc finger protein 479</td>\n",
       "      <td>90827.0</td>\n",
       "      <td>7:57119614-57139864</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7p11.2</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lung cancer, bladder carcinoma, prostate carci...</td>\n",
       "      <td>...</td>\n",
       "      <td>E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCDS43590.1,ENSG00000185177.12,KR19,NM_033273....</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>740</th>\n",
       "      <td>ZNF521</td>\n",
       "      <td>zinc finger protein 521</td>\n",
       "      <td>25925.0</td>\n",
       "      <td>18:25061926-25352152</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18q11.2</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ALL</td>\n",
       "      <td>...</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>PAX5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCDS32806.1,EHZF,ENSG00000198795.10,Evi3,NM_01...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>741</th>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>zinc and ring finger 3</td>\n",
       "      <td>84133.0</td>\n",
       "      <td>22:28883592-29057487</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22q12.1</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>colorectal cancer, adrenocortical carcinoma, g...</td>\n",
       "      <td>...</td>\n",
       "      <td>E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TSG</td>\n",
       "      <td>N, F, Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BK747E2.3,CCDS56225.1,ENSG00000183579.15,FLJ22...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>742</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>zinc finger (CCCH type), RNA-binding motif and...</td>\n",
       "      <td>8233.0</td>\n",
       "      <td>X:15790472-15823260</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Xp22.2</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MDS, CLL</td>\n",
       "      <td>...</td>\n",
       "      <td>L</td>\n",
       "      <td>Rec</td>\n",
       "      <td>TSG</td>\n",
       "      <td>F, S, Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCDS14172.1,ENSG00000169249.12,NM_005089.3,NP_...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>743 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene Symbol                                               Name  \\\n",
       "0          A1CF                     APOBEC1 complementation factor   \n",
       "1          ABI1                                   abl-interactor 1   \n",
       "2          ABL1  v-abl Abelson murine leukemia viral oncogene h...   \n",
       "3          ABL2     c-abl oncogene 2, non-receptor tyrosine kinase   \n",
       "4         ACKR3                      atypical chemokine receptor 3   \n",
       "..          ...                                                ...   \n",
       "738      ZNF429                            zinc finger protein 429   \n",
       "739      ZNF479                            zinc finger protein 479   \n",
       "740      ZNF521                            zinc finger protein 521   \n",
       "741       ZNRF3                             zinc and ring finger 3   \n",
       "742       ZRSR2  zinc finger (CCCH type), RNA-binding motif and...   \n",
       "\n",
       "     Entrez GeneId        Genome Location  Tier Hallmark  Chr Band Somatic  \\\n",
       "0          29974.0   10:50799421-50885675     2      NaN  10q11.23     yes   \n",
       "1          10006.0   10:26746593-26860935     1      Yes   10p12.1     yes   \n",
       "2             25.0  9:130713946-130885683     1      Yes   9q34.12     yes   \n",
       "3             27.0  1:179099327-179229601     1      NaN    1q25.2     yes   \n",
       "4          57007.0  2:236569641-236582358     1      Yes    2q37.3     yes   \n",
       "..             ...                    ...   ...      ...       ...     ...   \n",
       "738       353088.0   19:21505564-21538078     2      NaN     19p12     yes   \n",
       "739        90827.0    7:57119614-57139864     2      NaN    7p11.2     yes   \n",
       "740        25925.0   18:25061926-25352152     1      NaN   18q11.2     yes   \n",
       "741        84133.0   22:28883592-29057487     2      NaN   22q12.1     yes   \n",
       "742         8233.0    X:15790472-15823260     1      Yes    Xp22.2     yes   \n",
       "\n",
       "    Germline                              Tumour Types(Somatic)  ...  \\\n",
       "0        NaN                                           melanoma  ...   \n",
       "1        NaN                                                AML  ...   \n",
       "2        NaN                                    CML, ALL, T-ALL  ...   \n",
       "3        NaN                                                AML  ...   \n",
       "4        NaN                                             lipoma  ...   \n",
       "..       ...                                                ...  ...   \n",
       "738      NaN                                                GBM  ...   \n",
       "739      NaN  lung cancer, bladder carcinoma, prostate carci...  ...   \n",
       "740      NaN                                                ALL  ...   \n",
       "741      NaN  colorectal cancer, adrenocortical carcinoma, g...  ...   \n",
       "742      NaN                                           MDS, CLL  ...   \n",
       "\n",
       "    Tissue Type Molecular Genetics    Role in Cancer Mutation Types  \\\n",
       "0             E                NaN          oncogene            Mis   \n",
       "1             L                Dom       TSG, fusion              T   \n",
       "2             L                Dom  oncogene, fusion         T, Mis   \n",
       "3             L                Dom  oncogene, fusion              T   \n",
       "4             M                Dom  oncogene, fusion              T   \n",
       "..          ...                ...               ...            ...   \n",
       "738           O                NaN               NaN            Mis   \n",
       "739           E                NaN               NaN            Mis   \n",
       "740           L                Dom  oncogene, fusion              T   \n",
       "741           E                NaN               TSG      N, F, Mis   \n",
       "742           L                Rec               TSG      F, S, Mis   \n",
       "\n",
       "    Translocation Partner Other Germline Mut Other Syndrome  \\\n",
       "0                     NaN                NaN            NaN   \n",
       "1                   KMT2A                NaN            NaN   \n",
       "2       BCR, ETV6, NUP214                NaN            NaN   \n",
       "3                    ETV6                NaN            NaN   \n",
       "4                   HMGA2                NaN            NaN   \n",
       "..                    ...                ...            ...   \n",
       "738                   NaN                NaN            NaN   \n",
       "739                   NaN                NaN            NaN   \n",
       "740                  PAX5                NaN            NaN   \n",
       "741                   NaN                NaN            NaN   \n",
       "742                   NaN                NaN            NaN   \n",
       "\n",
       "                                              Synonyms  \\\n",
       "0    ACF,ACF64,ACF65,APOBEC1CF,ASP,CCDS73133.1,ENSG...   \n",
       "1    ABI-1,CCDS7150.1,E3B1,ENSG00000136754.17,NM_00...   \n",
       "2    ABL,CCDS35165.1,ENSG00000097007.17,JTK7,NM_007...   \n",
       "3    ABLL,ARG,CCDS30947.1,ENSG00000143322.19,NM_007...   \n",
       "4    CCDS2516.1,CMKOR1,CXCR7,ENSG00000144476.5,GPR1...   \n",
       "..                                                 ...   \n",
       "738  CCDS42537.1,ENSG00000197013.9,NM_001001415.2,N...   \n",
       "739  CCDS43590.1,ENSG00000185177.12,KR19,NM_033273....   \n",
       "740  CCDS32806.1,EHZF,ENSG00000198795.10,Evi3,NM_01...   \n",
       "741  BK747E2.3,CCDS56225.1,ENSG00000183579.15,FLJ22...   \n",
       "742  CCDS14172.1,ENSG00000169249.12,NM_005089.3,NP_...   \n",
       "\n",
       "                                        PDB Structures n_structures  \n",
       "0                                             ['2CPD']            1  \n",
       "1                                             ['7LXE']            1  \n",
       "2    ['1AB2', '1AWO', '1BBZ', '1JU5', '1OPL', '1ZZP...           81  \n",
       "3    ['2ECD', '2KK1', '2XYN', '3GVU', '3HMI', '3ULR...            9  \n",
       "4    ['6K3F', '7SK3', '7SK4', '7SK5', '7SK6', '7SK7...            8  \n",
       "..                                                 ...          ...  \n",
       "738                                                NaN            0  \n",
       "739                                                NaN            0  \n",
       "740                                                NaN            0  \n",
       "741                                                NaN            0  \n",
       "742                                                NaN            0  \n",
       "\n",
       "[743 rows x 22 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cgc_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_uniprot_id(pdb_id, chain):\n",
    "    # read pdb file \n",
    "    with open(f'pdb_files/{pdb_id}.pdb', 'r') as f:\n",
    "        pdb_text = f.read()\n",
    "        for line in pdb_text.split('\\n'):\n",
    "            vals = line.split()\n",
    "            if line.startswith('DBREF') and chain in vals[2] and 'UNP' in vals:\n",
    "                return vals[6] # uniprot id is the 6th value\n",
    "\n",
    "\n",
    "def fetch_sequence(uniprot_id):\n",
    "    url = f\"https://www.uniprot.org/uniprot/{uniprot_id}.fasta\"\n",
    "    response = requests.get(url)\n",
    "    if response.ok:\n",
    "        # Extract the sequence from the FASTA format\n",
    "        lines = response.text.split('\\n')\n",
    "        sequence = ''.join(lines[1:])  # Skip the header line\n",
    "        return sequence\n",
    "    else:\n",
    "        print(\"Failed to fetch sequence\")\n",
    "        return None\n",
    "\n",
    "\n",
    "def get_gene_name(uniprot_id):\n",
    "    response = requests.get(f\"https://www.uniprot.org/uniprot/{uniprot_id}.fasta\")\n",
    "    response = response.text\n",
    "    match = re.search(r'GN=([^\\s]+)', response)\n",
    "    if match:\n",
    "        return match.group(1)\n",
    "    else:\n",
    "        return None\n",
    "    \n",
    "    \n",
    "def read_finalsum(pdb_id, chain):\n",
    "    if os.path.exists(f'sbna_results/{pdb_id}/{chain}/{pdb_id}_monomer/FinalSum'):\n",
    "        with open(f'sbna_results/{pdb_id}/{chain}/{pdb_id}_monomer/FinalSum', 'r') as f:\n",
    "            # convert final_sum to dataframe\n",
    "            final_sum = f.readlines()\n",
    "            final_sum = [i.split() for i in final_sum]\n",
    "            rows = []\n",
    "            for row in final_sum:\n",
    "                # get letters only from row[0]\n",
    "                aa_3 = ''.join(filter(str.isalpha, row[0]))\n",
    "                if row[1] == 'NA' or len(aa_3) != 3: \n",
    "                    # skip if residue number is NA or residue code is not 3 letters\n",
    "                    continue\n",
    "                \n",
    "                aa_1 = seq1(aa_3) # convert to 1 letter code\n",
    "                res_num = ''.join(filter(str.isnumeric, row[0]))\n",
    "                score = float(row[1])\n",
    "                # print(f\"{aa_3} {aa_1} {res_num} {score}\")\n",
    "                rows.append([aa_3, aa_1, int(res_num), score])\n",
    "            final_sum = pd.DataFrame(rows, columns=['res', 'res_code', 'num', 'score'])\n",
    "        return final_sum\n",
    "    else:\n",
    "        print(f\"FinalSum file not found for {pdb_id}\")\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp_genes = [\"KRAS\", \"EGFR\", \"TP53\", \"HRAS\"]\n",
    "# tmp_genes = [\"KRAS\", \"TP53\", \"HRAS\"]\n",
    "# tmp_genes = [\"EGFR\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_data = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_data = list(final_data.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_data = cgc_data[cgc_data['PDB Structures'].notna()] # pdb_structures not nan\n",
    "filtered_data = filtered_data[filtered_data['Gene Symbol'].isin(tmp_genes)]\n",
    "\n",
    "for i, row in filtered_data.iterrows():\n",
    "    gene_symbol = row['Gene Symbol']\n",
    "    pdb_structures = row['PDB Structures']\n",
    "    for pdb_id in ast.literal_eval(pdb_structures):\n",
    "        if not os.path.exists(f'pdb_files/{pdb_id}.pdb'):\n",
    "            continue\n",
    "        else:\n",
    "            data = get_dbref_data(pdb_id)\n",
    "            resolution = get_resolution(pdb_id)\n",
    "            for i in data:\n",
    "                chain = i['chain']\n",
    "\n",
    "                # check if finalsum exists or if data is already in final_data\n",
    "                if not os.path.exists(f'sbna_results/{pdb_id}/{chain}/{pdb_id}_monomer/FinalSum') or any([pdb_id == i[1] and chain == i[2] for i in final_data]):\n",
    "                    continue    \n",
    "                      \n",
    "                uniprot_id = i['uniprot']\n",
    "                assoc_gene = get_gene_name(uniprot_id)\n",
    "                if assoc_gene != gene_symbol:\n",
    "                    # sometimes protein chain extends multiple genes, skip if not the same\n",
    "                    # in future can link if gene is in cgc_data/tcga_data\n",
    "                    continue\n",
    "\n",
    "                # read finalSum file and add to table\n",
    "                final_sum = read_finalsum(pdb_id, chain)\n",
    "\n",
    "                # check if seqeunce of final sum matches uniprot\n",
    "                uniprot_seq = fetch_sequence(uniprot_id)\n",
    "                \n",
    "                for _, row in final_sum.iterrows():\n",
    "                    num = row['num']\n",
    "                    start = i['start']\n",
    "                    end = i['end']\n",
    "                    outside_range = False\n",
    "                    residue_match = False\n",
    "                    try:\n",
    "                        # get residue from uniprot sequence\n",
    "                        uniprot_res = uniprot_seq[num-1]\n",
    "                    except:\n",
    "                        uniprot_res = None\n",
    "\n",
    "                    if num < start or num > end: # check if residue number is within range\n",
    "                        outside_range = True\n",
    "                        # print(f\"Residue {num} not in range {start} - {end}\")\n",
    "                    if row['res_code'] == uniprot_res: # check if residue code matches\n",
    "                        residue_match = True\n",
    "\n",
    "                    final_data.append([gene_symbol, pdb_id, chain, uniprot_id, assoc_gene, resolution, num, row['res_code'], uniprot_res, row['score'], outside_range, residue_match])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>chain</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>assoc_gene</th>\n",
       "      <th>resolution</th>\n",
       "      <th>res_num</th>\n",
       "      <th>pdb_res</th>\n",
       "      <th>uniprot_res</th>\n",
       "      <th>network_score</th>\n",
       "      <th>outside_range</th>\n",
       "      <th>residue_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>3POZ</td>\n",
       "      <td>A</td>\n",
       "      <td>P00533</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>1.5</td>\n",
       "      <td>701</td>\n",
       "      <td>Q</td>\n",
       "      <td>Q</td>\n",
       "      <td>-3.791355</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>3POZ</td>\n",
       "      <td>A</td>\n",
       "      <td>P00533</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>1.5</td>\n",
       "      <td>702</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>-2.634055</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>3POZ</td>\n",
       "      <td>A</td>\n",
       "      <td>P00533</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>1.5</td>\n",
       "      <td>703</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>-1.231055</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>3POZ</td>\n",
       "      <td>A</td>\n",
       "      <td>P00533</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>1.5</td>\n",
       "      <td>704</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>0.452193</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>3POZ</td>\n",
       "      <td>A</td>\n",
       "      <td>P00533</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>1.5</td>\n",
       "      <td>705</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>-1.086163</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52394</th>\n",
       "      <td>TP53</td>\n",
       "      <td>8A92</td>\n",
       "      <td>B</td>\n",
       "      <td>P04637</td>\n",
       "      <td>TP53</td>\n",
       "      <td>1.37</td>\n",
       "      <td>288</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>-1.791900</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52395</th>\n",
       "      <td>TP53</td>\n",
       "      <td>8A92</td>\n",
       "      <td>B</td>\n",
       "      <td>P04637</td>\n",
       "      <td>TP53</td>\n",
       "      <td>1.37</td>\n",
       "      <td>289</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>-1.017026</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52396</th>\n",
       "      <td>TP53</td>\n",
       "      <td>8A92</td>\n",
       "      <td>B</td>\n",
       "      <td>P04637</td>\n",
       "      <td>TP53</td>\n",
       "      <td>1.37</td>\n",
       "      <td>290</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>-1.560145</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52397</th>\n",
       "      <td>TP53</td>\n",
       "      <td>8A92</td>\n",
       "      <td>B</td>\n",
       "      <td>P04637</td>\n",
       "      <td>TP53</td>\n",
       "      <td>1.37</td>\n",
       "      <td>291</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "      <td>-2.160481</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52398</th>\n",
       "      <td>TP53</td>\n",
       "      <td>8A92</td>\n",
       "      <td>B</td>\n",
       "      <td>P04637</td>\n",
       "      <td>TP53</td>\n",
       "      <td>1.37</td>\n",
       "      <td>292</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "      <td>-1.710703</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>52399 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_symbol pdb_id chain uniprot_id assoc_gene resolution  res_num  \\\n",
       "0            EGFR   3POZ     A     P00533       EGFR        1.5      701   \n",
       "1            EGFR   3POZ     A     P00533       EGFR        1.5      702   \n",
       "2            EGFR   3POZ     A     P00533       EGFR        1.5      703   \n",
       "3            EGFR   3POZ     A     P00533       EGFR        1.5      704   \n",
       "4            EGFR   3POZ     A     P00533       EGFR        1.5      705   \n",
       "...           ...    ...   ...        ...        ...        ...      ...   \n",
       "52394        TP53   8A92     B     P04637       TP53       1.37      288   \n",
       "52395        TP53   8A92     B     P04637       TP53       1.37      289   \n",
       "52396        TP53   8A92     B     P04637       TP53       1.37      290   \n",
       "52397        TP53   8A92     B     P04637       TP53       1.37      291   \n",
       "52398        TP53   8A92     B     P04637       TP53       1.37      292   \n",
       "\n",
       "      pdb_res uniprot_res  network_score  outside_range  residue_match  \n",
       "0           Q           Q      -3.791355          False           True  \n",
       "1           A           A      -2.634055          False           True  \n",
       "2           L           L      -1.231055          False           True  \n",
       "3           L           L       0.452193          False           True  \n",
       "4           R           R      -1.086163          False           True  \n",
       "...       ...         ...            ...            ...            ...  \n",
       "52394       N           N      -1.791900          False           True  \n",
       "52395       L           L      -1.017026          False           True  \n",
       "52396       R           R      -1.560145          False           True  \n",
       "52397       K           K      -2.160481          False           True  \n",
       "52398       K           K      -1.710703          False           True  \n",
       "\n",
       "[52399 rows x 12 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_data = pd.DataFrame(final_data, columns=['gene_symbol', 'pdb_id', 'chain', 'uniprot_id', 'assoc_gene', 'resolution', 'res_num', 'pdb_res', 'uniprot_res', 'network_score', 'outside_range', 'residue_match'])\n",
    "final_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_data = final_data.sort_values([\"assoc_gene\", \"pdb_id\", \"chain\", \"uniprot_id\"], ascending=True)\n",
    "final_data.to_csv('final_data.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_data = pd.read_csv('final_data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "assoc_gene\n",
       "EGFR     28\n",
       "HRAS     30\n",
       "KRAS     31\n",
       "TP53    109\n",
       "Name: pdb_id, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count number of structures for each gene\n",
    "final_data.groupby('assoc_gene')['pdb_id'].nunique()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
