{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\liyoa\\anaconda3\\envs\\sbna\\Lib\\site-packages\\paramiko\\transport.py:219: CryptographyDeprecationWarning: Blowfish has been deprecated and will be removed in a future release\n",
      "  \"class\": algorithms.Blowfish,\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 50)\n",
    "import ast\n",
    "import os\n",
    "import re\n",
    "\n",
    "# import from ../utils.py\n",
    "import sys\n",
    "sys.path.insert(1, os.path.join(sys.path[0], '..'))\n",
    "from utils import dwl_pdb_file,  get_resolution, get_dbref_data, get_gene_name, calculate_sbna_and_download, read_finalsum_decomp, read_finalsum, align_finalsum_with_uniprot\n",
    "\n",
    "import paramiko # for ssh\n",
    "import traceback\n",
    "\n",
    "import seaborn  as sns\n",
    "sns.set_style(\"darkgrid\")\n",
    "from matplotlib import pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Name</th>\n",
       "      <th>Entrez GeneId</th>\n",
       "      <th>Genome Location</th>\n",
       "      <th>Tier</th>\n",
       "      <th>Hallmark</th>\n",
       "      <th>Chr Band</th>\n",
       "      <th>Somatic</th>\n",
       "      <th>Germline</th>\n",
       "      <th>Tumour Types(Somatic)</th>\n",
       "      <th>Tumour Types(Germline)</th>\n",
       "      <th>Cancer Syndrome</th>\n",
       "      <th>Tissue Type</th>\n",
       "      <th>Molecular Genetics</th>\n",
       "      <th>Role in Cancer</th>\n",
       "      <th>Mutation Types</th>\n",
       "      <th>Translocation Partner</th>\n",
       "      <th>Other Germline Mut</th>\n",
       "      <th>Other Syndrome</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>PDB Structures</th>\n",
       "      <th>n_structures</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>29974.0</td>\n",
       "      <td>10:50799421-50885675</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>melanoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACF,ACF64,ACF65,APOBEC1CF,ASP,CCDS73133.1,ENSG...</td>\n",
       "      <td>['2CPD']</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>abl-interactor 1</td>\n",
       "      <td>10006.0</td>\n",
       "      <td>10:26746593-26860935</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>10p12.1</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>TSG, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>KMT2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABI-1,CCDS7150.1,E3B1,ENSG00000136754.17,NM_00...</td>\n",
       "      <td>['7LXE']</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>v-abl Abelson murine leukemia viral oncogene h...</td>\n",
       "      <td>25.0</td>\n",
       "      <td>9:130713946-130885683</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>9q34.12</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CML, ALL, T-ALL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T, Mis</td>\n",
       "      <td>BCR, ETV6, NUP214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL,CCDS35165.1,ENSG00000097007.17,JTK7,NM_007...</td>\n",
       "      <td>['1AB2', '1AWO', '1BBZ', '1JU5', '1OPL', '1ZZP...</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene Symbol                                               Name  \\\n",
       "0        A1CF                     APOBEC1 complementation factor   \n",
       "1        ABI1                                   abl-interactor 1   \n",
       "2        ABL1  v-abl Abelson murine leukemia viral oncogene h...   \n",
       "\n",
       "   Entrez GeneId        Genome Location  Tier Hallmark  Chr Band Somatic  \\\n",
       "0        29974.0   10:50799421-50885675     2      NaN  10q11.23     yes   \n",
       "1        10006.0   10:26746593-26860935     1      Yes   10p12.1     yes   \n",
       "2           25.0  9:130713946-130885683     1      Yes   9q34.12     yes   \n",
       "\n",
       "  Germline Tumour Types(Somatic) Tumour Types(Germline) Cancer Syndrome  \\\n",
       "0      NaN              melanoma                    NaN             NaN   \n",
       "1      NaN                   AML                    NaN             NaN   \n",
       "2      NaN       CML, ALL, T-ALL                    NaN             NaN   \n",
       "\n",
       "  Tissue Type Molecular Genetics    Role in Cancer Mutation Types  \\\n",
       "0           E                NaN          oncogene            Mis   \n",
       "1           L                Dom       TSG, fusion              T   \n",
       "2           L                Dom  oncogene, fusion         T, Mis   \n",
       "\n",
       "  Translocation Partner Other Germline Mut Other Syndrome  \\\n",
       "0                   NaN                NaN            NaN   \n",
       "1                 KMT2A                NaN            NaN   \n",
       "2     BCR, ETV6, NUP214                NaN            NaN   \n",
       "\n",
       "                                            Synonyms  \\\n",
       "0  ACF,ACF64,ACF65,APOBEC1CF,ASP,CCDS73133.1,ENSG...   \n",
       "1  ABI-1,CCDS7150.1,E3B1,ENSG00000136754.17,NM_00...   \n",
       "2  ABL,CCDS35165.1,ENSG00000097007.17,JTK7,NM_007...   \n",
       "\n",
       "                                      PDB Structures  n_structures  \n",
       "0                                           ['2CPD']             1  \n",
       "1                                           ['7LXE']             1  \n",
       "2  ['1AB2', '1AWO', '1BBZ', '1JU5', '1OPL', '1ZZP...            81  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cgc_data = pd.read_csv(\"../data_prep/Census_all_with_pdb.csv\")\n",
    "cgc_data.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes with PDB structures associated with lung cancer:  54\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Name</th>\n",
       "      <th>Entrez GeneId</th>\n",
       "      <th>Genome Location</th>\n",
       "      <th>Tier</th>\n",
       "      <th>Hallmark</th>\n",
       "      <th>Chr Band</th>\n",
       "      <th>Somatic</th>\n",
       "      <th>Germline</th>\n",
       "      <th>Tumour Types(Somatic)</th>\n",
       "      <th>Tumour Types(Germline)</th>\n",
       "      <th>Cancer Syndrome</th>\n",
       "      <th>Tissue Type</th>\n",
       "      <th>Molecular Genetics</th>\n",
       "      <th>Role in Cancer</th>\n",
       "      <th>Mutation Types</th>\n",
       "      <th>Translocation Partner</th>\n",
       "      <th>Other Germline Mut</th>\n",
       "      <th>Other Syndrome</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>PDB Structures</th>\n",
       "      <th>n_structures</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>v-akt murine thymoma viral oncogene homolog 1</td>\n",
       "      <td>207.0</td>\n",
       "      <td>14:104769349-104794124</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>14q32.33</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>breast, colorectal, ovarian, NSCLC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AKT,CCDS9994.1,ENSG00000142208.15,P31749,PKB,P...</td>\n",
       "      <td>['1H10', '1UNP', '1UNQ', '1UNR', '2UVM', '2UZR...</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ALK</td>\n",
       "      <td>anaplastic lymphoma kinase (Ki-1)</td>\n",
       "      <td>238.0</td>\n",
       "      <td>2:29192774-29921566</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2p23.2</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>ALCL, NSCLC, neuroblastoma, inflammatory myofi...</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>familial neuroblastoma</td>\n",
       "      <td>L, E, M</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T, Mis, A</td>\n",
       "      <td>NPM1, TPM3, TFG, TPM4, ATIC, CLTC, MSN, RNF213...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCDS33172.1,CD246,ENSG00000171094.17,NM_004304...</td>\n",
       "      <td>['2KUP', '2KUQ', '2XB7', '2XBA', '2XP2', '2YFX...</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARAF</td>\n",
       "      <td>A-Raf proto-oncogene, serine/threonine kinase</td>\n",
       "      <td>369.0</td>\n",
       "      <td>X:47561100-47571920</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Xp11.3</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cholangiocarcinoma, lung adenocarcinoma, Lange...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E, L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARAF1,CCDS75970.1,ENSG00000078061.12,NM_001256...</td>\n",
       "      <td>['1WXM', '2MSE']</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene Symbol                                           Name  Entrez GeneId  \\\n",
       "0        AKT1  v-akt murine thymoma viral oncogene homolog 1          207.0   \n",
       "1         ALK              anaplastic lymphoma kinase (Ki-1)          238.0   \n",
       "2        ARAF  A-Raf proto-oncogene, serine/threonine kinase          369.0   \n",
       "\n",
       "          Genome Location  Tier Hallmark  Chr Band Somatic Germline  \\\n",
       "0  14:104769349-104794124     1      Yes  14q32.33     yes      NaN   \n",
       "1     2:29192774-29921566     1      Yes    2p23.2     yes      yes   \n",
       "2     X:47561100-47571920     2      NaN    Xp11.3     yes      NaN   \n",
       "\n",
       "                               Tumour Types(Somatic) Tumour Types(Germline)  \\\n",
       "0                 breast, colorectal, ovarian, NSCLC                    NaN   \n",
       "1  ALCL, NSCLC, neuroblastoma, inflammatory myofi...          neuroblastoma   \n",
       "2  cholangiocarcinoma, lung adenocarcinoma, Lange...                    NaN   \n",
       "\n",
       "          Cancer Syndrome Tissue Type Molecular Genetics    Role in Cancer  \\\n",
       "0                     NaN           E                Dom          oncogene   \n",
       "1  familial neuroblastoma     L, E, M                Dom  oncogene, fusion   \n",
       "2                     NaN        E, L                NaN          oncogene   \n",
       "\n",
       "  Mutation Types                              Translocation Partner  \\\n",
       "0            Mis                                                NaN   \n",
       "1      T, Mis, A  NPM1, TPM3, TFG, TPM4, ATIC, CLTC, MSN, RNF213...   \n",
       "2            Mis                                                NaN   \n",
       "\n",
       "  Other Germline Mut Other Syndrome  \\\n",
       "0                NaN            NaN   \n",
       "1                NaN            NaN   \n",
       "2                NaN            NaN   \n",
       "\n",
       "                                            Synonyms  \\\n",
       "0  AKT,CCDS9994.1,ENSG00000142208.15,P31749,PKB,P...   \n",
       "1  CCDS33172.1,CD246,ENSG00000171094.17,NM_004304...   \n",
       "2  ARAF1,CCDS75970.1,ENSG00000078061.12,NM_001256...   \n",
       "\n",
       "                                      PDB Structures  n_structures  \n",
       "0  ['1H10', '1UNP', '1UNQ', '1UNR', '2UVM', '2UZR...            36  \n",
       "1  ['2KUP', '2KUQ', '2XB7', '2XBA', '2XP2', '2YFX...            77  \n",
       "2                                   ['1WXM', '2MSE']             2  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# if any contains \"lung\" or \"NSCLC\"\n",
    "lung_genes = cgc_data[(cgc_data['Tumour Types(Somatic)'].str.contains(\"lung|NSCLC\", case=False, na=False)) & (cgc_data['PDB Structures'].notnull())].reset_index(drop=True)\n",
    "print(\"Number of genes with PDB structures associated with lung cancer: \", lung_genes.shape[0])\n",
    "lung_genes.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download respective pdbs and read them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dwl_pdb_file(\"2OCJ\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "pdbs = ['7SI1', \"7UKV\", \"3POZ\", \"8F1X\", \"4LQM\", \"2ITZ\", \"6JWL\", \"5X27\", \"6GOD\", \"5UK9\", \\\n",
    "        \"6MBU\", \"4LDJ\", \"4L8G\", \"8AFB\", \"8AZX\", \"6GOF\", \"8TXG\", \"6GJ7\", \"6GOE\", \"8AZZ\", \\\n",
    "            \"7C40\", \"6MBQ\", \"4EPR\", \"3II5\", \"3Q4C\", \"3PRF\", \"3PPK\", \"3D4Q\", \"5FD2\", \"4FC0\", \\\n",
    "                \"4E26\", \"5C9C\", \"3IDP\", \"4MNF\", \"4G9R\", '4LMN', '4U7Z', '3E8N', '3DY7', '3V04', '4U81', '4U80', '3PP1', '7SI5', '7SI4', '3JPX', '3JZG', '3K26', '3K27', '5U5H', '5H13', '5U5K', '1Y6A', '1Y6B', '3VHK', '3VHE', '3VID']\n",
    "\n",
    "pdbs = ['6LVL', '6LVK', '2A91', '2QOO', '2QO2', '3FXX', '1UNR', '1UNQ', '2UZS', '1BG2', '2QOL', '2QO9', '2QOQ', '2QOI', '4J98', '4J96', '2GSF', '2UZR', '2QOF', '2UVM', '1MKJ', '2QOD', '2QOK', '3ETO', '2QOC', '2QO7', '5LT0', '3HMS']\n",
    "pdbs = ['2YJR', '3AOX', '4CMT', '4CTB', '4FNW', '4FNX', '4FNY', ]\n",
    "\n",
    "# all pdbs\n",
    "# pdbs =\n",
    "# for pdb in pdbs:\n",
    "for pdb in set(t):\n",
    "    print(dwl_pdb_file(pdb))\n",
    "\n",
    "# for pdb in [i[0] for i in t]:\n",
    "#     print(dwl_pdb_file(pdb))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading PDB file 1H10 for AKT1 (1911 structures remaining)\r"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[6], line 12\u001b[0m\n\u001b[0;32m     10\u001b[0m \u001b[38;5;66;03m# check if the file already exists\u001b[39;00m\n\u001b[0;32m     11\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mexists(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpdb_files/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mpdb_id\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.pdb\u001b[39m\u001b[38;5;124m'\u001b[39m):\n\u001b[1;32m---> 12\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m dwl_pdb_file(pdb_id): \u001b[38;5;66;03m# returns True or False if the file was downloaded\u001b[39;00m\n\u001b[0;32m     13\u001b[0m         \u001b[38;5;28;01mcontinue\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\liyoa\\OneDrive\\Documents\\MDS\\FIT5210\\code\\NetworkAnalysis\\lung_cancer\\..\\utils.py:28\u001b[0m, in \u001b[0;36mdwl_pdb_file\u001b[1;34m(pdb_id)\u001b[0m\n\u001b[0;32m     26\u001b[0m base_url \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhttps://files.rcsb.org/download/\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m     27\u001b[0m pdb_url \u001b[38;5;241m=\u001b[39m base_url \u001b[38;5;241m+\u001b[39m pdb_id \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.pdb\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m---> 28\u001b[0m response \u001b[38;5;241m=\u001b[39m requests\u001b[38;5;241m.\u001b[39mget(pdb_url)\n\u001b[0;32m     29\u001b[0m \u001b[38;5;66;03m# Check if the request was successful\u001b[39;00m\n\u001b[0;32m     30\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m response\u001b[38;5;241m.\u001b[39mstatus_code \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m200\u001b[39m:\n\u001b[0;32m     31\u001b[0m     \u001b[38;5;66;03m# Save the PDB file to pdb_files folder\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\liyoa\\anaconda3\\envs\\sbna\\Lib\\site-packages\\requests\\api.py:73\u001b[0m, in \u001b[0;36mget\u001b[1;34m(url, params, **kwargs)\u001b[0m\n\u001b[0;32m     62\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget\u001b[39m(url, params\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m     63\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[0;32m     64\u001b[0m \n\u001b[0;32m     65\u001b[0m \u001b[38;5;124;03m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m     70\u001b[0m \u001b[38;5;124;03m    :rtype: requests.Response\u001b[39;00m\n\u001b[0;32m     71\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m---> 73\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m request(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mget\u001b[39m\u001b[38;5;124m\"\u001b[39m, url, params\u001b[38;5;241m=\u001b[39mparams, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[1;32mc:\\Users\\liyoa\\anaconda3\\envs\\sbna\\Lib\\site-packages\\requests\\api.py:59\u001b[0m, in \u001b[0;36mrequest\u001b[1;34m(method, url, **kwargs)\u001b[0m\n\u001b[0;32m     55\u001b[0m \u001b[38;5;66;03m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[0;32m     56\u001b[0m \u001b[38;5;66;03m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[0;32m     57\u001b[0m \u001b[38;5;66;03m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[0;32m     58\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m sessions\u001b[38;5;241m.\u001b[39mSession() \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[1;32m---> 59\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m session\u001b[38;5;241m.\u001b[39mrequest(method\u001b[38;5;241m=\u001b[39mmethod, url\u001b[38;5;241m=\u001b[39murl, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[1;32mc:\\Users\\liyoa\\anaconda3\\envs\\sbna\\Lib\\site-packages\\requests\\sessions.py:589\u001b[0m, in \u001b[0;36mSession.request\u001b[1;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[0;32m    584\u001b[0m send_kwargs \u001b[38;5;241m=\u001b[39m {\n\u001b[0;32m    585\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m\"\u001b[39m: timeout,\n\u001b[0;32m    586\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m\"\u001b[39m: allow_redirects,\n\u001b[0;32m    587\u001b[0m }\n\u001b[0;32m    588\u001b[0m send_kwargs\u001b[38;5;241m.\u001b[39mupdate(settings)\n\u001b[1;32m--> 589\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msend(prep, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39msend_kwargs)\n\u001b[0;32m    591\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "File \u001b[1;32mc:\\Users\\liyoa\\anaconda3\\envs\\sbna\\Lib\\site-packages\\requests\\sessions.py:703\u001b[0m, in \u001b[0;36mSession.send\u001b[1;34m(self, request, **kwargs)\u001b[0m\n\u001b[0;32m    700\u001b[0m start \u001b[38;5;241m=\u001b[39m preferred_clock()\n\u001b[0;32m    702\u001b[0m \u001b[38;5;66;03m# Send the request\u001b[39;00m\n\u001b[1;32m--> 703\u001b[0m r \u001b[38;5;241m=\u001b[39m adapter\u001b[38;5;241m.\u001b[39msend(request, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[0;32m    705\u001b[0m \u001b[38;5;66;03m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[0;32m    706\u001b[0m elapsed \u001b[38;5;241m=\u001b[39m preferred_clock() \u001b[38;5;241m-\u001b[39m start\n",
      "File \u001b[1;32mc:\\Users\\liyoa\\anaconda3\\envs\\sbna\\Lib\\site-packages\\requests\\adapters.py:486\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[1;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[0;32m    483\u001b[0m     timeout \u001b[38;5;241m=\u001b[39m TimeoutSauce(connect\u001b[38;5;241m=\u001b[39mtimeout, read\u001b[38;5;241m=\u001b[39mtimeout)\n\u001b[0;32m    485\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m--> 486\u001b[0m     resp \u001b[38;5;241m=\u001b[39m conn\u001b[38;5;241m.\u001b[39murlopen(\n\u001b[0;32m    487\u001b[0m         method\u001b[38;5;241m=\u001b[39mrequest\u001b[38;5;241m.\u001b[39mmethod,\n\u001b[0;32m    488\u001b[0m         url\u001b[38;5;241m=\u001b[39murl,\n\u001b[0;32m    489\u001b[0m         body\u001b[38;5;241m=\u001b[39mrequest\u001b[38;5;241m.\u001b[39mbody,\n\u001b[0;32m    490\u001b[0m         headers\u001b[38;5;241m=\u001b[39mrequest\u001b[38;5;241m.\u001b[39mheaders,\n\u001b[0;32m    491\u001b[0m         redirect\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[0;32m    492\u001b[0m         assert_same_host\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[0;32m    493\u001b[0m         preload_content\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[0;32m    494\u001b[0m         decode_content\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[0;32m    495\u001b[0m         retries\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmax_retries,\n\u001b[0;32m    496\u001b[0m         timeout\u001b[38;5;241m=\u001b[39mtimeout,\n\u001b[0;32m    497\u001b[0m         chunked\u001b[38;5;241m=\u001b[39mchunked,\n\u001b[0;32m    498\u001b[0m     )\n\u001b[0;32m    500\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ProtocolError, \u001b[38;5;167;01mOSError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m    501\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m(err, request\u001b[38;5;241m=\u001b[39mrequest)\n",
      "File \u001b[1;32mc:\\Users\\liyoa\\anaconda3\\envs\\sbna\\Lib\\site-packages\\urllib3\\connectionpool.py:790\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[1;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, preload_content, decode_content, **response_kw)\u001b[0m\n\u001b[0;32m    787\u001b[0m response_conn \u001b[38;5;241m=\u001b[39m conn \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m release_conn \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m    789\u001b[0m \u001b[38;5;66;03m# Make the request on the HTTPConnection object\u001b[39;00m\n\u001b[1;32m--> 790\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_make_request(\n\u001b[0;32m    791\u001b[0m     conn,\n\u001b[0;32m    792\u001b[0m     method,\n\u001b[0;32m    793\u001b[0m     url,\n\u001b[0;32m    794\u001b[0m     timeout\u001b[38;5;241m=\u001b[39mtimeout_obj,\n\u001b[0;32m    795\u001b[0m     body\u001b[38;5;241m=\u001b[39mbody,\n\u001b[0;32m    796\u001b[0m     headers\u001b[38;5;241m=\u001b[39mheaders,\n\u001b[0;32m    797\u001b[0m     chunked\u001b[38;5;241m=\u001b[39mchunked,\n\u001b[0;32m    798\u001b[0m     retries\u001b[38;5;241m=\u001b[39mretries,\n\u001b[0;32m    799\u001b[0m     response_conn\u001b[38;5;241m=\u001b[39mresponse_conn,\n\u001b[0;32m    800\u001b[0m     preload_content\u001b[38;5;241m=\u001b[39mpreload_content,\n\u001b[0;32m    801\u001b[0m     decode_content\u001b[38;5;241m=\u001b[39mdecode_content,\n\u001b[0;32m    802\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mresponse_kw,\n\u001b[0;32m    803\u001b[0m )\n\u001b[0;32m    805\u001b[0m \u001b[38;5;66;03m# Everything went great!\u001b[39;00m\n\u001b[0;32m    806\u001b[0m clean_exit \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\liyoa\\anaconda3\\envs\\sbna\\Lib\\site-packages\\urllib3\\connectionpool.py:536\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[1;34m(self, conn, method, url, body, headers, retries, timeout, chunked, response_conn, preload_content, decode_content, enforce_content_length)\u001b[0m\n\u001b[0;32m    534\u001b[0m \u001b[38;5;66;03m# Receive the response from the server\u001b[39;00m\n\u001b[0;32m    535\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m--> 536\u001b[0m     response \u001b[38;5;241m=\u001b[39m conn\u001b[38;5;241m.\u001b[39mgetresponse()\n\u001b[0;32m    537\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (BaseSSLError, \u001b[38;5;167;01mOSError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m    538\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_raise_timeout(err\u001b[38;5;241m=\u001b[39me, url\u001b[38;5;241m=\u001b[39murl, timeout_value\u001b[38;5;241m=\u001b[39mread_timeout)\n",
      "File \u001b[1;32mc:\\Users\\liyoa\\anaconda3\\envs\\sbna\\Lib\\site-packages\\urllib3\\connection.py:454\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    451\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mresponse\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m HTTPResponse\n\u001b[0;32m    453\u001b[0m \u001b[38;5;66;03m# Get the response from http.client.HTTPConnection\u001b[39;00m\n\u001b[1;32m--> 454\u001b[0m httplib_response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39mgetresponse()\n\u001b[0;32m    456\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m    457\u001b[0m     assert_header_parsing(httplib_response\u001b[38;5;241m.\u001b[39mmsg)\n",
      "File \u001b[1;32mc:\\Users\\liyoa\\anaconda3\\envs\\sbna\\Lib\\http\\client.py:1423\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m   1421\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m   1422\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m-> 1423\u001b[0m         response\u001b[38;5;241m.\u001b[39mbegin()\n\u001b[0;32m   1424\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m:\n\u001b[0;32m   1425\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[1;32mc:\\Users\\liyoa\\anaconda3\\envs\\sbna\\Lib\\http\\client.py:331\u001b[0m, in \u001b[0;36mHTTPResponse.begin\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    329\u001b[0m \u001b[38;5;66;03m# read until we get a non-100 response\u001b[39;00m\n\u001b[0;32m    330\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[1;32m--> 331\u001b[0m     version, status, reason \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_read_status()\n\u001b[0;32m    332\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m status \u001b[38;5;241m!=\u001b[39m CONTINUE:\n\u001b[0;32m    333\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\liyoa\\anaconda3\\envs\\sbna\\Lib\\http\\client.py:292\u001b[0m, in \u001b[0;36mHTTPResponse._read_status\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    291\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_read_status\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m--> 292\u001b[0m     line \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfp\u001b[38;5;241m.\u001b[39mreadline(_MAXLINE \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m), \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124miso-8859-1\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m    293\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(line) \u001b[38;5;241m>\u001b[39m _MAXLINE:\n\u001b[0;32m    294\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m LineTooLong(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstatus line\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\liyoa\\anaconda3\\envs\\sbna\\Lib\\socket.py:707\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[1;34m(self, b)\u001b[0m\n\u001b[0;32m    705\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m    706\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m--> 707\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sock\u001b[38;5;241m.\u001b[39mrecv_into(b)\n\u001b[0;32m    708\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m timeout:\n\u001b[0;32m    709\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_timeout_occurred \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\liyoa\\anaconda3\\envs\\sbna\\Lib\\ssl.py:1252\u001b[0m, in \u001b[0;36mSSLSocket.recv_into\u001b[1;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[0;32m   1248\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m flags \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m   1249\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[0;32m   1250\u001b[0m           \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnon-zero flags not allowed in calls to recv_into() on \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m\n\u001b[0;32m   1251\u001b[0m           \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m)\n\u001b[1;32m-> 1252\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mread(nbytes, buffer)\n\u001b[0;32m   1253\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m   1254\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39mrecv_into(buffer, nbytes, flags)\n",
      "File \u001b[1;32mc:\\Users\\liyoa\\anaconda3\\envs\\sbna\\Lib\\ssl.py:1104\u001b[0m, in \u001b[0;36mSSLSocket.read\u001b[1;34m(self, len, buffer)\u001b[0m\n\u001b[0;32m   1102\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m   1103\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m buffer \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m-> 1104\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sslobj\u001b[38;5;241m.\u001b[39mread(\u001b[38;5;28mlen\u001b[39m, buffer)\n\u001b[0;32m   1105\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m   1106\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sslobj\u001b[38;5;241m.\u001b[39mread(\u001b[38;5;28mlen\u001b[39m)\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# count number of structures for all gene\n",
    "n_total = lung_genes['n_structures'].sum()\n",
    "\n",
    "for _, row in lung_genes.iterrows():\n",
    "    gene = row['Gene Symbol']\n",
    "    pdb_structures = ast.literal_eval(row['PDB Structures'])\n",
    "    for pdb_id in pdb_structures:\n",
    "        n_total -= 1\n",
    "        print(f\"Downloading PDB file {pdb_id} for {gene} ({n_total} structures remaining)\", end='\\r')\n",
    "        # check if the file already exists\n",
    "        if not os.path.exists(f'pdb_files/{pdb_id}.pdb'):\n",
    "            if not dwl_pdb_file(pdb_id): # returns True or False if the file was downloaded\n",
    "                continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdb_data = []\n",
    "uniprot_map = {}\n",
    "\n",
    "# count number of structures for all gene\n",
    "n_total = lung_genes['n_structures'].sum()\n",
    "\n",
    "for _, row in lung_genes.iterrows():\n",
    "    gene = row['Gene Symbol']\n",
    "    pdb_structures = ast.literal_eval(row['PDB Structures'])\n",
    "    for pdb_id in pdb_structures:\n",
    "        n_total -= 1\n",
    "        # check if the file already exists\n",
    "        if not os.path.exists(f'../pdb_files/{pdb_id}.pdb'):\n",
    "            continue\n",
    "        # read the file\n",
    "        resolution = get_resolution(pdb_id)\n",
    "        dbref = get_dbref_data(pdb_id)\n",
    "        for i in dbref:\n",
    "            chain = i['chain']\n",
    "            start = i['start']\n",
    "            end = i['end']\n",
    "            uniprot = i['uniprot']\n",
    "            if uniprot not in uniprot_map:\n",
    "                uniprot_map[uniprot] = get_gene_name(uniprot)\n",
    "            assoc_gene = uniprot_map[uniprot]\n",
    "            pdb_data.append([gene, pdb_id, resolution, chain, start, end, uniprot, assoc_gene])\n",
    "\n",
    "pdb_data = pd.DataFrame(pdb_data, columns=['gene', 'pdb_id', 'resolution', 'chain', 'seq_start', 'seq_end', 'uniprot_id', 'assoc_gene'])\n",
    "pdb_data.head()\n",
    "pdb_data['length'] = pdb_data['seq_end'] - pdb_data['seq_start'] + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sort gene ascending, length descending and resolution ascneding\n",
    "pdb_data.sort_values(by=['gene', 'length', 'resolution', \"pdb_id\"], ascending=[True, False, True, True]).to_csv(\"best_pdbs.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  Calculate SBNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "lung_genes = pd.read_excel('lung_genes.xlsx', sheet_name=\"final\")\n",
    "\n",
    "# groupby pdb and concatenate the chain in a list\n",
    "tmp = lung_genes.groupby('PDB')['Chain'].apply(list).reset_index()\n",
    "# lung_genes = lung_genes[:23]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'chain': 'A', 'uniprot': 'P04637', 'start': 94, 'end': 293}, {'chain': 'B', 'uniprot': 'P04637', 'start': 94, 'end': 293}, {'chain': 'C', 'uniprot': 'P04637', 'start': 94, 'end': 293}, {'chain': 'D', 'uniprot': 'P04637', 'start': 94, 'end': 293}]\n",
      "[{'chain': 'A', 'uniprot': 'P04637', 'start': 94, 'end': 312}, {'chain': 'B', 'uniprot': 'P04637', 'start': 94, 'end': 312}, {'chain': 'C', 'uniprot': 'P04637', 'start': 94, 'end': 312}, {'chain': 'D', 'uniprot': 'P04637', 'start': 94, 'end': 312}]\n",
      "[{'chain': 'A', 'uniprot': 'P04637', 'start': 94, 'end': 293}, {'chain': 'B', 'uniprot': 'P04637', 'start': 94, 'end': 293}, {'chain': 'C', 'uniprot': 'P04637', 'start': 94, 'end': 293}, {'chain': 'D', 'uniprot': 'P04637', 'start': 94, 'end': 293}]\n",
      "[{'chain': 'A', 'uniprot': 'P04637', 'start': 94, 'end': 293}, {'chain': 'B', 'uniprot': 'P04637', 'start': 94, 'end': 293}, {'chain': 'C', 'uniprot': 'P04637', 'start': 94, 'end': 293}, {'chain': 'D', 'uniprot': 'P04637', 'start': 94, 'end': 293}]\n",
      "[{'chain': 'A', 'uniprot': 'P04637', 'start': 94, 'end': 312}, {'chain': 'B', 'uniprot': 'P04637', 'start': 94, 'end': 312}, {'chain': 'C', 'uniprot': 'P04637', 'start': 94, 'end': 312}, {'chain': 'D', 'uniprot': 'P04637', 'start': 94, 'end': 312}]\n",
      "[{'chain': 'A', 'uniprot': 'P04637', 'start': 94, 'end': 293}, {'chain': 'B', 'uniprot': 'P04637', 'start': 94, 'end': 293}, {'chain': 'C', 'uniprot': 'P04637', 'start': 94, 'end': 293}, {'chain': 'D', 'uniprot': 'P04637', 'start': 94, 'end': 293}]\n",
      "[{'chain': 'A', 'uniprot': 'P04637', 'start': 97, 'end': 289}, {'chain': 'B', 'uniprot': 'P04637', 'start': 97, 'end': 289}, {'chain': 'C', 'uniprot': 'P04637', 'start': 97, 'end': 289}, {'chain': 'D', 'uniprot': 'P04637', 'start': 97, 'end': 289}]\n",
      "[{'chain': 'A', 'uniprot': 'P04637', 'start': 94, 'end': 293}, {'chain': 'B', 'uniprot': 'P04637', 'start': 94, 'end': 293}, {'chain': 'C', 'uniprot': 'P04637', 'start': 94, 'end': 293}, {'chain': 'D', 'uniprot': 'P04637', 'start': 94, 'end': 293}]\n",
      "[{'chain': 'A', 'uniprot': 'P04637', 'start': 94, 'end': 312}, {'chain': 'B', 'uniprot': 'P04637', 'start': 94, 'end': 312}, {'chain': 'C', 'uniprot': 'P04637', 'start': 94, 'end': 312}, {'chain': 'D', 'uniprot': 'P04637', 'start': 94, 'end': 312}]\n",
      "[{'chain': 'A', 'uniprot': 'P04637', 'start': 94, 'end': 293}, {'chain': 'B', 'uniprot': 'P04637', 'start': 94, 'end': 293}, {'chain': 'C', 'uniprot': 'P04637', 'start': 94, 'end': 293}, {'chain': 'D', 'uniprot': 'P04637', 'start': 94, 'end': 293}]\n"
     ]
    }
   ],
   "source": [
    "for pdb in set(t):\n",
    "    print(get_dbref_data(pdb))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['2YJR', '3AOX', '4CMT', '4CTB', '4FNW', '4FNX', '4FNY']"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdbs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under the terms of the\n",
      "GNU General Public License versions 2 or 3.\n",
      "For more information about these matters see\n",
      "https://www.gnu.org/licenses/.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Processing 7DHY with chain A...\n",
      "Starting phenix.pdbtools\n",
      "on Sun Jun  2 13:22:47 2024 by yliy0004\n",
      "===============================================================================\n",
      "\n",
      "Processing files:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Found model, 7DHY.pdb\n",
      "\n",
      "Processing PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Adding command-line PHIL:\n",
      "  -------------------------\n",
      "    keep=Chain A\n",
      "\n",
      "Final processed PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "  data_manager {\n",
      "    model {\n",
      "      file = \"7DHY.pdb\"\n",
      "    }\n",
      "    default_model = \"7DHY.pdb\"\n",
      "  }\n",
      "  modify {\n",
      "    keep = Chain A\n",
      "  }\n",
      "\n",
      "\n",
      "Starting job\n",
      "===============================================================================\n",
      "Validating inputs\n",
      "Performing manipulations\n",
      "Size before: 6723 size after: 1629\n",
      "Writing output model\n",
      "\n",
      "===============================================================================\n",
      "Job complete\n",
      "usr+sys time: 2.47 seconds\n",
      "wall clock time: 5.93 seconds\n",
      "Cleaning PDB file\n",
      "Starting phenix.pdbtools\n",
      "on Sun Jun  2 13:22:51 2024 by yliy0004\n",
      "===============================================================================\n",
      "\n",
      "Processing files:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Found model, 7DHY_monomer.pdb\n",
      "\n",
      "Processing PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Adding command-line PHIL:\n",
      "  -------------------------\n",
      "    remove=water\n",
      "\n",
      "Final processed PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "  data_manager {\n",
      "    model {\n",
      "      file = \"7DHY_monomer.pdb\"\n",
      "    }\n",
      "    default_model = \"7DHY_monomer.pdb\"\n",
      "  }\n",
      "  modify {\n",
      "    remove = water\n",
      "  }\n",
      "\n",
      "\n",
      "Starting job\n",
      "===============================================================================\n",
      "Validating inputs\n",
      "Performing manipulations\n",
      "Size before: 1629 size after: 1527\n",
      "Writing output model\n",
      "\n",
      "===============================================================================\n",
      "Job complete\n",
      "usr+sys time: 1.64 seconds\n",
      "wall clock time: 1.66 seconds\n",
      "Writing out polypeptide file...\n",
      "Adding hydrogens...\n",
      "Protonating waters...\n",
      "Making .phi file\n",
      "Running full bond calculation script\n",
      "The ligands identified in this pdb are ARS-301A\n",
      "Screening for atoms that are close.\n",
      "Dictionaries loaded.\n",
      "Starting salt bridges.\n",
      "Starting hydrogen bonds.\n",
      "Starting pi-pi version 2.\n",
      "Running pi-cation version 2\n",
      "Starting disulfide\n",
      "Starting van der waals.\n",
      "Starting metals\n",
      "Starting DNA bonds...\n",
      "Starting Ligand centroids\n",
      "Starting residue centroids...\n",
      "Removing duplicates from BFactor file\n",
      "Making 7DHY_monomer_nowaters_net\n",
      "Removing any negative value edges\n",
      "Creating QC file\n",
      "Creating Centroid directory and moving files\n",
      "Running energetic R script\n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> ##Load libraries\n",
      "> library(igraph)\n",
      "> library(stats)\n",
      "> \n",
      "> ##Initialize keywords\n",
      "> # keyword is the base filename\n",
      "> keyword = args[1]\n",
      "> args=c(paste(args[1],\"_net\",sep=\"\"),\"weighted\",paste(keyword,\"_secondaryStructure\",sep=\"\"))\n",
      "> atomCorrection=FALSE\n",
      "> baseKeyword = unlist(strsplit(keyword,\"/\"))[1]\n",
      "> runVariableSelection = TRUE\n",
      "> \n",
      "> ##Load inversal data\n",
      "> #Unique atoms file\n",
      "> terminalAtomsFile = as.matrix(read.table(\"../../terminalAtoms\",sep=\"\\t\"))\n",
      "> terminalAtoms = {}\n",
      "> for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+     terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\", \"))\n",
      "+ }\n",
      "> \n",
      "> #Secondary structure file\n",
      "> secStructure = as.matrix(read.table(args[3]))[,1:2]\n",
      "> rownames(secStructure) = secStructure[,1]\n",
      "> \n",
      "> #RSA\n",
      "> rsa = as.matrix(read.table(paste(keyword,\".rsa\",sep=\"\")))\n",
      "> \n",
      "> ##Define functions\n",
      "> \n",
      "> #Add terminal atom details\n",
      "> addTermDetails = function(data) {\n",
      "+     out_details = c()\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+         if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+             code1 = 1\n",
      "+         } else {\n",
      "+             code1 = 0\n",
      "+         }\n",
      "+         if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+             code2 = 1\n",
      "+         } else {\n",
      "+             code2 = 0\n",
      "+         }\n",
      "+     out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+     }\n",
      "+     return(out_details)\n",
      "+ }\n",
      "> \n",
      "> collapse = function(data) {\n",
      "+     if (nrow(data)==0) {\n",
      "+         out = matrix(ncol = 8,nrow=0)\n",
      "+         colnames(out) = c(\"AA1\", \"AA2\", \"Type\", \"SumEdges\", \"percentMC\", \"influence1\", \"influence2\",\"degree\")\n",
      "+         results = c(); results$out = out\n",
      "+         return(results)\n",
      "+         break\n",
      "+     }\n",
      "+     #Edges should be listed bi-directionally\n",
      "+     x=unlist(lapply(data[,3],function(x){if(x==\"MCSC\"){return(\"SCMC\")}else if(x==\"SCMC\"){return(\"MCSC\")} else {return(x)}}))\n",
      "+     if (dim(data)[2]==7) {\n",
      "+         data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6]))\n",
      "+     } else {\n",
      "+         data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6],data[,9],data[,8]))\n",
      "+     }\n",
      "+     data = unique(data)\n",
      "+     out = c()\n",
      "+     out_noPP = c()\n",
      "+     completed = c()\n",
      "+     if (dim(data)[2]==7) {\n",
      "+         out_details = c()\n",
      "+         for (i in 1:nrow(data)) {\n",
      "+             acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+             acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+             if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+                 code1 = 1\n",
      "+             } else {\n",
      "+                 code1 = 0\n",
      "+             }\n",
      "+             if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+                 code2 = 1\n",
      "+             } else {\n",
      "+                 code2 = 0\n",
      "+             }\n",
      "+             out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+         }\n",
      "+         data = out_details\n",
      "+     } else {\n",
      "+         out_details = data\n",
      "+     }\n",
      "+     colnames(out_details) = c(\"AA1\",\"AA2\",\"Type1\",\"Weight\",\"Type2\",\"Atom1\",\"Atom2\",\"influence1\",\"influence2\")\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         if (i %in% completed == FALSE) {\n",
      "+             x=which((data[,1]==data[i,1] & data[,2] == data[i,2]))\n",
      "+             subdata = matrix(data[x,],ncol=9)\n",
      "+             edgeSum = 0\n",
      "+             MCMCcount = 0\n",
      "+             MCSCcount = 0\n",
      "+             SCMCcount = 0\n",
      "+             SCSCcount = 0\n",
      "+             influence1 = sum(as.numeric(subdata[,8]))\n",
      "+             influence2 = sum(as.numeric(subdata[,9]))\n",
      "+             for (j in 1:nrow(subdata)) {\n",
      "+                 edgeSum = edgeSum + as.numeric(subdata[j,4])\n",
      "+                 if (subdata[j,3]==\"MCMC\" | subdata[j,3]==\"MCSC\") {\n",
      "+                     MCMCcount = MCMCcount+1\n",
      "+                     MCSCcount = MCSCcount+1\n",
      "+                 } else if (subdata[j,3]==\"SCMC\") {\n",
      "+                     SCMCcount = SCMCcount+1\n",
      "+                 } else if (subdata[j,3]==\"SCSC\") {\n",
      "+                     SCSCcount = SCSCcount+1\n",
      "+                 }\n",
      "+             }\n",
      "+             #Header is AA1, AA2, mixed, sum of edges, percent MC, influence1, influence2, degreeAA1\n",
      "+             out = rbind(out,c(data[i,1],data[i,2],\"mixed\",edgeSum,(MCMCcount+MCSCcount)/(MCMCcount+MCSCcount+SCMCcount+SCSCcount),influence1,influence2,sum(data[,1]==data[i,1])))\n",
      "+             if (any(subdata[,5]==\"PP\")==FALSE) {out_noPP = rbind(out_noPP,c(data[i,1],data[i,2],\"mixed\",edgeSum,(MCMCcount+MCSCcount)/(MCMCcount+MCSCcount+SCMCcount+SCSCcount),influence1,influence2,sum(data[,1]==data[i,1])))}\n",
      "+             completed = c(completed,x)\n",
      "+         }\n",
      "+     }\n",
      "+     out = out[out[,4]>0,]\n",
      "+     out_noPP = out_noPP[out_noPP[,4]>0,]\n",
      "+     colnames(out) = c(\"AA1\", \"AA2\", \"Type\", \"SumEdges\", \"percentMC\", \"influence1\", \"influence2\",\"degree\")\n",
      "+     results = c()\n",
      "+     results$out = out\n",
      "+     results$out_noPP = out_noPP\n",
      "+     results$influence_all = out_details[as.numeric(out_details[,4])>0 & (out_details[,8]==\"1\" | out_details[,9]==\"1\"),]\n",
      "+     out_details = out_details[as.numeric(out_details[,4])>0,]\n",
      "+     out_subtract = out_details; out_subtract[out_subtract[,8]==0 & out_subtract[,9]==0,4] = (-1)*as.numeric(out_subtract[out_subtract[,8]==0 & out_subtract[,9]==0,4])\n",
      "+     out_subtract = out_subtract[out_subtract[,5]!=\"PP\",]\n",
      "+     results$out_subtract = out_subtract\n",
      "+     return(results)\n",
      "+ }\n",
      "> \n",
      "> removeRedundancy = function(dataTmp) {\n",
      "+     for (i in 1:nrow(dataTmp)) {\n",
      "+         if (order(dataTmp[i,1:2])[1]==2) {\n",
      "+             dataTmp[i,1:2] = c(dataTmp[i,2],dataTmp[i,1])\n",
      "+             dataTmp[i,3] = paste(unlist(strsplit(dataTmp[i,3],split=\"\"))[c(3,4,1,2)],collapse=\"\")\n",
      "+             dataTmp[i,c(6,7)] = c(dataTmp[i,7],dataTmp[i,6])\n",
      "+         }\n",
      "+     }\n",
      "+     dataTmp = unique(dataTmp)\n",
      "+     pairs = unique(dataTmp[,1:2])\n",
      "+     out = c()\n",
      "+     for (j in 1:nrow(pairs)) {\n",
      "+         pair = pairs[j,]\n",
      "+         subData = matrix(dataTmp[dataTmp[,1]==pair[1] & dataTmp[,2]==pair[2],],ncol=7)\n",
      "+         out = rbind(out, c(pair[1],pair[2],\"mixed\",0.5*sum(as.numeric(subData[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> collapse_directed = function(data) {\n",
      "+     pairs = unique(data[,1:2])\n",
      "+     out = c()\n",
      "+     for (i in 1:nrow(pairs)) {\n",
      "+         subdat = matrix(data[data[,1]==pairs[i,1]&data[,2]==pairs[i,2],],ncol=4)\n",
      "+         out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> collapse_agnostic = function(data) {\n",
      "+     pairs = unique(t(apply(data[,1:2],1,function(x){return(sort(x))})))\n",
      "+     out = c()\n",
      "+     for (i in 1:nrow(pairs)) {\n",
      "+         subdat = matrix(data[(data[,1]==pairs[i,1]&data[,2]==pairs[i,2]) | (data[,2]==pairs[i,1]&data[,1]==pairs[i,2]),],ncol=4)\n",
      "+         out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> distributCalc = function(set,nodes,net) {\n",
      "+     #x is assumed to be a set but can also just be 1 node\n",
      "+     distanceNet = distances(net,weights=net$weight)\n",
      "+     distanceTmp = c()\n",
      "+     Dr = c()\n",
      "+     for (node in nodes) {\n",
      "+         distanceTmp = c()\n",
      "+         for (item in set) {\n",
      "+             if (item!=node) {\n",
      "+                 if(item%in%rownames(distanceNet)==FALSE){distanceTmp=100;next}\n",
      "+                 if(node%in%rownames(distanceNet)==FALSE){distanceTmp=100;next}\n",
      "+                 distanceTmp = c(distanceTmp,distanceNet[item,node])\n",
      "+             }\n",
      "+         }\n",
      "+         if (length(distanceTmp)>0) {\n",
      "+             Dr = c(Dr, 1/min(distanceTmp))\n",
      "+         }\n",
      "+     }\n",
      "+     return(sum(Dr)/length(nodes))\n",
      "+ }\n",
      "> \n",
      "> ##Read in user data\n",
      "> data_all = as.matrix(read.table(args[1]))\n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> originalNodes = nodes\n",
      "> basedata = collapse(data_all)\n",
      "> basedata_noPP = removeRedundancy(basedata$out[,1:7])\n",
      "> basedata_allsidechain = collapse_agnostic(data_all[data_all[,3]!=\"MCMC\",c(1,2,3,4)])\n",
      "> influencedata_all = basedata$influence_all\n",
      "> influencedata_collapse = collapse(influencedata_all)\n",
      "> influencedata = removeRedundancy(influencedata_collapse$out[,1:7])\n",
      "> influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "> influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "> \n",
      "> influencedata_directed = collapse_directed(influencedata_all[influencedata_all[,8]==1,c(1,2,3,4)])\n",
      "> influencenet_directed = graph.edgelist(influencedata_directed[,c(2,1)],directed=TRUE)\n",
      "> influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "> \n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> \n",
      "> #Write out edgelist files\n",
      "> write.table(influencedata,file=paste(keyword,\"_allEdges\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=FALSE,row.names=FALSE)\n",
      "> write.table(influencedata_directed,file=paste(keyword,\"_uniqueEdges\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=FALSE,row.names=FALSE)\n",
      "> \n",
      "> ###Deal with missing nodes\n",
      "> if (any(nodes %in% secStructure[,1]==FALSE)) {\n",
      "+     for (node in nodes) {\n",
      "+         if (node %in% secStructure[,1]==FALSE) {\n",
      "+             secStructure = rbind(secStructure,c(node,\"xxx\",360))\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> for (node in nodes) {\n",
      "+     if (node %in% rownames(rsa) == FALSE) {\n",
      "+         rsa = rbind(rsa,c(0,0))\n",
      "+         rownames(rsa)[nrow(rsa)] = node\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> ##Begin centrality calculations\n",
      "> \n",
      "> #Modular calculations\n",
      "> \n",
      "> ##WALKTRAP\n",
      "> net = graph.edgelist(basedata_noPP[,1:2],directed=FALSE)\n",
      "> net$weight = 1/as.numeric(basedata_noPP[,4])\n",
      "> net_community=walktrap.community(net,weights=net$weight); net_community_vec = net_community$membership; names(net_community_vec)=net_community$names\n",
      "> net_community_vec_wt = net_community_vec\n",
      "> nodes = V(net)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> nodes.color = c()\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(influencedata)) {\n",
      "+     j = which((basedata_noPP[,1]==influencedata[i,1]&basedata_noPP[,2]==influencedata[i,2])|(basedata_noPP[,1]==influencedata[i,2]&basedata_noPP[,2]==influencedata[i,1]))\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) { edgecolors[j]=\"grey40\" }\n",
      "+ }\n",
      "> if(length(nodes)>200) {textsize=.25} else {textsize=.45}\n",
      "> plot(net,edge.color = edgecolors, vertex.color = nodes.color,vertex.size=4,vertex.label.cex=textsize,main=\"WALKTRAP\")\n",
      ">     \n",
      "> ##Weighted edge betweenness WALKTRAP\n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP - WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for WALKTRAP\n",
      "> firstOrderDegree = degree(influencenet)\n",
      "> firstOrderDegree_sidechain = degree(influencenet_directed,mode=c(\"in\"))\n",
      "> secondOrderDegree = c(); secondOrderDegree_sidechain=c()\n",
      "> nodes = V(influencenet)$name\n",
      "> for (node in nodes) {\n",
      "+     firstorder = neighbors(influencenet,node)\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet,neighbor)))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree = c(secondOrderDegree,length(secondorder))\n",
      "+     firstorder = neighbors(influencenet_directed,node,mode=c(\"in\"))\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet_directed,neighbor,mode=c(\"in\"))))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree_sidechain = c(secondOrderDegree_sidechain,length(secondorder))\n",
      "+ }\n",
      "> names(secondOrderDegree) = nodes; names(secondOrderDegree_sidechain) = nodes\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])])) - 1\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])])) - 1\n",
      "+ }\n",
      "> \n",
      "> #SAVE ALL AS WALKTRAP\n",
      "> node_edge_betweenness_wt = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_wt = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_wt = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_wt = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_wt = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_wt = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> ##2ARY STRUCTURE\n",
      "> net_community_vec = as.numeric(as.factor(secStructure[,2]))\n",
      "> names(net_community_vec) = secStructure[,1]\n",
      "> nodes = V(net)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> nodes.color = c()\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(basedata_noPP)) {\n",
      "+     if (net_community_vec[basedata_noPP[i,1]]!=net_community_vec[basedata_noPP[i,2]]) { edgecolors[i]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for 2ARY STRUCTURE\n",
      "> nodes = V(influencenet)$name\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])])) - 1\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])])) - 1\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #SAVE ALL AS STRIDE\n",
      "> node_edge_betweenness_stride = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_stride = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_stride = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_stride = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_stride = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_stride = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> \n",
      "> ### Create final dataset\n",
      "> \n",
      "> out = cbind(\n",
      "+ firstOrderDegree[nodes],\n",
      "+ firstOrderDegree_sidechain[nodes],\n",
      "+ secondOrderDegree[nodes],\n",
      "+ secondOrderDegree_sidechain[nodes],\n",
      "+ node_edge_betweenness_stride[nodes],\n",
      "+ node_edge_betweenness_sidechain_stride[nodes],\n",
      "+ node_intermodular_degree_stride[nodes],\n",
      "+ node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ node_edge_betweenness_wt[nodes],\n",
      "+ node_edge_betweenness_sidechain_wt[nodes],\n",
      "+ node_intermodular_degree_wt[nodes],\n",
      "+ node_intermodular_degree_sidechain_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_wt[nodes]\n",
      "+ )\n",
      "> rownames(out) = nodes\n",
      "> colnames(out) = c(\"Degree\",\"Degree_sidechain\",\"SecondOrderDegree\",\"SecondOrderDegree_sidechain\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_sidechain\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_sidechain\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_sidechain\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_sidechain\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_sidechain\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_sidechain\")\n",
      "> \n",
      "> #add back in nodes that were not networked as zeros\n",
      "> zeroMat = matrix(0,nrow=length(which(originalNodes%in%nodes==FALSE)),ncol=ncol(out));rownames(zeroMat)=originalNodes[which(originalNodes%in%nodes==FALSE)]\n",
      "> out = rbind(out,zeroMat)\n",
      "> \n",
      "> #Add 1 to everything to avoid zeros, except ligand\n",
      "> out[is.na(out)]=0\n",
      "> out = out+1\n",
      "> \n",
      "> #Write out\n",
      "> write.table(out,file=paste(keyword,\"_scoresEnergetics\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "> #Standard normalization\n",
      "> outZ = apply(out,2,function(x){return(scale(x))})\n",
      "> rownames(outZ) = rownames(out)\n",
      "> write.table(outZ,file=paste(keyword,\"_scoresEnergeticsZ\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "Running centroid R script\n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(igraph)\n",
      "> library(stats)\n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> ##read in data\n",
      "> ## arg1: col1 and col2 are edges, col3 is weight of edge\n",
      "> ## arg2: weighted, unweighted <-- no longer using\n",
      "> ## arg3: forced module definition <-- no longer using\n",
      "> \n",
      "> # Set desired parameters\n",
      "> removeMCMC = FALSE\n",
      "> linearcutoff = 1\n",
      "> directed = FALSE\n",
      "> secondaryStructure2 = FALSE\n",
      "> sidechainMode = TRUE ##build network of just SCSC, SCMC and MCSC. Still use unique atoms for the _uniqueAtoms directed plots\n",
      "> uniqueAtomsMode = FALSE ##build network of just unique atoms (each edge has to have at least 1). Still use unique atoms for the _uniqueAtoms directed plots\n",
      "> useDNA=TRUE\n",
      "> uniqueAtomsUnbiased=FALSE\n",
      "> uniqueAtomsGaurav=FALSE\n",
      "> uniqueAtomsOLD=FALSE\n",
      "> atomCorrection = FALSE\n",
      "> uniqueAtomsGauravPP = FALSE\n",
      "> weighted = TRUE\n",
      "> removeWaters=TRUE\n",
      "> if (tail(unlist(strsplit(getwd(),\"/\")),n=1)==\"Centroid\") {\n",
      "+     ligandCentroidMode = TRUE\n",
      "+ } else {\n",
      "+     ligandCentroidMode = FALSE\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> # Read in ligand file\n",
      "> ligandfile = scan(paste(args[1],\"_ligand\",sep=\"\"),what=\"character\")\n",
      "> if (length(ligandfile)==0){\n",
      "+     ligandmode = FALSE\n",
      "+ } else {\n",
      "+     ligandmode = TRUE\n",
      "+ }\n",
      "> \n",
      "> ligandCentroidFile = scan(paste(args[1],\"NetLigand\",sep=\"\"),what=\"character\")\n",
      "> \n",
      "> # keyword is the base filename\n",
      "> # method is \"frag\" or \"module\" or \"both\"\n",
      "> keyword = args[1]\n",
      "> args=c(paste(args[1],\"_net\",sep=\"\"),\"weighted\",dir(\"./\",\"_secondaryStructure$\"))\n",
      "> \n",
      "> ##Load data\n",
      "> #Unique atoms file\n",
      "> if(uniqueAtomsUnbiased) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"/Users/vjpatel/Dropbox/Gaurav/uniqueAtomsUnbiased\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else if (uniqueAtomsGaurav) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"/Users/vjpatel/Dropbox/Gaurav/uniqueAtomsGaurav\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else if (uniqueAtomsOLD) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"../../../terminalAtoms\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"../../../uniqueAtoms\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     otherTA = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+         #otherTA[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,4],split=\",\"))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> #Secondary structure file\n",
      "> #Ligands are all one secondary structure\n",
      "> secStructure = as.matrix(read.table(args[3]))\n",
      "> phiAngleDB = {}\n",
      "> for (i in 1:nrow(secStructure)) {\n",
      "+     phiAngleDB[[secStructure[i,1]]] = as.numeric(secStructure[i,3])\n",
      "+ }\n",
      "> \n",
      "> if(ligandmode) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",\"_ligand\"),sep=\"\\t\"))\n",
      "+     liganddat = unique(liganddat[,2])\n",
      "+     for (ligand in liganddat) {\n",
      "+         secStructure = rbind(secStructure,c(sub(\"-\",\"\",ligand),\"ligand\",360))\n",
      "+         phiAngleDB[[sub(\"-\",\"\",ligand)]]=0\n",
      "+     }\n",
      "+ } else {\n",
      "+     liganddat = c()\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> ##Define functions\n",
      "> \n",
      "> ##Collapse matrix functions\n",
      "> ##Will return redundant matrix\n",
      "> ##Input needs to be 7 columns\n",
      "> ##Second input is different if the main input is a subset of the larger data\n",
      "> addTerminalDetails = function(data,largerdata) {\n",
      "+     #List edges bi-directionally\n",
      "+     x=unlist(lapply(data[,3],function(x){if(x==\"MCSC\"){return(\"SCMC\")}else if(x==\"SCMC\"){return(\"MCSC\")} else {return(x)}}))\n",
      "+     data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6]))\n",
      "+     data = unique(data)\n",
      "+     out = c()\n",
      "+     out_noPP = c()\n",
      "+     out_details = c()\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+         subdata = matrix(data[data[,1]==data[i,1]&data[,2]==data[i,2],],ncol=7)\n",
      "+         allsubdata = matrix(largerdata[largerdata[,1]==data[i,1],],ncol=7)\n",
      "+         code1=0;code2=0\n",
      "+         if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+             code1 = 1\n",
      "+         }\n",
      "+         if (data[i,5] == \"DNA\" & useDNA==TRUE) {\n",
      "+             code1 = 1\n",
      "+         }\n",
      "+         allsubdata = matrix(largerdata[largerdata[,2]==data[i,2],],ncol=7)\n",
      "+         if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+             code2 = 1\n",
      "+         }\n",
      "+         if (data[i,5] == \"DNA\" & useDNA==TRUE) {\n",
      "+             code2 = 1\n",
      "+         }\n",
      "+         if (data[i,5]==\"PICATION\"|data[i,5]==\"PIPI\") {\n",
      "+             code1=1\n",
      "+             code2=1\n",
      "+         }\n",
      "+         if (data[i,5]==\"PP\") {\n",
      "+             if (uniqueAtomsGaurav) {\n",
      "+                 if((acid1) == \"GLY\") {\n",
      "+                     code1 = 1\n",
      "+                     data[i,3]==\"SCSC\"\n",
      "+                 }\n",
      "+                 if ((acid2) == \"GLY\") {\n",
      "+                     code2 = 1\n",
      "+                     data[i,3] == \"SCSC\"\n",
      "+                 }\n",
      "+             }\n",
      "+         }\n",
      "+         out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+     }\n",
      "+     return(out_details)\n",
      "+ }\n",
      "> \n",
      "> ##Will return non-redundant matrix\n",
      "> collapse_agnostic = function(data) {\n",
      "+     if(nrow(data)==0) {\n",
      "+         return(matrix(ncol=4,nrow=0))\n",
      "+     } else if (ncol(data)==4 & data[1,3]==\"mixed\") {\n",
      "+         return(data)\n",
      "+     } else {\n",
      "+         pairs = unique(t(apply(matrix(data[,1:2],ncol=2),1,function(x){return(sort(x))})))\n",
      "+         out = c()\n",
      "+         for (i in 1:nrow(pairs)) {\n",
      "+             subdat = matrix(data[(data[,1]==pairs[i,1]&data[,2]==pairs[i,2]) | (data[,2]==pairs[i,1]&data[,1]==pairs[i,2]),1:4],ncol=4)\n",
      "+             subdat = unique(t(apply(subdat,1,function(x){a=which(x==sort(x[1:2])[1]);if(a==1){return(x)}else{return(c(x[2],x[1],paste(unlist(strsplit(x[3],\"\"))[c(3,4,1,2)],collapse=\"\"),x[4]))}})))\n",
      "+             if (weighted==FALSE) {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",1))\n",
      "+             } else {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+             }\n",
      "+         }\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> ##Will return directed collapsed matrix\n",
      "> collapse_directed = function(data) {\n",
      "+     if (length(data)==0) {\n",
      "+         return(matrix(nrow=0,ncol=4))\n",
      "+     } else {\n",
      "+         pairs = matrix(unique(data[,1:2]),ncol=2)\n",
      "+         out = c()\n",
      "+         for (i in 1:nrow(pairs)) {\n",
      "+             subdat = matrix(data[data[,1]==pairs[i,1]&data[,2]==pairs[i,2],1:4],ncol=4)\n",
      "+             if (weighted==FALSE) {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",1))\n",
      "+             } else {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+             }\n",
      "+         }\n",
      "+         return(matrix(out,ncol=4))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> rsa = as.matrix(read.table(dir(\"./\",\".rsa$\")))\n",
      "> rownames(rsa) = rsa[,1]\n",
      "> rsa[,1] = rsa[,2]\n",
      "> \n",
      "> #add ligand as -1\n",
      "> for (ligand in liganddat) {\n",
      "+     rsa = rbind(rsa,c(-1,-1))\n",
      "+     rownames(rsa)[nrow(rsa)] = sub(\"-\",\"\",ligand)\n",
      "+ }\n",
      "> \n",
      "> #Function to get amino acid\n",
      "> getAcid = function(residue) {\n",
      "+     if(all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"A\")) == TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"G\"))==TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"C\"))==TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"T\"))==TRUE)) {return(paste(unlist(strsplit(residue,\"\"))[1:2],collapse=\"\"))\n",
      "+     } else {\n",
      "+     return(paste(unlist(strsplit(residue,\"\"))[1:3],collapse=\"\"))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> ##################\n",
      "> ## read in data ##\n",
      "> ##################\n",
      "> data_all = as.matrix(read.table(args[1]))[,1:7]\n",
      "> if(removeWaters) {\n",
      "+     y=c(grep(\"HOH\",data_all[,1]),grep(\"HOH\",data_all[,2]))\n",
      "+     if(length(y)>0) {\n",
      "+         data_all = data_all[-y,]\n",
      "+     }\n",
      "+ }\n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> originalNodes = nodes\n",
      "> \n",
      "> data_all_original = rbind(data_all,data_all[,c(2,1,3,4,5,7,6)])\n",
      "> data_all = data_all[as.numeric(data_all[,4])>0,]\n",
      "> \n",
      "> basedata = addTerminalDetails(data_all,data_all_original)\n",
      "> basedata_noPP = collapse_agnostic(basedata[basedata[,5]!=\"PP\",])\n",
      "> sidechaindata = basedata[basedata[,3]!=\"MCMC\",1:4]\n",
      "> sidechaindata_detailed = basedata[basedata[,3]!=\"MCMC\",]\n",
      "> \n",
      "> influencedata_detailed = basedata\n",
      "> if (uniqueAtomsMode) {\n",
      "+     influencedata_detailed = basedata[basedata[,8]==1 | basedata[,9]==1,]\n",
      "+ }\n",
      "> influencedata = collapse_agnostic(influencedata_detailed)\n",
      "> influencedataGLY = rbind(influencedata[grep(\"GLY\",influencedata[,1]),],influencedata[grep(\"GLY\",influencedata[,1]),])\n",
      "> influencedata = influencedata[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,2])))>linearcutoff,]\n",
      "> \n",
      "> \n",
      "> influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "> influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "> \n",
      "> influencedata_directed = collapse_directed(matrix(influencedata_detailed[influencedata_detailed[,8]==1,1:4],ncol=4))\n",
      "> influencedata_directedGLY = rbind(influencedata_directed[grep(\"GLY\",influencedata_directed[,1]),],influencedata_directed[grep(\"GLY\",influencedata_directed[,2]),])\n",
      "> influencedata_directed = influencedata_directed[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,2])))>linearcutoff,]\n",
      "> \n",
      "> influencedata_directed = matrix(influencedata_directed,ncol=4)\n",
      "> influencenet_directed = graph.edgelist(matrix(influencedata_directed[,c(2,1)],ncol=2),directed=TRUE)\n",
      "> influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "> \n",
      "> \n",
      "> if(sidechainMode) {\n",
      "+     influencedata_detailed = sidechaindata_detailed\n",
      "+     influencedata = collapse_agnostic(influencedata_detailed)\n",
      "+     influencedataGLY = rbind(influencedata[grep(\"GLY\",influencedata[,1]),],influencedata[grep(\"GLY\",influencedata[,2]),])\n",
      "+     influencedata = influencedata[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,2])))>linearcutoff,]\n",
      "+     #Add back for GLY if uniqueAtomsGaurav\n",
      "+     if(uniqueAtomsGauravPP) {\n",
      "+         influencedata = rbind(influencedata,influencedataGLY)\n",
      "+     }\n",
      "+     influencedata = influencedata[as.numeric(influencedata[,4])>0,]\n",
      "+     \n",
      "+     influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "+     influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "+     \n",
      "+     influencedata_directed = collapse_directed(matrix(influencedata_detailed[influencedata_detailed[,8]==1,1:4],ncol=4))\n",
      "+     influencedata_directedGLY = rbind(influencedata_directed[grep(\"GLY\",influencedata_directed[,1]),],influencedata_directed[grep(\"GLY\",influencedata_directed[,2]),])\n",
      "+ \n",
      "+     influencedata_directed = influencedata_directed[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,2])))>linearcutoff,]\n",
      "+     #Add back for GLY if uniqueAtomsGaurav\n",
      "+     if(uniqueAtomsGauravPP) {\n",
      "+         influencedata_directed = rbind(influencedata_directed,influencedata_directedGLY)\n",
      "+     }\n",
      "+     influencedata_directed = matrix(influencedata_directed,ncol=4)\n",
      "+     influencenet_directed = graph.edgelist(matrix(influencedata_directed[,c(2,1)],ncol=2),directed=TRUE)\n",
      "+     influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "+     \n",
      "+ }\n",
      "> \n",
      "> #Replace missing nodes in the igraph net\n",
      "> nodesMissing = nodes[nodes %in% influencedata[,1]==FALSE & nodes %in% influencedata[,2]==FALSE]\n",
      "> \n",
      "> ##B-factor\n",
      "> bfactor = as.matrix(read.table(dir(\"./\",\"_Bfactor\"),row.names=1))\n",
      "> \n",
      "> #Replace missing nodes in secStructure matrix\n",
      "> if (any(nodes %in% secStructure[,1]==FALSE)) {\n",
      "+     for (node in nodes) {\n",
      "+         if (node %in% secStructure[,1]==FALSE) {\n",
      "+             secStructure = rbind(secStructure,c(node,\"xxx\",360))\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> for (node in nodes) {\n",
      "+     if (node %in% rownames(rsa) == FALSE) {\n",
      "+         rsa = rbind(rsa,c(0,0))\n",
      "+         rownames(rsa)[nrow(rsa)] = node\n",
      "+     }\n",
      "+     if (node %in% rownames(bfactor) == FALSE) {\n",
      "+         bfactor = rbind(bfactor,1)\n",
      "+         rownames(bfactor)[nrow(bfactor)] = node\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> \n",
      "> \n",
      "> #Give DNA attributes\n",
      "> if (\"DNA\" %in% data_all[,5]) {\n",
      "+     for (i in 1:nrow(data_all)) {\n",
      "+         if (data_all[i,5]==\"DNA\") {\n",
      "+             phiAngleDB[[data_all[i,1]]] = 90\n",
      "+             if (data_all[i,1] %in% rownames(rsa) == FALSE) {\n",
      "+                 rsa = rbind(rsa,c(0,0))\n",
      "+                 rownames(rsa)[nrow(rsa)] = data_all[i,1]\n",
      "+             }\n",
      "+             if (data_all[i,1] %in% rownames(secStructure) == FALSE) {\n",
      "+                 secStructure = rbind(secStructure,c(data_all[i,1],\"DNA\",360))\n",
      "+             }\n",
      "+             if (data_all[i,1] %in% rownames(bfactor) == FALSE) {\n",
      "+                 bfactor = rbind(bfactor,1)\n",
      "+                 rownames(bfactor)[nrow(bfactor)] = data_all[i,1]\n",
      "+             }\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #Betweenness\n",
      "> node_betweenness = betweenness(influencenet,weights=influencenet$weight,directed=directed)\n",
      "> node_betweenness_unique = betweenness(influencenet_directed,weights=influencenet_directed$weight,directed=directed)\n",
      "> \n",
      "> #Modular calculations\n",
      "> \n",
      "> ##WALKTRAP\n",
      "> net = graph.edgelist(basedata_noPP[,1:2],directed=FALSE)\n",
      "> net$weight = 1/as.numeric(basedata_noPP[,4])\n",
      "> net_community=walktrap.community(net,weights=net$weight); net_community_vec = net_community$membership; names(net_community_vec)=net_community$names\n",
      "> net_community_vec_wt = net_community_vec\n",
      "> #net_community_vec[buriednodes] = max(net_community_vec)+1\n",
      "> nodes = V(influencenet)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> nodes.color = c()\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(influencedata)) {\n",
      "+     j = which((basedata_noPP[,1]==influencedata[i,1]&basedata_noPP[,2]==influencedata[i,2])|(basedata_noPP[,1]==influencedata[i,2]&basedata_noPP[,2]==influencedata[i,1]))\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) { edgecolors[j]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP\n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP - WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         if(weight1==0 & weight2==0) {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]] + 0\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]] + 0\n",
      "+         } else {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for WALKTRAP\n",
      "> firstOrderDegree = degree(influencenet)\n",
      "> firstOrderDegree_sidechain = degree(influencenet_directed,mode=c(\"in\"))\n",
      "> \n",
      "> secondOrderDegree = c(); secondOrderDegree_sidechain=c()\n",
      "> nodes = V(influencenet)$name\n",
      "> for (node in nodes) {\n",
      "+     firstorder = neighbors(influencenet,node)\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet,neighbor)))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree = c(secondOrderDegree,length(secondorder))\n",
      "+     if(node %in% V(influencenet_directed)$name) {\n",
      "+         firstorder_sidechain = neighbors(influencenet_directed,node,mode=c(\"in\"))\n",
      "+     } else {\n",
      "+         firstorder_sidechain = c()\n",
      "+     }\n",
      "+     secondorder = c(); for (neighbor in firstorder_sidechain){secondorder = c(secondorder,names(neighbors(influencenet_directed,neighbor,mode=c(\"in\"))))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree_sidechain = c(secondOrderDegree_sidechain,length(secondorder))\n",
      "+ }\n",
      "> names(secondOrderDegree) = nodes; names(secondOrderDegree_sidechain) = nodes\n",
      "> \n",
      "> \n",
      "> #second order intermodular degree\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])]))\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])]))\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #SAVE ALL AS WALKTRAP\n",
      "> node_edge_betweenness_wt = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_wt = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_wt = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_wt = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_wt = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_wt = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> \n",
      "> ##2ARY STRUCTURE\n",
      "> secStructure = secStructure[,1:2]\n",
      "> rownames(secStructure) = secStructure[,1]\n",
      "> net_community_vec = as.numeric(as.factor(secStructure[,2]))\n",
      "> names(net_community_vec) = secStructure[,1]\n",
      "> #net_community_vec[buriednodes] = max(net_community_vec)+1\n",
      "> nodes = V(influencenet)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> nodes.color = c()\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(basedata_noPP)) {\n",
      "+     if (net_community_vec[basedata_noPP[i,1]]!=net_community_vec[basedata_noPP[i,2]]) { edgecolors[i]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         if(weight1==0 & weight2==0) {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]] + 0\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]] + 0\n",
      "+         } else {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> ##Degree and second order degree for 2ARY STRUCTURE\n",
      "> \n",
      "> #second order intermodular degree\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     #secondOrder_node_intermodular_degree[node] = length(unique(unlist(node_modules[names(node_modules[[node]])])))\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])]))\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])]))\n",
      "+ }\n",
      "> \n",
      "> #SAVE ALL AS STRIDE\n",
      "> node_edge_betweenness_stride = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_stride = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_stride = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_stride = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_stride = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_stride = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> ### Create final dataset\n",
      "> nodes = nodes[nodes!=\"DNA1000A\"]\n",
      "> nodes = nodes[nodes %in% c(\"DA\",\"DG\",\"DC\",\"DT\") == FALSE]\n",
      "> \n",
      "> ### Remove ligand amino acids from final dataset\n",
      "> ligands = sub(\"-\",\"\",liganddat)\n",
      "> nodes = nodes[nodes %in% ligands == FALSE]\n",
      "> \n",
      "> nodes[which(nodes %in% rownames(bfactor)==FALSE)]\n",
      "character(0)\n",
      "> \n",
      "> out = cbind(\n",
      "+ rsa[nodes,1],\n",
      "+ firstOrderDegree[nodes],\n",
      "+ firstOrderDegree_sidechain[nodes],\n",
      "+ secondOrderDegree[nodes],\n",
      "+ secondOrderDegree_sidechain[nodes],\n",
      "+ node_edge_betweenness_stride[nodes],\n",
      "+ node_edge_betweenness_sidechain_stride[nodes],\n",
      "+ node_intermodular_degree_stride[nodes],\n",
      "+ node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ node_edge_betweenness_wt[nodes],\n",
      "+ node_edge_betweenness_sidechain_wt[nodes],\n",
      "+ node_intermodular_degree_wt[nodes],\n",
      "+ node_intermodular_degree_sidechain_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_wt[nodes]\n",
      "+ )\n",
      "> \n",
      "> #write column names\n",
      "> colnames(out) = c(\"RSA\",\"Degree\",\"Degree_uniqueAtoms\",\"SecondOrderDegree\",\"SecondOrderDegree_uniqueAtoms\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_unqiueAtoms\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_uniqueAtoms\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_uniqueAtoms\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_uniqueAtoms\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_uniqueAtoms\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_uniqueAtoms\")\n",
      "> \n",
      "> \n",
      "> ## LIGANDS\n",
      "> if(ligandmode==TRUE & ligandCentroidMode==FALSE) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",\"_ligand\"),sep=\"\\t\"))\n",
      "+     \n",
      "+     tmp = c()\n",
      "+     i=1\n",
      "+     for (i in 1:nrow(liganddat)) {\n",
      "+         acid=paste(unlist(strsplit(liganddat[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         if (unlist(strsplit(liganddat[i,1],\"-\"))[2] %in% terminalAtoms[[acid]]) {\n",
      "+             tmp = rbind(tmp,c(unlist(strsplit(liganddat[i,1],\"-\"))[1],liganddat[i,2],liganddat[i,3]))\n",
      "+         }\n",
      "+     }\n",
      "+     \n",
      "+     #ligandvec = tapply(liganddat[,3],as.factor(liganddat[,1]),min)\n",
      "+     ligandvec = tapply(as.numeric(tmp[,3]),as.factor(tmp[,1]),min)\n",
      "+ }\n",
      "> \n",
      "> if(ligandmode==TRUE & ligandCentroidMode) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",paste(\"_centroidNetLigand$\",sep=\"\")),sep=\"\\t\"))\n",
      "+     \n",
      "+     tmp = c()\n",
      "+     i=1\n",
      "+     for (i in 1:nrow(liganddat)) {\n",
      "+         tmp = rbind(tmp,c(liganddat[i,1],liganddat[i,2],liganddat[i,3]))\n",
      "+     }\n",
      "+     \n",
      "+     ligandvec = tapply(tmp[,3],as.factor(tmp[,1]),min)\n",
      "+ }\n",
      "> \n",
      "> if (ligandmode==FALSE) {\n",
      "+     ligandvec = rep(0,length(nodes))\n",
      "+ }\n",
      "> \n",
      "> out = cbind(out,ligandvec[rownames(out)])\n",
      "> colnames(out) = c(\"RSA\",\"Degree\",\"Degree_uniqueAtoms\",\"SecondOrderDegree\",\"SecondOrderDegree_uniqueAtoms\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_unqiueAtoms\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_uniqueAtoms\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_uniqueAtoms\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_uniqueAtoms\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_uniqueAtoms\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_uniqueAtoms\",\"Ligand\")\n",
      "> \n",
      "> #add back in nodes that were not networked as zeros\n",
      "> zeroMat = matrix(0,nrow=length(which(originalNodes%in%nodes==FALSE)),ncol=ncol(out));rownames(zeroMat)=originalNodes[which(originalNodes%in%nodes==FALSE)]\n",
      "> out = rbind(out,zeroMat)\n",
      "> \n",
      "> #fix RSA and Bfactor and ligand\n",
      "> out[,\"RSA\"] = rsa[rownames(out),1]\n",
      "> out[,\"Ligand\"] = ligandvec[rownames(out)]\n",
      "> \n",
      "> #Add 1 to everything to avoid zeros, except ligand\n",
      "> out[is.na(out)]=0\n",
      "> out[out[,\"Ligand\"]==0,\"Ligand\"] = 150\n",
      "> out[1:nrow(out),2:ncol(out)] = as.numeric(out[1:nrow(out),2:ncol(out)])+1\n",
      "> \n",
      "> ##create Z score file\n",
      "> outZ = apply(out,2,function(x){return(scale(as.numeric(x)))})\n",
      "> rownames(outZ) = rownames(out)\n",
      "> colnames(outZ) = colnames(out)\n",
      "> outZ[outZ==NA]=0\n",
      "> outZ[outZ==\"NaN\"]=0\n",
      "> \n",
      "> #Write out scores and z-normed scores files\n",
      "> write.table(out,file=paste(keyword,\"_scoresCentroid\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> write.table(outZ,file=paste(keyword,\"_scoresCentroidZ\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "> \n",
      "Cleaning PDB file\n",
      "Starting phenix.pdbtools\n",
      "on Sun Jun  2 13:24:22 2024 by yliy0004\n",
      "===============================================================================\n",
      "\n",
      "Processing files:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Found model, 7DHY_multimer.pdb\n",
      "\n",
      "Processing PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Adding command-line PHIL:\n",
      "  -------------------------\n",
      "    remove=water\n",
      "\n",
      "Final processed PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "  data_manager {\n",
      "    model {\n",
      "      file = \"7DHY_multimer.pdb\"\n",
      "    }\n",
      "    default_model = \"7DHY_multimer.pdb\"\n",
      "  }\n",
      "  modify {\n",
      "    remove = water\n",
      "  }\n",
      "\n",
      "\n",
      "Starting job\n",
      "===============================================================================\n",
      "Validating inputs\n",
      "Performing manipulations\n",
      "Size before: 6723 size after: 6206\n",
      "Writing output model\n",
      "\n",
      "===============================================================================\n",
      "Job complete\n",
      "usr+sys time: 2.26 seconds\n",
      "wall clock time: 11.96 seconds\n",
      "Writing out polypeptide file...\n",
      "Adding hydrogens...\n",
      "Protonating waters...\n",
      "Making .phi file\n",
      "Running full bond calculation script\n",
      "The ligands identified in this pdb are GOL-306D, GOL-303C, GOL-303B, SO4-303D, ARS-301D, ARS-301C, ARS-301B, ARS-301A, GOL-304B, GOL-305D, GOL-304D, GOL-305B\n",
      "Screening for atoms that are close.\n",
      "Dictionaries loaded.\n",
      "Starting salt bridges.\n",
      "Starting hydrogen bonds.\n",
      "Starting pi-pi version 2.\n",
      "Running pi-cation version 2\n",
      "Starting disulfide\n",
      "Starting van der waals.\n",
      "Starting metals\n",
      "Starting DNA bonds...\n",
      "Starting Ligand centroids\n",
      "Starting residue centroids...\n",
      "Removing duplicates from BFactor file\n",
      "Making 7DHY_multimer_nowaters_net\n",
      "Removing any negative value edges\n",
      "Creating QC file\n",
      "Creating Centroid directory and moving files\n",
      "Running energetic R script\n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> ##Load libraries\n",
      "> library(igraph)\n",
      "> library(stats)\n",
      "> \n",
      "> ##Initialize keywords\n",
      "> # keyword is the base filename\n",
      "> keyword = args[1]\n",
      "> args=c(paste(args[1],\"_net\",sep=\"\"),\"weighted\",paste(keyword,\"_secondaryStructure\",sep=\"\"))\n",
      "> atomCorrection=FALSE\n",
      "> baseKeyword = unlist(strsplit(keyword,\"/\"))[1]\n",
      "> runVariableSelection = TRUE\n",
      "> \n",
      "> ##Load inversal data\n",
      "> #Unique atoms file\n",
      "> terminalAtomsFile = as.matrix(read.table(\"../../terminalAtoms\",sep=\"\\t\"))\n",
      "> terminalAtoms = {}\n",
      "> for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+     terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\", \"))\n",
      "+ }\n",
      "> \n",
      "> #Secondary structure file\n",
      "> secStructure = as.matrix(read.table(args[3]))[,1:2]\n",
      "> rownames(secStructure) = secStructure[,1]\n",
      "> \n",
      "> #RSA\n",
      "> rsa = as.matrix(read.table(paste(keyword,\".rsa\",sep=\"\")))\n",
      "> \n",
      "> ##Define functions\n",
      "> \n",
      "> #Add terminal atom details\n",
      "> addTermDetails = function(data) {\n",
      "+     out_details = c()\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+         if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+             code1 = 1\n",
      "+         } else {\n",
      "+             code1 = 0\n",
      "+         }\n",
      "+         if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+             code2 = 1\n",
      "+         } else {\n",
      "+             code2 = 0\n",
      "+         }\n",
      "+     out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+     }\n",
      "+     return(out_details)\n",
      "+ }\n",
      "> \n",
      "> collapse = function(data) {\n",
      "+     if (nrow(data)==0) {\n",
      "+         out = matrix(ncol = 8,nrow=0)\n",
      "+         colnames(out) = c(\"AA1\", \"AA2\", \"Type\", \"SumEdges\", \"percentMC\", \"influence1\", \"influence2\",\"degree\")\n",
      "+         results = c(); results$out = out\n",
      "+         return(results)\n",
      "+         break\n",
      "+     }\n",
      "+     #Edges should be listed bi-directionally\n",
      "+     x=unlist(lapply(data[,3],function(x){if(x==\"MCSC\"){return(\"SCMC\")}else if(x==\"SCMC\"){return(\"MCSC\")} else {return(x)}}))\n",
      "+     if (dim(data)[2]==7) {\n",
      "+         data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6]))\n",
      "+     } else {\n",
      "+         data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6],data[,9],data[,8]))\n",
      "+     }\n",
      "+     data = unique(data)\n",
      "+     out = c()\n",
      "+     out_noPP = c()\n",
      "+     completed = c()\n",
      "+     if (dim(data)[2]==7) {\n",
      "+         out_details = c()\n",
      "+         for (i in 1:nrow(data)) {\n",
      "+             acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+             acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+             if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+                 code1 = 1\n",
      "+             } else {\n",
      "+                 code1 = 0\n",
      "+             }\n",
      "+             if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+                 code2 = 1\n",
      "+             } else {\n",
      "+                 code2 = 0\n",
      "+             }\n",
      "+             out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+         }\n",
      "+         data = out_details\n",
      "+     } else {\n",
      "+         out_details = data\n",
      "+     }\n",
      "+     colnames(out_details) = c(\"AA1\",\"AA2\",\"Type1\",\"Weight\",\"Type2\",\"Atom1\",\"Atom2\",\"influence1\",\"influence2\")\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         if (i %in% completed == FALSE) {\n",
      "+             x=which((data[,1]==data[i,1] & data[,2] == data[i,2]))\n",
      "+             subdata = matrix(data[x,],ncol=9)\n",
      "+             edgeSum = 0\n",
      "+             MCMCcount = 0\n",
      "+             MCSCcount = 0\n",
      "+             SCMCcount = 0\n",
      "+             SCSCcount = 0\n",
      "+             influence1 = sum(as.numeric(subdata[,8]))\n",
      "+             influence2 = sum(as.numeric(subdata[,9]))\n",
      "+             for (j in 1:nrow(subdata)) {\n",
      "+                 edgeSum = edgeSum + as.numeric(subdata[j,4])\n",
      "+                 if (subdata[j,3]==\"MCMC\" | subdata[j,3]==\"MCSC\") {\n",
      "+                     MCMCcount = MCMCcount+1\n",
      "+                     MCSCcount = MCSCcount+1\n",
      "+                 } else if (subdata[j,3]==\"SCMC\") {\n",
      "+                     SCMCcount = SCMCcount+1\n",
      "+                 } else if (subdata[j,3]==\"SCSC\") {\n",
      "+                     SCSCcount = SCSCcount+1\n",
      "+                 }\n",
      "+             }\n",
      "+             #Header is AA1, AA2, mixed, sum of edges, percent MC, influence1, influence2, degreeAA1\n",
      "+             out = rbind(out,c(data[i,1],data[i,2],\"mixed\",edgeSum,(MCMCcount+MCSCcount)/(MCMCcount+MCSCcount+SCMCcount+SCSCcount),influence1,influence2,sum(data[,1]==data[i,1])))\n",
      "+             if (any(subdata[,5]==\"PP\")==FALSE) {out_noPP = rbind(out_noPP,c(data[i,1],data[i,2],\"mixed\",edgeSum,(MCMCcount+MCSCcount)/(MCMCcount+MCSCcount+SCMCcount+SCSCcount),influence1,influence2,sum(data[,1]==data[i,1])))}\n",
      "+             completed = c(completed,x)\n",
      "+         }\n",
      "+     }\n",
      "+     out = out[out[,4]>0,]\n",
      "+     out_noPP = out_noPP[out_noPP[,4]>0,]\n",
      "+     colnames(out) = c(\"AA1\", \"AA2\", \"Type\", \"SumEdges\", \"percentMC\", \"influence1\", \"influence2\",\"degree\")\n",
      "+     results = c()\n",
      "+     results$out = out\n",
      "+     results$out_noPP = out_noPP\n",
      "+     results$influence_all = out_details[as.numeric(out_details[,4])>0 & (out_details[,8]==\"1\" | out_details[,9]==\"1\"),]\n",
      "+     out_details = out_details[as.numeric(out_details[,4])>0,]\n",
      "+     out_subtract = out_details; out_subtract[out_subtract[,8]==0 & out_subtract[,9]==0,4] = (-1)*as.numeric(out_subtract[out_subtract[,8]==0 & out_subtract[,9]==0,4])\n",
      "+     out_subtract = out_subtract[out_subtract[,5]!=\"PP\",]\n",
      "+     results$out_subtract = out_subtract\n",
      "+     return(results)\n",
      "+ }\n",
      "> \n",
      "> removeRedundancy = function(dataTmp) {\n",
      "+     for (i in 1:nrow(dataTmp)) {\n",
      "+         if (order(dataTmp[i,1:2])[1]==2) {\n",
      "+             dataTmp[i,1:2] = c(dataTmp[i,2],dataTmp[i,1])\n",
      "+             dataTmp[i,3] = paste(unlist(strsplit(dataTmp[i,3],split=\"\"))[c(3,4,1,2)],collapse=\"\")\n",
      "+             dataTmp[i,c(6,7)] = c(dataTmp[i,7],dataTmp[i,6])\n",
      "+         }\n",
      "+     }\n",
      "+     dataTmp = unique(dataTmp)\n",
      "+     pairs = unique(dataTmp[,1:2])\n",
      "+     out = c()\n",
      "+     for (j in 1:nrow(pairs)) {\n",
      "+         pair = pairs[j,]\n",
      "+         subData = matrix(dataTmp[dataTmp[,1]==pair[1] & dataTmp[,2]==pair[2],],ncol=7)\n",
      "+         out = rbind(out, c(pair[1],pair[2],\"mixed\",0.5*sum(as.numeric(subData[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> collapse_directed = function(data) {\n",
      "+     pairs = unique(data[,1:2])\n",
      "+     out = c()\n",
      "+     for (i in 1:nrow(pairs)) {\n",
      "+         subdat = matrix(data[data[,1]==pairs[i,1]&data[,2]==pairs[i,2],],ncol=4)\n",
      "+         out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> collapse_agnostic = function(data) {\n",
      "+     pairs = unique(t(apply(data[,1:2],1,function(x){return(sort(x))})))\n",
      "+     out = c()\n",
      "+     for (i in 1:nrow(pairs)) {\n",
      "+         subdat = matrix(data[(data[,1]==pairs[i,1]&data[,2]==pairs[i,2]) | (data[,2]==pairs[i,1]&data[,1]==pairs[i,2]),],ncol=4)\n",
      "+         out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> distributCalc = function(set,nodes,net) {\n",
      "+     #x is assumed to be a set but can also just be 1 node\n",
      "+     distanceNet = distances(net,weights=net$weight)\n",
      "+     distanceTmp = c()\n",
      "+     Dr = c()\n",
      "+     for (node in nodes) {\n",
      "+         distanceTmp = c()\n",
      "+         for (item in set) {\n",
      "+             if (item!=node) {\n",
      "+                 if(item%in%rownames(distanceNet)==FALSE){distanceTmp=100;next}\n",
      "+                 if(node%in%rownames(distanceNet)==FALSE){distanceTmp=100;next}\n",
      "+                 distanceTmp = c(distanceTmp,distanceNet[item,node])\n",
      "+             }\n",
      "+         }\n",
      "+         if (length(distanceTmp)>0) {\n",
      "+             Dr = c(Dr, 1/min(distanceTmp))\n",
      "+         }\n",
      "+     }\n",
      "+     return(sum(Dr)/length(nodes))\n",
      "+ }\n",
      "> \n",
      "> ##Read in user data\n",
      "> data_all = as.matrix(read.table(args[1]))\n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> originalNodes = nodes\n",
      "> basedata = collapse(data_all)\n",
      "> basedata_noPP = removeRedundancy(basedata$out[,1:7])\n",
      "> basedata_allsidechain = collapse_agnostic(data_all[data_all[,3]!=\"MCMC\",c(1,2,3,4)])\n",
      "> influencedata_all = basedata$influence_all\n",
      "> influencedata_collapse = collapse(influencedata_all)\n",
      "> influencedata = removeRedundancy(influencedata_collapse$out[,1:7])\n",
      "> influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "> influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "> \n",
      "> influencedata_directed = collapse_directed(influencedata_all[influencedata_all[,8]==1,c(1,2,3,4)])\n",
      "> influencenet_directed = graph.edgelist(influencedata_directed[,c(2,1)],directed=TRUE)\n",
      "> influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "> \n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> \n",
      "> #Write out edgelist files\n",
      "> write.table(influencedata,file=paste(keyword,\"_allEdges\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=FALSE,row.names=FALSE)\n",
      "> write.table(influencedata_directed,file=paste(keyword,\"_uniqueEdges\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=FALSE,row.names=FALSE)\n",
      "> \n",
      "> ###Deal with missing nodes\n",
      "> if (any(nodes %in% secStructure[,1]==FALSE)) {\n",
      "+     for (node in nodes) {\n",
      "+         if (node %in% secStructure[,1]==FALSE) {\n",
      "+             secStructure = rbind(secStructure,c(node,\"xxx\",360))\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> for (node in nodes) {\n",
      "+     if (node %in% rownames(rsa) == FALSE) {\n",
      "+         rsa = rbind(rsa,c(0,0))\n",
      "+         rownames(rsa)[nrow(rsa)] = node\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> ##Begin centrality calculations\n",
      "> \n",
      "> #Modular calculations\n",
      "> \n",
      "> ##WALKTRAP\n",
      "> net = graph.edgelist(basedata_noPP[,1:2],directed=FALSE)\n",
      "> net$weight = 1/as.numeric(basedata_noPP[,4])\n",
      "> net_community=walktrap.community(net,weights=net$weight); net_community_vec = net_community$membership; names(net_community_vec)=net_community$names\n",
      "> net_community_vec_wt = net_community_vec\n",
      "> nodes = V(net)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> nodes.color = c()\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(influencedata)) {\n",
      "+     j = which((basedata_noPP[,1]==influencedata[i,1]&basedata_noPP[,2]==influencedata[i,2])|(basedata_noPP[,1]==influencedata[i,2]&basedata_noPP[,2]==influencedata[i,1]))\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) { edgecolors[j]=\"grey40\" }\n",
      "+ }\n",
      "> if(length(nodes)>200) {textsize=.25} else {textsize=.45}\n",
      "> plot(net,edge.color = edgecolors, vertex.color = nodes.color,vertex.size=4,vertex.label.cex=textsize,main=\"WALKTRAP\")\n",
      ">     \n",
      "> ##Weighted edge betweenness WALKTRAP\n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP - WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for WALKTRAP\n",
      "> firstOrderDegree = degree(influencenet)\n",
      "> firstOrderDegree_sidechain = degree(influencenet_directed,mode=c(\"in\"))\n",
      "> secondOrderDegree = c(); secondOrderDegree_sidechain=c()\n",
      "> nodes = V(influencenet)$name\n",
      "> for (node in nodes) {\n",
      "+     firstorder = neighbors(influencenet,node)\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet,neighbor)))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree = c(secondOrderDegree,length(secondorder))\n",
      "+     firstorder = neighbors(influencenet_directed,node,mode=c(\"in\"))\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet_directed,neighbor,mode=c(\"in\"))))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree_sidechain = c(secondOrderDegree_sidechain,length(secondorder))\n",
      "+ }\n",
      "> names(secondOrderDegree) = nodes; names(secondOrderDegree_sidechain) = nodes\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])])) - 1\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])])) - 1\n",
      "+ }\n",
      "> \n",
      "> #SAVE ALL AS WALKTRAP\n",
      "> node_edge_betweenness_wt = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_wt = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_wt = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_wt = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_wt = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_wt = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> ##2ARY STRUCTURE\n",
      "> net_community_vec = as.numeric(as.factor(secStructure[,2]))\n",
      "> names(net_community_vec) = secStructure[,1]\n",
      "> nodes = V(net)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> nodes.color = c()\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(basedata_noPP)) {\n",
      "+     if (net_community_vec[basedata_noPP[i,1]]!=net_community_vec[basedata_noPP[i,2]]) { edgecolors[i]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for 2ARY STRUCTURE\n",
      "> nodes = V(influencenet)$name\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])])) - 1\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])])) - 1\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #SAVE ALL AS STRIDE\n",
      "> node_edge_betweenness_stride = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_stride = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_stride = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_stride = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_stride = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_stride = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> \n",
      "> ### Create final dataset\n",
      "> \n",
      "> out = cbind(\n",
      "+ firstOrderDegree[nodes],\n",
      "+ firstOrderDegree_sidechain[nodes],\n",
      "+ secondOrderDegree[nodes],\n",
      "+ secondOrderDegree_sidechain[nodes],\n",
      "+ node_edge_betweenness_stride[nodes],\n",
      "+ node_edge_betweenness_sidechain_stride[nodes],\n",
      "+ node_intermodular_degree_stride[nodes],\n",
      "+ node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ node_edge_betweenness_wt[nodes],\n",
      "+ node_edge_betweenness_sidechain_wt[nodes],\n",
      "+ node_intermodular_degree_wt[nodes],\n",
      "+ node_intermodular_degree_sidechain_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_wt[nodes]\n",
      "+ )\n",
      "> rownames(out) = nodes\n",
      "> colnames(out) = c(\"Degree\",\"Degree_sidechain\",\"SecondOrderDegree\",\"SecondOrderDegree_sidechain\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_sidechain\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_sidechain\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_sidechain\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_sidechain\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_sidechain\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_sidechain\")\n",
      "> \n",
      "> #add back in nodes that were not networked as zeros\n",
      "> zeroMat = matrix(0,nrow=length(which(originalNodes%in%nodes==FALSE)),ncol=ncol(out));rownames(zeroMat)=originalNodes[which(originalNodes%in%nodes==FALSE)]\n",
      "> out = rbind(out,zeroMat)\n",
      "> \n",
      "> #Add 1 to everything to avoid zeros, except ligand\n",
      "> out[is.na(out)]=0\n",
      "> out = out+1\n",
      "> \n",
      "> #Write out\n",
      "> write.table(out,file=paste(keyword,\"_scoresEnergetics\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "> #Standard normalization\n",
      "> outZ = apply(out,2,function(x){return(scale(x))})\n",
      "> rownames(outZ) = rownames(out)\n",
      "> write.table(outZ,file=paste(keyword,\"_scoresEnergeticsZ\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "Running centroid R script\n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(igraph)\n",
      "> library(stats)\n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> ##read in data\n",
      "> ## arg1: col1 and col2 are edges, col3 is weight of edge\n",
      "> ## arg2: weighted, unweighted <-- no longer using\n",
      "> ## arg3: forced module definition <-- no longer using\n",
      "> \n",
      "> # Set desired parameters\n",
      "> removeMCMC = FALSE\n",
      "> linearcutoff = 1\n",
      "> directed = FALSE\n",
      "> secondaryStructure2 = FALSE\n",
      "> sidechainMode = TRUE ##build network of just SCSC, SCMC and MCSC. Still use unique atoms for the _uniqueAtoms directed plots\n",
      "> uniqueAtomsMode = FALSE ##build network of just unique atoms (each edge has to have at least 1). Still use unique atoms for the _uniqueAtoms directed plots\n",
      "> useDNA=TRUE\n",
      "> uniqueAtomsUnbiased=FALSE\n",
      "> uniqueAtomsGaurav=FALSE\n",
      "> uniqueAtomsOLD=FALSE\n",
      "> atomCorrection = FALSE\n",
      "> uniqueAtomsGauravPP = FALSE\n",
      "> weighted = TRUE\n",
      "> removeWaters=TRUE\n",
      "> if (tail(unlist(strsplit(getwd(),\"/\")),n=1)==\"Centroid\") {\n",
      "+     ligandCentroidMode = TRUE\n",
      "+ } else {\n",
      "+     ligandCentroidMode = FALSE\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> # Read in ligand file\n",
      "> ligandfile = scan(paste(args[1],\"_ligand\",sep=\"\"),what=\"character\")\n",
      "> if (length(ligandfile)==0){\n",
      "+     ligandmode = FALSE\n",
      "+ } else {\n",
      "+     ligandmode = TRUE\n",
      "+ }\n",
      "> \n",
      "> ligandCentroidFile = scan(paste(args[1],\"NetLigand\",sep=\"\"),what=\"character\")\n",
      "> \n",
      "> # keyword is the base filename\n",
      "> # method is \"frag\" or \"module\" or \"both\"\n",
      "> keyword = args[1]\n",
      "> args=c(paste(args[1],\"_net\",sep=\"\"),\"weighted\",dir(\"./\",\"_secondaryStructure$\"))\n",
      "> \n",
      "> ##Load data\n",
      "> #Unique atoms file\n",
      "> if(uniqueAtomsUnbiased) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"/Users/vjpatel/Dropbox/Gaurav/uniqueAtomsUnbiased\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else if (uniqueAtomsGaurav) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"/Users/vjpatel/Dropbox/Gaurav/uniqueAtomsGaurav\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else if (uniqueAtomsOLD) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"../../../terminalAtoms\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"../../../uniqueAtoms\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     otherTA = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+         #otherTA[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,4],split=\",\"))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> #Secondary structure file\n",
      "> #Ligands are all one secondary structure\n",
      "> secStructure = as.matrix(read.table(args[3]))\n",
      "> phiAngleDB = {}\n",
      "> for (i in 1:nrow(secStructure)) {\n",
      "+     phiAngleDB[[secStructure[i,1]]] = as.numeric(secStructure[i,3])\n",
      "+ }\n",
      "> \n",
      "> if(ligandmode) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",\"_ligand\"),sep=\"\\t\"))\n",
      "+     liganddat = unique(liganddat[,2])\n",
      "+     for (ligand in liganddat) {\n",
      "+         secStructure = rbind(secStructure,c(sub(\"-\",\"\",ligand),\"ligand\",360))\n",
      "+         phiAngleDB[[sub(\"-\",\"\",ligand)]]=0\n",
      "+     }\n",
      "+ } else {\n",
      "+     liganddat = c()\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> ##Define functions\n",
      "> \n",
      "> ##Collapse matrix functions\n",
      "> ##Will return redundant matrix\n",
      "> ##Input needs to be 7 columns\n",
      "> ##Second input is different if the main input is a subset of the larger data\n",
      "> addTerminalDetails = function(data,largerdata) {\n",
      "+     #List edges bi-directionally\n",
      "+     x=unlist(lapply(data[,3],function(x){if(x==\"MCSC\"){return(\"SCMC\")}else if(x==\"SCMC\"){return(\"MCSC\")} else {return(x)}}))\n",
      "+     data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6]))\n",
      "+     data = unique(data)\n",
      "+     out = c()\n",
      "+     out_noPP = c()\n",
      "+     out_details = c()\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+         subdata = matrix(data[data[,1]==data[i,1]&data[,2]==data[i,2],],ncol=7)\n",
      "+         allsubdata = matrix(largerdata[largerdata[,1]==data[i,1],],ncol=7)\n",
      "+         code1=0;code2=0\n",
      "+         if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+             code1 = 1\n",
      "+         }\n",
      "+         if (data[i,5] == \"DNA\" & useDNA==TRUE) {\n",
      "+             code1 = 1\n",
      "+         }\n",
      "+         allsubdata = matrix(largerdata[largerdata[,2]==data[i,2],],ncol=7)\n",
      "+         if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+             code2 = 1\n",
      "+         }\n",
      "+         if (data[i,5] == \"DNA\" & useDNA==TRUE) {\n",
      "+             code2 = 1\n",
      "+         }\n",
      "+         if (data[i,5]==\"PICATION\"|data[i,5]==\"PIPI\") {\n",
      "+             code1=1\n",
      "+             code2=1\n",
      "+         }\n",
      "+         if (data[i,5]==\"PP\") {\n",
      "+             if (uniqueAtomsGaurav) {\n",
      "+                 if((acid1) == \"GLY\") {\n",
      "+                     code1 = 1\n",
      "+                     data[i,3]==\"SCSC\"\n",
      "+                 }\n",
      "+                 if ((acid2) == \"GLY\") {\n",
      "+                     code2 = 1\n",
      "+                     data[i,3] == \"SCSC\"\n",
      "+                 }\n",
      "+             }\n",
      "+         }\n",
      "+         out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+     }\n",
      "+     return(out_details)\n",
      "+ }\n",
      "> \n",
      "> ##Will return non-redundant matrix\n",
      "> collapse_agnostic = function(data) {\n",
      "+     if(nrow(data)==0) {\n",
      "+         return(matrix(ncol=4,nrow=0))\n",
      "+     } else if (ncol(data)==4 & data[1,3]==\"mixed\") {\n",
      "+         return(data)\n",
      "+     } else {\n",
      "+         pairs = unique(t(apply(matrix(data[,1:2],ncol=2),1,function(x){return(sort(x))})))\n",
      "+         out = c()\n",
      "+         for (i in 1:nrow(pairs)) {\n",
      "+             subdat = matrix(data[(data[,1]==pairs[i,1]&data[,2]==pairs[i,2]) | (data[,2]==pairs[i,1]&data[,1]==pairs[i,2]),1:4],ncol=4)\n",
      "+             subdat = unique(t(apply(subdat,1,function(x){a=which(x==sort(x[1:2])[1]);if(a==1){return(x)}else{return(c(x[2],x[1],paste(unlist(strsplit(x[3],\"\"))[c(3,4,1,2)],collapse=\"\"),x[4]))}})))\n",
      "+             if (weighted==FALSE) {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",1))\n",
      "+             } else {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+             }\n",
      "+         }\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> ##Will return directed collapsed matrix\n",
      "> collapse_directed = function(data) {\n",
      "+     if (length(data)==0) {\n",
      "+         return(matrix(nrow=0,ncol=4))\n",
      "+     } else {\n",
      "+         pairs = matrix(unique(data[,1:2]),ncol=2)\n",
      "+         out = c()\n",
      "+         for (i in 1:nrow(pairs)) {\n",
      "+             subdat = matrix(data[data[,1]==pairs[i,1]&data[,2]==pairs[i,2],1:4],ncol=4)\n",
      "+             if (weighted==FALSE) {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",1))\n",
      "+             } else {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+             }\n",
      "+         }\n",
      "+         return(matrix(out,ncol=4))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> rsa = as.matrix(read.table(dir(\"./\",\".rsa$\")))\n",
      "> rownames(rsa) = rsa[,1]\n",
      "> rsa[,1] = rsa[,2]\n",
      "> \n",
      "> #add ligand as -1\n",
      "> for (ligand in liganddat) {\n",
      "+     rsa = rbind(rsa,c(-1,-1))\n",
      "+     rownames(rsa)[nrow(rsa)] = sub(\"-\",\"\",ligand)\n",
      "+ }\n",
      "> \n",
      "> #Function to get amino acid\n",
      "> getAcid = function(residue) {\n",
      "+     if(all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"A\")) == TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"G\"))==TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"C\"))==TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"T\"))==TRUE)) {return(paste(unlist(strsplit(residue,\"\"))[1:2],collapse=\"\"))\n",
      "+     } else {\n",
      "+     return(paste(unlist(strsplit(residue,\"\"))[1:3],collapse=\"\"))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> ##################\n",
      "> ## read in data ##\n",
      "> ##################\n",
      "> data_all = as.matrix(read.table(args[1]))[,1:7]\n",
      "> if(removeWaters) {\n",
      "+     y=c(grep(\"HOH\",data_all[,1]),grep(\"HOH\",data_all[,2]))\n",
      "+     if(length(y)>0) {\n",
      "+         data_all = data_all[-y,]\n",
      "+     }\n",
      "+ }\n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> originalNodes = nodes\n",
      "> \n",
      "> data_all_original = rbind(data_all,data_all[,c(2,1,3,4,5,7,6)])\n",
      "> data_all = data_all[as.numeric(data_all[,4])>0,]\n",
      "> \n",
      "> basedata = addTerminalDetails(data_all,data_all_original)\n",
      "> basedata_noPP = collapse_agnostic(basedata[basedata[,5]!=\"PP\",])\n",
      "> sidechaindata = basedata[basedata[,3]!=\"MCMC\",1:4]\n",
      "> sidechaindata_detailed = basedata[basedata[,3]!=\"MCMC\",]\n",
      "> \n",
      "> influencedata_detailed = basedata\n",
      "> if (uniqueAtomsMode) {\n",
      "+     influencedata_detailed = basedata[basedata[,8]==1 | basedata[,9]==1,]\n",
      "+ }\n",
      "> influencedata = collapse_agnostic(influencedata_detailed)\n",
      "> influencedataGLY = rbind(influencedata[grep(\"GLY\",influencedata[,1]),],influencedata[grep(\"GLY\",influencedata[,1]),])\n",
      "> influencedata = influencedata[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,2])))>linearcutoff,]\n",
      "> \n",
      "> \n",
      "> influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "> influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "> \n",
      "> influencedata_directed = collapse_directed(matrix(influencedata_detailed[influencedata_detailed[,8]==1,1:4],ncol=4))\n",
      "> influencedata_directedGLY = rbind(influencedata_directed[grep(\"GLY\",influencedata_directed[,1]),],influencedata_directed[grep(\"GLY\",influencedata_directed[,2]),])\n",
      "> influencedata_directed = influencedata_directed[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,2])))>linearcutoff,]\n",
      "> \n",
      "> influencedata_directed = matrix(influencedata_directed,ncol=4)\n",
      "> influencenet_directed = graph.edgelist(matrix(influencedata_directed[,c(2,1)],ncol=2),directed=TRUE)\n",
      "> influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "> \n",
      "> \n",
      "> if(sidechainMode) {\n",
      "+     influencedata_detailed = sidechaindata_detailed\n",
      "+     influencedata = collapse_agnostic(influencedata_detailed)\n",
      "+     influencedataGLY = rbind(influencedata[grep(\"GLY\",influencedata[,1]),],influencedata[grep(\"GLY\",influencedata[,2]),])\n",
      "+     influencedata = influencedata[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,2])))>linearcutoff,]\n",
      "+     #Add back for GLY if uniqueAtomsGaurav\n",
      "+     if(uniqueAtomsGauravPP) {\n",
      "+         influencedata = rbind(influencedata,influencedataGLY)\n",
      "+     }\n",
      "+     influencedata = influencedata[as.numeric(influencedata[,4])>0,]\n",
      "+     \n",
      "+     influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "+     influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "+     \n",
      "+     influencedata_directed = collapse_directed(matrix(influencedata_detailed[influencedata_detailed[,8]==1,1:4],ncol=4))\n",
      "+     influencedata_directedGLY = rbind(influencedata_directed[grep(\"GLY\",influencedata_directed[,1]),],influencedata_directed[grep(\"GLY\",influencedata_directed[,2]),])\n",
      "+ \n",
      "+     influencedata_directed = influencedata_directed[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,2])))>linearcutoff,]\n",
      "+     #Add back for GLY if uniqueAtomsGaurav\n",
      "+     if(uniqueAtomsGauravPP) {\n",
      "+         influencedata_directed = rbind(influencedata_directed,influencedata_directedGLY)\n",
      "+     }\n",
      "+     influencedata_directed = matrix(influencedata_directed,ncol=4)\n",
      "+     influencenet_directed = graph.edgelist(matrix(influencedata_directed[,c(2,1)],ncol=2),directed=TRUE)\n",
      "+     influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "+     \n",
      "+ }\n",
      "> \n",
      "> #Replace missing nodes in the igraph net\n",
      "> nodesMissing = nodes[nodes %in% influencedata[,1]==FALSE & nodes %in% influencedata[,2]==FALSE]\n",
      "> \n",
      "> ##B-factor\n",
      "> bfactor = as.matrix(read.table(dir(\"./\",\"_Bfactor\"),row.names=1))\n",
      "> \n",
      "> #Replace missing nodes in secStructure matrix\n",
      "> if (any(nodes %in% secStructure[,1]==FALSE)) {\n",
      "+     for (node in nodes) {\n",
      "+         if (node %in% secStructure[,1]==FALSE) {\n",
      "+             secStructure = rbind(secStructure,c(node,\"xxx\",360))\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> for (node in nodes) {\n",
      "+     if (node %in% rownames(rsa) == FALSE) {\n",
      "+         rsa = rbind(rsa,c(0,0))\n",
      "+         rownames(rsa)[nrow(rsa)] = node\n",
      "+     }\n",
      "+     if (node %in% rownames(bfactor) == FALSE) {\n",
      "+         bfactor = rbind(bfactor,1)\n",
      "+         rownames(bfactor)[nrow(bfactor)] = node\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> \n",
      "> \n",
      "> #Give DNA attributes\n",
      "> if (\"DNA\" %in% data_all[,5]) {\n",
      "+     for (i in 1:nrow(data_all)) {\n",
      "+         if (data_all[i,5]==\"DNA\") {\n",
      "+             phiAngleDB[[data_all[i,1]]] = 90\n",
      "+             if (data_all[i,1] %in% rownames(rsa) == FALSE) {\n",
      "+                 rsa = rbind(rsa,c(0,0))\n",
      "+                 rownames(rsa)[nrow(rsa)] = data_all[i,1]\n",
      "+             }\n",
      "+             if (data_all[i,1] %in% rownames(secStructure) == FALSE) {\n",
      "+                 secStructure = rbind(secStructure,c(data_all[i,1],\"DNA\",360))\n",
      "+             }\n",
      "+             if (data_all[i,1] %in% rownames(bfactor) == FALSE) {\n",
      "+                 bfactor = rbind(bfactor,1)\n",
      "+                 rownames(bfactor)[nrow(bfactor)] = data_all[i,1]\n",
      "+             }\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #Betweenness\n",
      "> node_betweenness = betweenness(influencenet,weights=influencenet$weight,directed=directed)\n",
      "> node_betweenness_unique = betweenness(influencenet_directed,weights=influencenet_directed$weight,directed=directed)\n",
      "> \n",
      "> #Modular calculations\n",
      "> \n",
      "> ##WALKTRAP\n",
      "> net = graph.edgelist(basedata_noPP[,1:2],directed=FALSE)\n",
      "> net$weight = 1/as.numeric(basedata_noPP[,4])\n",
      "> net_community=walktrap.community(net,weights=net$weight); net_community_vec = net_community$membership; names(net_community_vec)=net_community$names\n",
      "> net_community_vec_wt = net_community_vec\n",
      "> #net_community_vec[buriednodes] = max(net_community_vec)+1\n",
      "> nodes = V(influencenet)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> nodes.color = c()\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(influencedata)) {\n",
      "+     j = which((basedata_noPP[,1]==influencedata[i,1]&basedata_noPP[,2]==influencedata[i,2])|(basedata_noPP[,1]==influencedata[i,2]&basedata_noPP[,2]==influencedata[i,1]))\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) { edgecolors[j]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP\n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP - WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         if(weight1==0 & weight2==0) {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]] + 0\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]] + 0\n",
      "+         } else {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for WALKTRAP\n",
      "> firstOrderDegree = degree(influencenet)\n",
      "> firstOrderDegree_sidechain = degree(influencenet_directed,mode=c(\"in\"))\n",
      "> \n",
      "> secondOrderDegree = c(); secondOrderDegree_sidechain=c()\n",
      "> nodes = V(influencenet)$name\n",
      "> for (node in nodes) {\n",
      "+     firstorder = neighbors(influencenet,node)\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet,neighbor)))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree = c(secondOrderDegree,length(secondorder))\n",
      "+     if(node %in% V(influencenet_directed)$name) {\n",
      "+         firstorder_sidechain = neighbors(influencenet_directed,node,mode=c(\"in\"))\n",
      "+     } else {\n",
      "+         firstorder_sidechain = c()\n",
      "+     }\n",
      "+     secondorder = c(); for (neighbor in firstorder_sidechain){secondorder = c(secondorder,names(neighbors(influencenet_directed,neighbor,mode=c(\"in\"))))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree_sidechain = c(secondOrderDegree_sidechain,length(secondorder))\n",
      "+ }\n",
      "> names(secondOrderDegree) = nodes; names(secondOrderDegree_sidechain) = nodes\n",
      "> \n",
      "> \n",
      "> #second order intermodular degree\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])]))\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])]))\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #SAVE ALL AS WALKTRAP\n",
      "> node_edge_betweenness_wt = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_wt = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_wt = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_wt = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_wt = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_wt = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> \n",
      "> ##2ARY STRUCTURE\n",
      "> secStructure = secStructure[,1:2]\n",
      "> rownames(secStructure) = secStructure[,1]\n",
      "> net_community_vec = as.numeric(as.factor(secStructure[,2]))\n",
      "> names(net_community_vec) = secStructure[,1]\n",
      "> #net_community_vec[buriednodes] = max(net_community_vec)+1\n",
      "> nodes = V(influencenet)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> nodes.color = c()\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(basedata_noPP)) {\n",
      "+     if (net_community_vec[basedata_noPP[i,1]]!=net_community_vec[basedata_noPP[i,2]]) { edgecolors[i]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         if(weight1==0 & weight2==0) {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]] + 0\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]] + 0\n",
      "+         } else {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> ##Degree and second order degree for 2ARY STRUCTURE\n",
      "> \n",
      "> #second order intermodular degree\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     #secondOrder_node_intermodular_degree[node] = length(unique(unlist(node_modules[names(node_modules[[node]])])))\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])]))\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])]))\n",
      "+ }\n",
      "> \n",
      "> #SAVE ALL AS STRIDE\n",
      "> node_edge_betweenness_stride = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_stride = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_stride = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_stride = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_stride = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_stride = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> ### Create final dataset\n",
      "> nodes = nodes[nodes!=\"DNA1000A\"]\n",
      "> nodes = nodes[nodes %in% c(\"DA\",\"DG\",\"DC\",\"DT\") == FALSE]\n",
      "> \n",
      "> ### Remove ligand amino acids from final dataset\n",
      "> ligands = sub(\"-\",\"\",liganddat)\n",
      "> nodes = nodes[nodes %in% ligands == FALSE]\n",
      "> \n",
      "> nodes[which(nodes %in% rownames(bfactor)==FALSE)]\n",
      "character(0)\n",
      "> \n",
      "> out = cbind(\n",
      "+ rsa[nodes,1],\n",
      "+ firstOrderDegree[nodes],\n",
      "+ firstOrderDegree_sidechain[nodes],\n",
      "+ secondOrderDegree[nodes],\n",
      "+ secondOrderDegree_sidechain[nodes],\n",
      "+ node_edge_betweenness_stride[nodes],\n",
      "+ node_edge_betweenness_sidechain_stride[nodes],\n",
      "+ node_intermodular_degree_stride[nodes],\n",
      "+ node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ node_edge_betweenness_wt[nodes],\n",
      "+ node_edge_betweenness_sidechain_wt[nodes],\n",
      "+ node_intermodular_degree_wt[nodes],\n",
      "+ node_intermodular_degree_sidechain_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_wt[nodes]\n",
      "+ )\n",
      "> \n",
      "> #write column names\n",
      "> colnames(out) = c(\"RSA\",\"Degree\",\"Degree_uniqueAtoms\",\"SecondOrderDegree\",\"SecondOrderDegree_uniqueAtoms\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_unqiueAtoms\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_uniqueAtoms\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_uniqueAtoms\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_uniqueAtoms\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_uniqueAtoms\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_uniqueAtoms\")\n",
      "> \n",
      "> \n",
      "> ## LIGANDS\n",
      "> if(ligandmode==TRUE & ligandCentroidMode==FALSE) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",\"_ligand\"),sep=\"\\t\"))\n",
      "+     \n",
      "+     tmp = c()\n",
      "+     i=1\n",
      "+     for (i in 1:nrow(liganddat)) {\n",
      "+         acid=paste(unlist(strsplit(liganddat[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         if (unlist(strsplit(liganddat[i,1],\"-\"))[2] %in% terminalAtoms[[acid]]) {\n",
      "+             tmp = rbind(tmp,c(unlist(strsplit(liganddat[i,1],\"-\"))[1],liganddat[i,2],liganddat[i,3]))\n",
      "+         }\n",
      "+     }\n",
      "+     \n",
      "+     #ligandvec = tapply(liganddat[,3],as.factor(liganddat[,1]),min)\n",
      "+     ligandvec = tapply(as.numeric(tmp[,3]),as.factor(tmp[,1]),min)\n",
      "+ }\n",
      "> \n",
      "> if(ligandmode==TRUE & ligandCentroidMode) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",paste(\"_centroidNetLigand$\",sep=\"\")),sep=\"\\t\"))\n",
      "+     \n",
      "+     tmp = c()\n",
      "+     i=1\n",
      "+     for (i in 1:nrow(liganddat)) {\n",
      "+         tmp = rbind(tmp,c(liganddat[i,1],liganddat[i,2],liganddat[i,3]))\n",
      "+     }\n",
      "+     \n",
      "+     ligandvec = tapply(tmp[,3],as.factor(tmp[,1]),min)\n",
      "+ }\n",
      "> \n",
      "> if (ligandmode==FALSE) {\n",
      "+     ligandvec = rep(0,length(nodes))\n",
      "+ }\n",
      "> \n",
      "> out = cbind(out,ligandvec[rownames(out)])\n",
      "> colnames(out) = c(\"RSA\",\"Degree\",\"Degree_uniqueAtoms\",\"SecondOrderDegree\",\"SecondOrderDegree_uniqueAtoms\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_unqiueAtoms\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_uniqueAtoms\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_uniqueAtoms\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_uniqueAtoms\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_uniqueAtoms\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_uniqueAtoms\",\"Ligand\")\n",
      "> \n",
      "> #add back in nodes that were not networked as zeros\n",
      "> zeroMat = matrix(0,nrow=length(which(originalNodes%in%nodes==FALSE)),ncol=ncol(out));rownames(zeroMat)=originalNodes[which(originalNodes%in%nodes==FALSE)]\n",
      "> out = rbind(out,zeroMat)\n",
      "> \n",
      "> #fix RSA and Bfactor and ligand\n",
      "> out[,\"RSA\"] = rsa[rownames(out),1]\n",
      "> out[,\"Ligand\"] = ligandvec[rownames(out)]\n",
      "> \n",
      "> #Add 1 to everything to avoid zeros, except ligand\n",
      "> out[is.na(out)]=0\n",
      "> out[out[,\"Ligand\"]==0,\"Ligand\"] = 150\n",
      "> out[1:nrow(out),2:ncol(out)] = as.numeric(out[1:nrow(out),2:ncol(out)])+1\n",
      "> \n",
      "> ##create Z score file\n",
      "> outZ = apply(out,2,function(x){return(scale(as.numeric(x)))})\n",
      "> rownames(outZ) = rownames(out)\n",
      "> colnames(outZ) = colnames(out)\n",
      "> outZ[outZ==NA]=0\n",
      "> outZ[outZ==\"NaN\"]=0\n",
      "> \n",
      "> #Write out scores and z-normed scores files\n",
      "> write.table(out,file=paste(keyword,\"_scoresCentroid\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> write.table(outZ,file=paste(keyword,\"_scoresCentroidZ\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "> \n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> ## run from within the monomer folder\n",
      "> ## there can only be one *_scoresEnergetics and one Centroid/*_scoresCentroid in this folder\n",
      "> ## args[1] is the folder for the multimer\n",
      "> ## NOTE: results may differ slightly from publication because of missing data in the mutagenesis experiments used in teh paper\n",
      "> \n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> #Load monomer\n",
      "> data1 = read.table(dir(\"./\",\"_scoresEnergetics$\"),header=T,row.names=1)\n",
      "> data1Z = as.data.frame(apply(data1,2,function(x){return(scale(as.numeric(x)))}))\n",
      "> rownames(data1Z) = rownames(data1)\n",
      "> data1Z[is.na(data1Z)] = 0\n",
      "> data2 = read.table(dir(\"./Centroid\",\"_scoresCentroid$\", full.names=TRUE),header=T,row.names=1)\n",
      "> data2Z = as.data.frame(apply(data2,2,function(x){return(scale(as.numeric(x)))}))\n",
      "> rownames(data2Z) = rownames(data2)\n",
      "> data2Z[is.na(data2Z)] = 0\n",
      "> colnames(data2Z) = paste(colnames(data2Z),\"CENTROIDSC\",sep=\"\")\n",
      "> data2Z = data2Z[rownames(data1Z),]\n",
      "> data2 = data2[rownames(data1),]\n",
      "> \n",
      "> #Load multimer\n",
      "> data1Multimer = read.table(dir(args[1],\"_scoresEnergetics$\",full.names=TRUE),header=T,row.names=1)\n",
      "> data2Multimer = read.table(dir(paste(args[1],\"Centroid\",sep=\"/\"),\"_scoresCentroid$\",full.names=TRUE),header=T,row.names=1)\n",
      "> data2Multimer = data2Multimer[rownames(data1Multimer),]\n",
      "> data1MultimerZ = as.data.frame(apply(data1Multimer,2,function(x){return(scale(as.numeric(x)))}))\n",
      "> rownames(data1MultimerZ) = rownames(data1Multimer)\n",
      "> data2MultimerZ = as.data.frame(apply(data2Multimer,2,function(x){return(scale(as.numeric(x)))}))\n",
      "> rownames(data2MultimerZ) = rownames(data2Multimer)\n",
      "> colnames(data1MultimerZ) = paste(colnames(data1MultimerZ),\"MULTIMER\",sep=\"\")\n",
      "> colnames(data2MultimerZ) = paste(colnames(data2MultimerZ),\"MULTIMERCENTROIDSC\",sep=\"\")\n",
      "> data2MultimerZ = data2MultimerZ[rownames(data1MultimerZ),]\n",
      "> \n",
      "> #Add RSA\n",
      "> data1 = cbind(data2[,\"RSA\"],data1)\n",
      "> colnames(data1)[1]=\"RSA\"\n",
      "> dataZ = cbind(data1Z,data2Z)\n",
      "> \n",
      "> #Identify max NodeEdgeBetweennessSTRIDE_sidechain\n",
      "> acidsMonomer = c()\n",
      "> for (i in 1:length(rownames(dataZ))) {\n",
      "+     node = rownames(dataZ)[i]\n",
      "+     node = unlist(strsplit(node,split=\"\"))\n",
      "+     node = paste(node[1:(length(node)-1)],collapse=\"\")\n",
      "+     acidsMonomer = c(acidsMonomer,node)\n",
      "+ }\n",
      "> \n",
      "> acidsMultimer = c()\n",
      "> for (i in 1:length(rownames(data1MultimerZ))) {\n",
      "+     node = rownames(data1MultimerZ)[i]\n",
      "+     node = unlist(strsplit(node,split=\"\"))\n",
      "+     node = paste(node[1:(length(node)-1)],collapse=\"\")\n",
      "+     acidsMultimer = c(acidsMultimer,node)\n",
      "+ }\n",
      "> \n",
      "> allAcids = unique(c(acidsMonomer,acidsMultimer))\n",
      "> \n",
      "> #SecondOrderIntermodularDegree is the average of the multimer only\n",
      "> SecondOrderIntermodularDegree_ALL = (data1MultimerZ[,\"SecondOrderIntermodularDegreeSTRIDE_sidechainMULTIMER\"]+data1MultimerZ[,\"SecondOrderIntermodularDegreeSTRIDEMULTIMER\"]+data2MultimerZ[,\"SecondOrderIntermodularDegreeSTRIDEMULTIMERCENTROIDSC\"]+data2MultimerZ[,\"SecondOrderIntermodularDegreeWALKTRAPMULTIMERCENTROIDSC\"])/4\n",
      "> SecondOrderIntermodularDegree_AVERAGE = tapply(SecondOrderIntermodularDegree_ALL,as.factor(acidsMultimer),mean)\n",
      "> SecondOrderIntermodularDegree_AVERAGE = SecondOrderIntermodularDegree_AVERAGE[allAcids]\n",
      "> missing = which(allAcids%in%acidsMultimer == FALSE)\n",
      "> SecondOrderIntermodularDegree_AVERAGE[missing] = (data1Z[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"SecondOrderIntermodularDegreeSTRIDE_sidechain\"]+data1Z[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"SecondOrderIntermodularDegreeSTRIDE\"]+data2Z[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"SecondOrderIntermodularDegreeSTRIDECENTROIDSC\"]+data2Z[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"SecondOrderIntermodularDegreeWALKTRAPCENTROIDSC\"])/4\n",
      "> names(SecondOrderIntermodularDegree_AVERAGE)[missing] = allAcids[which(allAcids%in%acidsMultimer == FALSE)]\n",
      "> \n",
      "> #NodeEdgeBetweenness is the max of the monomer/multimer\n",
      "> NodeEdgeBetweennessSTRIDE_AVERAGE_MONOMER = (dataZ[,\"NodeEdgeBetweennessSTRIDE\"]+dataZ[,\"NodeEdgeBetweennessSTRIDE_sidechain\"])/2\n",
      "> NodeEdgeBetweennessSTRIDE_AVERAGE_MULTIMER = (data1MultimerZ[,\"NodeEdgeBetweennessSTRIDEMULTIMER\"]+data1MultimerZ[,\"NodeEdgeBetweennessSTRIDE_sidechainMULTIMER\"])/2\n",
      "> NodeEdgeBetweennessSTRIDE_sidechain_ALL = c(NodeEdgeBetweennessSTRIDE_AVERAGE_MONOMER,NodeEdgeBetweennessSTRIDE_AVERAGE_MULTIMER)\n",
      "> NodeEdgeBetweennessSTRIDE_sidechain_MAX = tapply(NodeEdgeBetweennessSTRIDE_sidechain_ALL,as.factor(c(acidsMonomer,acidsMultimer)),max,na.rm=TRUE)\n",
      "> NodeEdgeBetweennessSTRIDE_sidechain_MAX = NodeEdgeBetweennessSTRIDE_sidechain_MAX[allAcids]\n",
      "> \n",
      "> #Ligand is the min of the multimer only\n",
      "> LigandMULTIMERCENTROIDSC_ALL = data2MultimerZ[,\"LigandMULTIMERCENTROIDSC\"]\n",
      "> LigandMULTIMERCENTROIDSC_MIN = tapply(LigandMULTIMERCENTROIDSC_ALL,as.factor(acidsMultimer),min)\n",
      "> LigandMULTIMERCENTROIDSC_MIN = LigandMULTIMERCENTROIDSC_MIN[allAcids]\n",
      "> missing = which(allAcids%in%acidsMultimer == FALSE)\n",
      "> LigandMULTIMERCENTROIDSC_MIN[missing] = dataZ[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"LigandCENTROIDSC\"]\n",
      "> names(LigandMULTIMERCENTROIDSC_MIN)[missing] = allAcids[which(allAcids%in%acidsMultimer == FALSE)]\n",
      "> LigandMULTIMERCENTROIDSC_MIN[is.na(LigandMULTIMERCENTROIDSC_MIN)] = 0\n",
      "> \n",
      "> #RSA\n",
      "> RSA_ALL = c(data1[,\"RSA\"],data2Multimer[,\"RSA\"])\n",
      "> RSA_MIN = tapply(RSA_ALL,as.factor(c(acidsMonomer,acidsMultimer)),min,na.rm=TRUE)\n",
      "> RSA_MIN = RSA_MIN[allAcids]\n",
      "> \n",
      "> y = SecondOrderIntermodularDegree_AVERAGE+NodeEdgeBetweennessSTRIDE_sidechain_MAX-LigandMULTIMERCENTROIDSC_MIN\n",
      "> \n",
      "> out=cbind(names(y),y)\n",
      "> acidNumbers = sapply(out[,1],substring,4)\n",
      "> out = out[order(as.numeric(acidNumbers)),]\n",
      "> write.table(out,file=\"FinalSum\",sep=\"\\t\",quote=FALSE,row.names=FALSE,col.names=FALSE)\n",
      "> \n",
      "> # also write the 3 variables to a file\n",
      "> # Concatenating all three variables into a single data frame\n",
      "> result <- data.frame(\n",
      "+   Acid = names(SecondOrderIntermodularDegree_AVERAGE),\n",
      "+   SecondOrderIntermodularDegree_AVERAGE,\n",
      "+   NodeEdgeBetweennessSTRIDE_sidechain_MAX,\n",
      "+   LigandMULTIMERCENTROIDSC_MIN\n",
      "+ )\n",
      "> \n",
      "> # Writing the data frame to a file\n",
      "> write.table(result, file = \"FinalSum_Decomp\", sep = \"\\t\", quote = FALSE, row.names = FALSE, col.names = TRUE)\n",
      "> \n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "Read 579 items\n",
      "Read 579 items\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "Read 7713 items\n",
      "Read 7713 items\n",
      "\n",
      "SBNA results for 7DHY succesfully downloaded\n",
      "\n",
      "Processing 7DHY with chain B...\n",
      "Starting phenix.pdbtools\n",
      "on Sun Jun  2 13:34:06 2024 by yliy0004\n",
      "===============================================================================\n",
      "\n",
      "Processing files:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Found model, 7DHY.pdb\n",
      "\n",
      "Processing PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Adding command-line PHIL:\n",
      "  -------------------------\n",
      "    keep=Chain B\n",
      "\n",
      "Final processed PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "  data_manager {\n",
      "    model {\n",
      "      file = \"7DHY.pdb\"\n",
      "    }\n",
      "    default_model = \"7DHY.pdb\"\n",
      "  }\n",
      "  modify {\n",
      "    keep = Chain B\n",
      "  }\n",
      "\n",
      "\n",
      "Starting job\n",
      "===============================================================================\n",
      "Validating inputs\n",
      "Performing manipulations\n",
      "Size before: 6723 size after: 1720\n",
      "Writing output model\n",
      "\n",
      "===============================================================================\n",
      "Job complete\n",
      "usr+sys time: 2.53 seconds\n",
      "wall clock time: 9.21 seconds\n",
      "Cleaning PDB file\n",
      "Starting phenix.pdbtools\n",
      "on Sun Jun  2 13:35:13 2024 by yliy0004\n",
      "===============================================================================\n",
      "\n",
      "Processing files:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Found model, 7DHY_monomer.pdb\n",
      "\n",
      "Processing PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Adding command-line PHIL:\n",
      "  -------------------------\n",
      "    remove=water\n",
      "\n",
      "Final processed PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "  data_manager {\n",
      "    model {\n",
      "      file = \"7DHY_monomer.pdb\"\n",
      "    }\n",
      "    default_model = \"7DHY_monomer.pdb\"\n",
      "  }\n",
      "  modify {\n",
      "    remove = water\n",
      "  }\n",
      "\n",
      "\n",
      "Starting job\n",
      "===============================================================================\n",
      "Validating inputs\n",
      "Performing manipulations\n",
      "Size before: 1720 size after: 1571\n",
      "Writing output model\n",
      "\n",
      "===============================================================================\n",
      "Job complete\n",
      "usr+sys time: 1.99 seconds\n",
      "wall clock time: 7.09 seconds\n",
      "Writing out polypeptide file...\n",
      "Adding hydrogens...\n",
      "Protonating waters...\n",
      "Making .phi file\n",
      "Running full bond calculation script\n",
      "The ligands identified in this pdb are GOL-304B, GOL-303B, ARS-301B, GOL-305B\n",
      "Screening for atoms that are close.\n",
      "Dictionaries loaded.\n",
      "Starting salt bridges.\n",
      "Starting hydrogen bonds.\n",
      "Starting pi-pi version 2.\n",
      "Running pi-cation version 2\n",
      "Starting disulfide\n",
      "Starting van der waals.\n",
      "Starting metals\n",
      "Starting DNA bonds...\n",
      "Starting Ligand centroids\n",
      "Starting residue centroids...\n",
      "Removing duplicates from BFactor file\n",
      "Making 7DHY_monomer_nowaters_net\n",
      "Removing any negative value edges\n",
      "Creating QC file\n",
      "Creating Centroid directory and moving files\n",
      "Running energetic R script\n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> ##Load libraries\n",
      "> library(igraph)\n",
      "> library(stats)\n",
      "> \n",
      "> ##Initialize keywords\n",
      "> # keyword is the base filename\n",
      "> keyword = args[1]\n",
      "> args=c(paste(args[1],\"_net\",sep=\"\"),\"weighted\",paste(keyword,\"_secondaryStructure\",sep=\"\"))\n",
      "> atomCorrection=FALSE\n",
      "> baseKeyword = unlist(strsplit(keyword,\"/\"))[1]\n",
      "> runVariableSelection = TRUE\n",
      "> \n",
      "> ##Load inversal data\n",
      "> #Unique atoms file\n",
      "> terminalAtomsFile = as.matrix(read.table(\"../../terminalAtoms\",sep=\"\\t\"))\n",
      "> terminalAtoms = {}\n",
      "> for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+     terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\", \"))\n",
      "+ }\n",
      "> \n",
      "> #Secondary structure file\n",
      "> secStructure = as.matrix(read.table(args[3]))[,1:2]\n",
      "> rownames(secStructure) = secStructure[,1]\n",
      "> \n",
      "> #RSA\n",
      "> rsa = as.matrix(read.table(paste(keyword,\".rsa\",sep=\"\")))\n",
      "> \n",
      "> ##Define functions\n",
      "> \n",
      "> #Add terminal atom details\n",
      "> addTermDetails = function(data) {\n",
      "+     out_details = c()\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+         if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+             code1 = 1\n",
      "+         } else {\n",
      "+             code1 = 0\n",
      "+         }\n",
      "+         if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+             code2 = 1\n",
      "+         } else {\n",
      "+             code2 = 0\n",
      "+         }\n",
      "+     out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+     }\n",
      "+     return(out_details)\n",
      "+ }\n",
      "> \n",
      "> collapse = function(data) {\n",
      "+     if (nrow(data)==0) {\n",
      "+         out = matrix(ncol = 8,nrow=0)\n",
      "+         colnames(out) = c(\"AA1\", \"AA2\", \"Type\", \"SumEdges\", \"percentMC\", \"influence1\", \"influence2\",\"degree\")\n",
      "+         results = c(); results$out = out\n",
      "+         return(results)\n",
      "+         break\n",
      "+     }\n",
      "+     #Edges should be listed bi-directionally\n",
      "+     x=unlist(lapply(data[,3],function(x){if(x==\"MCSC\"){return(\"SCMC\")}else if(x==\"SCMC\"){return(\"MCSC\")} else {return(x)}}))\n",
      "+     if (dim(data)[2]==7) {\n",
      "+         data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6]))\n",
      "+     } else {\n",
      "+         data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6],data[,9],data[,8]))\n",
      "+     }\n",
      "+     data = unique(data)\n",
      "+     out = c()\n",
      "+     out_noPP = c()\n",
      "+     completed = c()\n",
      "+     if (dim(data)[2]==7) {\n",
      "+         out_details = c()\n",
      "+         for (i in 1:nrow(data)) {\n",
      "+             acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+             acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+             if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+                 code1 = 1\n",
      "+             } else {\n",
      "+                 code1 = 0\n",
      "+             }\n",
      "+             if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+                 code2 = 1\n",
      "+             } else {\n",
      "+                 code2 = 0\n",
      "+             }\n",
      "+             out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+         }\n",
      "+         data = out_details\n",
      "+     } else {\n",
      "+         out_details = data\n",
      "+     }\n",
      "+     colnames(out_details) = c(\"AA1\",\"AA2\",\"Type1\",\"Weight\",\"Type2\",\"Atom1\",\"Atom2\",\"influence1\",\"influence2\")\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         if (i %in% completed == FALSE) {\n",
      "+             x=which((data[,1]==data[i,1] & data[,2] == data[i,2]))\n",
      "+             subdata = matrix(data[x,],ncol=9)\n",
      "+             edgeSum = 0\n",
      "+             MCMCcount = 0\n",
      "+             MCSCcount = 0\n",
      "+             SCMCcount = 0\n",
      "+             SCSCcount = 0\n",
      "+             influence1 = sum(as.numeric(subdata[,8]))\n",
      "+             influence2 = sum(as.numeric(subdata[,9]))\n",
      "+             for (j in 1:nrow(subdata)) {\n",
      "+                 edgeSum = edgeSum + as.numeric(subdata[j,4])\n",
      "+                 if (subdata[j,3]==\"MCMC\" | subdata[j,3]==\"MCSC\") {\n",
      "+                     MCMCcount = MCMCcount+1\n",
      "+                     MCSCcount = MCSCcount+1\n",
      "+                 } else if (subdata[j,3]==\"SCMC\") {\n",
      "+                     SCMCcount = SCMCcount+1\n",
      "+                 } else if (subdata[j,3]==\"SCSC\") {\n",
      "+                     SCSCcount = SCSCcount+1\n",
      "+                 }\n",
      "+             }\n",
      "+             #Header is AA1, AA2, mixed, sum of edges, percent MC, influence1, influence2, degreeAA1\n",
      "+             out = rbind(out,c(data[i,1],data[i,2],\"mixed\",edgeSum,(MCMCcount+MCSCcount)/(MCMCcount+MCSCcount+SCMCcount+SCSCcount),influence1,influence2,sum(data[,1]==data[i,1])))\n",
      "+             if (any(subdata[,5]==\"PP\")==FALSE) {out_noPP = rbind(out_noPP,c(data[i,1],data[i,2],\"mixed\",edgeSum,(MCMCcount+MCSCcount)/(MCMCcount+MCSCcount+SCMCcount+SCSCcount),influence1,influence2,sum(data[,1]==data[i,1])))}\n",
      "+             completed = c(completed,x)\n",
      "+         }\n",
      "+     }\n",
      "+     out = out[out[,4]>0,]\n",
      "+     out_noPP = out_noPP[out_noPP[,4]>0,]\n",
      "+     colnames(out) = c(\"AA1\", \"AA2\", \"Type\", \"SumEdges\", \"percentMC\", \"influence1\", \"influence2\",\"degree\")\n",
      "+     results = c()\n",
      "+     results$out = out\n",
      "+     results$out_noPP = out_noPP\n",
      "+     results$influence_all = out_details[as.numeric(out_details[,4])>0 & (out_details[,8]==\"1\" | out_details[,9]==\"1\"),]\n",
      "+     out_details = out_details[as.numeric(out_details[,4])>0,]\n",
      "+     out_subtract = out_details; out_subtract[out_subtract[,8]==0 & out_subtract[,9]==0,4] = (-1)*as.numeric(out_subtract[out_subtract[,8]==0 & out_subtract[,9]==0,4])\n",
      "+     out_subtract = out_subtract[out_subtract[,5]!=\"PP\",]\n",
      "+     results$out_subtract = out_subtract\n",
      "+     return(results)\n",
      "+ }\n",
      "> \n",
      "> removeRedundancy = function(dataTmp) {\n",
      "+     for (i in 1:nrow(dataTmp)) {\n",
      "+         if (order(dataTmp[i,1:2])[1]==2) {\n",
      "+             dataTmp[i,1:2] = c(dataTmp[i,2],dataTmp[i,1])\n",
      "+             dataTmp[i,3] = paste(unlist(strsplit(dataTmp[i,3],split=\"\"))[c(3,4,1,2)],collapse=\"\")\n",
      "+             dataTmp[i,c(6,7)] = c(dataTmp[i,7],dataTmp[i,6])\n",
      "+         }\n",
      "+     }\n",
      "+     dataTmp = unique(dataTmp)\n",
      "+     pairs = unique(dataTmp[,1:2])\n",
      "+     out = c()\n",
      "+     for (j in 1:nrow(pairs)) {\n",
      "+         pair = pairs[j,]\n",
      "+         subData = matrix(dataTmp[dataTmp[,1]==pair[1] & dataTmp[,2]==pair[2],],ncol=7)\n",
      "+         out = rbind(out, c(pair[1],pair[2],\"mixed\",0.5*sum(as.numeric(subData[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> collapse_directed = function(data) {\n",
      "+     pairs = unique(data[,1:2])\n",
      "+     out = c()\n",
      "+     for (i in 1:nrow(pairs)) {\n",
      "+         subdat = matrix(data[data[,1]==pairs[i,1]&data[,2]==pairs[i,2],],ncol=4)\n",
      "+         out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> collapse_agnostic = function(data) {\n",
      "+     pairs = unique(t(apply(data[,1:2],1,function(x){return(sort(x))})))\n",
      "+     out = c()\n",
      "+     for (i in 1:nrow(pairs)) {\n",
      "+         subdat = matrix(data[(data[,1]==pairs[i,1]&data[,2]==pairs[i,2]) | (data[,2]==pairs[i,1]&data[,1]==pairs[i,2]),],ncol=4)\n",
      "+         out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> distributCalc = function(set,nodes,net) {\n",
      "+     #x is assumed to be a set but can also just be 1 node\n",
      "+     distanceNet = distances(net,weights=net$weight)\n",
      "+     distanceTmp = c()\n",
      "+     Dr = c()\n",
      "+     for (node in nodes) {\n",
      "+         distanceTmp = c()\n",
      "+         for (item in set) {\n",
      "+             if (item!=node) {\n",
      "+                 if(item%in%rownames(distanceNet)==FALSE){distanceTmp=100;next}\n",
      "+                 if(node%in%rownames(distanceNet)==FALSE){distanceTmp=100;next}\n",
      "+                 distanceTmp = c(distanceTmp,distanceNet[item,node])\n",
      "+             }\n",
      "+         }\n",
      "+         if (length(distanceTmp)>0) {\n",
      "+             Dr = c(Dr, 1/min(distanceTmp))\n",
      "+         }\n",
      "+     }\n",
      "+     return(sum(Dr)/length(nodes))\n",
      "+ }\n",
      "> \n",
      "> ##Read in user data\n",
      "> data_all = as.matrix(read.table(args[1]))\n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> originalNodes = nodes\n",
      "> basedata = collapse(data_all)\n",
      "> basedata_noPP = removeRedundancy(basedata$out[,1:7])\n",
      "> basedata_allsidechain = collapse_agnostic(data_all[data_all[,3]!=\"MCMC\",c(1,2,3,4)])\n",
      "> influencedata_all = basedata$influence_all\n",
      "> influencedata_collapse = collapse(influencedata_all)\n",
      "> influencedata = removeRedundancy(influencedata_collapse$out[,1:7])\n",
      "> influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "> influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "> \n",
      "> influencedata_directed = collapse_directed(influencedata_all[influencedata_all[,8]==1,c(1,2,3,4)])\n",
      "> influencenet_directed = graph.edgelist(influencedata_directed[,c(2,1)],directed=TRUE)\n",
      "> influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "> \n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> \n",
      "> #Write out edgelist files\n",
      "> write.table(influencedata,file=paste(keyword,\"_allEdges\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=FALSE,row.names=FALSE)\n",
      "> write.table(influencedata_directed,file=paste(keyword,\"_uniqueEdges\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=FALSE,row.names=FALSE)\n",
      "> \n",
      "> ###Deal with missing nodes\n",
      "> if (any(nodes %in% secStructure[,1]==FALSE)) {\n",
      "+     for (node in nodes) {\n",
      "+         if (node %in% secStructure[,1]==FALSE) {\n",
      "+             secStructure = rbind(secStructure,c(node,\"xxx\",360))\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> for (node in nodes) {\n",
      "+     if (node %in% rownames(rsa) == FALSE) {\n",
      "+         rsa = rbind(rsa,c(0,0))\n",
      "+         rownames(rsa)[nrow(rsa)] = node\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> ##Begin centrality calculations\n",
      "> \n",
      "> #Modular calculations\n",
      "> \n",
      "> ##WALKTRAP\n",
      "> net = graph.edgelist(basedata_noPP[,1:2],directed=FALSE)\n",
      "> net$weight = 1/as.numeric(basedata_noPP[,4])\n",
      "> net_community=walktrap.community(net,weights=net$weight); net_community_vec = net_community$membership; names(net_community_vec)=net_community$names\n",
      "> net_community_vec_wt = net_community_vec\n",
      "> nodes = V(net)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> nodes.color = c()\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(influencedata)) {\n",
      "+     j = which((basedata_noPP[,1]==influencedata[i,1]&basedata_noPP[,2]==influencedata[i,2])|(basedata_noPP[,1]==influencedata[i,2]&basedata_noPP[,2]==influencedata[i,1]))\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) { edgecolors[j]=\"grey40\" }\n",
      "+ }\n",
      "> if(length(nodes)>200) {textsize=.25} else {textsize=.45}\n",
      "> plot(net,edge.color = edgecolors, vertex.color = nodes.color,vertex.size=4,vertex.label.cex=textsize,main=\"WALKTRAP\")\n",
      ">     \n",
      "> ##Weighted edge betweenness WALKTRAP\n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP - WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for WALKTRAP\n",
      "> firstOrderDegree = degree(influencenet)\n",
      "> firstOrderDegree_sidechain = degree(influencenet_directed,mode=c(\"in\"))\n",
      "> secondOrderDegree = c(); secondOrderDegree_sidechain=c()\n",
      "> nodes = V(influencenet)$name\n",
      "> for (node in nodes) {\n",
      "+     firstorder = neighbors(influencenet,node)\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet,neighbor)))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree = c(secondOrderDegree,length(secondorder))\n",
      "+     firstorder = neighbors(influencenet_directed,node,mode=c(\"in\"))\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet_directed,neighbor,mode=c(\"in\"))))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree_sidechain = c(secondOrderDegree_sidechain,length(secondorder))\n",
      "+ }\n",
      "> names(secondOrderDegree) = nodes; names(secondOrderDegree_sidechain) = nodes\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])])) - 1\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])])) - 1\n",
      "+ }\n",
      "> \n",
      "> #SAVE ALL AS WALKTRAP\n",
      "> node_edge_betweenness_wt = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_wt = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_wt = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_wt = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_wt = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_wt = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> ##2ARY STRUCTURE\n",
      "> net_community_vec = as.numeric(as.factor(secStructure[,2]))\n",
      "> names(net_community_vec) = secStructure[,1]\n",
      "> nodes = V(net)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> nodes.color = c()\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(basedata_noPP)) {\n",
      "+     if (net_community_vec[basedata_noPP[i,1]]!=net_community_vec[basedata_noPP[i,2]]) { edgecolors[i]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for 2ARY STRUCTURE\n",
      "> nodes = V(influencenet)$name\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])])) - 1\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])])) - 1\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #SAVE ALL AS STRIDE\n",
      "> node_edge_betweenness_stride = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_stride = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_stride = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_stride = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_stride = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_stride = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> \n",
      "> ### Create final dataset\n",
      "> \n",
      "> out = cbind(\n",
      "+ firstOrderDegree[nodes],\n",
      "+ firstOrderDegree_sidechain[nodes],\n",
      "+ secondOrderDegree[nodes],\n",
      "+ secondOrderDegree_sidechain[nodes],\n",
      "+ node_edge_betweenness_stride[nodes],\n",
      "+ node_edge_betweenness_sidechain_stride[nodes],\n",
      "+ node_intermodular_degree_stride[nodes],\n",
      "+ node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ node_edge_betweenness_wt[nodes],\n",
      "+ node_edge_betweenness_sidechain_wt[nodes],\n",
      "+ node_intermodular_degree_wt[nodes],\n",
      "+ node_intermodular_degree_sidechain_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_wt[nodes]\n",
      "+ )\n",
      "> rownames(out) = nodes\n",
      "> colnames(out) = c(\"Degree\",\"Degree_sidechain\",\"SecondOrderDegree\",\"SecondOrderDegree_sidechain\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_sidechain\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_sidechain\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_sidechain\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_sidechain\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_sidechain\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_sidechain\")\n",
      "> \n",
      "> #add back in nodes that were not networked as zeros\n",
      "> zeroMat = matrix(0,nrow=length(which(originalNodes%in%nodes==FALSE)),ncol=ncol(out));rownames(zeroMat)=originalNodes[which(originalNodes%in%nodes==FALSE)]\n",
      "> out = rbind(out,zeroMat)\n",
      "> \n",
      "> #Add 1 to everything to avoid zeros, except ligand\n",
      "> out[is.na(out)]=0\n",
      "> out = out+1\n",
      "> \n",
      "> #Write out\n",
      "> write.table(out,file=paste(keyword,\"_scoresEnergetics\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "> #Standard normalization\n",
      "> outZ = apply(out,2,function(x){return(scale(x))})\n",
      "> rownames(outZ) = rownames(out)\n",
      "> write.table(outZ,file=paste(keyword,\"_scoresEnergeticsZ\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "Running centroid R script\n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(igraph)\n",
      "> library(stats)\n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> ##read in data\n",
      "> ## arg1: col1 and col2 are edges, col3 is weight of edge\n",
      "> ## arg2: weighted, unweighted <-- no longer using\n",
      "> ## arg3: forced module definition <-- no longer using\n",
      "> \n",
      "> # Set desired parameters\n",
      "> removeMCMC = FALSE\n",
      "> linearcutoff = 1\n",
      "> directed = FALSE\n",
      "> secondaryStructure2 = FALSE\n",
      "> sidechainMode = TRUE ##build network of just SCSC, SCMC and MCSC. Still use unique atoms for the _uniqueAtoms directed plots\n",
      "> uniqueAtomsMode = FALSE ##build network of just unique atoms (each edge has to have at least 1). Still use unique atoms for the _uniqueAtoms directed plots\n",
      "> useDNA=TRUE\n",
      "> uniqueAtomsUnbiased=FALSE\n",
      "> uniqueAtomsGaurav=FALSE\n",
      "> uniqueAtomsOLD=FALSE\n",
      "> atomCorrection = FALSE\n",
      "> uniqueAtomsGauravPP = FALSE\n",
      "> weighted = TRUE\n",
      "> removeWaters=TRUE\n",
      "> if (tail(unlist(strsplit(getwd(),\"/\")),n=1)==\"Centroid\") {\n",
      "+     ligandCentroidMode = TRUE\n",
      "+ } else {\n",
      "+     ligandCentroidMode = FALSE\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> # Read in ligand file\n",
      "> ligandfile = scan(paste(args[1],\"_ligand\",sep=\"\"),what=\"character\")\n",
      "> if (length(ligandfile)==0){\n",
      "+     ligandmode = FALSE\n",
      "+ } else {\n",
      "+     ligandmode = TRUE\n",
      "+ }\n",
      "> \n",
      "> ligandCentroidFile = scan(paste(args[1],\"NetLigand\",sep=\"\"),what=\"character\")\n",
      "> \n",
      "> # keyword is the base filename\n",
      "> # method is \"frag\" or \"module\" or \"both\"\n",
      "> keyword = args[1]\n",
      "> args=c(paste(args[1],\"_net\",sep=\"\"),\"weighted\",dir(\"./\",\"_secondaryStructure$\"))\n",
      "> \n",
      "> ##Load data\n",
      "> #Unique atoms file\n",
      "> if(uniqueAtomsUnbiased) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"/Users/vjpatel/Dropbox/Gaurav/uniqueAtomsUnbiased\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else if (uniqueAtomsGaurav) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"/Users/vjpatel/Dropbox/Gaurav/uniqueAtomsGaurav\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else if (uniqueAtomsOLD) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"../../../terminalAtoms\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"../../../uniqueAtoms\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     otherTA = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+         #otherTA[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,4],split=\",\"))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> #Secondary structure file\n",
      "> #Ligands are all one secondary structure\n",
      "> secStructure = as.matrix(read.table(args[3]))\n",
      "> phiAngleDB = {}\n",
      "> for (i in 1:nrow(secStructure)) {\n",
      "+     phiAngleDB[[secStructure[i,1]]] = as.numeric(secStructure[i,3])\n",
      "+ }\n",
      "> \n",
      "> if(ligandmode) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",\"_ligand\"),sep=\"\\t\"))\n",
      "+     liganddat = unique(liganddat[,2])\n",
      "+     for (ligand in liganddat) {\n",
      "+         secStructure = rbind(secStructure,c(sub(\"-\",\"\",ligand),\"ligand\",360))\n",
      "+         phiAngleDB[[sub(\"-\",\"\",ligand)]]=0\n",
      "+     }\n",
      "+ } else {\n",
      "+     liganddat = c()\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> ##Define functions\n",
      "> \n",
      "> ##Collapse matrix functions\n",
      "> ##Will return redundant matrix\n",
      "> ##Input needs to be 7 columns\n",
      "> ##Second input is different if the main input is a subset of the larger data\n",
      "> addTerminalDetails = function(data,largerdata) {\n",
      "+     #List edges bi-directionally\n",
      "+     x=unlist(lapply(data[,3],function(x){if(x==\"MCSC\"){return(\"SCMC\")}else if(x==\"SCMC\"){return(\"MCSC\")} else {return(x)}}))\n",
      "+     data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6]))\n",
      "+     data = unique(data)\n",
      "+     out = c()\n",
      "+     out_noPP = c()\n",
      "+     out_details = c()\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+         subdata = matrix(data[data[,1]==data[i,1]&data[,2]==data[i,2],],ncol=7)\n",
      "+         allsubdata = matrix(largerdata[largerdata[,1]==data[i,1],],ncol=7)\n",
      "+         code1=0;code2=0\n",
      "+         if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+             code1 = 1\n",
      "+         }\n",
      "+         if (data[i,5] == \"DNA\" & useDNA==TRUE) {\n",
      "+             code1 = 1\n",
      "+         }\n",
      "+         allsubdata = matrix(largerdata[largerdata[,2]==data[i,2],],ncol=7)\n",
      "+         if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+             code2 = 1\n",
      "+         }\n",
      "+         if (data[i,5] == \"DNA\" & useDNA==TRUE) {\n",
      "+             code2 = 1\n",
      "+         }\n",
      "+         if (data[i,5]==\"PICATION\"|data[i,5]==\"PIPI\") {\n",
      "+             code1=1\n",
      "+             code2=1\n",
      "+         }\n",
      "+         if (data[i,5]==\"PP\") {\n",
      "+             if (uniqueAtomsGaurav) {\n",
      "+                 if((acid1) == \"GLY\") {\n",
      "+                     code1 = 1\n",
      "+                     data[i,3]==\"SCSC\"\n",
      "+                 }\n",
      "+                 if ((acid2) == \"GLY\") {\n",
      "+                     code2 = 1\n",
      "+                     data[i,3] == \"SCSC\"\n",
      "+                 }\n",
      "+             }\n",
      "+         }\n",
      "+         out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+     }\n",
      "+     return(out_details)\n",
      "+ }\n",
      "> \n",
      "> ##Will return non-redundant matrix\n",
      "> collapse_agnostic = function(data) {\n",
      "+     if(nrow(data)==0) {\n",
      "+         return(matrix(ncol=4,nrow=0))\n",
      "+     } else if (ncol(data)==4 & data[1,3]==\"mixed\") {\n",
      "+         return(data)\n",
      "+     } else {\n",
      "+         pairs = unique(t(apply(matrix(data[,1:2],ncol=2),1,function(x){return(sort(x))})))\n",
      "+         out = c()\n",
      "+         for (i in 1:nrow(pairs)) {\n",
      "+             subdat = matrix(data[(data[,1]==pairs[i,1]&data[,2]==pairs[i,2]) | (data[,2]==pairs[i,1]&data[,1]==pairs[i,2]),1:4],ncol=4)\n",
      "+             subdat = unique(t(apply(subdat,1,function(x){a=which(x==sort(x[1:2])[1]);if(a==1){return(x)}else{return(c(x[2],x[1],paste(unlist(strsplit(x[3],\"\"))[c(3,4,1,2)],collapse=\"\"),x[4]))}})))\n",
      "+             if (weighted==FALSE) {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",1))\n",
      "+             } else {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+             }\n",
      "+         }\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> ##Will return directed collapsed matrix\n",
      "> collapse_directed = function(data) {\n",
      "+     if (length(data)==0) {\n",
      "+         return(matrix(nrow=0,ncol=4))\n",
      "+     } else {\n",
      "+         pairs = matrix(unique(data[,1:2]),ncol=2)\n",
      "+         out = c()\n",
      "+         for (i in 1:nrow(pairs)) {\n",
      "+             subdat = matrix(data[data[,1]==pairs[i,1]&data[,2]==pairs[i,2],1:4],ncol=4)\n",
      "+             if (weighted==FALSE) {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",1))\n",
      "+             } else {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+             }\n",
      "+         }\n",
      "+         return(matrix(out,ncol=4))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> rsa = as.matrix(read.table(dir(\"./\",\".rsa$\")))\n",
      "> rownames(rsa) = rsa[,1]\n",
      "> rsa[,1] = rsa[,2]\n",
      "> \n",
      "> #add ligand as -1\n",
      "> for (ligand in liganddat) {\n",
      "+     rsa = rbind(rsa,c(-1,-1))\n",
      "+     rownames(rsa)[nrow(rsa)] = sub(\"-\",\"\",ligand)\n",
      "+ }\n",
      "> \n",
      "> #Function to get amino acid\n",
      "> getAcid = function(residue) {\n",
      "+     if(all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"A\")) == TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"G\"))==TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"C\"))==TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"T\"))==TRUE)) {return(paste(unlist(strsplit(residue,\"\"))[1:2],collapse=\"\"))\n",
      "+     } else {\n",
      "+     return(paste(unlist(strsplit(residue,\"\"))[1:3],collapse=\"\"))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> ##################\n",
      "> ## read in data ##\n",
      "> ##################\n",
      "> data_all = as.matrix(read.table(args[1]))[,1:7]\n",
      "> if(removeWaters) {\n",
      "+     y=c(grep(\"HOH\",data_all[,1]),grep(\"HOH\",data_all[,2]))\n",
      "+     if(length(y)>0) {\n",
      "+         data_all = data_all[-y,]\n",
      "+     }\n",
      "+ }\n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> originalNodes = nodes\n",
      "> \n",
      "> data_all_original = rbind(data_all,data_all[,c(2,1,3,4,5,7,6)])\n",
      "> data_all = data_all[as.numeric(data_all[,4])>0,]\n",
      "> \n",
      "> basedata = addTerminalDetails(data_all,data_all_original)\n",
      "> basedata_noPP = collapse_agnostic(basedata[basedata[,5]!=\"PP\",])\n",
      "> sidechaindata = basedata[basedata[,3]!=\"MCMC\",1:4]\n",
      "> sidechaindata_detailed = basedata[basedata[,3]!=\"MCMC\",]\n",
      "> \n",
      "> influencedata_detailed = basedata\n",
      "> if (uniqueAtomsMode) {\n",
      "+     influencedata_detailed = basedata[basedata[,8]==1 | basedata[,9]==1,]\n",
      "+ }\n",
      "> influencedata = collapse_agnostic(influencedata_detailed)\n",
      "> influencedataGLY = rbind(influencedata[grep(\"GLY\",influencedata[,1]),],influencedata[grep(\"GLY\",influencedata[,1]),])\n",
      "> influencedata = influencedata[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,2])))>linearcutoff,]\n",
      "> \n",
      "> \n",
      "> influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "> influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "> \n",
      "> influencedata_directed = collapse_directed(matrix(influencedata_detailed[influencedata_detailed[,8]==1,1:4],ncol=4))\n",
      "> influencedata_directedGLY = rbind(influencedata_directed[grep(\"GLY\",influencedata_directed[,1]),],influencedata_directed[grep(\"GLY\",influencedata_directed[,2]),])\n",
      "> influencedata_directed = influencedata_directed[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,2])))>linearcutoff,]\n",
      "> \n",
      "> influencedata_directed = matrix(influencedata_directed,ncol=4)\n",
      "> influencenet_directed = graph.edgelist(matrix(influencedata_directed[,c(2,1)],ncol=2),directed=TRUE)\n",
      "> influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "> \n",
      "> \n",
      "> if(sidechainMode) {\n",
      "+     influencedata_detailed = sidechaindata_detailed\n",
      "+     influencedata = collapse_agnostic(influencedata_detailed)\n",
      "+     influencedataGLY = rbind(influencedata[grep(\"GLY\",influencedata[,1]),],influencedata[grep(\"GLY\",influencedata[,2]),])\n",
      "+     influencedata = influencedata[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,2])))>linearcutoff,]\n",
      "+     #Add back for GLY if uniqueAtomsGaurav\n",
      "+     if(uniqueAtomsGauravPP) {\n",
      "+         influencedata = rbind(influencedata,influencedataGLY)\n",
      "+     }\n",
      "+     influencedata = influencedata[as.numeric(influencedata[,4])>0,]\n",
      "+     \n",
      "+     influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "+     influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "+     \n",
      "+     influencedata_directed = collapse_directed(matrix(influencedata_detailed[influencedata_detailed[,8]==1,1:4],ncol=4))\n",
      "+     influencedata_directedGLY = rbind(influencedata_directed[grep(\"GLY\",influencedata_directed[,1]),],influencedata_directed[grep(\"GLY\",influencedata_directed[,2]),])\n",
      "+ \n",
      "+     influencedata_directed = influencedata_directed[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,2])))>linearcutoff,]\n",
      "+     #Add back for GLY if uniqueAtomsGaurav\n",
      "+     if(uniqueAtomsGauravPP) {\n",
      "+         influencedata_directed = rbind(influencedata_directed,influencedata_directedGLY)\n",
      "+     }\n",
      "+     influencedata_directed = matrix(influencedata_directed,ncol=4)\n",
      "+     influencenet_directed = graph.edgelist(matrix(influencedata_directed[,c(2,1)],ncol=2),directed=TRUE)\n",
      "+     influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "+     \n",
      "+ }\n",
      "> \n",
      "> #Replace missing nodes in the igraph net\n",
      "> nodesMissing = nodes[nodes %in% influencedata[,1]==FALSE & nodes %in% influencedata[,2]==FALSE]\n",
      "> \n",
      "> ##B-factor\n",
      "> bfactor = as.matrix(read.table(dir(\"./\",\"_Bfactor\"),row.names=1))\n",
      "> \n",
      "> #Replace missing nodes in secStructure matrix\n",
      "> if (any(nodes %in% secStructure[,1]==FALSE)) {\n",
      "+     for (node in nodes) {\n",
      "+         if (node %in% secStructure[,1]==FALSE) {\n",
      "+             secStructure = rbind(secStructure,c(node,\"xxx\",360))\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> for (node in nodes) {\n",
      "+     if (node %in% rownames(rsa) == FALSE) {\n",
      "+         rsa = rbind(rsa,c(0,0))\n",
      "+         rownames(rsa)[nrow(rsa)] = node\n",
      "+     }\n",
      "+     if (node %in% rownames(bfactor) == FALSE) {\n",
      "+         bfactor = rbind(bfactor,1)\n",
      "+         rownames(bfactor)[nrow(bfactor)] = node\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> \n",
      "> \n",
      "> #Give DNA attributes\n",
      "> if (\"DNA\" %in% data_all[,5]) {\n",
      "+     for (i in 1:nrow(data_all)) {\n",
      "+         if (data_all[i,5]==\"DNA\") {\n",
      "+             phiAngleDB[[data_all[i,1]]] = 90\n",
      "+             if (data_all[i,1] %in% rownames(rsa) == FALSE) {\n",
      "+                 rsa = rbind(rsa,c(0,0))\n",
      "+                 rownames(rsa)[nrow(rsa)] = data_all[i,1]\n",
      "+             }\n",
      "+             if (data_all[i,1] %in% rownames(secStructure) == FALSE) {\n",
      "+                 secStructure = rbind(secStructure,c(data_all[i,1],\"DNA\",360))\n",
      "+             }\n",
      "+             if (data_all[i,1] %in% rownames(bfactor) == FALSE) {\n",
      "+                 bfactor = rbind(bfactor,1)\n",
      "+                 rownames(bfactor)[nrow(bfactor)] = data_all[i,1]\n",
      "+             }\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #Betweenness\n",
      "> node_betweenness = betweenness(influencenet,weights=influencenet$weight,directed=directed)\n",
      "> node_betweenness_unique = betweenness(influencenet_directed,weights=influencenet_directed$weight,directed=directed)\n",
      "> \n",
      "> #Modular calculations\n",
      "> \n",
      "> ##WALKTRAP\n",
      "> net = graph.edgelist(basedata_noPP[,1:2],directed=FALSE)\n",
      "> net$weight = 1/as.numeric(basedata_noPP[,4])\n",
      "> net_community=walktrap.community(net,weights=net$weight); net_community_vec = net_community$membership; names(net_community_vec)=net_community$names\n",
      "> net_community_vec_wt = net_community_vec\n",
      "> #net_community_vec[buriednodes] = max(net_community_vec)+1\n",
      "> nodes = V(influencenet)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> nodes.color = c()\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(influencedata)) {\n",
      "+     j = which((basedata_noPP[,1]==influencedata[i,1]&basedata_noPP[,2]==influencedata[i,2])|(basedata_noPP[,1]==influencedata[i,2]&basedata_noPP[,2]==influencedata[i,1]))\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) { edgecolors[j]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP\n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP - WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         if(weight1==0 & weight2==0) {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]] + 0\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]] + 0\n",
      "+         } else {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for WALKTRAP\n",
      "> firstOrderDegree = degree(influencenet)\n",
      "> firstOrderDegree_sidechain = degree(influencenet_directed,mode=c(\"in\"))\n",
      "> \n",
      "> secondOrderDegree = c(); secondOrderDegree_sidechain=c()\n",
      "> nodes = V(influencenet)$name\n",
      "> for (node in nodes) {\n",
      "+     firstorder = neighbors(influencenet,node)\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet,neighbor)))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree = c(secondOrderDegree,length(secondorder))\n",
      "+     if(node %in% V(influencenet_directed)$name) {\n",
      "+         firstorder_sidechain = neighbors(influencenet_directed,node,mode=c(\"in\"))\n",
      "+     } else {\n",
      "+         firstorder_sidechain = c()\n",
      "+     }\n",
      "+     secondorder = c(); for (neighbor in firstorder_sidechain){secondorder = c(secondorder,names(neighbors(influencenet_directed,neighbor,mode=c(\"in\"))))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree_sidechain = c(secondOrderDegree_sidechain,length(secondorder))\n",
      "+ }\n",
      "> names(secondOrderDegree) = nodes; names(secondOrderDegree_sidechain) = nodes\n",
      "> \n",
      "> \n",
      "> #second order intermodular degree\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])]))\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])]))\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #SAVE ALL AS WALKTRAP\n",
      "> node_edge_betweenness_wt = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_wt = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_wt = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_wt = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_wt = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_wt = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> \n",
      "> ##2ARY STRUCTURE\n",
      "> secStructure = secStructure[,1:2]\n",
      "> rownames(secStructure) = secStructure[,1]\n",
      "> net_community_vec = as.numeric(as.factor(secStructure[,2]))\n",
      "> names(net_community_vec) = secStructure[,1]\n",
      "> #net_community_vec[buriednodes] = max(net_community_vec)+1\n",
      "> nodes = V(influencenet)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> nodes.color = c()\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(basedata_noPP)) {\n",
      "+     if (net_community_vec[basedata_noPP[i,1]]!=net_community_vec[basedata_noPP[i,2]]) { edgecolors[i]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         if(weight1==0 & weight2==0) {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]] + 0\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]] + 0\n",
      "+         } else {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> ##Degree and second order degree for 2ARY STRUCTURE\n",
      "> \n",
      "> #second order intermodular degree\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     #secondOrder_node_intermodular_degree[node] = length(unique(unlist(node_modules[names(node_modules[[node]])])))\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])]))\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])]))\n",
      "+ }\n",
      "> \n",
      "> #SAVE ALL AS STRIDE\n",
      "> node_edge_betweenness_stride = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_stride = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_stride = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_stride = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_stride = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_stride = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> ### Create final dataset\n",
      "> nodes = nodes[nodes!=\"DNA1000A\"]\n",
      "> nodes = nodes[nodes %in% c(\"DA\",\"DG\",\"DC\",\"DT\") == FALSE]\n",
      "> \n",
      "> ### Remove ligand amino acids from final dataset\n",
      "> ligands = sub(\"-\",\"\",liganddat)\n",
      "> nodes = nodes[nodes %in% ligands == FALSE]\n",
      "> \n",
      "> nodes[which(nodes %in% rownames(bfactor)==FALSE)]\n",
      "character(0)\n",
      "> \n",
      "> out = cbind(\n",
      "+ rsa[nodes,1],\n",
      "+ firstOrderDegree[nodes],\n",
      "+ firstOrderDegree_sidechain[nodes],\n",
      "+ secondOrderDegree[nodes],\n",
      "+ secondOrderDegree_sidechain[nodes],\n",
      "+ node_edge_betweenness_stride[nodes],\n",
      "+ node_edge_betweenness_sidechain_stride[nodes],\n",
      "+ node_intermodular_degree_stride[nodes],\n",
      "+ node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ node_edge_betweenness_wt[nodes],\n",
      "+ node_edge_betweenness_sidechain_wt[nodes],\n",
      "+ node_intermodular_degree_wt[nodes],\n",
      "+ node_intermodular_degree_sidechain_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_wt[nodes]\n",
      "+ )\n",
      "> \n",
      "> #write column names\n",
      "> colnames(out) = c(\"RSA\",\"Degree\",\"Degree_uniqueAtoms\",\"SecondOrderDegree\",\"SecondOrderDegree_uniqueAtoms\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_unqiueAtoms\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_uniqueAtoms\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_uniqueAtoms\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_uniqueAtoms\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_uniqueAtoms\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_uniqueAtoms\")\n",
      "> \n",
      "> \n",
      "> ## LIGANDS\n",
      "> if(ligandmode==TRUE & ligandCentroidMode==FALSE) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",\"_ligand\"),sep=\"\\t\"))\n",
      "+     \n",
      "+     tmp = c()\n",
      "+     i=1\n",
      "+     for (i in 1:nrow(liganddat)) {\n",
      "+         acid=paste(unlist(strsplit(liganddat[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         if (unlist(strsplit(liganddat[i,1],\"-\"))[2] %in% terminalAtoms[[acid]]) {\n",
      "+             tmp = rbind(tmp,c(unlist(strsplit(liganddat[i,1],\"-\"))[1],liganddat[i,2],liganddat[i,3]))\n",
      "+         }\n",
      "+     }\n",
      "+     \n",
      "+     #ligandvec = tapply(liganddat[,3],as.factor(liganddat[,1]),min)\n",
      "+     ligandvec = tapply(as.numeric(tmp[,3]),as.factor(tmp[,1]),min)\n",
      "+ }\n",
      "> \n",
      "> if(ligandmode==TRUE & ligandCentroidMode) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",paste(\"_centroidNetLigand$\",sep=\"\")),sep=\"\\t\"))\n",
      "+     \n",
      "+     tmp = c()\n",
      "+     i=1\n",
      "+     for (i in 1:nrow(liganddat)) {\n",
      "+         tmp = rbind(tmp,c(liganddat[i,1],liganddat[i,2],liganddat[i,3]))\n",
      "+     }\n",
      "+     \n",
      "+     ligandvec = tapply(tmp[,3],as.factor(tmp[,1]),min)\n",
      "+ }\n",
      "> \n",
      "> if (ligandmode==FALSE) {\n",
      "+     ligandvec = rep(0,length(nodes))\n",
      "+ }\n",
      "> \n",
      "> out = cbind(out,ligandvec[rownames(out)])\n",
      "> colnames(out) = c(\"RSA\",\"Degree\",\"Degree_uniqueAtoms\",\"SecondOrderDegree\",\"SecondOrderDegree_uniqueAtoms\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_unqiueAtoms\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_uniqueAtoms\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_uniqueAtoms\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_uniqueAtoms\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_uniqueAtoms\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_uniqueAtoms\",\"Ligand\")\n",
      "> \n",
      "> #add back in nodes that were not networked as zeros\n",
      "> zeroMat = matrix(0,nrow=length(which(originalNodes%in%nodes==FALSE)),ncol=ncol(out));rownames(zeroMat)=originalNodes[which(originalNodes%in%nodes==FALSE)]\n",
      "> out = rbind(out,zeroMat)\n",
      "> \n",
      "> #fix RSA and Bfactor and ligand\n",
      "> out[,\"RSA\"] = rsa[rownames(out),1]\n",
      "> out[,\"Ligand\"] = ligandvec[rownames(out)]\n",
      "> \n",
      "> #Add 1 to everything to avoid zeros, except ligand\n",
      "> out[is.na(out)]=0\n",
      "> out[out[,\"Ligand\"]==0,\"Ligand\"] = 150\n",
      "> out[1:nrow(out),2:ncol(out)] = as.numeric(out[1:nrow(out),2:ncol(out)])+1\n",
      "> \n",
      "> ##create Z score file\n",
      "> outZ = apply(out,2,function(x){return(scale(as.numeric(x)))})\n",
      "> rownames(outZ) = rownames(out)\n",
      "> colnames(outZ) = colnames(out)\n",
      "> outZ[outZ==NA]=0\n",
      "> outZ[outZ==\"NaN\"]=0\n",
      "> \n",
      "> #Write out scores and z-normed scores files\n",
      "> write.table(out,file=paste(keyword,\"_scoresCentroid\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> write.table(outZ,file=paste(keyword,\"_scoresCentroidZ\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "> \n",
      "Cleaning PDB file\n",
      "Starting phenix.pdbtools\n",
      "on Sun Jun  2 13:36:32 2024 by yliy0004\n",
      "===============================================================================\n",
      "\n",
      "Processing files:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Found model, 7DHY_multimer.pdb\n",
      "\n",
      "Processing PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Adding command-line PHIL:\n",
      "  -------------------------\n",
      "    remove=water\n",
      "\n",
      "Final processed PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "  data_manager {\n",
      "    model {\n",
      "      file = \"7DHY_multimer.pdb\"\n",
      "    }\n",
      "    default_model = \"7DHY_multimer.pdb\"\n",
      "  }\n",
      "  modify {\n",
      "    remove = water\n",
      "  }\n",
      "\n",
      "\n",
      "Starting job\n",
      "===============================================================================\n",
      "Validating inputs\n",
      "Performing manipulations\n",
      "Size before: 6723 size after: 6206\n",
      "Writing output model\n",
      "\n",
      "===============================================================================\n",
      "Job complete\n",
      "usr+sys time: 2.16 seconds\n",
      "wall clock time: 2.33 seconds\n",
      "Writing out polypeptide file...\n",
      "Adding hydrogens...\n",
      "Protonating waters...\n",
      "Making .phi file\n",
      "Running full bond calculation script\n",
      "The ligands identified in this pdb are GOL-306D, GOL-303C, GOL-303B, SO4-303D, ARS-301D, ARS-301C, ARS-301B, ARS-301A, GOL-304B, GOL-305D, GOL-304D, GOL-305B\n",
      "Screening for atoms that are close.\n",
      "Dictionaries loaded.\n",
      "Starting salt bridges.\n",
      "Starting hydrogen bonds.\n",
      "Starting pi-pi version 2.\n",
      "Running pi-cation version 2\n",
      "Starting disulfide\n",
      "Starting van der waals.\n",
      "Starting metals\n",
      "Starting DNA bonds...\n",
      "Starting Ligand centroids\n",
      "Starting residue centroids...\n",
      "Removing duplicates from BFactor file\n",
      "Making 7DHY_multimer_nowaters_net\n",
      "Removing any negative value edges\n",
      "Creating QC file\n",
      "Creating Centroid directory and moving files\n",
      "Running energetic R script\n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> ##Load libraries\n",
      "> library(igraph)\n",
      "> library(stats)\n",
      "> \n",
      "> ##Initialize keywords\n",
      "> # keyword is the base filename\n",
      "> keyword = args[1]\n",
      "> args=c(paste(args[1],\"_net\",sep=\"\"),\"weighted\",paste(keyword,\"_secondaryStructure\",sep=\"\"))\n",
      "> atomCorrection=FALSE\n",
      "> baseKeyword = unlist(strsplit(keyword,\"/\"))[1]\n",
      "> runVariableSelection = TRUE\n",
      "> \n",
      "> ##Load inversal data\n",
      "> #Unique atoms file\n",
      "> terminalAtomsFile = as.matrix(read.table(\"../../terminalAtoms\",sep=\"\\t\"))\n",
      "> terminalAtoms = {}\n",
      "> for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+     terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\", \"))\n",
      "+ }\n",
      "> \n",
      "> #Secondary structure file\n",
      "> secStructure = as.matrix(read.table(args[3]))[,1:2]\n",
      "> rownames(secStructure) = secStructure[,1]\n",
      "> \n",
      "> #RSA\n",
      "> rsa = as.matrix(read.table(paste(keyword,\".rsa\",sep=\"\")))\n",
      "> \n",
      "> ##Define functions\n",
      "> \n",
      "> #Add terminal atom details\n",
      "> addTermDetails = function(data) {\n",
      "+     out_details = c()\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+         if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+             code1 = 1\n",
      "+         } else {\n",
      "+             code1 = 0\n",
      "+         }\n",
      "+         if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+             code2 = 1\n",
      "+         } else {\n",
      "+             code2 = 0\n",
      "+         }\n",
      "+     out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+     }\n",
      "+     return(out_details)\n",
      "+ }\n",
      "> \n",
      "> collapse = function(data) {\n",
      "+     if (nrow(data)==0) {\n",
      "+         out = matrix(ncol = 8,nrow=0)\n",
      "+         colnames(out) = c(\"AA1\", \"AA2\", \"Type\", \"SumEdges\", \"percentMC\", \"influence1\", \"influence2\",\"degree\")\n",
      "+         results = c(); results$out = out\n",
      "+         return(results)\n",
      "+         break\n",
      "+     }\n",
      "+     #Edges should be listed bi-directionally\n",
      "+     x=unlist(lapply(data[,3],function(x){if(x==\"MCSC\"){return(\"SCMC\")}else if(x==\"SCMC\"){return(\"MCSC\")} else {return(x)}}))\n",
      "+     if (dim(data)[2]==7) {\n",
      "+         data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6]))\n",
      "+     } else {\n",
      "+         data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6],data[,9],data[,8]))\n",
      "+     }\n",
      "+     data = unique(data)\n",
      "+     out = c()\n",
      "+     out_noPP = c()\n",
      "+     completed = c()\n",
      "+     if (dim(data)[2]==7) {\n",
      "+         out_details = c()\n",
      "+         for (i in 1:nrow(data)) {\n",
      "+             acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+             acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+             if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+                 code1 = 1\n",
      "+             } else {\n",
      "+                 code1 = 0\n",
      "+             }\n",
      "+             if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+                 code2 = 1\n",
      "+             } else {\n",
      "+                 code2 = 0\n",
      "+             }\n",
      "+             out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+         }\n",
      "+         data = out_details\n",
      "+     } else {\n",
      "+         out_details = data\n",
      "+     }\n",
      "+     colnames(out_details) = c(\"AA1\",\"AA2\",\"Type1\",\"Weight\",\"Type2\",\"Atom1\",\"Atom2\",\"influence1\",\"influence2\")\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         if (i %in% completed == FALSE) {\n",
      "+             x=which((data[,1]==data[i,1] & data[,2] == data[i,2]))\n",
      "+             subdata = matrix(data[x,],ncol=9)\n",
      "+             edgeSum = 0\n",
      "+             MCMCcount = 0\n",
      "+             MCSCcount = 0\n",
      "+             SCMCcount = 0\n",
      "+             SCSCcount = 0\n",
      "+             influence1 = sum(as.numeric(subdata[,8]))\n",
      "+             influence2 = sum(as.numeric(subdata[,9]))\n",
      "+             for (j in 1:nrow(subdata)) {\n",
      "+                 edgeSum = edgeSum + as.numeric(subdata[j,4])\n",
      "+                 if (subdata[j,3]==\"MCMC\" | subdata[j,3]==\"MCSC\") {\n",
      "+                     MCMCcount = MCMCcount+1\n",
      "+                     MCSCcount = MCSCcount+1\n",
      "+                 } else if (subdata[j,3]==\"SCMC\") {\n",
      "+                     SCMCcount = SCMCcount+1\n",
      "+                 } else if (subdata[j,3]==\"SCSC\") {\n",
      "+                     SCSCcount = SCSCcount+1\n",
      "+                 }\n",
      "+             }\n",
      "+             #Header is AA1, AA2, mixed, sum of edges, percent MC, influence1, influence2, degreeAA1\n",
      "+             out = rbind(out,c(data[i,1],data[i,2],\"mixed\",edgeSum,(MCMCcount+MCSCcount)/(MCMCcount+MCSCcount+SCMCcount+SCSCcount),influence1,influence2,sum(data[,1]==data[i,1])))\n",
      "+             if (any(subdata[,5]==\"PP\")==FALSE) {out_noPP = rbind(out_noPP,c(data[i,1],data[i,2],\"mixed\",edgeSum,(MCMCcount+MCSCcount)/(MCMCcount+MCSCcount+SCMCcount+SCSCcount),influence1,influence2,sum(data[,1]==data[i,1])))}\n",
      "+             completed = c(completed,x)\n",
      "+         }\n",
      "+     }\n",
      "+     out = out[out[,4]>0,]\n",
      "+     out_noPP = out_noPP[out_noPP[,4]>0,]\n",
      "+     colnames(out) = c(\"AA1\", \"AA2\", \"Type\", \"SumEdges\", \"percentMC\", \"influence1\", \"influence2\",\"degree\")\n",
      "+     results = c()\n",
      "+     results$out = out\n",
      "+     results$out_noPP = out_noPP\n",
      "+     results$influence_all = out_details[as.numeric(out_details[,4])>0 & (out_details[,8]==\"1\" | out_details[,9]==\"1\"),]\n",
      "+     out_details = out_details[as.numeric(out_details[,4])>0,]\n",
      "+     out_subtract = out_details; out_subtract[out_subtract[,8]==0 & out_subtract[,9]==0,4] = (-1)*as.numeric(out_subtract[out_subtract[,8]==0 & out_subtract[,9]==0,4])\n",
      "+     out_subtract = out_subtract[out_subtract[,5]!=\"PP\",]\n",
      "+     results$out_subtract = out_subtract\n",
      "+     return(results)\n",
      "+ }\n",
      "> \n",
      "> removeRedundancy = function(dataTmp) {\n",
      "+     for (i in 1:nrow(dataTmp)) {\n",
      "+         if (order(dataTmp[i,1:2])[1]==2) {\n",
      "+             dataTmp[i,1:2] = c(dataTmp[i,2],dataTmp[i,1])\n",
      "+             dataTmp[i,3] = paste(unlist(strsplit(dataTmp[i,3],split=\"\"))[c(3,4,1,2)],collapse=\"\")\n",
      "+             dataTmp[i,c(6,7)] = c(dataTmp[i,7],dataTmp[i,6])\n",
      "+         }\n",
      "+     }\n",
      "+     dataTmp = unique(dataTmp)\n",
      "+     pairs = unique(dataTmp[,1:2])\n",
      "+     out = c()\n",
      "+     for (j in 1:nrow(pairs)) {\n",
      "+         pair = pairs[j,]\n",
      "+         subData = matrix(dataTmp[dataTmp[,1]==pair[1] & dataTmp[,2]==pair[2],],ncol=7)\n",
      "+         out = rbind(out, c(pair[1],pair[2],\"mixed\",0.5*sum(as.numeric(subData[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> collapse_directed = function(data) {\n",
      "+     pairs = unique(data[,1:2])\n",
      "+     out = c()\n",
      "+     for (i in 1:nrow(pairs)) {\n",
      "+         subdat = matrix(data[data[,1]==pairs[i,1]&data[,2]==pairs[i,2],],ncol=4)\n",
      "+         out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> collapse_agnostic = function(data) {\n",
      "+     pairs = unique(t(apply(data[,1:2],1,function(x){return(sort(x))})))\n",
      "+     out = c()\n",
      "+     for (i in 1:nrow(pairs)) {\n",
      "+         subdat = matrix(data[(data[,1]==pairs[i,1]&data[,2]==pairs[i,2]) | (data[,2]==pairs[i,1]&data[,1]==pairs[i,2]),],ncol=4)\n",
      "+         out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> distributCalc = function(set,nodes,net) {\n",
      "+     #x is assumed to be a set but can also just be 1 node\n",
      "+     distanceNet = distances(net,weights=net$weight)\n",
      "+     distanceTmp = c()\n",
      "+     Dr = c()\n",
      "+     for (node in nodes) {\n",
      "+         distanceTmp = c()\n",
      "+         for (item in set) {\n",
      "+             if (item!=node) {\n",
      "+                 if(item%in%rownames(distanceNet)==FALSE){distanceTmp=100;next}\n",
      "+                 if(node%in%rownames(distanceNet)==FALSE){distanceTmp=100;next}\n",
      "+                 distanceTmp = c(distanceTmp,distanceNet[item,node])\n",
      "+             }\n",
      "+         }\n",
      "+         if (length(distanceTmp)>0) {\n",
      "+             Dr = c(Dr, 1/min(distanceTmp))\n",
      "+         }\n",
      "+     }\n",
      "+     return(sum(Dr)/length(nodes))\n",
      "+ }\n",
      "> \n",
      "> ##Read in user data\n",
      "> data_all = as.matrix(read.table(args[1]))\n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> originalNodes = nodes\n",
      "> basedata = collapse(data_all)\n",
      "> basedata_noPP = removeRedundancy(basedata$out[,1:7])\n",
      "> basedata_allsidechain = collapse_agnostic(data_all[data_all[,3]!=\"MCMC\",c(1,2,3,4)])\n",
      "> influencedata_all = basedata$influence_all\n",
      "> influencedata_collapse = collapse(influencedata_all)\n",
      "> influencedata = removeRedundancy(influencedata_collapse$out[,1:7])\n",
      "> influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "> influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "> \n",
      "> influencedata_directed = collapse_directed(influencedata_all[influencedata_all[,8]==1,c(1,2,3,4)])\n",
      "> influencenet_directed = graph.edgelist(influencedata_directed[,c(2,1)],directed=TRUE)\n",
      "> influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "> \n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> \n",
      "> #Write out edgelist files\n",
      "> write.table(influencedata,file=paste(keyword,\"_allEdges\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=FALSE,row.names=FALSE)\n",
      "> write.table(influencedata_directed,file=paste(keyword,\"_uniqueEdges\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=FALSE,row.names=FALSE)\n",
      "> \n",
      "> ###Deal with missing nodes\n",
      "> if (any(nodes %in% secStructure[,1]==FALSE)) {\n",
      "+     for (node in nodes) {\n",
      "+         if (node %in% secStructure[,1]==FALSE) {\n",
      "+             secStructure = rbind(secStructure,c(node,\"xxx\",360))\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> for (node in nodes) {\n",
      "+     if (node %in% rownames(rsa) == FALSE) {\n",
      "+         rsa = rbind(rsa,c(0,0))\n",
      "+         rownames(rsa)[nrow(rsa)] = node\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> ##Begin centrality calculations\n",
      "> \n",
      "> #Modular calculations\n",
      "> \n",
      "> ##WALKTRAP\n",
      "> net = graph.edgelist(basedata_noPP[,1:2],directed=FALSE)\n",
      "> net$weight = 1/as.numeric(basedata_noPP[,4])\n",
      "> net_community=walktrap.community(net,weights=net$weight); net_community_vec = net_community$membership; names(net_community_vec)=net_community$names\n",
      "> net_community_vec_wt = net_community_vec\n",
      "> nodes = V(net)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> nodes.color = c()\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(influencedata)) {\n",
      "+     j = which((basedata_noPP[,1]==influencedata[i,1]&basedata_noPP[,2]==influencedata[i,2])|(basedata_noPP[,1]==influencedata[i,2]&basedata_noPP[,2]==influencedata[i,1]))\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) { edgecolors[j]=\"grey40\" }\n",
      "+ }\n",
      "> if(length(nodes)>200) {textsize=.25} else {textsize=.45}\n",
      "> plot(net,edge.color = edgecolors, vertex.color = nodes.color,vertex.size=4,vertex.label.cex=textsize,main=\"WALKTRAP\")\n",
      ">     \n",
      "> ##Weighted edge betweenness WALKTRAP\n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP - WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for WALKTRAP\n",
      "> firstOrderDegree = degree(influencenet)\n",
      "> firstOrderDegree_sidechain = degree(influencenet_directed,mode=c(\"in\"))\n",
      "> secondOrderDegree = c(); secondOrderDegree_sidechain=c()\n",
      "> nodes = V(influencenet)$name\n",
      "> for (node in nodes) {\n",
      "+     firstorder = neighbors(influencenet,node)\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet,neighbor)))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree = c(secondOrderDegree,length(secondorder))\n",
      "+     firstorder = neighbors(influencenet_directed,node,mode=c(\"in\"))\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet_directed,neighbor,mode=c(\"in\"))))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree_sidechain = c(secondOrderDegree_sidechain,length(secondorder))\n",
      "+ }\n",
      "> names(secondOrderDegree) = nodes; names(secondOrderDegree_sidechain) = nodes\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])])) - 1\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])])) - 1\n",
      "+ }\n",
      "> \n",
      "> #SAVE ALL AS WALKTRAP\n",
      "> node_edge_betweenness_wt = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_wt = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_wt = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_wt = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_wt = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_wt = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> ##2ARY STRUCTURE\n",
      "> net_community_vec = as.numeric(as.factor(secStructure[,2]))\n",
      "> names(net_community_vec) = secStructure[,1]\n",
      "> nodes = V(net)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> nodes.color = c()\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(basedata_noPP)) {\n",
      "+     if (net_community_vec[basedata_noPP[i,1]]!=net_community_vec[basedata_noPP[i,2]]) { edgecolors[i]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for 2ARY STRUCTURE\n",
      "> nodes = V(influencenet)$name\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])])) - 1\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])])) - 1\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #SAVE ALL AS STRIDE\n",
      "> node_edge_betweenness_stride = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_stride = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_stride = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_stride = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_stride = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_stride = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> \n",
      "> ### Create final dataset\n",
      "> \n",
      "> out = cbind(\n",
      "+ firstOrderDegree[nodes],\n",
      "+ firstOrderDegree_sidechain[nodes],\n",
      "+ secondOrderDegree[nodes],\n",
      "+ secondOrderDegree_sidechain[nodes],\n",
      "+ node_edge_betweenness_stride[nodes],\n",
      "+ node_edge_betweenness_sidechain_stride[nodes],\n",
      "+ node_intermodular_degree_stride[nodes],\n",
      "+ node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ node_edge_betweenness_wt[nodes],\n",
      "+ node_edge_betweenness_sidechain_wt[nodes],\n",
      "+ node_intermodular_degree_wt[nodes],\n",
      "+ node_intermodular_degree_sidechain_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_wt[nodes]\n",
      "+ )\n",
      "> rownames(out) = nodes\n",
      "> colnames(out) = c(\"Degree\",\"Degree_sidechain\",\"SecondOrderDegree\",\"SecondOrderDegree_sidechain\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_sidechain\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_sidechain\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_sidechain\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_sidechain\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_sidechain\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_sidechain\")\n",
      "> \n",
      "> #add back in nodes that were not networked as zeros\n",
      "> zeroMat = matrix(0,nrow=length(which(originalNodes%in%nodes==FALSE)),ncol=ncol(out));rownames(zeroMat)=originalNodes[which(originalNodes%in%nodes==FALSE)]\n",
      "> out = rbind(out,zeroMat)\n",
      "> \n",
      "> #Add 1 to everything to avoid zeros, except ligand\n",
      "> out[is.na(out)]=0\n",
      "> out = out+1\n",
      "> \n",
      "> #Write out\n",
      "> write.table(out,file=paste(keyword,\"_scoresEnergetics\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "> #Standard normalization\n",
      "> outZ = apply(out,2,function(x){return(scale(x))})\n",
      "> rownames(outZ) = rownames(out)\n",
      "> write.table(outZ,file=paste(keyword,\"_scoresEnergeticsZ\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "Running centroid R script\n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(igraph)\n",
      "> library(stats)\n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> ##read in data\n",
      "> ## arg1: col1 and col2 are edges, col3 is weight of edge\n",
      "> ## arg2: weighted, unweighted <-- no longer using\n",
      "> ## arg3: forced module definition <-- no longer using\n",
      "> \n",
      "> # Set desired parameters\n",
      "> removeMCMC = FALSE\n",
      "> linearcutoff = 1\n",
      "> directed = FALSE\n",
      "> secondaryStructure2 = FALSE\n",
      "> sidechainMode = TRUE ##build network of just SCSC, SCMC and MCSC. Still use unique atoms for the _uniqueAtoms directed plots\n",
      "> uniqueAtomsMode = FALSE ##build network of just unique atoms (each edge has to have at least 1). Still use unique atoms for the _uniqueAtoms directed plots\n",
      "> useDNA=TRUE\n",
      "> uniqueAtomsUnbiased=FALSE\n",
      "> uniqueAtomsGaurav=FALSE\n",
      "> uniqueAtomsOLD=FALSE\n",
      "> atomCorrection = FALSE\n",
      "> uniqueAtomsGauravPP = FALSE\n",
      "> weighted = TRUE\n",
      "> removeWaters=TRUE\n",
      "> if (tail(unlist(strsplit(getwd(),\"/\")),n=1)==\"Centroid\") {\n",
      "+     ligandCentroidMode = TRUE\n",
      "+ } else {\n",
      "+     ligandCentroidMode = FALSE\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> # Read in ligand file\n",
      "> ligandfile = scan(paste(args[1],\"_ligand\",sep=\"\"),what=\"character\")\n",
      "> if (length(ligandfile)==0){\n",
      "+     ligandmode = FALSE\n",
      "+ } else {\n",
      "+     ligandmode = TRUE\n",
      "+ }\n",
      "> \n",
      "> ligandCentroidFile = scan(paste(args[1],\"NetLigand\",sep=\"\"),what=\"character\")\n",
      "> \n",
      "> # keyword is the base filename\n",
      "> # method is \"frag\" or \"module\" or \"both\"\n",
      "> keyword = args[1]\n",
      "> args=c(paste(args[1],\"_net\",sep=\"\"),\"weighted\",dir(\"./\",\"_secondaryStructure$\"))\n",
      "> \n",
      "> ##Load data\n",
      "> #Unique atoms file\n",
      "> if(uniqueAtomsUnbiased) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"/Users/vjpatel/Dropbox/Gaurav/uniqueAtomsUnbiased\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else if (uniqueAtomsGaurav) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"/Users/vjpatel/Dropbox/Gaurav/uniqueAtomsGaurav\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else if (uniqueAtomsOLD) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"../../../terminalAtoms\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"../../../uniqueAtoms\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     otherTA = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+         #otherTA[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,4],split=\",\"))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> #Secondary structure file\n",
      "> #Ligands are all one secondary structure\n",
      "> secStructure = as.matrix(read.table(args[3]))\n",
      "> phiAngleDB = {}\n",
      "> for (i in 1:nrow(secStructure)) {\n",
      "+     phiAngleDB[[secStructure[i,1]]] = as.numeric(secStructure[i,3])\n",
      "+ }\n",
      "> \n",
      "> if(ligandmode) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",\"_ligand\"),sep=\"\\t\"))\n",
      "+     liganddat = unique(liganddat[,2])\n",
      "+     for (ligand in liganddat) {\n",
      "+         secStructure = rbind(secStructure,c(sub(\"-\",\"\",ligand),\"ligand\",360))\n",
      "+         phiAngleDB[[sub(\"-\",\"\",ligand)]]=0\n",
      "+     }\n",
      "+ } else {\n",
      "+     liganddat = c()\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> ##Define functions\n",
      "> \n",
      "> ##Collapse matrix functions\n",
      "> ##Will return redundant matrix\n",
      "> ##Input needs to be 7 columns\n",
      "> ##Second input is different if the main input is a subset of the larger data\n",
      "> addTerminalDetails = function(data,largerdata) {\n",
      "+     #List edges bi-directionally\n",
      "+     x=unlist(lapply(data[,3],function(x){if(x==\"MCSC\"){return(\"SCMC\")}else if(x==\"SCMC\"){return(\"MCSC\")} else {return(x)}}))\n",
      "+     data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6]))\n",
      "+     data = unique(data)\n",
      "+     out = c()\n",
      "+     out_noPP = c()\n",
      "+     out_details = c()\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+         subdata = matrix(data[data[,1]==data[i,1]&data[,2]==data[i,2],],ncol=7)\n",
      "+         allsubdata = matrix(largerdata[largerdata[,1]==data[i,1],],ncol=7)\n",
      "+         code1=0;code2=0\n",
      "+         if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+             code1 = 1\n",
      "+         }\n",
      "+         if (data[i,5] == \"DNA\" & useDNA==TRUE) {\n",
      "+             code1 = 1\n",
      "+         }\n",
      "+         allsubdata = matrix(largerdata[largerdata[,2]==data[i,2],],ncol=7)\n",
      "+         if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+             code2 = 1\n",
      "+         }\n",
      "+         if (data[i,5] == \"DNA\" & useDNA==TRUE) {\n",
      "+             code2 = 1\n",
      "+         }\n",
      "+         if (data[i,5]==\"PICATION\"|data[i,5]==\"PIPI\") {\n",
      "+             code1=1\n",
      "+             code2=1\n",
      "+         }\n",
      "+         if (data[i,5]==\"PP\") {\n",
      "+             if (uniqueAtomsGaurav) {\n",
      "+                 if((acid1) == \"GLY\") {\n",
      "+                     code1 = 1\n",
      "+                     data[i,3]==\"SCSC\"\n",
      "+                 }\n",
      "+                 if ((acid2) == \"GLY\") {\n",
      "+                     code2 = 1\n",
      "+                     data[i,3] == \"SCSC\"\n",
      "+                 }\n",
      "+             }\n",
      "+         }\n",
      "+         out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+     }\n",
      "+     return(out_details)\n",
      "+ }\n",
      "> \n",
      "> ##Will return non-redundant matrix\n",
      "> collapse_agnostic = function(data) {\n",
      "+     if(nrow(data)==0) {\n",
      "+         return(matrix(ncol=4,nrow=0))\n",
      "+     } else if (ncol(data)==4 & data[1,3]==\"mixed\") {\n",
      "+         return(data)\n",
      "+     } else {\n",
      "+         pairs = unique(t(apply(matrix(data[,1:2],ncol=2),1,function(x){return(sort(x))})))\n",
      "+         out = c()\n",
      "+         for (i in 1:nrow(pairs)) {\n",
      "+             subdat = matrix(data[(data[,1]==pairs[i,1]&data[,2]==pairs[i,2]) | (data[,2]==pairs[i,1]&data[,1]==pairs[i,2]),1:4],ncol=4)\n",
      "+             subdat = unique(t(apply(subdat,1,function(x){a=which(x==sort(x[1:2])[1]);if(a==1){return(x)}else{return(c(x[2],x[1],paste(unlist(strsplit(x[3],\"\"))[c(3,4,1,2)],collapse=\"\"),x[4]))}})))\n",
      "+             if (weighted==FALSE) {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",1))\n",
      "+             } else {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+             }\n",
      "+         }\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> ##Will return directed collapsed matrix\n",
      "> collapse_directed = function(data) {\n",
      "+     if (length(data)==0) {\n",
      "+         return(matrix(nrow=0,ncol=4))\n",
      "+     } else {\n",
      "+         pairs = matrix(unique(data[,1:2]),ncol=2)\n",
      "+         out = c()\n",
      "+         for (i in 1:nrow(pairs)) {\n",
      "+             subdat = matrix(data[data[,1]==pairs[i,1]&data[,2]==pairs[i,2],1:4],ncol=4)\n",
      "+             if (weighted==FALSE) {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",1))\n",
      "+             } else {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+             }\n",
      "+         }\n",
      "+         return(matrix(out,ncol=4))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> rsa = as.matrix(read.table(dir(\"./\",\".rsa$\")))\n",
      "> rownames(rsa) = rsa[,1]\n",
      "> rsa[,1] = rsa[,2]\n",
      "> \n",
      "> #add ligand as -1\n",
      "> for (ligand in liganddat) {\n",
      "+     rsa = rbind(rsa,c(-1,-1))\n",
      "+     rownames(rsa)[nrow(rsa)] = sub(\"-\",\"\",ligand)\n",
      "+ }\n",
      "> \n",
      "> #Function to get amino acid\n",
      "> getAcid = function(residue) {\n",
      "+     if(all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"A\")) == TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"G\"))==TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"C\"))==TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"T\"))==TRUE)) {return(paste(unlist(strsplit(residue,\"\"))[1:2],collapse=\"\"))\n",
      "+     } else {\n",
      "+     return(paste(unlist(strsplit(residue,\"\"))[1:3],collapse=\"\"))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> ##################\n",
      "> ## read in data ##\n",
      "> ##################\n",
      "> data_all = as.matrix(read.table(args[1]))[,1:7]\n",
      "> if(removeWaters) {\n",
      "+     y=c(grep(\"HOH\",data_all[,1]),grep(\"HOH\",data_all[,2]))\n",
      "+     if(length(y)>0) {\n",
      "+         data_all = data_all[-y,]\n",
      "+     }\n",
      "+ }\n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> originalNodes = nodes\n",
      "> \n",
      "> data_all_original = rbind(data_all,data_all[,c(2,1,3,4,5,7,6)])\n",
      "> data_all = data_all[as.numeric(data_all[,4])>0,]\n",
      "> \n",
      "> basedata = addTerminalDetails(data_all,data_all_original)\n",
      "> basedata_noPP = collapse_agnostic(basedata[basedata[,5]!=\"PP\",])\n",
      "> sidechaindata = basedata[basedata[,3]!=\"MCMC\",1:4]\n",
      "> sidechaindata_detailed = basedata[basedata[,3]!=\"MCMC\",]\n",
      "> \n",
      "> influencedata_detailed = basedata\n",
      "> if (uniqueAtomsMode) {\n",
      "+     influencedata_detailed = basedata[basedata[,8]==1 | basedata[,9]==1,]\n",
      "+ }\n",
      "> influencedata = collapse_agnostic(influencedata_detailed)\n",
      "> influencedataGLY = rbind(influencedata[grep(\"GLY\",influencedata[,1]),],influencedata[grep(\"GLY\",influencedata[,1]),])\n",
      "> influencedata = influencedata[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,2])))>linearcutoff,]\n",
      "> \n",
      "> \n",
      "> influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "> influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "> \n",
      "> influencedata_directed = collapse_directed(matrix(influencedata_detailed[influencedata_detailed[,8]==1,1:4],ncol=4))\n",
      "> influencedata_directedGLY = rbind(influencedata_directed[grep(\"GLY\",influencedata_directed[,1]),],influencedata_directed[grep(\"GLY\",influencedata_directed[,2]),])\n",
      "> influencedata_directed = influencedata_directed[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,2])))>linearcutoff,]\n",
      "> \n",
      "> influencedata_directed = matrix(influencedata_directed,ncol=4)\n",
      "> influencenet_directed = graph.edgelist(matrix(influencedata_directed[,c(2,1)],ncol=2),directed=TRUE)\n",
      "> influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "> \n",
      "> \n",
      "> if(sidechainMode) {\n",
      "+     influencedata_detailed = sidechaindata_detailed\n",
      "+     influencedata = collapse_agnostic(influencedata_detailed)\n",
      "+     influencedataGLY = rbind(influencedata[grep(\"GLY\",influencedata[,1]),],influencedata[grep(\"GLY\",influencedata[,2]),])\n",
      "+     influencedata = influencedata[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,2])))>linearcutoff,]\n",
      "+     #Add back for GLY if uniqueAtomsGaurav\n",
      "+     if(uniqueAtomsGauravPP) {\n",
      "+         influencedata = rbind(influencedata,influencedataGLY)\n",
      "+     }\n",
      "+     influencedata = influencedata[as.numeric(influencedata[,4])>0,]\n",
      "+     \n",
      "+     influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "+     influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "+     \n",
      "+     influencedata_directed = collapse_directed(matrix(influencedata_detailed[influencedata_detailed[,8]==1,1:4],ncol=4))\n",
      "+     influencedata_directedGLY = rbind(influencedata_directed[grep(\"GLY\",influencedata_directed[,1]),],influencedata_directed[grep(\"GLY\",influencedata_directed[,2]),])\n",
      "+ \n",
      "+     influencedata_directed = influencedata_directed[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,2])))>linearcutoff,]\n",
      "+     #Add back for GLY if uniqueAtomsGaurav\n",
      "+     if(uniqueAtomsGauravPP) {\n",
      "+         influencedata_directed = rbind(influencedata_directed,influencedata_directedGLY)\n",
      "+     }\n",
      "+     influencedata_directed = matrix(influencedata_directed,ncol=4)\n",
      "+     influencenet_directed = graph.edgelist(matrix(influencedata_directed[,c(2,1)],ncol=2),directed=TRUE)\n",
      "+     influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "+     \n",
      "+ }\n",
      "> \n",
      "> #Replace missing nodes in the igraph net\n",
      "> nodesMissing = nodes[nodes %in% influencedata[,1]==FALSE & nodes %in% influencedata[,2]==FALSE]\n",
      "> \n",
      "> ##B-factor\n",
      "> bfactor = as.matrix(read.table(dir(\"./\",\"_Bfactor\"),row.names=1))\n",
      "> \n",
      "> #Replace missing nodes in secStructure matrix\n",
      "> if (any(nodes %in% secStructure[,1]==FALSE)) {\n",
      "+     for (node in nodes) {\n",
      "+         if (node %in% secStructure[,1]==FALSE) {\n",
      "+             secStructure = rbind(secStructure,c(node,\"xxx\",360))\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> for (node in nodes) {\n",
      "+     if (node %in% rownames(rsa) == FALSE) {\n",
      "+         rsa = rbind(rsa,c(0,0))\n",
      "+         rownames(rsa)[nrow(rsa)] = node\n",
      "+     }\n",
      "+     if (node %in% rownames(bfactor) == FALSE) {\n",
      "+         bfactor = rbind(bfactor,1)\n",
      "+         rownames(bfactor)[nrow(bfactor)] = node\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> \n",
      "> \n",
      "> #Give DNA attributes\n",
      "> if (\"DNA\" %in% data_all[,5]) {\n",
      "+     for (i in 1:nrow(data_all)) {\n",
      "+         if (data_all[i,5]==\"DNA\") {\n",
      "+             phiAngleDB[[data_all[i,1]]] = 90\n",
      "+             if (data_all[i,1] %in% rownames(rsa) == FALSE) {\n",
      "+                 rsa = rbind(rsa,c(0,0))\n",
      "+                 rownames(rsa)[nrow(rsa)] = data_all[i,1]\n",
      "+             }\n",
      "+             if (data_all[i,1] %in% rownames(secStructure) == FALSE) {\n",
      "+                 secStructure = rbind(secStructure,c(data_all[i,1],\"DNA\",360))\n",
      "+             }\n",
      "+             if (data_all[i,1] %in% rownames(bfactor) == FALSE) {\n",
      "+                 bfactor = rbind(bfactor,1)\n",
      "+                 rownames(bfactor)[nrow(bfactor)] = data_all[i,1]\n",
      "+             }\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #Betweenness\n",
      "> node_betweenness = betweenness(influencenet,weights=influencenet$weight,directed=directed)\n",
      "> node_betweenness_unique = betweenness(influencenet_directed,weights=influencenet_directed$weight,directed=directed)\n",
      "> \n",
      "> #Modular calculations\n",
      "> \n",
      "> ##WALKTRAP\n",
      "> net = graph.edgelist(basedata_noPP[,1:2],directed=FALSE)\n",
      "> net$weight = 1/as.numeric(basedata_noPP[,4])\n",
      "> net_community=walktrap.community(net,weights=net$weight); net_community_vec = net_community$membership; names(net_community_vec)=net_community$names\n",
      "> net_community_vec_wt = net_community_vec\n",
      "> #net_community_vec[buriednodes] = max(net_community_vec)+1\n",
      "> nodes = V(influencenet)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> nodes.color = c()\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(influencedata)) {\n",
      "+     j = which((basedata_noPP[,1]==influencedata[i,1]&basedata_noPP[,2]==influencedata[i,2])|(basedata_noPP[,1]==influencedata[i,2]&basedata_noPP[,2]==influencedata[i,1]))\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) { edgecolors[j]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP\n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP - WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         if(weight1==0 & weight2==0) {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]] + 0\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]] + 0\n",
      "+         } else {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for WALKTRAP\n",
      "> firstOrderDegree = degree(influencenet)\n",
      "> firstOrderDegree_sidechain = degree(influencenet_directed,mode=c(\"in\"))\n",
      "> \n",
      "> secondOrderDegree = c(); secondOrderDegree_sidechain=c()\n",
      "> nodes = V(influencenet)$name\n",
      "> for (node in nodes) {\n",
      "+     firstorder = neighbors(influencenet,node)\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet,neighbor)))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree = c(secondOrderDegree,length(secondorder))\n",
      "+     if(node %in% V(influencenet_directed)$name) {\n",
      "+         firstorder_sidechain = neighbors(influencenet_directed,node,mode=c(\"in\"))\n",
      "+     } else {\n",
      "+         firstorder_sidechain = c()\n",
      "+     }\n",
      "+     secondorder = c(); for (neighbor in firstorder_sidechain){secondorder = c(secondorder,names(neighbors(influencenet_directed,neighbor,mode=c(\"in\"))))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree_sidechain = c(secondOrderDegree_sidechain,length(secondorder))\n",
      "+ }\n",
      "> names(secondOrderDegree) = nodes; names(secondOrderDegree_sidechain) = nodes\n",
      "> \n",
      "> \n",
      "> #second order intermodular degree\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])]))\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])]))\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #SAVE ALL AS WALKTRAP\n",
      "> node_edge_betweenness_wt = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_wt = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_wt = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_wt = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_wt = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_wt = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> \n",
      "> ##2ARY STRUCTURE\n",
      "> secStructure = secStructure[,1:2]\n",
      "> rownames(secStructure) = secStructure[,1]\n",
      "> net_community_vec = as.numeric(as.factor(secStructure[,2]))\n",
      "> names(net_community_vec) = secStructure[,1]\n",
      "> #net_community_vec[buriednodes] = max(net_community_vec)+1\n",
      "> nodes = V(influencenet)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> nodes.color = c()\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(basedata_noPP)) {\n",
      "+     if (net_community_vec[basedata_noPP[i,1]]!=net_community_vec[basedata_noPP[i,2]]) { edgecolors[i]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         if(weight1==0 & weight2==0) {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]] + 0\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]] + 0\n",
      "+         } else {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> ##Degree and second order degree for 2ARY STRUCTURE\n",
      "> \n",
      "> #second order intermodular degree\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     #secondOrder_node_intermodular_degree[node] = length(unique(unlist(node_modules[names(node_modules[[node]])])))\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])]))\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])]))\n",
      "+ }\n",
      "> \n",
      "> #SAVE ALL AS STRIDE\n",
      "> node_edge_betweenness_stride = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_stride = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_stride = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_stride = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_stride = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_stride = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> ### Create final dataset\n",
      "> nodes = nodes[nodes!=\"DNA1000A\"]\n",
      "> nodes = nodes[nodes %in% c(\"DA\",\"DG\",\"DC\",\"DT\") == FALSE]\n",
      "> \n",
      "> ### Remove ligand amino acids from final dataset\n",
      "> ligands = sub(\"-\",\"\",liganddat)\n",
      "> nodes = nodes[nodes %in% ligands == FALSE]\n",
      "> \n",
      "> nodes[which(nodes %in% rownames(bfactor)==FALSE)]\n",
      "character(0)\n",
      "> \n",
      "> out = cbind(\n",
      "+ rsa[nodes,1],\n",
      "+ firstOrderDegree[nodes],\n",
      "+ firstOrderDegree_sidechain[nodes],\n",
      "+ secondOrderDegree[nodes],\n",
      "+ secondOrderDegree_sidechain[nodes],\n",
      "+ node_edge_betweenness_stride[nodes],\n",
      "+ node_edge_betweenness_sidechain_stride[nodes],\n",
      "+ node_intermodular_degree_stride[nodes],\n",
      "+ node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ node_edge_betweenness_wt[nodes],\n",
      "+ node_edge_betweenness_sidechain_wt[nodes],\n",
      "+ node_intermodular_degree_wt[nodes],\n",
      "+ node_intermodular_degree_sidechain_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_wt[nodes]\n",
      "+ )\n",
      "> \n",
      "> #write column names\n",
      "> colnames(out) = c(\"RSA\",\"Degree\",\"Degree_uniqueAtoms\",\"SecondOrderDegree\",\"SecondOrderDegree_uniqueAtoms\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_unqiueAtoms\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_uniqueAtoms\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_uniqueAtoms\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_uniqueAtoms\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_uniqueAtoms\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_uniqueAtoms\")\n",
      "> \n",
      "> \n",
      "> ## LIGANDS\n",
      "> if(ligandmode==TRUE & ligandCentroidMode==FALSE) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",\"_ligand\"),sep=\"\\t\"))\n",
      "+     \n",
      "+     tmp = c()\n",
      "+     i=1\n",
      "+     for (i in 1:nrow(liganddat)) {\n",
      "+         acid=paste(unlist(strsplit(liganddat[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         if (unlist(strsplit(liganddat[i,1],\"-\"))[2] %in% terminalAtoms[[acid]]) {\n",
      "+             tmp = rbind(tmp,c(unlist(strsplit(liganddat[i,1],\"-\"))[1],liganddat[i,2],liganddat[i,3]))\n",
      "+         }\n",
      "+     }\n",
      "+     \n",
      "+     #ligandvec = tapply(liganddat[,3],as.factor(liganddat[,1]),min)\n",
      "+     ligandvec = tapply(as.numeric(tmp[,3]),as.factor(tmp[,1]),min)\n",
      "+ }\n",
      "> \n",
      "> if(ligandmode==TRUE & ligandCentroidMode) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",paste(\"_centroidNetLigand$\",sep=\"\")),sep=\"\\t\"))\n",
      "+     \n",
      "+     tmp = c()\n",
      "+     i=1\n",
      "+     for (i in 1:nrow(liganddat)) {\n",
      "+         tmp = rbind(tmp,c(liganddat[i,1],liganddat[i,2],liganddat[i,3]))\n",
      "+     }\n",
      "+     \n",
      "+     ligandvec = tapply(tmp[,3],as.factor(tmp[,1]),min)\n",
      "+ }\n",
      "> \n",
      "> if (ligandmode==FALSE) {\n",
      "+     ligandvec = rep(0,length(nodes))\n",
      "+ }\n",
      "> \n",
      "> out = cbind(out,ligandvec[rownames(out)])\n",
      "> colnames(out) = c(\"RSA\",\"Degree\",\"Degree_uniqueAtoms\",\"SecondOrderDegree\",\"SecondOrderDegree_uniqueAtoms\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_unqiueAtoms\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_uniqueAtoms\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_uniqueAtoms\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_uniqueAtoms\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_uniqueAtoms\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_uniqueAtoms\",\"Ligand\")\n",
      "> \n",
      "> #add back in nodes that were not networked as zeros\n",
      "> zeroMat = matrix(0,nrow=length(which(originalNodes%in%nodes==FALSE)),ncol=ncol(out));rownames(zeroMat)=originalNodes[which(originalNodes%in%nodes==FALSE)]\n",
      "> out = rbind(out,zeroMat)\n",
      "> \n",
      "> #fix RSA and Bfactor and ligand\n",
      "> out[,\"RSA\"] = rsa[rownames(out),1]\n",
      "> out[,\"Ligand\"] = ligandvec[rownames(out)]\n",
      "> \n",
      "> #Add 1 to everything to avoid zeros, except ligand\n",
      "> out[is.na(out)]=0\n",
      "> out[out[,\"Ligand\"]==0,\"Ligand\"] = 150\n",
      "> out[1:nrow(out),2:ncol(out)] = as.numeric(out[1:nrow(out),2:ncol(out)])+1\n",
      "> \n",
      "> ##create Z score file\n",
      "> outZ = apply(out,2,function(x){return(scale(as.numeric(x)))})\n",
      "> rownames(outZ) = rownames(out)\n",
      "> colnames(outZ) = colnames(out)\n",
      "> outZ[outZ==NA]=0\n",
      "> outZ[outZ==\"NaN\"]=0\n",
      "> \n",
      "> #Write out scores and z-normed scores files\n",
      "> write.table(out,file=paste(keyword,\"_scoresCentroid\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> write.table(outZ,file=paste(keyword,\"_scoresCentroidZ\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "> \n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> ## run from within the monomer folder\n",
      "> ## there can only be one *_scoresEnergetics and one Centroid/*_scoresCentroid in this folder\n",
      "> ## args[1] is the folder for the multimer\n",
      "> ## NOTE: results may differ slightly from publication because of missing data in the mutagenesis experiments used in teh paper\n",
      "> \n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> #Load monomer\n",
      "> data1 = read.table(dir(\"./\",\"_scoresEnergetics$\"),header=T,row.names=1)\n",
      "> data1Z = as.data.frame(apply(data1,2,function(x){return(scale(as.numeric(x)))}))\n",
      "> rownames(data1Z) = rownames(data1)\n",
      "> data1Z[is.na(data1Z)] = 0\n",
      "> data2 = read.table(dir(\"./Centroid\",\"_scoresCentroid$\", full.names=TRUE),header=T,row.names=1)\n",
      "> data2Z = as.data.frame(apply(data2,2,function(x){return(scale(as.numeric(x)))}))\n",
      "> rownames(data2Z) = rownames(data2)\n",
      "> data2Z[is.na(data2Z)] = 0\n",
      "> colnames(data2Z) = paste(colnames(data2Z),\"CENTROIDSC\",sep=\"\")\n",
      "> data2Z = data2Z[rownames(data1Z),]\n",
      "> data2 = data2[rownames(data1),]\n",
      "> \n",
      "> #Load multimer\n",
      "> data1Multimer = read.table(dir(args[1],\"_scoresEnergetics$\",full.names=TRUE),header=T,row.names=1)\n",
      "> data2Multimer = read.table(dir(paste(args[1],\"Centroid\",sep=\"/\"),\"_scoresCentroid$\",full.names=TRUE),header=T,row.names=1)\n",
      "> data2Multimer = data2Multimer[rownames(data1Multimer),]\n",
      "> data1MultimerZ = as.data.frame(apply(data1Multimer,2,function(x){return(scale(as.numeric(x)))}))\n",
      "> rownames(data1MultimerZ) = rownames(data1Multimer)\n",
      "> data2MultimerZ = as.data.frame(apply(data2Multimer,2,function(x){return(scale(as.numeric(x)))}))\n",
      "> rownames(data2MultimerZ) = rownames(data2Multimer)\n",
      "> colnames(data1MultimerZ) = paste(colnames(data1MultimerZ),\"MULTIMER\",sep=\"\")\n",
      "> colnames(data2MultimerZ) = paste(colnames(data2MultimerZ),\"MULTIMERCENTROIDSC\",sep=\"\")\n",
      "> data2MultimerZ = data2MultimerZ[rownames(data1MultimerZ),]\n",
      "> \n",
      "> #Add RSA\n",
      "> data1 = cbind(data2[,\"RSA\"],data1)\n",
      "> colnames(data1)[1]=\"RSA\"\n",
      "> dataZ = cbind(data1Z,data2Z)\n",
      "> \n",
      "> #Identify max NodeEdgeBetweennessSTRIDE_sidechain\n",
      "> acidsMonomer = c()\n",
      "> for (i in 1:length(rownames(dataZ))) {\n",
      "+     node = rownames(dataZ)[i]\n",
      "+     node = unlist(strsplit(node,split=\"\"))\n",
      "+     node = paste(node[1:(length(node)-1)],collapse=\"\")\n",
      "+     acidsMonomer = c(acidsMonomer,node)\n",
      "+ }\n",
      "> \n",
      "> acidsMultimer = c()\n",
      "> for (i in 1:length(rownames(data1MultimerZ))) {\n",
      "+     node = rownames(data1MultimerZ)[i]\n",
      "+     node = unlist(strsplit(node,split=\"\"))\n",
      "+     node = paste(node[1:(length(node)-1)],collapse=\"\")\n",
      "+     acidsMultimer = c(acidsMultimer,node)\n",
      "+ }\n",
      "> \n",
      "> allAcids = unique(c(acidsMonomer,acidsMultimer))\n",
      "> \n",
      "> #SecondOrderIntermodularDegree is the average of the multimer only\n",
      "> SecondOrderIntermodularDegree_ALL = (data1MultimerZ[,\"SecondOrderIntermodularDegreeSTRIDE_sidechainMULTIMER\"]+data1MultimerZ[,\"SecondOrderIntermodularDegreeSTRIDEMULTIMER\"]+data2MultimerZ[,\"SecondOrderIntermodularDegreeSTRIDEMULTIMERCENTROIDSC\"]+data2MultimerZ[,\"SecondOrderIntermodularDegreeWALKTRAPMULTIMERCENTROIDSC\"])/4\n",
      "> SecondOrderIntermodularDegree_AVERAGE = tapply(SecondOrderIntermodularDegree_ALL,as.factor(acidsMultimer),mean)\n",
      "> SecondOrderIntermodularDegree_AVERAGE = SecondOrderIntermodularDegree_AVERAGE[allAcids]\n",
      "> missing = which(allAcids%in%acidsMultimer == FALSE)\n",
      "> SecondOrderIntermodularDegree_AVERAGE[missing] = (data1Z[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"SecondOrderIntermodularDegreeSTRIDE_sidechain\"]+data1Z[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"SecondOrderIntermodularDegreeSTRIDE\"]+data2Z[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"SecondOrderIntermodularDegreeSTRIDECENTROIDSC\"]+data2Z[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"SecondOrderIntermodularDegreeWALKTRAPCENTROIDSC\"])/4\n",
      "> names(SecondOrderIntermodularDegree_AVERAGE)[missing] = allAcids[which(allAcids%in%acidsMultimer == FALSE)]\n",
      "> \n",
      "> #NodeEdgeBetweenness is the max of the monomer/multimer\n",
      "> NodeEdgeBetweennessSTRIDE_AVERAGE_MONOMER = (dataZ[,\"NodeEdgeBetweennessSTRIDE\"]+dataZ[,\"NodeEdgeBetweennessSTRIDE_sidechain\"])/2\n",
      "> NodeEdgeBetweennessSTRIDE_AVERAGE_MULTIMER = (data1MultimerZ[,\"NodeEdgeBetweennessSTRIDEMULTIMER\"]+data1MultimerZ[,\"NodeEdgeBetweennessSTRIDE_sidechainMULTIMER\"])/2\n",
      "> NodeEdgeBetweennessSTRIDE_sidechain_ALL = c(NodeEdgeBetweennessSTRIDE_AVERAGE_MONOMER,NodeEdgeBetweennessSTRIDE_AVERAGE_MULTIMER)\n",
      "> NodeEdgeBetweennessSTRIDE_sidechain_MAX = tapply(NodeEdgeBetweennessSTRIDE_sidechain_ALL,as.factor(c(acidsMonomer,acidsMultimer)),max,na.rm=TRUE)\n",
      "> NodeEdgeBetweennessSTRIDE_sidechain_MAX = NodeEdgeBetweennessSTRIDE_sidechain_MAX[allAcids]\n",
      "> \n",
      "> #Ligand is the min of the multimer only\n",
      "> LigandMULTIMERCENTROIDSC_ALL = data2MultimerZ[,\"LigandMULTIMERCENTROIDSC\"]\n",
      "> LigandMULTIMERCENTROIDSC_MIN = tapply(LigandMULTIMERCENTROIDSC_ALL,as.factor(acidsMultimer),min)\n",
      "> LigandMULTIMERCENTROIDSC_MIN = LigandMULTIMERCENTROIDSC_MIN[allAcids]\n",
      "> missing = which(allAcids%in%acidsMultimer == FALSE)\n",
      "> LigandMULTIMERCENTROIDSC_MIN[missing] = dataZ[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"LigandCENTROIDSC\"]\n",
      "> names(LigandMULTIMERCENTROIDSC_MIN)[missing] = allAcids[which(allAcids%in%acidsMultimer == FALSE)]\n",
      "> LigandMULTIMERCENTROIDSC_MIN[is.na(LigandMULTIMERCENTROIDSC_MIN)] = 0\n",
      "> \n",
      "> #RSA\n",
      "> RSA_ALL = c(data1[,\"RSA\"],data2Multimer[,\"RSA\"])\n",
      "> RSA_MIN = tapply(RSA_ALL,as.factor(c(acidsMonomer,acidsMultimer)),min,na.rm=TRUE)\n",
      "> RSA_MIN = RSA_MIN[allAcids]\n",
      "> \n",
      "> y = SecondOrderIntermodularDegree_AVERAGE+NodeEdgeBetweennessSTRIDE_sidechain_MAX-LigandMULTIMERCENTROIDSC_MIN\n",
      "> \n",
      "> out=cbind(names(y),y)\n",
      "> acidNumbers = sapply(out[,1],substring,4)\n",
      "> out = out[order(as.numeric(acidNumbers)),]\n",
      "> write.table(out,file=\"FinalSum\",sep=\"\\t\",quote=FALSE,row.names=FALSE,col.names=FALSE)\n",
      "> \n",
      "> # also write the 3 variables to a file\n",
      "> # Concatenating all three variables into a single data frame\n",
      "> result <- data.frame(\n",
      "+   Acid = names(SecondOrderIntermodularDegree_AVERAGE),\n",
      "+   SecondOrderIntermodularDegree_AVERAGE,\n",
      "+   NodeEdgeBetweennessSTRIDE_sidechain_MAX,\n",
      "+   LigandMULTIMERCENTROIDSC_MIN\n",
      "+ )\n",
      "> \n",
      "> # Writing the data frame to a file\n",
      "> write.table(result, file = \"FinalSum_Decomp\", sep = \"\\t\", quote = FALSE, row.names = FALSE, col.names = TRUE)\n",
      "> \n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "Read 1152 items\n",
      "Read 1152 items\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "Read 7713 items\n",
      "Read 7713 items\n",
      "\n",
      "SBNA results for 7DHY succesfully downloaded\n",
      "\n",
      "Processing 7DHY with chain C...\n",
      "Starting phenix.pdbtools\n",
      "on Sun Jun  2 13:44:11 2024 by yliy0004\n",
      "===============================================================================\n",
      "\n",
      "Processing files:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Found model, 7DHY.pdb\n",
      "\n",
      "Processing PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Adding command-line PHIL:\n",
      "  -------------------------\n",
      "    keep=Chain C\n",
      "\n",
      "Final processed PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "  data_manager {\n",
      "    model {\n",
      "      file = \"7DHY.pdb\"\n",
      "    }\n",
      "    default_model = \"7DHY.pdb\"\n",
      "  }\n",
      "  modify {\n",
      "    keep = Chain C\n",
      "  }\n",
      "\n",
      "\n",
      "Starting job\n",
      "===============================================================================\n",
      "Validating inputs\n",
      "Performing manipulations\n",
      "Size before: 6723 size after: 1654\n",
      "Writing output model\n",
      "\n",
      "===============================================================================\n",
      "Job complete\n",
      "usr+sys time: 1.95 seconds\n",
      "wall clock time: 1.99 seconds\n",
      "Cleaning PDB file\n",
      "Starting phenix.pdbtools\n",
      "on Sun Jun  2 13:44:13 2024 by yliy0004\n",
      "===============================================================================\n",
      "\n",
      "Processing files:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Found model, 7DHY_monomer.pdb\n",
      "\n",
      "Processing PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Adding command-line PHIL:\n",
      "  -------------------------\n",
      "    remove=water\n",
      "\n",
      "Final processed PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "  data_manager {\n",
      "    model {\n",
      "      file = \"7DHY_monomer.pdb\"\n",
      "    }\n",
      "    default_model = \"7DHY_monomer.pdb\"\n",
      "  }\n",
      "  modify {\n",
      "    remove = water\n",
      "  }\n",
      "\n",
      "\n",
      "Starting job\n",
      "===============================================================================\n",
      "Validating inputs\n",
      "Performing manipulations\n",
      "Size before: 1654 size after: 1539\n",
      "Writing output model\n",
      "\n",
      "===============================================================================\n",
      "Job complete\n",
      "usr+sys time: 1.66 seconds\n",
      "wall clock time: 1.67 seconds\n",
      "Writing out polypeptide file...\n",
      "Adding hydrogens...\n",
      "Protonating waters...\n",
      "Making .phi file\n",
      "Running full bond calculation script\n",
      "The ligands identified in this pdb are GOL-303C, ARS-301C\n",
      "Screening for atoms that are close.\n",
      "Dictionaries loaded.\n",
      "Starting salt bridges.\n",
      "Starting hydrogen bonds.\n",
      "Starting pi-pi version 2.\n",
      "Running pi-cation version 2\n",
      "Starting disulfide\n",
      "Starting van der waals.\n",
      "Starting metals\n",
      "Starting DNA bonds...\n",
      "Starting Ligand centroids\n",
      "Starting residue centroids...\n",
      "Removing duplicates from BFactor file\n",
      "Making 7DHY_monomer_nowaters_net\n",
      "Removing any negative value edges\n",
      "Creating QC file\n",
      "Creating Centroid directory and moving files\n",
      "Running energetic R script\n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> ##Load libraries\n",
      "> library(igraph)\n",
      "> library(stats)\n",
      "> \n",
      "> ##Initialize keywords\n",
      "> # keyword is the base filename\n",
      "> keyword = args[1]\n",
      "> args=c(paste(args[1],\"_net\",sep=\"\"),\"weighted\",paste(keyword,\"_secondaryStructure\",sep=\"\"))\n",
      "> atomCorrection=FALSE\n",
      "> baseKeyword = unlist(strsplit(keyword,\"/\"))[1]\n",
      "> runVariableSelection = TRUE\n",
      "> \n",
      "> ##Load inversal data\n",
      "> #Unique atoms file\n",
      "> terminalAtomsFile = as.matrix(read.table(\"../../terminalAtoms\",sep=\"\\t\"))\n",
      "> terminalAtoms = {}\n",
      "> for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+     terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\", \"))\n",
      "+ }\n",
      "> \n",
      "> #Secondary structure file\n",
      "> secStructure = as.matrix(read.table(args[3]))[,1:2]\n",
      "> rownames(secStructure) = secStructure[,1]\n",
      "> \n",
      "> #RSA\n",
      "> rsa = as.matrix(read.table(paste(keyword,\".rsa\",sep=\"\")))\n",
      "> \n",
      "> ##Define functions\n",
      "> \n",
      "> #Add terminal atom details\n",
      "> addTermDetails = function(data) {\n",
      "+     out_details = c()\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+         if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+             code1 = 1\n",
      "+         } else {\n",
      "+             code1 = 0\n",
      "+         }\n",
      "+         if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+             code2 = 1\n",
      "+         } else {\n",
      "+             code2 = 0\n",
      "+         }\n",
      "+     out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+     }\n",
      "+     return(out_details)\n",
      "+ }\n",
      "> \n",
      "> collapse = function(data) {\n",
      "+     if (nrow(data)==0) {\n",
      "+         out = matrix(ncol = 8,nrow=0)\n",
      "+         colnames(out) = c(\"AA1\", \"AA2\", \"Type\", \"SumEdges\", \"percentMC\", \"influence1\", \"influence2\",\"degree\")\n",
      "+         results = c(); results$out = out\n",
      "+         return(results)\n",
      "+         break\n",
      "+     }\n",
      "+     #Edges should be listed bi-directionally\n",
      "+     x=unlist(lapply(data[,3],function(x){if(x==\"MCSC\"){return(\"SCMC\")}else if(x==\"SCMC\"){return(\"MCSC\")} else {return(x)}}))\n",
      "+     if (dim(data)[2]==7) {\n",
      "+         data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6]))\n",
      "+     } else {\n",
      "+         data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6],data[,9],data[,8]))\n",
      "+     }\n",
      "+     data = unique(data)\n",
      "+     out = c()\n",
      "+     out_noPP = c()\n",
      "+     completed = c()\n",
      "+     if (dim(data)[2]==7) {\n",
      "+         out_details = c()\n",
      "+         for (i in 1:nrow(data)) {\n",
      "+             acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+             acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+             if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+                 code1 = 1\n",
      "+             } else {\n",
      "+                 code1 = 0\n",
      "+             }\n",
      "+             if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+                 code2 = 1\n",
      "+             } else {\n",
      "+                 code2 = 0\n",
      "+             }\n",
      "+             out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+         }\n",
      "+         data = out_details\n",
      "+     } else {\n",
      "+         out_details = data\n",
      "+     }\n",
      "+     colnames(out_details) = c(\"AA1\",\"AA2\",\"Type1\",\"Weight\",\"Type2\",\"Atom1\",\"Atom2\",\"influence1\",\"influence2\")\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         if (i %in% completed == FALSE) {\n",
      "+             x=which((data[,1]==data[i,1] & data[,2] == data[i,2]))\n",
      "+             subdata = matrix(data[x,],ncol=9)\n",
      "+             edgeSum = 0\n",
      "+             MCMCcount = 0\n",
      "+             MCSCcount = 0\n",
      "+             SCMCcount = 0\n",
      "+             SCSCcount = 0\n",
      "+             influence1 = sum(as.numeric(subdata[,8]))\n",
      "+             influence2 = sum(as.numeric(subdata[,9]))\n",
      "+             for (j in 1:nrow(subdata)) {\n",
      "+                 edgeSum = edgeSum + as.numeric(subdata[j,4])\n",
      "+                 if (subdata[j,3]==\"MCMC\" | subdata[j,3]==\"MCSC\") {\n",
      "+                     MCMCcount = MCMCcount+1\n",
      "+                     MCSCcount = MCSCcount+1\n",
      "+                 } else if (subdata[j,3]==\"SCMC\") {\n",
      "+                     SCMCcount = SCMCcount+1\n",
      "+                 } else if (subdata[j,3]==\"SCSC\") {\n",
      "+                     SCSCcount = SCSCcount+1\n",
      "+                 }\n",
      "+             }\n",
      "+             #Header is AA1, AA2, mixed, sum of edges, percent MC, influence1, influence2, degreeAA1\n",
      "+             out = rbind(out,c(data[i,1],data[i,2],\"mixed\",edgeSum,(MCMCcount+MCSCcount)/(MCMCcount+MCSCcount+SCMCcount+SCSCcount),influence1,influence2,sum(data[,1]==data[i,1])))\n",
      "+             if (any(subdata[,5]==\"PP\")==FALSE) {out_noPP = rbind(out_noPP,c(data[i,1],data[i,2],\"mixed\",edgeSum,(MCMCcount+MCSCcount)/(MCMCcount+MCSCcount+SCMCcount+SCSCcount),influence1,influence2,sum(data[,1]==data[i,1])))}\n",
      "+             completed = c(completed,x)\n",
      "+         }\n",
      "+     }\n",
      "+     out = out[out[,4]>0,]\n",
      "+     out_noPP = out_noPP[out_noPP[,4]>0,]\n",
      "+     colnames(out) = c(\"AA1\", \"AA2\", \"Type\", \"SumEdges\", \"percentMC\", \"influence1\", \"influence2\",\"degree\")\n",
      "+     results = c()\n",
      "+     results$out = out\n",
      "+     results$out_noPP = out_noPP\n",
      "+     results$influence_all = out_details[as.numeric(out_details[,4])>0 & (out_details[,8]==\"1\" | out_details[,9]==\"1\"),]\n",
      "+     out_details = out_details[as.numeric(out_details[,4])>0,]\n",
      "+     out_subtract = out_details; out_subtract[out_subtract[,8]==0 & out_subtract[,9]==0,4] = (-1)*as.numeric(out_subtract[out_subtract[,8]==0 & out_subtract[,9]==0,4])\n",
      "+     out_subtract = out_subtract[out_subtract[,5]!=\"PP\",]\n",
      "+     results$out_subtract = out_subtract\n",
      "+     return(results)\n",
      "+ }\n",
      "> \n",
      "> removeRedundancy = function(dataTmp) {\n",
      "+     for (i in 1:nrow(dataTmp)) {\n",
      "+         if (order(dataTmp[i,1:2])[1]==2) {\n",
      "+             dataTmp[i,1:2] = c(dataTmp[i,2],dataTmp[i,1])\n",
      "+             dataTmp[i,3] = paste(unlist(strsplit(dataTmp[i,3],split=\"\"))[c(3,4,1,2)],collapse=\"\")\n",
      "+             dataTmp[i,c(6,7)] = c(dataTmp[i,7],dataTmp[i,6])\n",
      "+         }\n",
      "+     }\n",
      "+     dataTmp = unique(dataTmp)\n",
      "+     pairs = unique(dataTmp[,1:2])\n",
      "+     out = c()\n",
      "+     for (j in 1:nrow(pairs)) {\n",
      "+         pair = pairs[j,]\n",
      "+         subData = matrix(dataTmp[dataTmp[,1]==pair[1] & dataTmp[,2]==pair[2],],ncol=7)\n",
      "+         out = rbind(out, c(pair[1],pair[2],\"mixed\",0.5*sum(as.numeric(subData[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> collapse_directed = function(data) {\n",
      "+     pairs = unique(data[,1:2])\n",
      "+     out = c()\n",
      "+     for (i in 1:nrow(pairs)) {\n",
      "+         subdat = matrix(data[data[,1]==pairs[i,1]&data[,2]==pairs[i,2],],ncol=4)\n",
      "+         out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> collapse_agnostic = function(data) {\n",
      "+     pairs = unique(t(apply(data[,1:2],1,function(x){return(sort(x))})))\n",
      "+     out = c()\n",
      "+     for (i in 1:nrow(pairs)) {\n",
      "+         subdat = matrix(data[(data[,1]==pairs[i,1]&data[,2]==pairs[i,2]) | (data[,2]==pairs[i,1]&data[,1]==pairs[i,2]),],ncol=4)\n",
      "+         out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> distributCalc = function(set,nodes,net) {\n",
      "+     #x is assumed to be a set but can also just be 1 node\n",
      "+     distanceNet = distances(net,weights=net$weight)\n",
      "+     distanceTmp = c()\n",
      "+     Dr = c()\n",
      "+     for (node in nodes) {\n",
      "+         distanceTmp = c()\n",
      "+         for (item in set) {\n",
      "+             if (item!=node) {\n",
      "+                 if(item%in%rownames(distanceNet)==FALSE){distanceTmp=100;next}\n",
      "+                 if(node%in%rownames(distanceNet)==FALSE){distanceTmp=100;next}\n",
      "+                 distanceTmp = c(distanceTmp,distanceNet[item,node])\n",
      "+             }\n",
      "+         }\n",
      "+         if (length(distanceTmp)>0) {\n",
      "+             Dr = c(Dr, 1/min(distanceTmp))\n",
      "+         }\n",
      "+     }\n",
      "+     return(sum(Dr)/length(nodes))\n",
      "+ }\n",
      "> \n",
      "> ##Read in user data\n",
      "> data_all = as.matrix(read.table(args[1]))\n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> originalNodes = nodes\n",
      "> basedata = collapse(data_all)\n",
      "> basedata_noPP = removeRedundancy(basedata$out[,1:7])\n",
      "> basedata_allsidechain = collapse_agnostic(data_all[data_all[,3]!=\"MCMC\",c(1,2,3,4)])\n",
      "> influencedata_all = basedata$influence_all\n",
      "> influencedata_collapse = collapse(influencedata_all)\n",
      "> influencedata = removeRedundancy(influencedata_collapse$out[,1:7])\n",
      "> influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "> influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "> \n",
      "> influencedata_directed = collapse_directed(influencedata_all[influencedata_all[,8]==1,c(1,2,3,4)])\n",
      "> influencenet_directed = graph.edgelist(influencedata_directed[,c(2,1)],directed=TRUE)\n",
      "> influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "> \n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> \n",
      "> #Write out edgelist files\n",
      "> write.table(influencedata,file=paste(keyword,\"_allEdges\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=FALSE,row.names=FALSE)\n",
      "> write.table(influencedata_directed,file=paste(keyword,\"_uniqueEdges\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=FALSE,row.names=FALSE)\n",
      "> \n",
      "> ###Deal with missing nodes\n",
      "> if (any(nodes %in% secStructure[,1]==FALSE)) {\n",
      "+     for (node in nodes) {\n",
      "+         if (node %in% secStructure[,1]==FALSE) {\n",
      "+             secStructure = rbind(secStructure,c(node,\"xxx\",360))\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> for (node in nodes) {\n",
      "+     if (node %in% rownames(rsa) == FALSE) {\n",
      "+         rsa = rbind(rsa,c(0,0))\n",
      "+         rownames(rsa)[nrow(rsa)] = node\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> ##Begin centrality calculations\n",
      "> \n",
      "> #Modular calculations\n",
      "> \n",
      "> ##WALKTRAP\n",
      "> net = graph.edgelist(basedata_noPP[,1:2],directed=FALSE)\n",
      "> net$weight = 1/as.numeric(basedata_noPP[,4])\n",
      "> net_community=walktrap.community(net,weights=net$weight); net_community_vec = net_community$membership; names(net_community_vec)=net_community$names\n",
      "> net_community_vec_wt = net_community_vec\n",
      "> nodes = V(net)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> nodes.color = c()\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(influencedata)) {\n",
      "+     j = which((basedata_noPP[,1]==influencedata[i,1]&basedata_noPP[,2]==influencedata[i,2])|(basedata_noPP[,1]==influencedata[i,2]&basedata_noPP[,2]==influencedata[i,1]))\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) { edgecolors[j]=\"grey40\" }\n",
      "+ }\n",
      "> if(length(nodes)>200) {textsize=.25} else {textsize=.45}\n",
      "> plot(net,edge.color = edgecolors, vertex.color = nodes.color,vertex.size=4,vertex.label.cex=textsize,main=\"WALKTRAP\")\n",
      ">     \n",
      "> ##Weighted edge betweenness WALKTRAP\n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP - WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for WALKTRAP\n",
      "> firstOrderDegree = degree(influencenet)\n",
      "> firstOrderDegree_sidechain = degree(influencenet_directed,mode=c(\"in\"))\n",
      "> secondOrderDegree = c(); secondOrderDegree_sidechain=c()\n",
      "> nodes = V(influencenet)$name\n",
      "> for (node in nodes) {\n",
      "+     firstorder = neighbors(influencenet,node)\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet,neighbor)))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree = c(secondOrderDegree,length(secondorder))\n",
      "+     firstorder = neighbors(influencenet_directed,node,mode=c(\"in\"))\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet_directed,neighbor,mode=c(\"in\"))))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree_sidechain = c(secondOrderDegree_sidechain,length(secondorder))\n",
      "+ }\n",
      "> names(secondOrderDegree) = nodes; names(secondOrderDegree_sidechain) = nodes\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])])) - 1\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])])) - 1\n",
      "+ }\n",
      "> \n",
      "> #SAVE ALL AS WALKTRAP\n",
      "> node_edge_betweenness_wt = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_wt = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_wt = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_wt = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_wt = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_wt = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> ##2ARY STRUCTURE\n",
      "> net_community_vec = as.numeric(as.factor(secStructure[,2]))\n",
      "> names(net_community_vec) = secStructure[,1]\n",
      "> nodes = V(net)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> nodes.color = c()\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(basedata_noPP)) {\n",
      "+     if (net_community_vec[basedata_noPP[i,1]]!=net_community_vec[basedata_noPP[i,2]]) { edgecolors[i]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for 2ARY STRUCTURE\n",
      "> nodes = V(influencenet)$name\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])])) - 1\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])])) - 1\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #SAVE ALL AS STRIDE\n",
      "> node_edge_betweenness_stride = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_stride = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_stride = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_stride = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_stride = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_stride = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> \n",
      "> ### Create final dataset\n",
      "> \n",
      "> out = cbind(\n",
      "+ firstOrderDegree[nodes],\n",
      "+ firstOrderDegree_sidechain[nodes],\n",
      "+ secondOrderDegree[nodes],\n",
      "+ secondOrderDegree_sidechain[nodes],\n",
      "+ node_edge_betweenness_stride[nodes],\n",
      "+ node_edge_betweenness_sidechain_stride[nodes],\n",
      "+ node_intermodular_degree_stride[nodes],\n",
      "+ node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ node_edge_betweenness_wt[nodes],\n",
      "+ node_edge_betweenness_sidechain_wt[nodes],\n",
      "+ node_intermodular_degree_wt[nodes],\n",
      "+ node_intermodular_degree_sidechain_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_wt[nodes]\n",
      "+ )\n",
      "> rownames(out) = nodes\n",
      "> colnames(out) = c(\"Degree\",\"Degree_sidechain\",\"SecondOrderDegree\",\"SecondOrderDegree_sidechain\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_sidechain\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_sidechain\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_sidechain\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_sidechain\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_sidechain\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_sidechain\")\n",
      "> \n",
      "> #add back in nodes that were not networked as zeros\n",
      "> zeroMat = matrix(0,nrow=length(which(originalNodes%in%nodes==FALSE)),ncol=ncol(out));rownames(zeroMat)=originalNodes[which(originalNodes%in%nodes==FALSE)]\n",
      "> out = rbind(out,zeroMat)\n",
      "> \n",
      "> #Add 1 to everything to avoid zeros, except ligand\n",
      "> out[is.na(out)]=0\n",
      "> out = out+1\n",
      "> \n",
      "> #Write out\n",
      "> write.table(out,file=paste(keyword,\"_scoresEnergetics\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "> #Standard normalization\n",
      "> outZ = apply(out,2,function(x){return(scale(x))})\n",
      "> rownames(outZ) = rownames(out)\n",
      "> write.table(outZ,file=paste(keyword,\"_scoresEnergeticsZ\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "Running centroid R script\n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(igraph)\n",
      "> library(stats)\n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> ##read in data\n",
      "> ## arg1: col1 and col2 are edges, col3 is weight of edge\n",
      "> ## arg2: weighted, unweighted <-- no longer using\n",
      "> ## arg3: forced module definition <-- no longer using\n",
      "> \n",
      "> # Set desired parameters\n",
      "> removeMCMC = FALSE\n",
      "> linearcutoff = 1\n",
      "> directed = FALSE\n",
      "> secondaryStructure2 = FALSE\n",
      "> sidechainMode = TRUE ##build network of just SCSC, SCMC and MCSC. Still use unique atoms for the _uniqueAtoms directed plots\n",
      "> uniqueAtomsMode = FALSE ##build network of just unique atoms (each edge has to have at least 1). Still use unique atoms for the _uniqueAtoms directed plots\n",
      "> useDNA=TRUE\n",
      "> uniqueAtomsUnbiased=FALSE\n",
      "> uniqueAtomsGaurav=FALSE\n",
      "> uniqueAtomsOLD=FALSE\n",
      "> atomCorrection = FALSE\n",
      "> uniqueAtomsGauravPP = FALSE\n",
      "> weighted = TRUE\n",
      "> removeWaters=TRUE\n",
      "> if (tail(unlist(strsplit(getwd(),\"/\")),n=1)==\"Centroid\") {\n",
      "+     ligandCentroidMode = TRUE\n",
      "+ } else {\n",
      "+     ligandCentroidMode = FALSE\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> # Read in ligand file\n",
      "> ligandfile = scan(paste(args[1],\"_ligand\",sep=\"\"),what=\"character\")\n",
      "> if (length(ligandfile)==0){\n",
      "+     ligandmode = FALSE\n",
      "+ } else {\n",
      "+     ligandmode = TRUE\n",
      "+ }\n",
      "> \n",
      "> ligandCentroidFile = scan(paste(args[1],\"NetLigand\",sep=\"\"),what=\"character\")\n",
      "> \n",
      "> # keyword is the base filename\n",
      "> # method is \"frag\" or \"module\" or \"both\"\n",
      "> keyword = args[1]\n",
      "> args=c(paste(args[1],\"_net\",sep=\"\"),\"weighted\",dir(\"./\",\"_secondaryStructure$\"))\n",
      "> \n",
      "> ##Load data\n",
      "> #Unique atoms file\n",
      "> if(uniqueAtomsUnbiased) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"/Users/vjpatel/Dropbox/Gaurav/uniqueAtomsUnbiased\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else if (uniqueAtomsGaurav) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"/Users/vjpatel/Dropbox/Gaurav/uniqueAtomsGaurav\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else if (uniqueAtomsOLD) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"../../../terminalAtoms\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"../../../uniqueAtoms\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     otherTA = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+         #otherTA[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,4],split=\",\"))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> #Secondary structure file\n",
      "> #Ligands are all one secondary structure\n",
      "> secStructure = as.matrix(read.table(args[3]))\n",
      "> phiAngleDB = {}\n",
      "> for (i in 1:nrow(secStructure)) {\n",
      "+     phiAngleDB[[secStructure[i,1]]] = as.numeric(secStructure[i,3])\n",
      "+ }\n",
      "> \n",
      "> if(ligandmode) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",\"_ligand\"),sep=\"\\t\"))\n",
      "+     liganddat = unique(liganddat[,2])\n",
      "+     for (ligand in liganddat) {\n",
      "+         secStructure = rbind(secStructure,c(sub(\"-\",\"\",ligand),\"ligand\",360))\n",
      "+         phiAngleDB[[sub(\"-\",\"\",ligand)]]=0\n",
      "+     }\n",
      "+ } else {\n",
      "+     liganddat = c()\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> ##Define functions\n",
      "> \n",
      "> ##Collapse matrix functions\n",
      "> ##Will return redundant matrix\n",
      "> ##Input needs to be 7 columns\n",
      "> ##Second input is different if the main input is a subset of the larger data\n",
      "> addTerminalDetails = function(data,largerdata) {\n",
      "+     #List edges bi-directionally\n",
      "+     x=unlist(lapply(data[,3],function(x){if(x==\"MCSC\"){return(\"SCMC\")}else if(x==\"SCMC\"){return(\"MCSC\")} else {return(x)}}))\n",
      "+     data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6]))\n",
      "+     data = unique(data)\n",
      "+     out = c()\n",
      "+     out_noPP = c()\n",
      "+     out_details = c()\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+         subdata = matrix(data[data[,1]==data[i,1]&data[,2]==data[i,2],],ncol=7)\n",
      "+         allsubdata = matrix(largerdata[largerdata[,1]==data[i,1],],ncol=7)\n",
      "+         code1=0;code2=0\n",
      "+         if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+             code1 = 1\n",
      "+         }\n",
      "+         if (data[i,5] == \"DNA\" & useDNA==TRUE) {\n",
      "+             code1 = 1\n",
      "+         }\n",
      "+         allsubdata = matrix(largerdata[largerdata[,2]==data[i,2],],ncol=7)\n",
      "+         if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+             code2 = 1\n",
      "+         }\n",
      "+         if (data[i,5] == \"DNA\" & useDNA==TRUE) {\n",
      "+             code2 = 1\n",
      "+         }\n",
      "+         if (data[i,5]==\"PICATION\"|data[i,5]==\"PIPI\") {\n",
      "+             code1=1\n",
      "+             code2=1\n",
      "+         }\n",
      "+         if (data[i,5]==\"PP\") {\n",
      "+             if (uniqueAtomsGaurav) {\n",
      "+                 if((acid1) == \"GLY\") {\n",
      "+                     code1 = 1\n",
      "+                     data[i,3]==\"SCSC\"\n",
      "+                 }\n",
      "+                 if ((acid2) == \"GLY\") {\n",
      "+                     code2 = 1\n",
      "+                     data[i,3] == \"SCSC\"\n",
      "+                 }\n",
      "+             }\n",
      "+         }\n",
      "+         out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+     }\n",
      "+     return(out_details)\n",
      "+ }\n",
      "> \n",
      "> ##Will return non-redundant matrix\n",
      "> collapse_agnostic = function(data) {\n",
      "+     if(nrow(data)==0) {\n",
      "+         return(matrix(ncol=4,nrow=0))\n",
      "+     } else if (ncol(data)==4 & data[1,3]==\"mixed\") {\n",
      "+         return(data)\n",
      "+     } else {\n",
      "+         pairs = unique(t(apply(matrix(data[,1:2],ncol=2),1,function(x){return(sort(x))})))\n",
      "+         out = c()\n",
      "+         for (i in 1:nrow(pairs)) {\n",
      "+             subdat = matrix(data[(data[,1]==pairs[i,1]&data[,2]==pairs[i,2]) | (data[,2]==pairs[i,1]&data[,1]==pairs[i,2]),1:4],ncol=4)\n",
      "+             subdat = unique(t(apply(subdat,1,function(x){a=which(x==sort(x[1:2])[1]);if(a==1){return(x)}else{return(c(x[2],x[1],paste(unlist(strsplit(x[3],\"\"))[c(3,4,1,2)],collapse=\"\"),x[4]))}})))\n",
      "+             if (weighted==FALSE) {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",1))\n",
      "+             } else {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+             }\n",
      "+         }\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> ##Will return directed collapsed matrix\n",
      "> collapse_directed = function(data) {\n",
      "+     if (length(data)==0) {\n",
      "+         return(matrix(nrow=0,ncol=4))\n",
      "+     } else {\n",
      "+         pairs = matrix(unique(data[,1:2]),ncol=2)\n",
      "+         out = c()\n",
      "+         for (i in 1:nrow(pairs)) {\n",
      "+             subdat = matrix(data[data[,1]==pairs[i,1]&data[,2]==pairs[i,2],1:4],ncol=4)\n",
      "+             if (weighted==FALSE) {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",1))\n",
      "+             } else {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+             }\n",
      "+         }\n",
      "+         return(matrix(out,ncol=4))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> rsa = as.matrix(read.table(dir(\"./\",\".rsa$\")))\n",
      "> rownames(rsa) = rsa[,1]\n",
      "> rsa[,1] = rsa[,2]\n",
      "> \n",
      "> #add ligand as -1\n",
      "> for (ligand in liganddat) {\n",
      "+     rsa = rbind(rsa,c(-1,-1))\n",
      "+     rownames(rsa)[nrow(rsa)] = sub(\"-\",\"\",ligand)\n",
      "+ }\n",
      "> \n",
      "> #Function to get amino acid\n",
      "> getAcid = function(residue) {\n",
      "+     if(all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"A\")) == TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"G\"))==TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"C\"))==TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"T\"))==TRUE)) {return(paste(unlist(strsplit(residue,\"\"))[1:2],collapse=\"\"))\n",
      "+     } else {\n",
      "+     return(paste(unlist(strsplit(residue,\"\"))[1:3],collapse=\"\"))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> ##################\n",
      "> ## read in data ##\n",
      "> ##################\n",
      "> data_all = as.matrix(read.table(args[1]))[,1:7]\n",
      "> if(removeWaters) {\n",
      "+     y=c(grep(\"HOH\",data_all[,1]),grep(\"HOH\",data_all[,2]))\n",
      "+     if(length(y)>0) {\n",
      "+         data_all = data_all[-y,]\n",
      "+     }\n",
      "+ }\n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> originalNodes = nodes\n",
      "> \n",
      "> data_all_original = rbind(data_all,data_all[,c(2,1,3,4,5,7,6)])\n",
      "> data_all = data_all[as.numeric(data_all[,4])>0,]\n",
      "> \n",
      "> basedata = addTerminalDetails(data_all,data_all_original)\n",
      "> basedata_noPP = collapse_agnostic(basedata[basedata[,5]!=\"PP\",])\n",
      "> sidechaindata = basedata[basedata[,3]!=\"MCMC\",1:4]\n",
      "> sidechaindata_detailed = basedata[basedata[,3]!=\"MCMC\",]\n",
      "> \n",
      "> influencedata_detailed = basedata\n",
      "> if (uniqueAtomsMode) {\n",
      "+     influencedata_detailed = basedata[basedata[,8]==1 | basedata[,9]==1,]\n",
      "+ }\n",
      "> influencedata = collapse_agnostic(influencedata_detailed)\n",
      "> influencedataGLY = rbind(influencedata[grep(\"GLY\",influencedata[,1]),],influencedata[grep(\"GLY\",influencedata[,1]),])\n",
      "> influencedata = influencedata[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,2])))>linearcutoff,]\n",
      "> \n",
      "> \n",
      "> influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "> influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "> \n",
      "> influencedata_directed = collapse_directed(matrix(influencedata_detailed[influencedata_detailed[,8]==1,1:4],ncol=4))\n",
      "> influencedata_directedGLY = rbind(influencedata_directed[grep(\"GLY\",influencedata_directed[,1]),],influencedata_directed[grep(\"GLY\",influencedata_directed[,2]),])\n",
      "> influencedata_directed = influencedata_directed[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,2])))>linearcutoff,]\n",
      "> \n",
      "> influencedata_directed = matrix(influencedata_directed,ncol=4)\n",
      "> influencenet_directed = graph.edgelist(matrix(influencedata_directed[,c(2,1)],ncol=2),directed=TRUE)\n",
      "> influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "> \n",
      "> \n",
      "> if(sidechainMode) {\n",
      "+     influencedata_detailed = sidechaindata_detailed\n",
      "+     influencedata = collapse_agnostic(influencedata_detailed)\n",
      "+     influencedataGLY = rbind(influencedata[grep(\"GLY\",influencedata[,1]),],influencedata[grep(\"GLY\",influencedata[,2]),])\n",
      "+     influencedata = influencedata[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,2])))>linearcutoff,]\n",
      "+     #Add back for GLY if uniqueAtomsGaurav\n",
      "+     if(uniqueAtomsGauravPP) {\n",
      "+         influencedata = rbind(influencedata,influencedataGLY)\n",
      "+     }\n",
      "+     influencedata = influencedata[as.numeric(influencedata[,4])>0,]\n",
      "+     \n",
      "+     influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "+     influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "+     \n",
      "+     influencedata_directed = collapse_directed(matrix(influencedata_detailed[influencedata_detailed[,8]==1,1:4],ncol=4))\n",
      "+     influencedata_directedGLY = rbind(influencedata_directed[grep(\"GLY\",influencedata_directed[,1]),],influencedata_directed[grep(\"GLY\",influencedata_directed[,2]),])\n",
      "+ \n",
      "+     influencedata_directed = influencedata_directed[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,2])))>linearcutoff,]\n",
      "+     #Add back for GLY if uniqueAtomsGaurav\n",
      "+     if(uniqueAtomsGauravPP) {\n",
      "+         influencedata_directed = rbind(influencedata_directed,influencedata_directedGLY)\n",
      "+     }\n",
      "+     influencedata_directed = matrix(influencedata_directed,ncol=4)\n",
      "+     influencenet_directed = graph.edgelist(matrix(influencedata_directed[,c(2,1)],ncol=2),directed=TRUE)\n",
      "+     influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "+     \n",
      "+ }\n",
      "> \n",
      "> #Replace missing nodes in the igraph net\n",
      "> nodesMissing = nodes[nodes %in% influencedata[,1]==FALSE & nodes %in% influencedata[,2]==FALSE]\n",
      "> \n",
      "> ##B-factor\n",
      "> bfactor = as.matrix(read.table(dir(\"./\",\"_Bfactor\"),row.names=1))\n",
      "> \n",
      "> #Replace missing nodes in secStructure matrix\n",
      "> if (any(nodes %in% secStructure[,1]==FALSE)) {\n",
      "+     for (node in nodes) {\n",
      "+         if (node %in% secStructure[,1]==FALSE) {\n",
      "+             secStructure = rbind(secStructure,c(node,\"xxx\",360))\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> for (node in nodes) {\n",
      "+     if (node %in% rownames(rsa) == FALSE) {\n",
      "+         rsa = rbind(rsa,c(0,0))\n",
      "+         rownames(rsa)[nrow(rsa)] = node\n",
      "+     }\n",
      "+     if (node %in% rownames(bfactor) == FALSE) {\n",
      "+         bfactor = rbind(bfactor,1)\n",
      "+         rownames(bfactor)[nrow(bfactor)] = node\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> \n",
      "> \n",
      "> #Give DNA attributes\n",
      "> if (\"DNA\" %in% data_all[,5]) {\n",
      "+     for (i in 1:nrow(data_all)) {\n",
      "+         if (data_all[i,5]==\"DNA\") {\n",
      "+             phiAngleDB[[data_all[i,1]]] = 90\n",
      "+             if (data_all[i,1] %in% rownames(rsa) == FALSE) {\n",
      "+                 rsa = rbind(rsa,c(0,0))\n",
      "+                 rownames(rsa)[nrow(rsa)] = data_all[i,1]\n",
      "+             }\n",
      "+             if (data_all[i,1] %in% rownames(secStructure) == FALSE) {\n",
      "+                 secStructure = rbind(secStructure,c(data_all[i,1],\"DNA\",360))\n",
      "+             }\n",
      "+             if (data_all[i,1] %in% rownames(bfactor) == FALSE) {\n",
      "+                 bfactor = rbind(bfactor,1)\n",
      "+                 rownames(bfactor)[nrow(bfactor)] = data_all[i,1]\n",
      "+             }\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #Betweenness\n",
      "> node_betweenness = betweenness(influencenet,weights=influencenet$weight,directed=directed)\n",
      "> node_betweenness_unique = betweenness(influencenet_directed,weights=influencenet_directed$weight,directed=directed)\n",
      "> \n",
      "> #Modular calculations\n",
      "> \n",
      "> ##WALKTRAP\n",
      "> net = graph.edgelist(basedata_noPP[,1:2],directed=FALSE)\n",
      "> net$weight = 1/as.numeric(basedata_noPP[,4])\n",
      "> net_community=walktrap.community(net,weights=net$weight); net_community_vec = net_community$membership; names(net_community_vec)=net_community$names\n",
      "> net_community_vec_wt = net_community_vec\n",
      "> #net_community_vec[buriednodes] = max(net_community_vec)+1\n",
      "> nodes = V(influencenet)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> nodes.color = c()\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(influencedata)) {\n",
      "+     j = which((basedata_noPP[,1]==influencedata[i,1]&basedata_noPP[,2]==influencedata[i,2])|(basedata_noPP[,1]==influencedata[i,2]&basedata_noPP[,2]==influencedata[i,1]))\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) { edgecolors[j]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP\n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP - WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         if(weight1==0 & weight2==0) {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]] + 0\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]] + 0\n",
      "+         } else {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for WALKTRAP\n",
      "> firstOrderDegree = degree(influencenet)\n",
      "> firstOrderDegree_sidechain = degree(influencenet_directed,mode=c(\"in\"))\n",
      "> \n",
      "> secondOrderDegree = c(); secondOrderDegree_sidechain=c()\n",
      "> nodes = V(influencenet)$name\n",
      "> for (node in nodes) {\n",
      "+     firstorder = neighbors(influencenet,node)\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet,neighbor)))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree = c(secondOrderDegree,length(secondorder))\n",
      "+     if(node %in% V(influencenet_directed)$name) {\n",
      "+         firstorder_sidechain = neighbors(influencenet_directed,node,mode=c(\"in\"))\n",
      "+     } else {\n",
      "+         firstorder_sidechain = c()\n",
      "+     }\n",
      "+     secondorder = c(); for (neighbor in firstorder_sidechain){secondorder = c(secondorder,names(neighbors(influencenet_directed,neighbor,mode=c(\"in\"))))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree_sidechain = c(secondOrderDegree_sidechain,length(secondorder))\n",
      "+ }\n",
      "> names(secondOrderDegree) = nodes; names(secondOrderDegree_sidechain) = nodes\n",
      "> \n",
      "> \n",
      "> #second order intermodular degree\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])]))\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])]))\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #SAVE ALL AS WALKTRAP\n",
      "> node_edge_betweenness_wt = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_wt = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_wt = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_wt = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_wt = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_wt = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> \n",
      "> ##2ARY STRUCTURE\n",
      "> secStructure = secStructure[,1:2]\n",
      "> rownames(secStructure) = secStructure[,1]\n",
      "> net_community_vec = as.numeric(as.factor(secStructure[,2]))\n",
      "> names(net_community_vec) = secStructure[,1]\n",
      "> #net_community_vec[buriednodes] = max(net_community_vec)+1\n",
      "> nodes = V(influencenet)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> nodes.color = c()\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(basedata_noPP)) {\n",
      "+     if (net_community_vec[basedata_noPP[i,1]]!=net_community_vec[basedata_noPP[i,2]]) { edgecolors[i]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         if(weight1==0 & weight2==0) {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]] + 0\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]] + 0\n",
      "+         } else {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> ##Degree and second order degree for 2ARY STRUCTURE\n",
      "> \n",
      "> #second order intermodular degree\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     #secondOrder_node_intermodular_degree[node] = length(unique(unlist(node_modules[names(node_modules[[node]])])))\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])]))\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])]))\n",
      "+ }\n",
      "> \n",
      "> #SAVE ALL AS STRIDE\n",
      "> node_edge_betweenness_stride = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_stride = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_stride = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_stride = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_stride = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_stride = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> ### Create final dataset\n",
      "> nodes = nodes[nodes!=\"DNA1000A\"]\n",
      "> nodes = nodes[nodes %in% c(\"DA\",\"DG\",\"DC\",\"DT\") == FALSE]\n",
      "> \n",
      "> ### Remove ligand amino acids from final dataset\n",
      "> ligands = sub(\"-\",\"\",liganddat)\n",
      "> nodes = nodes[nodes %in% ligands == FALSE]\n",
      "> \n",
      "> nodes[which(nodes %in% rownames(bfactor)==FALSE)]\n",
      "character(0)\n",
      "> \n",
      "> out = cbind(\n",
      "+ rsa[nodes,1],\n",
      "+ firstOrderDegree[nodes],\n",
      "+ firstOrderDegree_sidechain[nodes],\n",
      "+ secondOrderDegree[nodes],\n",
      "+ secondOrderDegree_sidechain[nodes],\n",
      "+ node_edge_betweenness_stride[nodes],\n",
      "+ node_edge_betweenness_sidechain_stride[nodes],\n",
      "+ node_intermodular_degree_stride[nodes],\n",
      "+ node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ node_edge_betweenness_wt[nodes],\n",
      "+ node_edge_betweenness_sidechain_wt[nodes],\n",
      "+ node_intermodular_degree_wt[nodes],\n",
      "+ node_intermodular_degree_sidechain_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_wt[nodes]\n",
      "+ )\n",
      "> \n",
      "> #write column names\n",
      "> colnames(out) = c(\"RSA\",\"Degree\",\"Degree_uniqueAtoms\",\"SecondOrderDegree\",\"SecondOrderDegree_uniqueAtoms\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_unqiueAtoms\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_uniqueAtoms\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_uniqueAtoms\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_uniqueAtoms\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_uniqueAtoms\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_uniqueAtoms\")\n",
      "> \n",
      "> \n",
      "> ## LIGANDS\n",
      "> if(ligandmode==TRUE & ligandCentroidMode==FALSE) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",\"_ligand\"),sep=\"\\t\"))\n",
      "+     \n",
      "+     tmp = c()\n",
      "+     i=1\n",
      "+     for (i in 1:nrow(liganddat)) {\n",
      "+         acid=paste(unlist(strsplit(liganddat[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         if (unlist(strsplit(liganddat[i,1],\"-\"))[2] %in% terminalAtoms[[acid]]) {\n",
      "+             tmp = rbind(tmp,c(unlist(strsplit(liganddat[i,1],\"-\"))[1],liganddat[i,2],liganddat[i,3]))\n",
      "+         }\n",
      "+     }\n",
      "+     \n",
      "+     #ligandvec = tapply(liganddat[,3],as.factor(liganddat[,1]),min)\n",
      "+     ligandvec = tapply(as.numeric(tmp[,3]),as.factor(tmp[,1]),min)\n",
      "+ }\n",
      "> \n",
      "> if(ligandmode==TRUE & ligandCentroidMode) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",paste(\"_centroidNetLigand$\",sep=\"\")),sep=\"\\t\"))\n",
      "+     \n",
      "+     tmp = c()\n",
      "+     i=1\n",
      "+     for (i in 1:nrow(liganddat)) {\n",
      "+         tmp = rbind(tmp,c(liganddat[i,1],liganddat[i,2],liganddat[i,3]))\n",
      "+     }\n",
      "+     \n",
      "+     ligandvec = tapply(tmp[,3],as.factor(tmp[,1]),min)\n",
      "+ }\n",
      "> \n",
      "> if (ligandmode==FALSE) {\n",
      "+     ligandvec = rep(0,length(nodes))\n",
      "+ }\n",
      "> \n",
      "> out = cbind(out,ligandvec[rownames(out)])\n",
      "> colnames(out) = c(\"RSA\",\"Degree\",\"Degree_uniqueAtoms\",\"SecondOrderDegree\",\"SecondOrderDegree_uniqueAtoms\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_unqiueAtoms\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_uniqueAtoms\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_uniqueAtoms\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_uniqueAtoms\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_uniqueAtoms\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_uniqueAtoms\",\"Ligand\")\n",
      "> \n",
      "> #add back in nodes that were not networked as zeros\n",
      "> zeroMat = matrix(0,nrow=length(which(originalNodes%in%nodes==FALSE)),ncol=ncol(out));rownames(zeroMat)=originalNodes[which(originalNodes%in%nodes==FALSE)]\n",
      "> out = rbind(out,zeroMat)\n",
      "> \n",
      "> #fix RSA and Bfactor and ligand\n",
      "> out[,\"RSA\"] = rsa[rownames(out),1]\n",
      "> out[,\"Ligand\"] = ligandvec[rownames(out)]\n",
      "> \n",
      "> #Add 1 to everything to avoid zeros, except ligand\n",
      "> out[is.na(out)]=0\n",
      "> out[out[,\"Ligand\"]==0,\"Ligand\"] = 150\n",
      "> out[1:nrow(out),2:ncol(out)] = as.numeric(out[1:nrow(out),2:ncol(out)])+1\n",
      "> \n",
      "> ##create Z score file\n",
      "> outZ = apply(out,2,function(x){return(scale(as.numeric(x)))})\n",
      "> rownames(outZ) = rownames(out)\n",
      "> colnames(outZ) = colnames(out)\n",
      "> outZ[outZ==NA]=0\n",
      "> outZ[outZ==\"NaN\"]=0\n",
      "> \n",
      "> #Write out scores and z-normed scores files\n",
      "> write.table(out,file=paste(keyword,\"_scoresCentroid\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> write.table(outZ,file=paste(keyword,\"_scoresCentroidZ\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "> \n",
      "Cleaning PDB file\n",
      "Starting phenix.pdbtools\n",
      "on Sun Jun  2 13:44:58 2024 by yliy0004\n",
      "===============================================================================\n",
      "\n",
      "Processing files:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Found model, 7DHY_multimer.pdb\n",
      "\n",
      "Processing PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Adding command-line PHIL:\n",
      "  -------------------------\n",
      "    remove=water\n",
      "\n",
      "Final processed PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "  data_manager {\n",
      "    model {\n",
      "      file = \"7DHY_multimer.pdb\"\n",
      "    }\n",
      "    default_model = \"7DHY_multimer.pdb\"\n",
      "  }\n",
      "  modify {\n",
      "    remove = water\n",
      "  }\n",
      "\n",
      "\n",
      "Starting job\n",
      "===============================================================================\n",
      "Validating inputs\n",
      "Performing manipulations\n",
      "Size before: 6723 size after: 6206\n",
      "Writing output model\n",
      "\n",
      "===============================================================================\n",
      "Job complete\n",
      "usr+sys time: 2.06 seconds\n",
      "wall clock time: 2.12 seconds\n",
      "Writing out polypeptide file...\n",
      "Adding hydrogens...\n",
      "Protonating waters...\n",
      "Making .phi file\n",
      "Running full bond calculation script\n",
      "The ligands identified in this pdb are GOL-306D, GOL-303C, GOL-303B, SO4-303D, ARS-301D, ARS-301C, ARS-301B, ARS-301A, GOL-304B, GOL-305D, GOL-304D, GOL-305B\n",
      "Screening for atoms that are close.\n",
      "Dictionaries loaded.\n",
      "Starting salt bridges.\n",
      "Starting hydrogen bonds.\n",
      "Starting pi-pi version 2.\n",
      "Running pi-cation version 2\n",
      "Starting disulfide\n",
      "Starting van der waals.\n",
      "Starting metals\n",
      "Starting DNA bonds...\n",
      "Starting Ligand centroids\n",
      "Starting residue centroids...\n",
      "Removing duplicates from BFactor file\n",
      "Making 7DHY_multimer_nowaters_net\n",
      "Removing any negative value edges\n",
      "Creating QC file\n",
      "Creating Centroid directory and moving files\n",
      "Running energetic R script\n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> ##Load libraries\n",
      "> library(igraph)\n",
      "> library(stats)\n",
      "> \n",
      "> ##Initialize keywords\n",
      "> # keyword is the base filename\n",
      "> keyword = args[1]\n",
      "> args=c(paste(args[1],\"_net\",sep=\"\"),\"weighted\",paste(keyword,\"_secondaryStructure\",sep=\"\"))\n",
      "> atomCorrection=FALSE\n",
      "> baseKeyword = unlist(strsplit(keyword,\"/\"))[1]\n",
      "> runVariableSelection = TRUE\n",
      "> \n",
      "> ##Load inversal data\n",
      "> #Unique atoms file\n",
      "> terminalAtomsFile = as.matrix(read.table(\"../../terminalAtoms\",sep=\"\\t\"))\n",
      "> terminalAtoms = {}\n",
      "> for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+     terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\", \"))\n",
      "+ }\n",
      "> \n",
      "> #Secondary structure file\n",
      "> secStructure = as.matrix(read.table(args[3]))[,1:2]\n",
      "> rownames(secStructure) = secStructure[,1]\n",
      "> \n",
      "> #RSA\n",
      "> rsa = as.matrix(read.table(paste(keyword,\".rsa\",sep=\"\")))\n",
      "> \n",
      "> ##Define functions\n",
      "> \n",
      "> #Add terminal atom details\n",
      "> addTermDetails = function(data) {\n",
      "+     out_details = c()\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+         if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+             code1 = 1\n",
      "+         } else {\n",
      "+             code1 = 0\n",
      "+         }\n",
      "+         if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+             code2 = 1\n",
      "+         } else {\n",
      "+             code2 = 0\n",
      "+         }\n",
      "+     out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+     }\n",
      "+     return(out_details)\n",
      "+ }\n",
      "> \n",
      "> collapse = function(data) {\n",
      "+     if (nrow(data)==0) {\n",
      "+         out = matrix(ncol = 8,nrow=0)\n",
      "+         colnames(out) = c(\"AA1\", \"AA2\", \"Type\", \"SumEdges\", \"percentMC\", \"influence1\", \"influence2\",\"degree\")\n",
      "+         results = c(); results$out = out\n",
      "+         return(results)\n",
      "+         break\n",
      "+     }\n",
      "+     #Edges should be listed bi-directionally\n",
      "+     x=unlist(lapply(data[,3],function(x){if(x==\"MCSC\"){return(\"SCMC\")}else if(x==\"SCMC\"){return(\"MCSC\")} else {return(x)}}))\n",
      "+     if (dim(data)[2]==7) {\n",
      "+         data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6]))\n",
      "+     } else {\n",
      "+         data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6],data[,9],data[,8]))\n",
      "+     }\n",
      "+     data = unique(data)\n",
      "+     out = c()\n",
      "+     out_noPP = c()\n",
      "+     completed = c()\n",
      "+     if (dim(data)[2]==7) {\n",
      "+         out_details = c()\n",
      "+         for (i in 1:nrow(data)) {\n",
      "+             acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+             acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+             if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+                 code1 = 1\n",
      "+             } else {\n",
      "+                 code1 = 0\n",
      "+             }\n",
      "+             if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+                 code2 = 1\n",
      "+             } else {\n",
      "+                 code2 = 0\n",
      "+             }\n",
      "+             out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+         }\n",
      "+         data = out_details\n",
      "+     } else {\n",
      "+         out_details = data\n",
      "+     }\n",
      "+     colnames(out_details) = c(\"AA1\",\"AA2\",\"Type1\",\"Weight\",\"Type2\",\"Atom1\",\"Atom2\",\"influence1\",\"influence2\")\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         if (i %in% completed == FALSE) {\n",
      "+             x=which((data[,1]==data[i,1] & data[,2] == data[i,2]))\n",
      "+             subdata = matrix(data[x,],ncol=9)\n",
      "+             edgeSum = 0\n",
      "+             MCMCcount = 0\n",
      "+             MCSCcount = 0\n",
      "+             SCMCcount = 0\n",
      "+             SCSCcount = 0\n",
      "+             influence1 = sum(as.numeric(subdata[,8]))\n",
      "+             influence2 = sum(as.numeric(subdata[,9]))\n",
      "+             for (j in 1:nrow(subdata)) {\n",
      "+                 edgeSum = edgeSum + as.numeric(subdata[j,4])\n",
      "+                 if (subdata[j,3]==\"MCMC\" | subdata[j,3]==\"MCSC\") {\n",
      "+                     MCMCcount = MCMCcount+1\n",
      "+                     MCSCcount = MCSCcount+1\n",
      "+                 } else if (subdata[j,3]==\"SCMC\") {\n",
      "+                     SCMCcount = SCMCcount+1\n",
      "+                 } else if (subdata[j,3]==\"SCSC\") {\n",
      "+                     SCSCcount = SCSCcount+1\n",
      "+                 }\n",
      "+             }\n",
      "+             #Header is AA1, AA2, mixed, sum of edges, percent MC, influence1, influence2, degreeAA1\n",
      "+             out = rbind(out,c(data[i,1],data[i,2],\"mixed\",edgeSum,(MCMCcount+MCSCcount)/(MCMCcount+MCSCcount+SCMCcount+SCSCcount),influence1,influence2,sum(data[,1]==data[i,1])))\n",
      "+             if (any(subdata[,5]==\"PP\")==FALSE) {out_noPP = rbind(out_noPP,c(data[i,1],data[i,2],\"mixed\",edgeSum,(MCMCcount+MCSCcount)/(MCMCcount+MCSCcount+SCMCcount+SCSCcount),influence1,influence2,sum(data[,1]==data[i,1])))}\n",
      "+             completed = c(completed,x)\n",
      "+         }\n",
      "+     }\n",
      "+     out = out[out[,4]>0,]\n",
      "+     out_noPP = out_noPP[out_noPP[,4]>0,]\n",
      "+     colnames(out) = c(\"AA1\", \"AA2\", \"Type\", \"SumEdges\", \"percentMC\", \"influence1\", \"influence2\",\"degree\")\n",
      "+     results = c()\n",
      "+     results$out = out\n",
      "+     results$out_noPP = out_noPP\n",
      "+     results$influence_all = out_details[as.numeric(out_details[,4])>0 & (out_details[,8]==\"1\" | out_details[,9]==\"1\"),]\n",
      "+     out_details = out_details[as.numeric(out_details[,4])>0,]\n",
      "+     out_subtract = out_details; out_subtract[out_subtract[,8]==0 & out_subtract[,9]==0,4] = (-1)*as.numeric(out_subtract[out_subtract[,8]==0 & out_subtract[,9]==0,4])\n",
      "+     out_subtract = out_subtract[out_subtract[,5]!=\"PP\",]\n",
      "+     results$out_subtract = out_subtract\n",
      "+     return(results)\n",
      "+ }\n",
      "> \n",
      "> removeRedundancy = function(dataTmp) {\n",
      "+     for (i in 1:nrow(dataTmp)) {\n",
      "+         if (order(dataTmp[i,1:2])[1]==2) {\n",
      "+             dataTmp[i,1:2] = c(dataTmp[i,2],dataTmp[i,1])\n",
      "+             dataTmp[i,3] = paste(unlist(strsplit(dataTmp[i,3],split=\"\"))[c(3,4,1,2)],collapse=\"\")\n",
      "+             dataTmp[i,c(6,7)] = c(dataTmp[i,7],dataTmp[i,6])\n",
      "+         }\n",
      "+     }\n",
      "+     dataTmp = unique(dataTmp)\n",
      "+     pairs = unique(dataTmp[,1:2])\n",
      "+     out = c()\n",
      "+     for (j in 1:nrow(pairs)) {\n",
      "+         pair = pairs[j,]\n",
      "+         subData = matrix(dataTmp[dataTmp[,1]==pair[1] & dataTmp[,2]==pair[2],],ncol=7)\n",
      "+         out = rbind(out, c(pair[1],pair[2],\"mixed\",0.5*sum(as.numeric(subData[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> collapse_directed = function(data) {\n",
      "+     pairs = unique(data[,1:2])\n",
      "+     out = c()\n",
      "+     for (i in 1:nrow(pairs)) {\n",
      "+         subdat = matrix(data[data[,1]==pairs[i,1]&data[,2]==pairs[i,2],],ncol=4)\n",
      "+         out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> collapse_agnostic = function(data) {\n",
      "+     pairs = unique(t(apply(data[,1:2],1,function(x){return(sort(x))})))\n",
      "+     out = c()\n",
      "+     for (i in 1:nrow(pairs)) {\n",
      "+         subdat = matrix(data[(data[,1]==pairs[i,1]&data[,2]==pairs[i,2]) | (data[,2]==pairs[i,1]&data[,1]==pairs[i,2]),],ncol=4)\n",
      "+         out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> distributCalc = function(set,nodes,net) {\n",
      "+     #x is assumed to be a set but can also just be 1 node\n",
      "+     distanceNet = distances(net,weights=net$weight)\n",
      "+     distanceTmp = c()\n",
      "+     Dr = c()\n",
      "+     for (node in nodes) {\n",
      "+         distanceTmp = c()\n",
      "+         for (item in set) {\n",
      "+             if (item!=node) {\n",
      "+                 if(item%in%rownames(distanceNet)==FALSE){distanceTmp=100;next}\n",
      "+                 if(node%in%rownames(distanceNet)==FALSE){distanceTmp=100;next}\n",
      "+                 distanceTmp = c(distanceTmp,distanceNet[item,node])\n",
      "+             }\n",
      "+         }\n",
      "+         if (length(distanceTmp)>0) {\n",
      "+             Dr = c(Dr, 1/min(distanceTmp))\n",
      "+         }\n",
      "+     }\n",
      "+     return(sum(Dr)/length(nodes))\n",
      "+ }\n",
      "> \n",
      "> ##Read in user data\n",
      "> data_all = as.matrix(read.table(args[1]))\n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> originalNodes = nodes\n",
      "> basedata = collapse(data_all)\n",
      "> basedata_noPP = removeRedundancy(basedata$out[,1:7])\n",
      "> basedata_allsidechain = collapse_agnostic(data_all[data_all[,3]!=\"MCMC\",c(1,2,3,4)])\n",
      "> influencedata_all = basedata$influence_all\n",
      "> influencedata_collapse = collapse(influencedata_all)\n",
      "> influencedata = removeRedundancy(influencedata_collapse$out[,1:7])\n",
      "> influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "> influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "> \n",
      "> influencedata_directed = collapse_directed(influencedata_all[influencedata_all[,8]==1,c(1,2,3,4)])\n",
      "> influencenet_directed = graph.edgelist(influencedata_directed[,c(2,1)],directed=TRUE)\n",
      "> influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "> \n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> \n",
      "> #Write out edgelist files\n",
      "> write.table(influencedata,file=paste(keyword,\"_allEdges\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=FALSE,row.names=FALSE)\n",
      "> write.table(influencedata_directed,file=paste(keyword,\"_uniqueEdges\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=FALSE,row.names=FALSE)\n",
      "> \n",
      "> ###Deal with missing nodes\n",
      "> if (any(nodes %in% secStructure[,1]==FALSE)) {\n",
      "+     for (node in nodes) {\n",
      "+         if (node %in% secStructure[,1]==FALSE) {\n",
      "+             secStructure = rbind(secStructure,c(node,\"xxx\",360))\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> for (node in nodes) {\n",
      "+     if (node %in% rownames(rsa) == FALSE) {\n",
      "+         rsa = rbind(rsa,c(0,0))\n",
      "+         rownames(rsa)[nrow(rsa)] = node\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> ##Begin centrality calculations\n",
      "> \n",
      "> #Modular calculations\n",
      "> \n",
      "> ##WALKTRAP\n",
      "> net = graph.edgelist(basedata_noPP[,1:2],directed=FALSE)\n",
      "> net$weight = 1/as.numeric(basedata_noPP[,4])\n",
      "> net_community=walktrap.community(net,weights=net$weight); net_community_vec = net_community$membership; names(net_community_vec)=net_community$names\n",
      "> net_community_vec_wt = net_community_vec\n",
      "> nodes = V(net)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> nodes.color = c()\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(influencedata)) {\n",
      "+     j = which((basedata_noPP[,1]==influencedata[i,1]&basedata_noPP[,2]==influencedata[i,2])|(basedata_noPP[,1]==influencedata[i,2]&basedata_noPP[,2]==influencedata[i,1]))\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) { edgecolors[j]=\"grey40\" }\n",
      "+ }\n",
      "> if(length(nodes)>200) {textsize=.25} else {textsize=.45}\n",
      "> plot(net,edge.color = edgecolors, vertex.color = nodes.color,vertex.size=4,vertex.label.cex=textsize,main=\"WALKTRAP\")\n",
      ">     \n",
      "> ##Weighted edge betweenness WALKTRAP\n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP - WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for WALKTRAP\n",
      "> firstOrderDegree = degree(influencenet)\n",
      "> firstOrderDegree_sidechain = degree(influencenet_directed,mode=c(\"in\"))\n",
      "> secondOrderDegree = c(); secondOrderDegree_sidechain=c()\n",
      "> nodes = V(influencenet)$name\n",
      "> for (node in nodes) {\n",
      "+     firstorder = neighbors(influencenet,node)\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet,neighbor)))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree = c(secondOrderDegree,length(secondorder))\n",
      "+     firstorder = neighbors(influencenet_directed,node,mode=c(\"in\"))\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet_directed,neighbor,mode=c(\"in\"))))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree_sidechain = c(secondOrderDegree_sidechain,length(secondorder))\n",
      "+ }\n",
      "> names(secondOrderDegree) = nodes; names(secondOrderDegree_sidechain) = nodes\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])])) - 1\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])])) - 1\n",
      "+ }\n",
      "> \n",
      "> #SAVE ALL AS WALKTRAP\n",
      "> node_edge_betweenness_wt = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_wt = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_wt = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_wt = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_wt = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_wt = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> ##2ARY STRUCTURE\n",
      "> net_community_vec = as.numeric(as.factor(secStructure[,2]))\n",
      "> names(net_community_vec) = secStructure[,1]\n",
      "> nodes = V(net)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> nodes.color = c()\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(basedata_noPP)) {\n",
      "+     if (net_community_vec[basedata_noPP[i,1]]!=net_community_vec[basedata_noPP[i,2]]) { edgecolors[i]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for 2ARY STRUCTURE\n",
      "> nodes = V(influencenet)$name\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])])) - 1\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])])) - 1\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #SAVE ALL AS STRIDE\n",
      "> node_edge_betweenness_stride = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_stride = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_stride = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_stride = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_stride = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_stride = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> \n",
      "> ### Create final dataset\n",
      "> \n",
      "> out = cbind(\n",
      "+ firstOrderDegree[nodes],\n",
      "+ firstOrderDegree_sidechain[nodes],\n",
      "+ secondOrderDegree[nodes],\n",
      "+ secondOrderDegree_sidechain[nodes],\n",
      "+ node_edge_betweenness_stride[nodes],\n",
      "+ node_edge_betweenness_sidechain_stride[nodes],\n",
      "+ node_intermodular_degree_stride[nodes],\n",
      "+ node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ node_edge_betweenness_wt[nodes],\n",
      "+ node_edge_betweenness_sidechain_wt[nodes],\n",
      "+ node_intermodular_degree_wt[nodes],\n",
      "+ node_intermodular_degree_sidechain_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_wt[nodes]\n",
      "+ )\n",
      "> rownames(out) = nodes\n",
      "> colnames(out) = c(\"Degree\",\"Degree_sidechain\",\"SecondOrderDegree\",\"SecondOrderDegree_sidechain\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_sidechain\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_sidechain\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_sidechain\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_sidechain\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_sidechain\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_sidechain\")\n",
      "> \n",
      "> #add back in nodes that were not networked as zeros\n",
      "> zeroMat = matrix(0,nrow=length(which(originalNodes%in%nodes==FALSE)),ncol=ncol(out));rownames(zeroMat)=originalNodes[which(originalNodes%in%nodes==FALSE)]\n",
      "> out = rbind(out,zeroMat)\n",
      "> \n",
      "> #Add 1 to everything to avoid zeros, except ligand\n",
      "> out[is.na(out)]=0\n",
      "> out = out+1\n",
      "> \n",
      "> #Write out\n",
      "> write.table(out,file=paste(keyword,\"_scoresEnergetics\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "> #Standard normalization\n",
      "> outZ = apply(out,2,function(x){return(scale(x))})\n",
      "> rownames(outZ) = rownames(out)\n",
      "> write.table(outZ,file=paste(keyword,\"_scoresEnergeticsZ\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "Running centroid R script\n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(igraph)\n",
      "> library(stats)\n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> ##read in data\n",
      "> ## arg1: col1 and col2 are edges, col3 is weight of edge\n",
      "> ## arg2: weighted, unweighted <-- no longer using\n",
      "> ## arg3: forced module definition <-- no longer using\n",
      "> \n",
      "> # Set desired parameters\n",
      "> removeMCMC = FALSE\n",
      "> linearcutoff = 1\n",
      "> directed = FALSE\n",
      "> secondaryStructure2 = FALSE\n",
      "> sidechainMode = TRUE ##build network of just SCSC, SCMC and MCSC. Still use unique atoms for the _uniqueAtoms directed plots\n",
      "> uniqueAtomsMode = FALSE ##build network of just unique atoms (each edge has to have at least 1). Still use unique atoms for the _uniqueAtoms directed plots\n",
      "> useDNA=TRUE\n",
      "> uniqueAtomsUnbiased=FALSE\n",
      "> uniqueAtomsGaurav=FALSE\n",
      "> uniqueAtomsOLD=FALSE\n",
      "> atomCorrection = FALSE\n",
      "> uniqueAtomsGauravPP = FALSE\n",
      "> weighted = TRUE\n",
      "> removeWaters=TRUE\n",
      "> if (tail(unlist(strsplit(getwd(),\"/\")),n=1)==\"Centroid\") {\n",
      "+     ligandCentroidMode = TRUE\n",
      "+ } else {\n",
      "+     ligandCentroidMode = FALSE\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> # Read in ligand file\n",
      "> ligandfile = scan(paste(args[1],\"_ligand\",sep=\"\"),what=\"character\")\n",
      "> if (length(ligandfile)==0){\n",
      "+     ligandmode = FALSE\n",
      "+ } else {\n",
      "+     ligandmode = TRUE\n",
      "+ }\n",
      "> \n",
      "> ligandCentroidFile = scan(paste(args[1],\"NetLigand\",sep=\"\"),what=\"character\")\n",
      "> \n",
      "> # keyword is the base filename\n",
      "> # method is \"frag\" or \"module\" or \"both\"\n",
      "> keyword = args[1]\n",
      "> args=c(paste(args[1],\"_net\",sep=\"\"),\"weighted\",dir(\"./\",\"_secondaryStructure$\"))\n",
      "> \n",
      "> ##Load data\n",
      "> #Unique atoms file\n",
      "> if(uniqueAtomsUnbiased) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"/Users/vjpatel/Dropbox/Gaurav/uniqueAtomsUnbiased\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else if (uniqueAtomsGaurav) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"/Users/vjpatel/Dropbox/Gaurav/uniqueAtomsGaurav\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else if (uniqueAtomsOLD) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"../../../terminalAtoms\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"../../../uniqueAtoms\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     otherTA = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+         #otherTA[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,4],split=\",\"))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> #Secondary structure file\n",
      "> #Ligands are all one secondary structure\n",
      "> secStructure = as.matrix(read.table(args[3]))\n",
      "> phiAngleDB = {}\n",
      "> for (i in 1:nrow(secStructure)) {\n",
      "+     phiAngleDB[[secStructure[i,1]]] = as.numeric(secStructure[i,3])\n",
      "+ }\n",
      "> \n",
      "> if(ligandmode) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",\"_ligand\"),sep=\"\\t\"))\n",
      "+     liganddat = unique(liganddat[,2])\n",
      "+     for (ligand in liganddat) {\n",
      "+         secStructure = rbind(secStructure,c(sub(\"-\",\"\",ligand),\"ligand\",360))\n",
      "+         phiAngleDB[[sub(\"-\",\"\",ligand)]]=0\n",
      "+     }\n",
      "+ } else {\n",
      "+     liganddat = c()\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> ##Define functions\n",
      "> \n",
      "> ##Collapse matrix functions\n",
      "> ##Will return redundant matrix\n",
      "> ##Input needs to be 7 columns\n",
      "> ##Second input is different if the main input is a subset of the larger data\n",
      "> addTerminalDetails = function(data,largerdata) {\n",
      "+     #List edges bi-directionally\n",
      "+     x=unlist(lapply(data[,3],function(x){if(x==\"MCSC\"){return(\"SCMC\")}else if(x==\"SCMC\"){return(\"MCSC\")} else {return(x)}}))\n",
      "+     data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6]))\n",
      "+     data = unique(data)\n",
      "+     out = c()\n",
      "+     out_noPP = c()\n",
      "+     out_details = c()\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+         subdata = matrix(data[data[,1]==data[i,1]&data[,2]==data[i,2],],ncol=7)\n",
      "+         allsubdata = matrix(largerdata[largerdata[,1]==data[i,1],],ncol=7)\n",
      "+         code1=0;code2=0\n",
      "+         if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+             code1 = 1\n",
      "+         }\n",
      "+         if (data[i,5] == \"DNA\" & useDNA==TRUE) {\n",
      "+             code1 = 1\n",
      "+         }\n",
      "+         allsubdata = matrix(largerdata[largerdata[,2]==data[i,2],],ncol=7)\n",
      "+         if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+             code2 = 1\n",
      "+         }\n",
      "+         if (data[i,5] == \"DNA\" & useDNA==TRUE) {\n",
      "+             code2 = 1\n",
      "+         }\n",
      "+         if (data[i,5]==\"PICATION\"|data[i,5]==\"PIPI\") {\n",
      "+             code1=1\n",
      "+             code2=1\n",
      "+         }\n",
      "+         if (data[i,5]==\"PP\") {\n",
      "+             if (uniqueAtomsGaurav) {\n",
      "+                 if((acid1) == \"GLY\") {\n",
      "+                     code1 = 1\n",
      "+                     data[i,3]==\"SCSC\"\n",
      "+                 }\n",
      "+                 if ((acid2) == \"GLY\") {\n",
      "+                     code2 = 1\n",
      "+                     data[i,3] == \"SCSC\"\n",
      "+                 }\n",
      "+             }\n",
      "+         }\n",
      "+         out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+     }\n",
      "+     return(out_details)\n",
      "+ }\n",
      "> \n",
      "> ##Will return non-redundant matrix\n",
      "> collapse_agnostic = function(data) {\n",
      "+     if(nrow(data)==0) {\n",
      "+         return(matrix(ncol=4,nrow=0))\n",
      "+     } else if (ncol(data)==4 & data[1,3]==\"mixed\") {\n",
      "+         return(data)\n",
      "+     } else {\n",
      "+         pairs = unique(t(apply(matrix(data[,1:2],ncol=2),1,function(x){return(sort(x))})))\n",
      "+         out = c()\n",
      "+         for (i in 1:nrow(pairs)) {\n",
      "+             subdat = matrix(data[(data[,1]==pairs[i,1]&data[,2]==pairs[i,2]) | (data[,2]==pairs[i,1]&data[,1]==pairs[i,2]),1:4],ncol=4)\n",
      "+             subdat = unique(t(apply(subdat,1,function(x){a=which(x==sort(x[1:2])[1]);if(a==1){return(x)}else{return(c(x[2],x[1],paste(unlist(strsplit(x[3],\"\"))[c(3,4,1,2)],collapse=\"\"),x[4]))}})))\n",
      "+             if (weighted==FALSE) {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",1))\n",
      "+             } else {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+             }\n",
      "+         }\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> ##Will return directed collapsed matrix\n",
      "> collapse_directed = function(data) {\n",
      "+     if (length(data)==0) {\n",
      "+         return(matrix(nrow=0,ncol=4))\n",
      "+     } else {\n",
      "+         pairs = matrix(unique(data[,1:2]),ncol=2)\n",
      "+         out = c()\n",
      "+         for (i in 1:nrow(pairs)) {\n",
      "+             subdat = matrix(data[data[,1]==pairs[i,1]&data[,2]==pairs[i,2],1:4],ncol=4)\n",
      "+             if (weighted==FALSE) {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",1))\n",
      "+             } else {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+             }\n",
      "+         }\n",
      "+         return(matrix(out,ncol=4))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> rsa = as.matrix(read.table(dir(\"./\",\".rsa$\")))\n",
      "> rownames(rsa) = rsa[,1]\n",
      "> rsa[,1] = rsa[,2]\n",
      "> \n",
      "> #add ligand as -1\n",
      "> for (ligand in liganddat) {\n",
      "+     rsa = rbind(rsa,c(-1,-1))\n",
      "+     rownames(rsa)[nrow(rsa)] = sub(\"-\",\"\",ligand)\n",
      "+ }\n",
      "> \n",
      "> #Function to get amino acid\n",
      "> getAcid = function(residue) {\n",
      "+     if(all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"A\")) == TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"G\"))==TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"C\"))==TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"T\"))==TRUE)) {return(paste(unlist(strsplit(residue,\"\"))[1:2],collapse=\"\"))\n",
      "+     } else {\n",
      "+     return(paste(unlist(strsplit(residue,\"\"))[1:3],collapse=\"\"))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> ##################\n",
      "> ## read in data ##\n",
      "> ##################\n",
      "> data_all = as.matrix(read.table(args[1]))[,1:7]\n",
      "> if(removeWaters) {\n",
      "+     y=c(grep(\"HOH\",data_all[,1]),grep(\"HOH\",data_all[,2]))\n",
      "+     if(length(y)>0) {\n",
      "+         data_all = data_all[-y,]\n",
      "+     }\n",
      "+ }\n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> originalNodes = nodes\n",
      "> \n",
      "> data_all_original = rbind(data_all,data_all[,c(2,1,3,4,5,7,6)])\n",
      "> data_all = data_all[as.numeric(data_all[,4])>0,]\n",
      "> \n",
      "> basedata = addTerminalDetails(data_all,data_all_original)\n",
      "> basedata_noPP = collapse_agnostic(basedata[basedata[,5]!=\"PP\",])\n",
      "> sidechaindata = basedata[basedata[,3]!=\"MCMC\",1:4]\n",
      "> sidechaindata_detailed = basedata[basedata[,3]!=\"MCMC\",]\n",
      "> \n",
      "> influencedata_detailed = basedata\n",
      "> if (uniqueAtomsMode) {\n",
      "+     influencedata_detailed = basedata[basedata[,8]==1 | basedata[,9]==1,]\n",
      "+ }\n",
      "> influencedata = collapse_agnostic(influencedata_detailed)\n",
      "> influencedataGLY = rbind(influencedata[grep(\"GLY\",influencedata[,1]),],influencedata[grep(\"GLY\",influencedata[,1]),])\n",
      "> influencedata = influencedata[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,2])))>linearcutoff,]\n",
      "> \n",
      "> \n",
      "> influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "> influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "> \n",
      "> influencedata_directed = collapse_directed(matrix(influencedata_detailed[influencedata_detailed[,8]==1,1:4],ncol=4))\n",
      "> influencedata_directedGLY = rbind(influencedata_directed[grep(\"GLY\",influencedata_directed[,1]),],influencedata_directed[grep(\"GLY\",influencedata_directed[,2]),])\n",
      "> influencedata_directed = influencedata_directed[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,2])))>linearcutoff,]\n",
      "> \n",
      "> influencedata_directed = matrix(influencedata_directed,ncol=4)\n",
      "> influencenet_directed = graph.edgelist(matrix(influencedata_directed[,c(2,1)],ncol=2),directed=TRUE)\n",
      "> influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "> \n",
      "> \n",
      "> if(sidechainMode) {\n",
      "+     influencedata_detailed = sidechaindata_detailed\n",
      "+     influencedata = collapse_agnostic(influencedata_detailed)\n",
      "+     influencedataGLY = rbind(influencedata[grep(\"GLY\",influencedata[,1]),],influencedata[grep(\"GLY\",influencedata[,2]),])\n",
      "+     influencedata = influencedata[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,2])))>linearcutoff,]\n",
      "+     #Add back for GLY if uniqueAtomsGaurav\n",
      "+     if(uniqueAtomsGauravPP) {\n",
      "+         influencedata = rbind(influencedata,influencedataGLY)\n",
      "+     }\n",
      "+     influencedata = influencedata[as.numeric(influencedata[,4])>0,]\n",
      "+     \n",
      "+     influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "+     influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "+     \n",
      "+     influencedata_directed = collapse_directed(matrix(influencedata_detailed[influencedata_detailed[,8]==1,1:4],ncol=4))\n",
      "+     influencedata_directedGLY = rbind(influencedata_directed[grep(\"GLY\",influencedata_directed[,1]),],influencedata_directed[grep(\"GLY\",influencedata_directed[,2]),])\n",
      "+ \n",
      "+     influencedata_directed = influencedata_directed[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,2])))>linearcutoff,]\n",
      "+     #Add back for GLY if uniqueAtomsGaurav\n",
      "+     if(uniqueAtomsGauravPP) {\n",
      "+         influencedata_directed = rbind(influencedata_directed,influencedata_directedGLY)\n",
      "+     }\n",
      "+     influencedata_directed = matrix(influencedata_directed,ncol=4)\n",
      "+     influencenet_directed = graph.edgelist(matrix(influencedata_directed[,c(2,1)],ncol=2),directed=TRUE)\n",
      "+     influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "+     \n",
      "+ }\n",
      "> \n",
      "> #Replace missing nodes in the igraph net\n",
      "> nodesMissing = nodes[nodes %in% influencedata[,1]==FALSE & nodes %in% influencedata[,2]==FALSE]\n",
      "> \n",
      "> ##B-factor\n",
      "> bfactor = as.matrix(read.table(dir(\"./\",\"_Bfactor\"),row.names=1))\n",
      "> \n",
      "> #Replace missing nodes in secStructure matrix\n",
      "> if (any(nodes %in% secStructure[,1]==FALSE)) {\n",
      "+     for (node in nodes) {\n",
      "+         if (node %in% secStructure[,1]==FALSE) {\n",
      "+             secStructure = rbind(secStructure,c(node,\"xxx\",360))\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> for (node in nodes) {\n",
      "+     if (node %in% rownames(rsa) == FALSE) {\n",
      "+         rsa = rbind(rsa,c(0,0))\n",
      "+         rownames(rsa)[nrow(rsa)] = node\n",
      "+     }\n",
      "+     if (node %in% rownames(bfactor) == FALSE) {\n",
      "+         bfactor = rbind(bfactor,1)\n",
      "+         rownames(bfactor)[nrow(bfactor)] = node\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> \n",
      "> \n",
      "> #Give DNA attributes\n",
      "> if (\"DNA\" %in% data_all[,5]) {\n",
      "+     for (i in 1:nrow(data_all)) {\n",
      "+         if (data_all[i,5]==\"DNA\") {\n",
      "+             phiAngleDB[[data_all[i,1]]] = 90\n",
      "+             if (data_all[i,1] %in% rownames(rsa) == FALSE) {\n",
      "+                 rsa = rbind(rsa,c(0,0))\n",
      "+                 rownames(rsa)[nrow(rsa)] = data_all[i,1]\n",
      "+             }\n",
      "+             if (data_all[i,1] %in% rownames(secStructure) == FALSE) {\n",
      "+                 secStructure = rbind(secStructure,c(data_all[i,1],\"DNA\",360))\n",
      "+             }\n",
      "+             if (data_all[i,1] %in% rownames(bfactor) == FALSE) {\n",
      "+                 bfactor = rbind(bfactor,1)\n",
      "+                 rownames(bfactor)[nrow(bfactor)] = data_all[i,1]\n",
      "+             }\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #Betweenness\n",
      "> node_betweenness = betweenness(influencenet,weights=influencenet$weight,directed=directed)\n",
      "> node_betweenness_unique = betweenness(influencenet_directed,weights=influencenet_directed$weight,directed=directed)\n",
      "> \n",
      "> #Modular calculations\n",
      "> \n",
      "> ##WALKTRAP\n",
      "> net = graph.edgelist(basedata_noPP[,1:2],directed=FALSE)\n",
      "> net$weight = 1/as.numeric(basedata_noPP[,4])\n",
      "> net_community=walktrap.community(net,weights=net$weight); net_community_vec = net_community$membership; names(net_community_vec)=net_community$names\n",
      "> net_community_vec_wt = net_community_vec\n",
      "> #net_community_vec[buriednodes] = max(net_community_vec)+1\n",
      "> nodes = V(influencenet)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> nodes.color = c()\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(influencedata)) {\n",
      "+     j = which((basedata_noPP[,1]==influencedata[i,1]&basedata_noPP[,2]==influencedata[i,2])|(basedata_noPP[,1]==influencedata[i,2]&basedata_noPP[,2]==influencedata[i,1]))\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) { edgecolors[j]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP\n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP - WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         if(weight1==0 & weight2==0) {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]] + 0\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]] + 0\n",
      "+         } else {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for WALKTRAP\n",
      "> firstOrderDegree = degree(influencenet)\n",
      "> firstOrderDegree_sidechain = degree(influencenet_directed,mode=c(\"in\"))\n",
      "> \n",
      "> secondOrderDegree = c(); secondOrderDegree_sidechain=c()\n",
      "> nodes = V(influencenet)$name\n",
      "> for (node in nodes) {\n",
      "+     firstorder = neighbors(influencenet,node)\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet,neighbor)))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree = c(secondOrderDegree,length(secondorder))\n",
      "+     if(node %in% V(influencenet_directed)$name) {\n",
      "+         firstorder_sidechain = neighbors(influencenet_directed,node,mode=c(\"in\"))\n",
      "+     } else {\n",
      "+         firstorder_sidechain = c()\n",
      "+     }\n",
      "+     secondorder = c(); for (neighbor in firstorder_sidechain){secondorder = c(secondorder,names(neighbors(influencenet_directed,neighbor,mode=c(\"in\"))))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree_sidechain = c(secondOrderDegree_sidechain,length(secondorder))\n",
      "+ }\n",
      "> names(secondOrderDegree) = nodes; names(secondOrderDegree_sidechain) = nodes\n",
      "> \n",
      "> \n",
      "> #second order intermodular degree\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])]))\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])]))\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #SAVE ALL AS WALKTRAP\n",
      "> node_edge_betweenness_wt = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_wt = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_wt = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_wt = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_wt = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_wt = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> \n",
      "> ##2ARY STRUCTURE\n",
      "> secStructure = secStructure[,1:2]\n",
      "> rownames(secStructure) = secStructure[,1]\n",
      "> net_community_vec = as.numeric(as.factor(secStructure[,2]))\n",
      "> names(net_community_vec) = secStructure[,1]\n",
      "> #net_community_vec[buriednodes] = max(net_community_vec)+1\n",
      "> nodes = V(influencenet)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> nodes.color = c()\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(basedata_noPP)) {\n",
      "+     if (net_community_vec[basedata_noPP[i,1]]!=net_community_vec[basedata_noPP[i,2]]) { edgecolors[i]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         if(weight1==0 & weight2==0) {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]] + 0\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]] + 0\n",
      "+         } else {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> ##Degree and second order degree for 2ARY STRUCTURE\n",
      "> \n",
      "> #second order intermodular degree\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     #secondOrder_node_intermodular_degree[node] = length(unique(unlist(node_modules[names(node_modules[[node]])])))\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])]))\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])]))\n",
      "+ }\n",
      "> \n",
      "> #SAVE ALL AS STRIDE\n",
      "> node_edge_betweenness_stride = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_stride = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_stride = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_stride = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_stride = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_stride = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> ### Create final dataset\n",
      "> nodes = nodes[nodes!=\"DNA1000A\"]\n",
      "> nodes = nodes[nodes %in% c(\"DA\",\"DG\",\"DC\",\"DT\") == FALSE]\n",
      "> \n",
      "> ### Remove ligand amino acids from final dataset\n",
      "> ligands = sub(\"-\",\"\",liganddat)\n",
      "> nodes = nodes[nodes %in% ligands == FALSE]\n",
      "> \n",
      "> nodes[which(nodes %in% rownames(bfactor)==FALSE)]\n",
      "character(0)\n",
      "> \n",
      "> out = cbind(\n",
      "+ rsa[nodes,1],\n",
      "+ firstOrderDegree[nodes],\n",
      "+ firstOrderDegree_sidechain[nodes],\n",
      "+ secondOrderDegree[nodes],\n",
      "+ secondOrderDegree_sidechain[nodes],\n",
      "+ node_edge_betweenness_stride[nodes],\n",
      "+ node_edge_betweenness_sidechain_stride[nodes],\n",
      "+ node_intermodular_degree_stride[nodes],\n",
      "+ node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ node_edge_betweenness_wt[nodes],\n",
      "+ node_edge_betweenness_sidechain_wt[nodes],\n",
      "+ node_intermodular_degree_wt[nodes],\n",
      "+ node_intermodular_degree_sidechain_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_wt[nodes]\n",
      "+ )\n",
      "> \n",
      "> #write column names\n",
      "> colnames(out) = c(\"RSA\",\"Degree\",\"Degree_uniqueAtoms\",\"SecondOrderDegree\",\"SecondOrderDegree_uniqueAtoms\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_unqiueAtoms\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_uniqueAtoms\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_uniqueAtoms\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_uniqueAtoms\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_uniqueAtoms\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_uniqueAtoms\")\n",
      "> \n",
      "> \n",
      "> ## LIGANDS\n",
      "> if(ligandmode==TRUE & ligandCentroidMode==FALSE) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",\"_ligand\"),sep=\"\\t\"))\n",
      "+     \n",
      "+     tmp = c()\n",
      "+     i=1\n",
      "+     for (i in 1:nrow(liganddat)) {\n",
      "+         acid=paste(unlist(strsplit(liganddat[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         if (unlist(strsplit(liganddat[i,1],\"-\"))[2] %in% terminalAtoms[[acid]]) {\n",
      "+             tmp = rbind(tmp,c(unlist(strsplit(liganddat[i,1],\"-\"))[1],liganddat[i,2],liganddat[i,3]))\n",
      "+         }\n",
      "+     }\n",
      "+     \n",
      "+     #ligandvec = tapply(liganddat[,3],as.factor(liganddat[,1]),min)\n",
      "+     ligandvec = tapply(as.numeric(tmp[,3]),as.factor(tmp[,1]),min)\n",
      "+ }\n",
      "> \n",
      "> if(ligandmode==TRUE & ligandCentroidMode) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",paste(\"_centroidNetLigand$\",sep=\"\")),sep=\"\\t\"))\n",
      "+     \n",
      "+     tmp = c()\n",
      "+     i=1\n",
      "+     for (i in 1:nrow(liganddat)) {\n",
      "+         tmp = rbind(tmp,c(liganddat[i,1],liganddat[i,2],liganddat[i,3]))\n",
      "+     }\n",
      "+     \n",
      "+     ligandvec = tapply(tmp[,3],as.factor(tmp[,1]),min)\n",
      "+ }\n",
      "> \n",
      "> if (ligandmode==FALSE) {\n",
      "+     ligandvec = rep(0,length(nodes))\n",
      "+ }\n",
      "> \n",
      "> out = cbind(out,ligandvec[rownames(out)])\n",
      "> colnames(out) = c(\"RSA\",\"Degree\",\"Degree_uniqueAtoms\",\"SecondOrderDegree\",\"SecondOrderDegree_uniqueAtoms\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_unqiueAtoms\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_uniqueAtoms\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_uniqueAtoms\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_uniqueAtoms\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_uniqueAtoms\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_uniqueAtoms\",\"Ligand\")\n",
      "> \n",
      "> #add back in nodes that were not networked as zeros\n",
      "> zeroMat = matrix(0,nrow=length(which(originalNodes%in%nodes==FALSE)),ncol=ncol(out));rownames(zeroMat)=originalNodes[which(originalNodes%in%nodes==FALSE)]\n",
      "> out = rbind(out,zeroMat)\n",
      "> \n",
      "> #fix RSA and Bfactor and ligand\n",
      "> out[,\"RSA\"] = rsa[rownames(out),1]\n",
      "> out[,\"Ligand\"] = ligandvec[rownames(out)]\n",
      "> \n",
      "> #Add 1 to everything to avoid zeros, except ligand\n",
      "> out[is.na(out)]=0\n",
      "> out[out[,\"Ligand\"]==0,\"Ligand\"] = 150\n",
      "> out[1:nrow(out),2:ncol(out)] = as.numeric(out[1:nrow(out),2:ncol(out)])+1\n",
      "> \n",
      "> ##create Z score file\n",
      "> outZ = apply(out,2,function(x){return(scale(as.numeric(x)))})\n",
      "> rownames(outZ) = rownames(out)\n",
      "> colnames(outZ) = colnames(out)\n",
      "> outZ[outZ==NA]=0\n",
      "> outZ[outZ==\"NaN\"]=0\n",
      "> \n",
      "> #Write out scores and z-normed scores files\n",
      "> write.table(out,file=paste(keyword,\"_scoresCentroid\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> write.table(outZ,file=paste(keyword,\"_scoresCentroidZ\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "> \n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> ## run from within the monomer folder\n",
      "> ## there can only be one *_scoresEnergetics and one Centroid/*_scoresCentroid in this folder\n",
      "> ## args[1] is the folder for the multimer\n",
      "> ## NOTE: results may differ slightly from publication because of missing data in the mutagenesis experiments used in teh paper\n",
      "> \n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> #Load monomer\n",
      "> data1 = read.table(dir(\"./\",\"_scoresEnergetics$\"),header=T,row.names=1)\n",
      "> data1Z = as.data.frame(apply(data1,2,function(x){return(scale(as.numeric(x)))}))\n",
      "> rownames(data1Z) = rownames(data1)\n",
      "> data1Z[is.na(data1Z)] = 0\n",
      "> data2 = read.table(dir(\"./Centroid\",\"_scoresCentroid$\", full.names=TRUE),header=T,row.names=1)\n",
      "> data2Z = as.data.frame(apply(data2,2,function(x){return(scale(as.numeric(x)))}))\n",
      "> rownames(data2Z) = rownames(data2)\n",
      "> data2Z[is.na(data2Z)] = 0\n",
      "> colnames(data2Z) = paste(colnames(data2Z),\"CENTROIDSC\",sep=\"\")\n",
      "> data2Z = data2Z[rownames(data1Z),]\n",
      "> data2 = data2[rownames(data1),]\n",
      "> \n",
      "> #Load multimer\n",
      "> data1Multimer = read.table(dir(args[1],\"_scoresEnergetics$\",full.names=TRUE),header=T,row.names=1)\n",
      "> data2Multimer = read.table(dir(paste(args[1],\"Centroid\",sep=\"/\"),\"_scoresCentroid$\",full.names=TRUE),header=T,row.names=1)\n",
      "> data2Multimer = data2Multimer[rownames(data1Multimer),]\n",
      "> data1MultimerZ = as.data.frame(apply(data1Multimer,2,function(x){return(scale(as.numeric(x)))}))\n",
      "> rownames(data1MultimerZ) = rownames(data1Multimer)\n",
      "> data2MultimerZ = as.data.frame(apply(data2Multimer,2,function(x){return(scale(as.numeric(x)))}))\n",
      "> rownames(data2MultimerZ) = rownames(data2Multimer)\n",
      "> colnames(data1MultimerZ) = paste(colnames(data1MultimerZ),\"MULTIMER\",sep=\"\")\n",
      "> colnames(data2MultimerZ) = paste(colnames(data2MultimerZ),\"MULTIMERCENTROIDSC\",sep=\"\")\n",
      "> data2MultimerZ = data2MultimerZ[rownames(data1MultimerZ),]\n",
      "> \n",
      "> #Add RSA\n",
      "> data1 = cbind(data2[,\"RSA\"],data1)\n",
      "> colnames(data1)[1]=\"RSA\"\n",
      "> dataZ = cbind(data1Z,data2Z)\n",
      "> \n",
      "> #Identify max NodeEdgeBetweennessSTRIDE_sidechain\n",
      "> acidsMonomer = c()\n",
      "> for (i in 1:length(rownames(dataZ))) {\n",
      "+     node = rownames(dataZ)[i]\n",
      "+     node = unlist(strsplit(node,split=\"\"))\n",
      "+     node = paste(node[1:(length(node)-1)],collapse=\"\")\n",
      "+     acidsMonomer = c(acidsMonomer,node)\n",
      "+ }\n",
      "> \n",
      "> acidsMultimer = c()\n",
      "> for (i in 1:length(rownames(data1MultimerZ))) {\n",
      "+     node = rownames(data1MultimerZ)[i]\n",
      "+     node = unlist(strsplit(node,split=\"\"))\n",
      "+     node = paste(node[1:(length(node)-1)],collapse=\"\")\n",
      "+     acidsMultimer = c(acidsMultimer,node)\n",
      "+ }\n",
      "> \n",
      "> allAcids = unique(c(acidsMonomer,acidsMultimer))\n",
      "> \n",
      "> #SecondOrderIntermodularDegree is the average of the multimer only\n",
      "> SecondOrderIntermodularDegree_ALL = (data1MultimerZ[,\"SecondOrderIntermodularDegreeSTRIDE_sidechainMULTIMER\"]+data1MultimerZ[,\"SecondOrderIntermodularDegreeSTRIDEMULTIMER\"]+data2MultimerZ[,\"SecondOrderIntermodularDegreeSTRIDEMULTIMERCENTROIDSC\"]+data2MultimerZ[,\"SecondOrderIntermodularDegreeWALKTRAPMULTIMERCENTROIDSC\"])/4\n",
      "> SecondOrderIntermodularDegree_AVERAGE = tapply(SecondOrderIntermodularDegree_ALL,as.factor(acidsMultimer),mean)\n",
      "> SecondOrderIntermodularDegree_AVERAGE = SecondOrderIntermodularDegree_AVERAGE[allAcids]\n",
      "> missing = which(allAcids%in%acidsMultimer == FALSE)\n",
      "> SecondOrderIntermodularDegree_AVERAGE[missing] = (data1Z[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"SecondOrderIntermodularDegreeSTRIDE_sidechain\"]+data1Z[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"SecondOrderIntermodularDegreeSTRIDE\"]+data2Z[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"SecondOrderIntermodularDegreeSTRIDECENTROIDSC\"]+data2Z[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"SecondOrderIntermodularDegreeWALKTRAPCENTROIDSC\"])/4\n",
      "> names(SecondOrderIntermodularDegree_AVERAGE)[missing] = allAcids[which(allAcids%in%acidsMultimer == FALSE)]\n",
      "> \n",
      "> #NodeEdgeBetweenness is the max of the monomer/multimer\n",
      "> NodeEdgeBetweennessSTRIDE_AVERAGE_MONOMER = (dataZ[,\"NodeEdgeBetweennessSTRIDE\"]+dataZ[,\"NodeEdgeBetweennessSTRIDE_sidechain\"])/2\n",
      "> NodeEdgeBetweennessSTRIDE_AVERAGE_MULTIMER = (data1MultimerZ[,\"NodeEdgeBetweennessSTRIDEMULTIMER\"]+data1MultimerZ[,\"NodeEdgeBetweennessSTRIDE_sidechainMULTIMER\"])/2\n",
      "> NodeEdgeBetweennessSTRIDE_sidechain_ALL = c(NodeEdgeBetweennessSTRIDE_AVERAGE_MONOMER,NodeEdgeBetweennessSTRIDE_AVERAGE_MULTIMER)\n",
      "> NodeEdgeBetweennessSTRIDE_sidechain_MAX = tapply(NodeEdgeBetweennessSTRIDE_sidechain_ALL,as.factor(c(acidsMonomer,acidsMultimer)),max,na.rm=TRUE)\n",
      "> NodeEdgeBetweennessSTRIDE_sidechain_MAX = NodeEdgeBetweennessSTRIDE_sidechain_MAX[allAcids]\n",
      "> \n",
      "> #Ligand is the min of the multimer only\n",
      "> LigandMULTIMERCENTROIDSC_ALL = data2MultimerZ[,\"LigandMULTIMERCENTROIDSC\"]\n",
      "> LigandMULTIMERCENTROIDSC_MIN = tapply(LigandMULTIMERCENTROIDSC_ALL,as.factor(acidsMultimer),min)\n",
      "> LigandMULTIMERCENTROIDSC_MIN = LigandMULTIMERCENTROIDSC_MIN[allAcids]\n",
      "> missing = which(allAcids%in%acidsMultimer == FALSE)\n",
      "> LigandMULTIMERCENTROIDSC_MIN[missing] = dataZ[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"LigandCENTROIDSC\"]\n",
      "> names(LigandMULTIMERCENTROIDSC_MIN)[missing] = allAcids[which(allAcids%in%acidsMultimer == FALSE)]\n",
      "> LigandMULTIMERCENTROIDSC_MIN[is.na(LigandMULTIMERCENTROIDSC_MIN)] = 0\n",
      "> \n",
      "> #RSA\n",
      "> RSA_ALL = c(data1[,\"RSA\"],data2Multimer[,\"RSA\"])\n",
      "> RSA_MIN = tapply(RSA_ALL,as.factor(c(acidsMonomer,acidsMultimer)),min,na.rm=TRUE)\n",
      "> RSA_MIN = RSA_MIN[allAcids]\n",
      "> \n",
      "> y = SecondOrderIntermodularDegree_AVERAGE+NodeEdgeBetweennessSTRIDE_sidechain_MAX-LigandMULTIMERCENTROIDSC_MIN\n",
      "> \n",
      "> out=cbind(names(y),y)\n",
      "> acidNumbers = sapply(out[,1],substring,4)\n",
      "> out = out[order(as.numeric(acidNumbers)),]\n",
      "> write.table(out,file=\"FinalSum\",sep=\"\\t\",quote=FALSE,row.names=FALSE,col.names=FALSE)\n",
      "> \n",
      "> # also write the 3 variables to a file\n",
      "> # Concatenating all three variables into a single data frame\n",
      "> result <- data.frame(\n",
      "+   Acid = names(SecondOrderIntermodularDegree_AVERAGE),\n",
      "+   SecondOrderIntermodularDegree_AVERAGE,\n",
      "+   NodeEdgeBetweennessSTRIDE_sidechain_MAX,\n",
      "+   LigandMULTIMERCENTROIDSC_MIN\n",
      "+ )\n",
      "> \n",
      "> # Writing the data frame to a file\n",
      "> write.table(result, file = \"FinalSum_Decomp\", sep = \"\\t\", quote = FALSE, row.names = FALSE, col.names = TRUE)\n",
      "> \n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "Read 1080 items\n",
      "Read 1080 items\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "Read 7713 items\n",
      "Read 7713 items\n",
      "\n",
      "SBNA results for 7DHY succesfully downloaded\n",
      "\n",
      "Processing 7DHY with chain D...\n",
      "Starting phenix.pdbtools\n",
      "on Sun Jun  2 13:52:37 2024 by yliy0004\n",
      "===============================================================================\n",
      "\n",
      "Processing files:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Found model, 7DHY.pdb\n",
      "\n",
      "Processing PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Adding command-line PHIL:\n",
      "  -------------------------\n",
      "    keep=Chain D\n",
      "\n",
      "Final processed PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "  data_manager {\n",
      "    model {\n",
      "      file = \"7DHY.pdb\"\n",
      "    }\n",
      "    default_model = \"7DHY.pdb\"\n",
      "  }\n",
      "  modify {\n",
      "    keep = Chain D\n",
      "  }\n",
      "\n",
      "\n",
      "Starting job\n",
      "===============================================================================\n",
      "Validating inputs\n",
      "Performing manipulations\n",
      "Size before: 6723 size after: 1720\n",
      "Writing output model\n",
      "\n",
      "===============================================================================\n",
      "Job complete\n",
      "usr+sys time: 1.92 seconds\n",
      "wall clock time: 1.94 seconds\n",
      "Cleaning PDB file\n",
      "Starting phenix.pdbtools\n",
      "on Sun Jun  2 13:52:39 2024 by yliy0004\n",
      "===============================================================================\n",
      "\n",
      "Processing files:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Found model, 7DHY_monomer.pdb\n",
      "\n",
      "Processing PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Adding command-line PHIL:\n",
      "  -------------------------\n",
      "    remove=water\n",
      "\n",
      "Final processed PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "  data_manager {\n",
      "    model {\n",
      "      file = \"7DHY_monomer.pdb\"\n",
      "    }\n",
      "    default_model = \"7DHY_monomer.pdb\"\n",
      "  }\n",
      "  modify {\n",
      "    remove = water\n",
      "  }\n",
      "\n",
      "\n",
      "Starting job\n",
      "===============================================================================\n",
      "Validating inputs\n",
      "Performing manipulations\n",
      "Size before: 1720 size after: 1569\n",
      "Writing output model\n",
      "\n",
      "===============================================================================\n",
      "Job complete\n",
      "usr+sys time: 1.68 seconds\n",
      "wall clock time: 1.70 seconds\n",
      "Writing out polypeptide file...\n",
      "Adding hydrogens...\n",
      "Protonating waters...\n",
      "Making .phi file\n",
      "Running full bond calculation script\n",
      "The ligands identified in this pdb are GOL-306D, GOL-305D, SO4-303D, ARS-301D, GOL-304D\n",
      "Screening for atoms that are close.\n",
      "Dictionaries loaded.\n",
      "Starting salt bridges.\n",
      "Starting hydrogen bonds.\n",
      "Starting pi-pi version 2.\n",
      "Running pi-cation version 2\n",
      "Starting disulfide\n",
      "Starting van der waals.\n",
      "Starting metals\n",
      "Starting DNA bonds...\n",
      "Starting Ligand centroids\n",
      "Starting residue centroids...\n",
      "Removing duplicates from BFactor file\n",
      "Making 7DHY_monomer_nowaters_net\n",
      "Removing any negative value edges\n",
      "Creating QC file\n",
      "Creating Centroid directory and moving files\n",
      "Running energetic R script\n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> ##Load libraries\n",
      "> library(igraph)\n",
      "> library(stats)\n",
      "> \n",
      "> ##Initialize keywords\n",
      "> # keyword is the base filename\n",
      "> keyword = args[1]\n",
      "> args=c(paste(args[1],\"_net\",sep=\"\"),\"weighted\",paste(keyword,\"_secondaryStructure\",sep=\"\"))\n",
      "> atomCorrection=FALSE\n",
      "> baseKeyword = unlist(strsplit(keyword,\"/\"))[1]\n",
      "> runVariableSelection = TRUE\n",
      "> \n",
      "> ##Load inversal data\n",
      "> #Unique atoms file\n",
      "> terminalAtomsFile = as.matrix(read.table(\"../../terminalAtoms\",sep=\"\\t\"))\n",
      "> terminalAtoms = {}\n",
      "> for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+     terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\", \"))\n",
      "+ }\n",
      "> \n",
      "> #Secondary structure file\n",
      "> secStructure = as.matrix(read.table(args[3]))[,1:2]\n",
      "> rownames(secStructure) = secStructure[,1]\n",
      "> \n",
      "> #RSA\n",
      "> rsa = as.matrix(read.table(paste(keyword,\".rsa\",sep=\"\")))\n",
      "> \n",
      "> ##Define functions\n",
      "> \n",
      "> #Add terminal atom details\n",
      "> addTermDetails = function(data) {\n",
      "+     out_details = c()\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+         if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+             code1 = 1\n",
      "+         } else {\n",
      "+             code1 = 0\n",
      "+         }\n",
      "+         if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+             code2 = 1\n",
      "+         } else {\n",
      "+             code2 = 0\n",
      "+         }\n",
      "+     out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+     }\n",
      "+     return(out_details)\n",
      "+ }\n",
      "> \n",
      "> collapse = function(data) {\n",
      "+     if (nrow(data)==0) {\n",
      "+         out = matrix(ncol = 8,nrow=0)\n",
      "+         colnames(out) = c(\"AA1\", \"AA2\", \"Type\", \"SumEdges\", \"percentMC\", \"influence1\", \"influence2\",\"degree\")\n",
      "+         results = c(); results$out = out\n",
      "+         return(results)\n",
      "+         break\n",
      "+     }\n",
      "+     #Edges should be listed bi-directionally\n",
      "+     x=unlist(lapply(data[,3],function(x){if(x==\"MCSC\"){return(\"SCMC\")}else if(x==\"SCMC\"){return(\"MCSC\")} else {return(x)}}))\n",
      "+     if (dim(data)[2]==7) {\n",
      "+         data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6]))\n",
      "+     } else {\n",
      "+         data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6],data[,9],data[,8]))\n",
      "+     }\n",
      "+     data = unique(data)\n",
      "+     out = c()\n",
      "+     out_noPP = c()\n",
      "+     completed = c()\n",
      "+     if (dim(data)[2]==7) {\n",
      "+         out_details = c()\n",
      "+         for (i in 1:nrow(data)) {\n",
      "+             acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+             acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+             if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+                 code1 = 1\n",
      "+             } else {\n",
      "+                 code1 = 0\n",
      "+             }\n",
      "+             if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+                 code2 = 1\n",
      "+             } else {\n",
      "+                 code2 = 0\n",
      "+             }\n",
      "+             out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+         }\n",
      "+         data = out_details\n",
      "+     } else {\n",
      "+         out_details = data\n",
      "+     }\n",
      "+     colnames(out_details) = c(\"AA1\",\"AA2\",\"Type1\",\"Weight\",\"Type2\",\"Atom1\",\"Atom2\",\"influence1\",\"influence2\")\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         if (i %in% completed == FALSE) {\n",
      "+             x=which((data[,1]==data[i,1] & data[,2] == data[i,2]))\n",
      "+             subdata = matrix(data[x,],ncol=9)\n",
      "+             edgeSum = 0\n",
      "+             MCMCcount = 0\n",
      "+             MCSCcount = 0\n",
      "+             SCMCcount = 0\n",
      "+             SCSCcount = 0\n",
      "+             influence1 = sum(as.numeric(subdata[,8]))\n",
      "+             influence2 = sum(as.numeric(subdata[,9]))\n",
      "+             for (j in 1:nrow(subdata)) {\n",
      "+                 edgeSum = edgeSum + as.numeric(subdata[j,4])\n",
      "+                 if (subdata[j,3]==\"MCMC\" | subdata[j,3]==\"MCSC\") {\n",
      "+                     MCMCcount = MCMCcount+1\n",
      "+                     MCSCcount = MCSCcount+1\n",
      "+                 } else if (subdata[j,3]==\"SCMC\") {\n",
      "+                     SCMCcount = SCMCcount+1\n",
      "+                 } else if (subdata[j,3]==\"SCSC\") {\n",
      "+                     SCSCcount = SCSCcount+1\n",
      "+                 }\n",
      "+             }\n",
      "+             #Header is AA1, AA2, mixed, sum of edges, percent MC, influence1, influence2, degreeAA1\n",
      "+             out = rbind(out,c(data[i,1],data[i,2],\"mixed\",edgeSum,(MCMCcount+MCSCcount)/(MCMCcount+MCSCcount+SCMCcount+SCSCcount),influence1,influence2,sum(data[,1]==data[i,1])))\n",
      "+             if (any(subdata[,5]==\"PP\")==FALSE) {out_noPP = rbind(out_noPP,c(data[i,1],data[i,2],\"mixed\",edgeSum,(MCMCcount+MCSCcount)/(MCMCcount+MCSCcount+SCMCcount+SCSCcount),influence1,influence2,sum(data[,1]==data[i,1])))}\n",
      "+             completed = c(completed,x)\n",
      "+         }\n",
      "+     }\n",
      "+     out = out[out[,4]>0,]\n",
      "+     out_noPP = out_noPP[out_noPP[,4]>0,]\n",
      "+     colnames(out) = c(\"AA1\", \"AA2\", \"Type\", \"SumEdges\", \"percentMC\", \"influence1\", \"influence2\",\"degree\")\n",
      "+     results = c()\n",
      "+     results$out = out\n",
      "+     results$out_noPP = out_noPP\n",
      "+     results$influence_all = out_details[as.numeric(out_details[,4])>0 & (out_details[,8]==\"1\" | out_details[,9]==\"1\"),]\n",
      "+     out_details = out_details[as.numeric(out_details[,4])>0,]\n",
      "+     out_subtract = out_details; out_subtract[out_subtract[,8]==0 & out_subtract[,9]==0,4] = (-1)*as.numeric(out_subtract[out_subtract[,8]==0 & out_subtract[,9]==0,4])\n",
      "+     out_subtract = out_subtract[out_subtract[,5]!=\"PP\",]\n",
      "+     results$out_subtract = out_subtract\n",
      "+     return(results)\n",
      "+ }\n",
      "> \n",
      "> removeRedundancy = function(dataTmp) {\n",
      "+     for (i in 1:nrow(dataTmp)) {\n",
      "+         if (order(dataTmp[i,1:2])[1]==2) {\n",
      "+             dataTmp[i,1:2] = c(dataTmp[i,2],dataTmp[i,1])\n",
      "+             dataTmp[i,3] = paste(unlist(strsplit(dataTmp[i,3],split=\"\"))[c(3,4,1,2)],collapse=\"\")\n",
      "+             dataTmp[i,c(6,7)] = c(dataTmp[i,7],dataTmp[i,6])\n",
      "+         }\n",
      "+     }\n",
      "+     dataTmp = unique(dataTmp)\n",
      "+     pairs = unique(dataTmp[,1:2])\n",
      "+     out = c()\n",
      "+     for (j in 1:nrow(pairs)) {\n",
      "+         pair = pairs[j,]\n",
      "+         subData = matrix(dataTmp[dataTmp[,1]==pair[1] & dataTmp[,2]==pair[2],],ncol=7)\n",
      "+         out = rbind(out, c(pair[1],pair[2],\"mixed\",0.5*sum(as.numeric(subData[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> collapse_directed = function(data) {\n",
      "+     pairs = unique(data[,1:2])\n",
      "+     out = c()\n",
      "+     for (i in 1:nrow(pairs)) {\n",
      "+         subdat = matrix(data[data[,1]==pairs[i,1]&data[,2]==pairs[i,2],],ncol=4)\n",
      "+         out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> collapse_agnostic = function(data) {\n",
      "+     pairs = unique(t(apply(data[,1:2],1,function(x){return(sort(x))})))\n",
      "+     out = c()\n",
      "+     for (i in 1:nrow(pairs)) {\n",
      "+         subdat = matrix(data[(data[,1]==pairs[i,1]&data[,2]==pairs[i,2]) | (data[,2]==pairs[i,1]&data[,1]==pairs[i,2]),],ncol=4)\n",
      "+         out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> distributCalc = function(set,nodes,net) {\n",
      "+     #x is assumed to be a set but can also just be 1 node\n",
      "+     distanceNet = distances(net,weights=net$weight)\n",
      "+     distanceTmp = c()\n",
      "+     Dr = c()\n",
      "+     for (node in nodes) {\n",
      "+         distanceTmp = c()\n",
      "+         for (item in set) {\n",
      "+             if (item!=node) {\n",
      "+                 if(item%in%rownames(distanceNet)==FALSE){distanceTmp=100;next}\n",
      "+                 if(node%in%rownames(distanceNet)==FALSE){distanceTmp=100;next}\n",
      "+                 distanceTmp = c(distanceTmp,distanceNet[item,node])\n",
      "+             }\n",
      "+         }\n",
      "+         if (length(distanceTmp)>0) {\n",
      "+             Dr = c(Dr, 1/min(distanceTmp))\n",
      "+         }\n",
      "+     }\n",
      "+     return(sum(Dr)/length(nodes))\n",
      "+ }\n",
      "> \n",
      "> ##Read in user data\n",
      "> data_all = as.matrix(read.table(args[1]))\n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> originalNodes = nodes\n",
      "> basedata = collapse(data_all)\n",
      "> basedata_noPP = removeRedundancy(basedata$out[,1:7])\n",
      "> basedata_allsidechain = collapse_agnostic(data_all[data_all[,3]!=\"MCMC\",c(1,2,3,4)])\n",
      "> influencedata_all = basedata$influence_all\n",
      "> influencedata_collapse = collapse(influencedata_all)\n",
      "> influencedata = removeRedundancy(influencedata_collapse$out[,1:7])\n",
      "> influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "> influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "> \n",
      "> influencedata_directed = collapse_directed(influencedata_all[influencedata_all[,8]==1,c(1,2,3,4)])\n",
      "> influencenet_directed = graph.edgelist(influencedata_directed[,c(2,1)],directed=TRUE)\n",
      "> influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "> \n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> \n",
      "> #Write out edgelist files\n",
      "> write.table(influencedata,file=paste(keyword,\"_allEdges\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=FALSE,row.names=FALSE)\n",
      "> write.table(influencedata_directed,file=paste(keyword,\"_uniqueEdges\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=FALSE,row.names=FALSE)\n",
      "> \n",
      "> ###Deal with missing nodes\n",
      "> if (any(nodes %in% secStructure[,1]==FALSE)) {\n",
      "+     for (node in nodes) {\n",
      "+         if (node %in% secStructure[,1]==FALSE) {\n",
      "+             secStructure = rbind(secStructure,c(node,\"xxx\",360))\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> for (node in nodes) {\n",
      "+     if (node %in% rownames(rsa) == FALSE) {\n",
      "+         rsa = rbind(rsa,c(0,0))\n",
      "+         rownames(rsa)[nrow(rsa)] = node\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> ##Begin centrality calculations\n",
      "> \n",
      "> #Modular calculations\n",
      "> \n",
      "> ##WALKTRAP\n",
      "> net = graph.edgelist(basedata_noPP[,1:2],directed=FALSE)\n",
      "> net$weight = 1/as.numeric(basedata_noPP[,4])\n",
      "> net_community=walktrap.community(net,weights=net$weight); net_community_vec = net_community$membership; names(net_community_vec)=net_community$names\n",
      "> net_community_vec_wt = net_community_vec\n",
      "> nodes = V(net)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> nodes.color = c()\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(influencedata)) {\n",
      "+     j = which((basedata_noPP[,1]==influencedata[i,1]&basedata_noPP[,2]==influencedata[i,2])|(basedata_noPP[,1]==influencedata[i,2]&basedata_noPP[,2]==influencedata[i,1]))\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) { edgecolors[j]=\"grey40\" }\n",
      "+ }\n",
      "> if(length(nodes)>200) {textsize=.25} else {textsize=.45}\n",
      "> plot(net,edge.color = edgecolors, vertex.color = nodes.color,vertex.size=4,vertex.label.cex=textsize,main=\"WALKTRAP\")\n",
      ">     \n",
      "> ##Weighted edge betweenness WALKTRAP\n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP - WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for WALKTRAP\n",
      "> firstOrderDegree = degree(influencenet)\n",
      "> firstOrderDegree_sidechain = degree(influencenet_directed,mode=c(\"in\"))\n",
      "> secondOrderDegree = c(); secondOrderDegree_sidechain=c()\n",
      "> nodes = V(influencenet)$name\n",
      "> for (node in nodes) {\n",
      "+     firstorder = neighbors(influencenet,node)\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet,neighbor)))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree = c(secondOrderDegree,length(secondorder))\n",
      "+     firstorder = neighbors(influencenet_directed,node,mode=c(\"in\"))\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet_directed,neighbor,mode=c(\"in\"))))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree_sidechain = c(secondOrderDegree_sidechain,length(secondorder))\n",
      "+ }\n",
      "> names(secondOrderDegree) = nodes; names(secondOrderDegree_sidechain) = nodes\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])])) - 1\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])])) - 1\n",
      "+ }\n",
      "> \n",
      "> #SAVE ALL AS WALKTRAP\n",
      "> node_edge_betweenness_wt = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_wt = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_wt = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_wt = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_wt = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_wt = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> ##2ARY STRUCTURE\n",
      "> net_community_vec = as.numeric(as.factor(secStructure[,2]))\n",
      "> names(net_community_vec) = secStructure[,1]\n",
      "> nodes = V(net)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> nodes.color = c()\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(basedata_noPP)) {\n",
      "+     if (net_community_vec[basedata_noPP[i,1]]!=net_community_vec[basedata_noPP[i,2]]) { edgecolors[i]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for 2ARY STRUCTURE\n",
      "> nodes = V(influencenet)$name\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])])) - 1\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])])) - 1\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #SAVE ALL AS STRIDE\n",
      "> node_edge_betweenness_stride = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_stride = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_stride = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_stride = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_stride = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_stride = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> \n",
      "> ### Create final dataset\n",
      "> \n",
      "> out = cbind(\n",
      "+ firstOrderDegree[nodes],\n",
      "+ firstOrderDegree_sidechain[nodes],\n",
      "+ secondOrderDegree[nodes],\n",
      "+ secondOrderDegree_sidechain[nodes],\n",
      "+ node_edge_betweenness_stride[nodes],\n",
      "+ node_edge_betweenness_sidechain_stride[nodes],\n",
      "+ node_intermodular_degree_stride[nodes],\n",
      "+ node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ node_edge_betweenness_wt[nodes],\n",
      "+ node_edge_betweenness_sidechain_wt[nodes],\n",
      "+ node_intermodular_degree_wt[nodes],\n",
      "+ node_intermodular_degree_sidechain_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_wt[nodes]\n",
      "+ )\n",
      "> rownames(out) = nodes\n",
      "> colnames(out) = c(\"Degree\",\"Degree_sidechain\",\"SecondOrderDegree\",\"SecondOrderDegree_sidechain\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_sidechain\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_sidechain\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_sidechain\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_sidechain\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_sidechain\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_sidechain\")\n",
      "> \n",
      "> #add back in nodes that were not networked as zeros\n",
      "> zeroMat = matrix(0,nrow=length(which(originalNodes%in%nodes==FALSE)),ncol=ncol(out));rownames(zeroMat)=originalNodes[which(originalNodes%in%nodes==FALSE)]\n",
      "> out = rbind(out,zeroMat)\n",
      "> \n",
      "> #Add 1 to everything to avoid zeros, except ligand\n",
      "> out[is.na(out)]=0\n",
      "> out = out+1\n",
      "> \n",
      "> #Write out\n",
      "> write.table(out,file=paste(keyword,\"_scoresEnergetics\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "> #Standard normalization\n",
      "> outZ = apply(out,2,function(x){return(scale(x))})\n",
      "> rownames(outZ) = rownames(out)\n",
      "> write.table(outZ,file=paste(keyword,\"_scoresEnergeticsZ\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "Running centroid R script\n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(igraph)\n",
      "> library(stats)\n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> ##read in data\n",
      "> ## arg1: col1 and col2 are edges, col3 is weight of edge\n",
      "> ## arg2: weighted, unweighted <-- no longer using\n",
      "> ## arg3: forced module definition <-- no longer using\n",
      "> \n",
      "> # Set desired parameters\n",
      "> removeMCMC = FALSE\n",
      "> linearcutoff = 1\n",
      "> directed = FALSE\n",
      "> secondaryStructure2 = FALSE\n",
      "> sidechainMode = TRUE ##build network of just SCSC, SCMC and MCSC. Still use unique atoms for the _uniqueAtoms directed plots\n",
      "> uniqueAtomsMode = FALSE ##build network of just unique atoms (each edge has to have at least 1). Still use unique atoms for the _uniqueAtoms directed plots\n",
      "> useDNA=TRUE\n",
      "> uniqueAtomsUnbiased=FALSE\n",
      "> uniqueAtomsGaurav=FALSE\n",
      "> uniqueAtomsOLD=FALSE\n",
      "> atomCorrection = FALSE\n",
      "> uniqueAtomsGauravPP = FALSE\n",
      "> weighted = TRUE\n",
      "> removeWaters=TRUE\n",
      "> if (tail(unlist(strsplit(getwd(),\"/\")),n=1)==\"Centroid\") {\n",
      "+     ligandCentroidMode = TRUE\n",
      "+ } else {\n",
      "+     ligandCentroidMode = FALSE\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> # Read in ligand file\n",
      "> ligandfile = scan(paste(args[1],\"_ligand\",sep=\"\"),what=\"character\")\n",
      "> if (length(ligandfile)==0){\n",
      "+     ligandmode = FALSE\n",
      "+ } else {\n",
      "+     ligandmode = TRUE\n",
      "+ }\n",
      "> \n",
      "> ligandCentroidFile = scan(paste(args[1],\"NetLigand\",sep=\"\"),what=\"character\")\n",
      "> \n",
      "> # keyword is the base filename\n",
      "> # method is \"frag\" or \"module\" or \"both\"\n",
      "> keyword = args[1]\n",
      "> args=c(paste(args[1],\"_net\",sep=\"\"),\"weighted\",dir(\"./\",\"_secondaryStructure$\"))\n",
      "> \n",
      "> ##Load data\n",
      "> #Unique atoms file\n",
      "> if(uniqueAtomsUnbiased) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"/Users/vjpatel/Dropbox/Gaurav/uniqueAtomsUnbiased\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else if (uniqueAtomsGaurav) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"/Users/vjpatel/Dropbox/Gaurav/uniqueAtomsGaurav\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else if (uniqueAtomsOLD) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"../../../terminalAtoms\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"../../../uniqueAtoms\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     otherTA = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+         #otherTA[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,4],split=\",\"))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> #Secondary structure file\n",
      "> #Ligands are all one secondary structure\n",
      "> secStructure = as.matrix(read.table(args[3]))\n",
      "> phiAngleDB = {}\n",
      "> for (i in 1:nrow(secStructure)) {\n",
      "+     phiAngleDB[[secStructure[i,1]]] = as.numeric(secStructure[i,3])\n",
      "+ }\n",
      "> \n",
      "> if(ligandmode) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",\"_ligand\"),sep=\"\\t\"))\n",
      "+     liganddat = unique(liganddat[,2])\n",
      "+     for (ligand in liganddat) {\n",
      "+         secStructure = rbind(secStructure,c(sub(\"-\",\"\",ligand),\"ligand\",360))\n",
      "+         phiAngleDB[[sub(\"-\",\"\",ligand)]]=0\n",
      "+     }\n",
      "+ } else {\n",
      "+     liganddat = c()\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> ##Define functions\n",
      "> \n",
      "> ##Collapse matrix functions\n",
      "> ##Will return redundant matrix\n",
      "> ##Input needs to be 7 columns\n",
      "> ##Second input is different if the main input is a subset of the larger data\n",
      "> addTerminalDetails = function(data,largerdata) {\n",
      "+     #List edges bi-directionally\n",
      "+     x=unlist(lapply(data[,3],function(x){if(x==\"MCSC\"){return(\"SCMC\")}else if(x==\"SCMC\"){return(\"MCSC\")} else {return(x)}}))\n",
      "+     data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6]))\n",
      "+     data = unique(data)\n",
      "+     out = c()\n",
      "+     out_noPP = c()\n",
      "+     out_details = c()\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+         subdata = matrix(data[data[,1]==data[i,1]&data[,2]==data[i,2],],ncol=7)\n",
      "+         allsubdata = matrix(largerdata[largerdata[,1]==data[i,1],],ncol=7)\n",
      "+         code1=0;code2=0\n",
      "+         if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+             code1 = 1\n",
      "+         }\n",
      "+         if (data[i,5] == \"DNA\" & useDNA==TRUE) {\n",
      "+             code1 = 1\n",
      "+         }\n",
      "+         allsubdata = matrix(largerdata[largerdata[,2]==data[i,2],],ncol=7)\n",
      "+         if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+             code2 = 1\n",
      "+         }\n",
      "+         if (data[i,5] == \"DNA\" & useDNA==TRUE) {\n",
      "+             code2 = 1\n",
      "+         }\n",
      "+         if (data[i,5]==\"PICATION\"|data[i,5]==\"PIPI\") {\n",
      "+             code1=1\n",
      "+             code2=1\n",
      "+         }\n",
      "+         if (data[i,5]==\"PP\") {\n",
      "+             if (uniqueAtomsGaurav) {\n",
      "+                 if((acid1) == \"GLY\") {\n",
      "+                     code1 = 1\n",
      "+                     data[i,3]==\"SCSC\"\n",
      "+                 }\n",
      "+                 if ((acid2) == \"GLY\") {\n",
      "+                     code2 = 1\n",
      "+                     data[i,3] == \"SCSC\"\n",
      "+                 }\n",
      "+             }\n",
      "+         }\n",
      "+         out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+     }\n",
      "+     return(out_details)\n",
      "+ }\n",
      "> \n",
      "> ##Will return non-redundant matrix\n",
      "> collapse_agnostic = function(data) {\n",
      "+     if(nrow(data)==0) {\n",
      "+         return(matrix(ncol=4,nrow=0))\n",
      "+     } else if (ncol(data)==4 & data[1,3]==\"mixed\") {\n",
      "+         return(data)\n",
      "+     } else {\n",
      "+         pairs = unique(t(apply(matrix(data[,1:2],ncol=2),1,function(x){return(sort(x))})))\n",
      "+         out = c()\n",
      "+         for (i in 1:nrow(pairs)) {\n",
      "+             subdat = matrix(data[(data[,1]==pairs[i,1]&data[,2]==pairs[i,2]) | (data[,2]==pairs[i,1]&data[,1]==pairs[i,2]),1:4],ncol=4)\n",
      "+             subdat = unique(t(apply(subdat,1,function(x){a=which(x==sort(x[1:2])[1]);if(a==1){return(x)}else{return(c(x[2],x[1],paste(unlist(strsplit(x[3],\"\"))[c(3,4,1,2)],collapse=\"\"),x[4]))}})))\n",
      "+             if (weighted==FALSE) {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",1))\n",
      "+             } else {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+             }\n",
      "+         }\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> ##Will return directed collapsed matrix\n",
      "> collapse_directed = function(data) {\n",
      "+     if (length(data)==0) {\n",
      "+         return(matrix(nrow=0,ncol=4))\n",
      "+     } else {\n",
      "+         pairs = matrix(unique(data[,1:2]),ncol=2)\n",
      "+         out = c()\n",
      "+         for (i in 1:nrow(pairs)) {\n",
      "+             subdat = matrix(data[data[,1]==pairs[i,1]&data[,2]==pairs[i,2],1:4],ncol=4)\n",
      "+             if (weighted==FALSE) {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",1))\n",
      "+             } else {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+             }\n",
      "+         }\n",
      "+         return(matrix(out,ncol=4))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> rsa = as.matrix(read.table(dir(\"./\",\".rsa$\")))\n",
      "> rownames(rsa) = rsa[,1]\n",
      "> rsa[,1] = rsa[,2]\n",
      "> \n",
      "> #add ligand as -1\n",
      "> for (ligand in liganddat) {\n",
      "+     rsa = rbind(rsa,c(-1,-1))\n",
      "+     rownames(rsa)[nrow(rsa)] = sub(\"-\",\"\",ligand)\n",
      "+ }\n",
      "> \n",
      "> #Function to get amino acid\n",
      "> getAcid = function(residue) {\n",
      "+     if(all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"A\")) == TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"G\"))==TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"C\"))==TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"T\"))==TRUE)) {return(paste(unlist(strsplit(residue,\"\"))[1:2],collapse=\"\"))\n",
      "+     } else {\n",
      "+     return(paste(unlist(strsplit(residue,\"\"))[1:3],collapse=\"\"))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> ##################\n",
      "> ## read in data ##\n",
      "> ##################\n",
      "> data_all = as.matrix(read.table(args[1]))[,1:7]\n",
      "> if(removeWaters) {\n",
      "+     y=c(grep(\"HOH\",data_all[,1]),grep(\"HOH\",data_all[,2]))\n",
      "+     if(length(y)>0) {\n",
      "+         data_all = data_all[-y,]\n",
      "+     }\n",
      "+ }\n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> originalNodes = nodes\n",
      "> \n",
      "> data_all_original = rbind(data_all,data_all[,c(2,1,3,4,5,7,6)])\n",
      "> data_all = data_all[as.numeric(data_all[,4])>0,]\n",
      "> \n",
      "> basedata = addTerminalDetails(data_all,data_all_original)\n",
      "> basedata_noPP = collapse_agnostic(basedata[basedata[,5]!=\"PP\",])\n",
      "> sidechaindata = basedata[basedata[,3]!=\"MCMC\",1:4]\n",
      "> sidechaindata_detailed = basedata[basedata[,3]!=\"MCMC\",]\n",
      "> \n",
      "> influencedata_detailed = basedata\n",
      "> if (uniqueAtomsMode) {\n",
      "+     influencedata_detailed = basedata[basedata[,8]==1 | basedata[,9]==1,]\n",
      "+ }\n",
      "> influencedata = collapse_agnostic(influencedata_detailed)\n",
      "> influencedataGLY = rbind(influencedata[grep(\"GLY\",influencedata[,1]),],influencedata[grep(\"GLY\",influencedata[,1]),])\n",
      "> influencedata = influencedata[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,2])))>linearcutoff,]\n",
      "> \n",
      "> \n",
      "> influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "> influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "> \n",
      "> influencedata_directed = collapse_directed(matrix(influencedata_detailed[influencedata_detailed[,8]==1,1:4],ncol=4))\n",
      "> influencedata_directedGLY = rbind(influencedata_directed[grep(\"GLY\",influencedata_directed[,1]),],influencedata_directed[grep(\"GLY\",influencedata_directed[,2]),])\n",
      "> influencedata_directed = influencedata_directed[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,2])))>linearcutoff,]\n",
      "> \n",
      "> influencedata_directed = matrix(influencedata_directed,ncol=4)\n",
      "> influencenet_directed = graph.edgelist(matrix(influencedata_directed[,c(2,1)],ncol=2),directed=TRUE)\n",
      "> influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "> \n",
      "> \n",
      "> if(sidechainMode) {\n",
      "+     influencedata_detailed = sidechaindata_detailed\n",
      "+     influencedata = collapse_agnostic(influencedata_detailed)\n",
      "+     influencedataGLY = rbind(influencedata[grep(\"GLY\",influencedata[,1]),],influencedata[grep(\"GLY\",influencedata[,2]),])\n",
      "+     influencedata = influencedata[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,2])))>linearcutoff,]\n",
      "+     #Add back for GLY if uniqueAtomsGaurav\n",
      "+     if(uniqueAtomsGauravPP) {\n",
      "+         influencedata = rbind(influencedata,influencedataGLY)\n",
      "+     }\n",
      "+     influencedata = influencedata[as.numeric(influencedata[,4])>0,]\n",
      "+     \n",
      "+     influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "+     influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "+     \n",
      "+     influencedata_directed = collapse_directed(matrix(influencedata_detailed[influencedata_detailed[,8]==1,1:4],ncol=4))\n",
      "+     influencedata_directedGLY = rbind(influencedata_directed[grep(\"GLY\",influencedata_directed[,1]),],influencedata_directed[grep(\"GLY\",influencedata_directed[,2]),])\n",
      "+ \n",
      "+     influencedata_directed = influencedata_directed[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,2])))>linearcutoff,]\n",
      "+     #Add back for GLY if uniqueAtomsGaurav\n",
      "+     if(uniqueAtomsGauravPP) {\n",
      "+         influencedata_directed = rbind(influencedata_directed,influencedata_directedGLY)\n",
      "+     }\n",
      "+     influencedata_directed = matrix(influencedata_directed,ncol=4)\n",
      "+     influencenet_directed = graph.edgelist(matrix(influencedata_directed[,c(2,1)],ncol=2),directed=TRUE)\n",
      "+     influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "+     \n",
      "+ }\n",
      "> \n",
      "> #Replace missing nodes in the igraph net\n",
      "> nodesMissing = nodes[nodes %in% influencedata[,1]==FALSE & nodes %in% influencedata[,2]==FALSE]\n",
      "> \n",
      "> ##B-factor\n",
      "> bfactor = as.matrix(read.table(dir(\"./\",\"_Bfactor\"),row.names=1))\n",
      "> \n",
      "> #Replace missing nodes in secStructure matrix\n",
      "> if (any(nodes %in% secStructure[,1]==FALSE)) {\n",
      "+     for (node in nodes) {\n",
      "+         if (node %in% secStructure[,1]==FALSE) {\n",
      "+             secStructure = rbind(secStructure,c(node,\"xxx\",360))\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> for (node in nodes) {\n",
      "+     if (node %in% rownames(rsa) == FALSE) {\n",
      "+         rsa = rbind(rsa,c(0,0))\n",
      "+         rownames(rsa)[nrow(rsa)] = node\n",
      "+     }\n",
      "+     if (node %in% rownames(bfactor) == FALSE) {\n",
      "+         bfactor = rbind(bfactor,1)\n",
      "+         rownames(bfactor)[nrow(bfactor)] = node\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> \n",
      "> \n",
      "> #Give DNA attributes\n",
      "> if (\"DNA\" %in% data_all[,5]) {\n",
      "+     for (i in 1:nrow(data_all)) {\n",
      "+         if (data_all[i,5]==\"DNA\") {\n",
      "+             phiAngleDB[[data_all[i,1]]] = 90\n",
      "+             if (data_all[i,1] %in% rownames(rsa) == FALSE) {\n",
      "+                 rsa = rbind(rsa,c(0,0))\n",
      "+                 rownames(rsa)[nrow(rsa)] = data_all[i,1]\n",
      "+             }\n",
      "+             if (data_all[i,1] %in% rownames(secStructure) == FALSE) {\n",
      "+                 secStructure = rbind(secStructure,c(data_all[i,1],\"DNA\",360))\n",
      "+             }\n",
      "+             if (data_all[i,1] %in% rownames(bfactor) == FALSE) {\n",
      "+                 bfactor = rbind(bfactor,1)\n",
      "+                 rownames(bfactor)[nrow(bfactor)] = data_all[i,1]\n",
      "+             }\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #Betweenness\n",
      "> node_betweenness = betweenness(influencenet,weights=influencenet$weight,directed=directed)\n",
      "> node_betweenness_unique = betweenness(influencenet_directed,weights=influencenet_directed$weight,directed=directed)\n",
      "> \n",
      "> #Modular calculations\n",
      "> \n",
      "> ##WALKTRAP\n",
      "> net = graph.edgelist(basedata_noPP[,1:2],directed=FALSE)\n",
      "> net$weight = 1/as.numeric(basedata_noPP[,4])\n",
      "> net_community=walktrap.community(net,weights=net$weight); net_community_vec = net_community$membership; names(net_community_vec)=net_community$names\n",
      "> net_community_vec_wt = net_community_vec\n",
      "> #net_community_vec[buriednodes] = max(net_community_vec)+1\n",
      "> nodes = V(influencenet)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> nodes.color = c()\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(influencedata)) {\n",
      "+     j = which((basedata_noPP[,1]==influencedata[i,1]&basedata_noPP[,2]==influencedata[i,2])|(basedata_noPP[,1]==influencedata[i,2]&basedata_noPP[,2]==influencedata[i,1]))\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) { edgecolors[j]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP\n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP - WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         if(weight1==0 & weight2==0) {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]] + 0\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]] + 0\n",
      "+         } else {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for WALKTRAP\n",
      "> firstOrderDegree = degree(influencenet)\n",
      "> firstOrderDegree_sidechain = degree(influencenet_directed,mode=c(\"in\"))\n",
      "> \n",
      "> secondOrderDegree = c(); secondOrderDegree_sidechain=c()\n",
      "> nodes = V(influencenet)$name\n",
      "> for (node in nodes) {\n",
      "+     firstorder = neighbors(influencenet,node)\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet,neighbor)))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree = c(secondOrderDegree,length(secondorder))\n",
      "+     if(node %in% V(influencenet_directed)$name) {\n",
      "+         firstorder_sidechain = neighbors(influencenet_directed,node,mode=c(\"in\"))\n",
      "+     } else {\n",
      "+         firstorder_sidechain = c()\n",
      "+     }\n",
      "+     secondorder = c(); for (neighbor in firstorder_sidechain){secondorder = c(secondorder,names(neighbors(influencenet_directed,neighbor,mode=c(\"in\"))))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree_sidechain = c(secondOrderDegree_sidechain,length(secondorder))\n",
      "+ }\n",
      "> names(secondOrderDegree) = nodes; names(secondOrderDegree_sidechain) = nodes\n",
      "> \n",
      "> \n",
      "> #second order intermodular degree\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])]))\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])]))\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #SAVE ALL AS WALKTRAP\n",
      "> node_edge_betweenness_wt = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_wt = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_wt = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_wt = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_wt = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_wt = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> \n",
      "> ##2ARY STRUCTURE\n",
      "> secStructure = secStructure[,1:2]\n",
      "> rownames(secStructure) = secStructure[,1]\n",
      "> net_community_vec = as.numeric(as.factor(secStructure[,2]))\n",
      "> names(net_community_vec) = secStructure[,1]\n",
      "> #net_community_vec[buriednodes] = max(net_community_vec)+1\n",
      "> nodes = V(influencenet)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> nodes.color = c()\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(basedata_noPP)) {\n",
      "+     if (net_community_vec[basedata_noPP[i,1]]!=net_community_vec[basedata_noPP[i,2]]) { edgecolors[i]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         if(weight1==0 & weight2==0) {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]] + 0\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]] + 0\n",
      "+         } else {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> ##Degree and second order degree for 2ARY STRUCTURE\n",
      "> \n",
      "> #second order intermodular degree\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     #secondOrder_node_intermodular_degree[node] = length(unique(unlist(node_modules[names(node_modules[[node]])])))\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])]))\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])]))\n",
      "+ }\n",
      "> \n",
      "> #SAVE ALL AS STRIDE\n",
      "> node_edge_betweenness_stride = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_stride = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_stride = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_stride = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_stride = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_stride = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> ### Create final dataset\n",
      "> nodes = nodes[nodes!=\"DNA1000A\"]\n",
      "> nodes = nodes[nodes %in% c(\"DA\",\"DG\",\"DC\",\"DT\") == FALSE]\n",
      "> \n",
      "> ### Remove ligand amino acids from final dataset\n",
      "> ligands = sub(\"-\",\"\",liganddat)\n",
      "> nodes = nodes[nodes %in% ligands == FALSE]\n",
      "> \n",
      "> nodes[which(nodes %in% rownames(bfactor)==FALSE)]\n",
      "character(0)\n",
      "> \n",
      "> out = cbind(\n",
      "+ rsa[nodes,1],\n",
      "+ firstOrderDegree[nodes],\n",
      "+ firstOrderDegree_sidechain[nodes],\n",
      "+ secondOrderDegree[nodes],\n",
      "+ secondOrderDegree_sidechain[nodes],\n",
      "+ node_edge_betweenness_stride[nodes],\n",
      "+ node_edge_betweenness_sidechain_stride[nodes],\n",
      "+ node_intermodular_degree_stride[nodes],\n",
      "+ node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ node_edge_betweenness_wt[nodes],\n",
      "+ node_edge_betweenness_sidechain_wt[nodes],\n",
      "+ node_intermodular_degree_wt[nodes],\n",
      "+ node_intermodular_degree_sidechain_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_wt[nodes]\n",
      "+ )\n",
      "> \n",
      "> #write column names\n",
      "> colnames(out) = c(\"RSA\",\"Degree\",\"Degree_uniqueAtoms\",\"SecondOrderDegree\",\"SecondOrderDegree_uniqueAtoms\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_unqiueAtoms\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_uniqueAtoms\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_uniqueAtoms\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_uniqueAtoms\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_uniqueAtoms\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_uniqueAtoms\")\n",
      "> \n",
      "> \n",
      "> ## LIGANDS\n",
      "> if(ligandmode==TRUE & ligandCentroidMode==FALSE) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",\"_ligand\"),sep=\"\\t\"))\n",
      "+     \n",
      "+     tmp = c()\n",
      "+     i=1\n",
      "+     for (i in 1:nrow(liganddat)) {\n",
      "+         acid=paste(unlist(strsplit(liganddat[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         if (unlist(strsplit(liganddat[i,1],\"-\"))[2] %in% terminalAtoms[[acid]]) {\n",
      "+             tmp = rbind(tmp,c(unlist(strsplit(liganddat[i,1],\"-\"))[1],liganddat[i,2],liganddat[i,3]))\n",
      "+         }\n",
      "+     }\n",
      "+     \n",
      "+     #ligandvec = tapply(liganddat[,3],as.factor(liganddat[,1]),min)\n",
      "+     ligandvec = tapply(as.numeric(tmp[,3]),as.factor(tmp[,1]),min)\n",
      "+ }\n",
      "> \n",
      "> if(ligandmode==TRUE & ligandCentroidMode) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",paste(\"_centroidNetLigand$\",sep=\"\")),sep=\"\\t\"))\n",
      "+     \n",
      "+     tmp = c()\n",
      "+     i=1\n",
      "+     for (i in 1:nrow(liganddat)) {\n",
      "+         tmp = rbind(tmp,c(liganddat[i,1],liganddat[i,2],liganddat[i,3]))\n",
      "+     }\n",
      "+     \n",
      "+     ligandvec = tapply(tmp[,3],as.factor(tmp[,1]),min)\n",
      "+ }\n",
      "> \n",
      "> if (ligandmode==FALSE) {\n",
      "+     ligandvec = rep(0,length(nodes))\n",
      "+ }\n",
      "> \n",
      "> out = cbind(out,ligandvec[rownames(out)])\n",
      "> colnames(out) = c(\"RSA\",\"Degree\",\"Degree_uniqueAtoms\",\"SecondOrderDegree\",\"SecondOrderDegree_uniqueAtoms\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_unqiueAtoms\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_uniqueAtoms\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_uniqueAtoms\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_uniqueAtoms\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_uniqueAtoms\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_uniqueAtoms\",\"Ligand\")\n",
      "> \n",
      "> #add back in nodes that were not networked as zeros\n",
      "> zeroMat = matrix(0,nrow=length(which(originalNodes%in%nodes==FALSE)),ncol=ncol(out));rownames(zeroMat)=originalNodes[which(originalNodes%in%nodes==FALSE)]\n",
      "> out = rbind(out,zeroMat)\n",
      "> \n",
      "> #fix RSA and Bfactor and ligand\n",
      "> out[,\"RSA\"] = rsa[rownames(out),1]\n",
      "> out[,\"Ligand\"] = ligandvec[rownames(out)]\n",
      "> \n",
      "> #Add 1 to everything to avoid zeros, except ligand\n",
      "> out[is.na(out)]=0\n",
      "> out[out[,\"Ligand\"]==0,\"Ligand\"] = 150\n",
      "> out[1:nrow(out),2:ncol(out)] = as.numeric(out[1:nrow(out),2:ncol(out)])+1\n",
      "> \n",
      "> ##create Z score file\n",
      "> outZ = apply(out,2,function(x){return(scale(as.numeric(x)))})\n",
      "> rownames(outZ) = rownames(out)\n",
      "> colnames(outZ) = colnames(out)\n",
      "> outZ[outZ==NA]=0\n",
      "> outZ[outZ==\"NaN\"]=0\n",
      "> \n",
      "> #Write out scores and z-normed scores files\n",
      "> write.table(out,file=paste(keyword,\"_scoresCentroid\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> write.table(outZ,file=paste(keyword,\"_scoresCentroidZ\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "> \n",
      "Cleaning PDB file\n",
      "Starting phenix.pdbtools\n",
      "on Sun Jun  2 13:53:19 2024 by yliy0004\n",
      "===============================================================================\n",
      "\n",
      "Processing files:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Found model, 7DHY_multimer.pdb\n",
      "\n",
      "Processing PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "  Adding command-line PHIL:\n",
      "  -------------------------\n",
      "    remove=water\n",
      "\n",
      "Final processed PHIL parameters:\n",
      "-------------------------------------------------------------------------------\n",
      "  data_manager {\n",
      "    model {\n",
      "      file = \"7DHY_multimer.pdb\"\n",
      "    }\n",
      "    default_model = \"7DHY_multimer.pdb\"\n",
      "  }\n",
      "  modify {\n",
      "    remove = water\n",
      "  }\n",
      "\n",
      "\n",
      "Starting job\n",
      "===============================================================================\n",
      "Validating inputs\n",
      "Performing manipulations\n",
      "Size before: 6723 size after: 6206\n",
      "Writing output model\n",
      "\n",
      "===============================================================================\n",
      "Job complete\n",
      "usr+sys time: 2.02 seconds\n",
      "wall clock time: 2.03 seconds\n",
      "Writing out polypeptide file...\n",
      "Adding hydrogens...\n",
      "Protonating waters...\n",
      "Making .phi file\n",
      "Running full bond calculation script\n",
      "The ligands identified in this pdb are GOL-306D, GOL-303C, GOL-303B, SO4-303D, ARS-301D, ARS-301C, ARS-301B, ARS-301A, GOL-304B, GOL-305D, GOL-304D, GOL-305B\n",
      "Screening for atoms that are close.\n",
      "Dictionaries loaded.\n",
      "Starting salt bridges.\n",
      "Starting hydrogen bonds.\n",
      "Starting pi-pi version 2.\n",
      "Running pi-cation version 2\n",
      "Starting disulfide\n",
      "Starting van der waals.\n",
      "Starting metals\n",
      "Starting DNA bonds...\n",
      "Starting Ligand centroids\n",
      "Starting residue centroids...\n",
      "Removing duplicates from BFactor file\n",
      "Making 7DHY_multimer_nowaters_net\n",
      "Removing any negative value edges\n",
      "Creating QC file\n",
      "Creating Centroid directory and moving files\n",
      "Running energetic R script\n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> ##Load libraries\n",
      "> library(igraph)\n",
      "> library(stats)\n",
      "> \n",
      "> ##Initialize keywords\n",
      "> # keyword is the base filename\n",
      "> keyword = args[1]\n",
      "> args=c(paste(args[1],\"_net\",sep=\"\"),\"weighted\",paste(keyword,\"_secondaryStructure\",sep=\"\"))\n",
      "> atomCorrection=FALSE\n",
      "> baseKeyword = unlist(strsplit(keyword,\"/\"))[1]\n",
      "> runVariableSelection = TRUE\n",
      "> \n",
      "> ##Load inversal data\n",
      "> #Unique atoms file\n",
      "> terminalAtomsFile = as.matrix(read.table(\"../../terminalAtoms\",sep=\"\\t\"))\n",
      "> terminalAtoms = {}\n",
      "> for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+     terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\", \"))\n",
      "+ }\n",
      "> \n",
      "> #Secondary structure file\n",
      "> secStructure = as.matrix(read.table(args[3]))[,1:2]\n",
      "> rownames(secStructure) = secStructure[,1]\n",
      "> \n",
      "> #RSA\n",
      "> rsa = as.matrix(read.table(paste(keyword,\".rsa\",sep=\"\")))\n",
      "> \n",
      "> ##Define functions\n",
      "> \n",
      "> #Add terminal atom details\n",
      "> addTermDetails = function(data) {\n",
      "+     out_details = c()\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+         if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+             code1 = 1\n",
      "+         } else {\n",
      "+             code1 = 0\n",
      "+         }\n",
      "+         if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+             code2 = 1\n",
      "+         } else {\n",
      "+             code2 = 0\n",
      "+         }\n",
      "+     out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+     }\n",
      "+     return(out_details)\n",
      "+ }\n",
      "> \n",
      "> collapse = function(data) {\n",
      "+     if (nrow(data)==0) {\n",
      "+         out = matrix(ncol = 8,nrow=0)\n",
      "+         colnames(out) = c(\"AA1\", \"AA2\", \"Type\", \"SumEdges\", \"percentMC\", \"influence1\", \"influence2\",\"degree\")\n",
      "+         results = c(); results$out = out\n",
      "+         return(results)\n",
      "+         break\n",
      "+     }\n",
      "+     #Edges should be listed bi-directionally\n",
      "+     x=unlist(lapply(data[,3],function(x){if(x==\"MCSC\"){return(\"SCMC\")}else if(x==\"SCMC\"){return(\"MCSC\")} else {return(x)}}))\n",
      "+     if (dim(data)[2]==7) {\n",
      "+         data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6]))\n",
      "+     } else {\n",
      "+         data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6],data[,9],data[,8]))\n",
      "+     }\n",
      "+     data = unique(data)\n",
      "+     out = c()\n",
      "+     out_noPP = c()\n",
      "+     completed = c()\n",
      "+     if (dim(data)[2]==7) {\n",
      "+         out_details = c()\n",
      "+         for (i in 1:nrow(data)) {\n",
      "+             acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+             acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+             if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+                 code1 = 1\n",
      "+             } else {\n",
      "+                 code1 = 0\n",
      "+             }\n",
      "+             if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+                 code2 = 1\n",
      "+             } else {\n",
      "+                 code2 = 0\n",
      "+             }\n",
      "+             out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+         }\n",
      "+         data = out_details\n",
      "+     } else {\n",
      "+         out_details = data\n",
      "+     }\n",
      "+     colnames(out_details) = c(\"AA1\",\"AA2\",\"Type1\",\"Weight\",\"Type2\",\"Atom1\",\"Atom2\",\"influence1\",\"influence2\")\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         if (i %in% completed == FALSE) {\n",
      "+             x=which((data[,1]==data[i,1] & data[,2] == data[i,2]))\n",
      "+             subdata = matrix(data[x,],ncol=9)\n",
      "+             edgeSum = 0\n",
      "+             MCMCcount = 0\n",
      "+             MCSCcount = 0\n",
      "+             SCMCcount = 0\n",
      "+             SCSCcount = 0\n",
      "+             influence1 = sum(as.numeric(subdata[,8]))\n",
      "+             influence2 = sum(as.numeric(subdata[,9]))\n",
      "+             for (j in 1:nrow(subdata)) {\n",
      "+                 edgeSum = edgeSum + as.numeric(subdata[j,4])\n",
      "+                 if (subdata[j,3]==\"MCMC\" | subdata[j,3]==\"MCSC\") {\n",
      "+                     MCMCcount = MCMCcount+1\n",
      "+                     MCSCcount = MCSCcount+1\n",
      "+                 } else if (subdata[j,3]==\"SCMC\") {\n",
      "+                     SCMCcount = SCMCcount+1\n",
      "+                 } else if (subdata[j,3]==\"SCSC\") {\n",
      "+                     SCSCcount = SCSCcount+1\n",
      "+                 }\n",
      "+             }\n",
      "+             #Header is AA1, AA2, mixed, sum of edges, percent MC, influence1, influence2, degreeAA1\n",
      "+             out = rbind(out,c(data[i,1],data[i,2],\"mixed\",edgeSum,(MCMCcount+MCSCcount)/(MCMCcount+MCSCcount+SCMCcount+SCSCcount),influence1,influence2,sum(data[,1]==data[i,1])))\n",
      "+             if (any(subdata[,5]==\"PP\")==FALSE) {out_noPP = rbind(out_noPP,c(data[i,1],data[i,2],\"mixed\",edgeSum,(MCMCcount+MCSCcount)/(MCMCcount+MCSCcount+SCMCcount+SCSCcount),influence1,influence2,sum(data[,1]==data[i,1])))}\n",
      "+             completed = c(completed,x)\n",
      "+         }\n",
      "+     }\n",
      "+     out = out[out[,4]>0,]\n",
      "+     out_noPP = out_noPP[out_noPP[,4]>0,]\n",
      "+     colnames(out) = c(\"AA1\", \"AA2\", \"Type\", \"SumEdges\", \"percentMC\", \"influence1\", \"influence2\",\"degree\")\n",
      "+     results = c()\n",
      "+     results$out = out\n",
      "+     results$out_noPP = out_noPP\n",
      "+     results$influence_all = out_details[as.numeric(out_details[,4])>0 & (out_details[,8]==\"1\" | out_details[,9]==\"1\"),]\n",
      "+     out_details = out_details[as.numeric(out_details[,4])>0,]\n",
      "+     out_subtract = out_details; out_subtract[out_subtract[,8]==0 & out_subtract[,9]==0,4] = (-1)*as.numeric(out_subtract[out_subtract[,8]==0 & out_subtract[,9]==0,4])\n",
      "+     out_subtract = out_subtract[out_subtract[,5]!=\"PP\",]\n",
      "+     results$out_subtract = out_subtract\n",
      "+     return(results)\n",
      "+ }\n",
      "> \n",
      "> removeRedundancy = function(dataTmp) {\n",
      "+     for (i in 1:nrow(dataTmp)) {\n",
      "+         if (order(dataTmp[i,1:2])[1]==2) {\n",
      "+             dataTmp[i,1:2] = c(dataTmp[i,2],dataTmp[i,1])\n",
      "+             dataTmp[i,3] = paste(unlist(strsplit(dataTmp[i,3],split=\"\"))[c(3,4,1,2)],collapse=\"\")\n",
      "+             dataTmp[i,c(6,7)] = c(dataTmp[i,7],dataTmp[i,6])\n",
      "+         }\n",
      "+     }\n",
      "+     dataTmp = unique(dataTmp)\n",
      "+     pairs = unique(dataTmp[,1:2])\n",
      "+     out = c()\n",
      "+     for (j in 1:nrow(pairs)) {\n",
      "+         pair = pairs[j,]\n",
      "+         subData = matrix(dataTmp[dataTmp[,1]==pair[1] & dataTmp[,2]==pair[2],],ncol=7)\n",
      "+         out = rbind(out, c(pair[1],pair[2],\"mixed\",0.5*sum(as.numeric(subData[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> collapse_directed = function(data) {\n",
      "+     pairs = unique(data[,1:2])\n",
      "+     out = c()\n",
      "+     for (i in 1:nrow(pairs)) {\n",
      "+         subdat = matrix(data[data[,1]==pairs[i,1]&data[,2]==pairs[i,2],],ncol=4)\n",
      "+         out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> collapse_agnostic = function(data) {\n",
      "+     pairs = unique(t(apply(data[,1:2],1,function(x){return(sort(x))})))\n",
      "+     out = c()\n",
      "+     for (i in 1:nrow(pairs)) {\n",
      "+         subdat = matrix(data[(data[,1]==pairs[i,1]&data[,2]==pairs[i,2]) | (data[,2]==pairs[i,1]&data[,1]==pairs[i,2]),],ncol=4)\n",
      "+         out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> distributCalc = function(set,nodes,net) {\n",
      "+     #x is assumed to be a set but can also just be 1 node\n",
      "+     distanceNet = distances(net,weights=net$weight)\n",
      "+     distanceTmp = c()\n",
      "+     Dr = c()\n",
      "+     for (node in nodes) {\n",
      "+         distanceTmp = c()\n",
      "+         for (item in set) {\n",
      "+             if (item!=node) {\n",
      "+                 if(item%in%rownames(distanceNet)==FALSE){distanceTmp=100;next}\n",
      "+                 if(node%in%rownames(distanceNet)==FALSE){distanceTmp=100;next}\n",
      "+                 distanceTmp = c(distanceTmp,distanceNet[item,node])\n",
      "+             }\n",
      "+         }\n",
      "+         if (length(distanceTmp)>0) {\n",
      "+             Dr = c(Dr, 1/min(distanceTmp))\n",
      "+         }\n",
      "+     }\n",
      "+     return(sum(Dr)/length(nodes))\n",
      "+ }\n",
      "> \n",
      "> ##Read in user data\n",
      "> data_all = as.matrix(read.table(args[1]))\n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> originalNodes = nodes\n",
      "> basedata = collapse(data_all)\n",
      "> basedata_noPP = removeRedundancy(basedata$out[,1:7])\n",
      "> basedata_allsidechain = collapse_agnostic(data_all[data_all[,3]!=\"MCMC\",c(1,2,3,4)])\n",
      "> influencedata_all = basedata$influence_all\n",
      "> influencedata_collapse = collapse(influencedata_all)\n",
      "> influencedata = removeRedundancy(influencedata_collapse$out[,1:7])\n",
      "> influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "> influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "> \n",
      "> influencedata_directed = collapse_directed(influencedata_all[influencedata_all[,8]==1,c(1,2,3,4)])\n",
      "> influencenet_directed = graph.edgelist(influencedata_directed[,c(2,1)],directed=TRUE)\n",
      "> influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "> \n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> \n",
      "> #Write out edgelist files\n",
      "> write.table(influencedata,file=paste(keyword,\"_allEdges\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=FALSE,row.names=FALSE)\n",
      "> write.table(influencedata_directed,file=paste(keyword,\"_uniqueEdges\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=FALSE,row.names=FALSE)\n",
      "> \n",
      "> ###Deal with missing nodes\n",
      "> if (any(nodes %in% secStructure[,1]==FALSE)) {\n",
      "+     for (node in nodes) {\n",
      "+         if (node %in% secStructure[,1]==FALSE) {\n",
      "+             secStructure = rbind(secStructure,c(node,\"xxx\",360))\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> for (node in nodes) {\n",
      "+     if (node %in% rownames(rsa) == FALSE) {\n",
      "+         rsa = rbind(rsa,c(0,0))\n",
      "+         rownames(rsa)[nrow(rsa)] = node\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> ##Begin centrality calculations\n",
      "> \n",
      "> #Modular calculations\n",
      "> \n",
      "> ##WALKTRAP\n",
      "> net = graph.edgelist(basedata_noPP[,1:2],directed=FALSE)\n",
      "> net$weight = 1/as.numeric(basedata_noPP[,4])\n",
      "> net_community=walktrap.community(net,weights=net$weight); net_community_vec = net_community$membership; names(net_community_vec)=net_community$names\n",
      "> net_community_vec_wt = net_community_vec\n",
      "> nodes = V(net)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> nodes.color = c()\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(influencedata)) {\n",
      "+     j = which((basedata_noPP[,1]==influencedata[i,1]&basedata_noPP[,2]==influencedata[i,2])|(basedata_noPP[,1]==influencedata[i,2]&basedata_noPP[,2]==influencedata[i,1]))\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) { edgecolors[j]=\"grey40\" }\n",
      "+ }\n",
      "> if(length(nodes)>200) {textsize=.25} else {textsize=.45}\n",
      "> plot(net,edge.color = edgecolors, vertex.color = nodes.color,vertex.size=4,vertex.label.cex=textsize,main=\"WALKTRAP\")\n",
      ">     \n",
      "> ##Weighted edge betweenness WALKTRAP\n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP - WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for WALKTRAP\n",
      "> firstOrderDegree = degree(influencenet)\n",
      "> firstOrderDegree_sidechain = degree(influencenet_directed,mode=c(\"in\"))\n",
      "> secondOrderDegree = c(); secondOrderDegree_sidechain=c()\n",
      "> nodes = V(influencenet)$name\n",
      "> for (node in nodes) {\n",
      "+     firstorder = neighbors(influencenet,node)\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet,neighbor)))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree = c(secondOrderDegree,length(secondorder))\n",
      "+     firstorder = neighbors(influencenet_directed,node,mode=c(\"in\"))\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet_directed,neighbor,mode=c(\"in\"))))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree_sidechain = c(secondOrderDegree_sidechain,length(secondorder))\n",
      "+ }\n",
      "> names(secondOrderDegree) = nodes; names(secondOrderDegree_sidechain) = nodes\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])])) - 1\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])])) - 1\n",
      "+ }\n",
      "> \n",
      "> #SAVE ALL AS WALKTRAP\n",
      "> node_edge_betweenness_wt = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_wt = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_wt = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_wt = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_wt = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_wt = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> ##2ARY STRUCTURE\n",
      "> net_community_vec = as.numeric(as.factor(secStructure[,2]))\n",
      "> names(net_community_vec) = secStructure[,1]\n",
      "> nodes = V(net)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> nodes.color = c()\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(basedata_noPP)) {\n",
      "+     if (net_community_vec[basedata_noPP[i,1]]!=net_community_vec[basedata_noPP[i,2]]) { edgecolors[i]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+         node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for 2ARY STRUCTURE\n",
      "> nodes = V(influencenet)$name\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])])) - 1\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])])) - 1\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #SAVE ALL AS STRIDE\n",
      "> node_edge_betweenness_stride = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_stride = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_stride = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_stride = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_stride = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_stride = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> \n",
      "> ### Create final dataset\n",
      "> \n",
      "> out = cbind(\n",
      "+ firstOrderDegree[nodes],\n",
      "+ firstOrderDegree_sidechain[nodes],\n",
      "+ secondOrderDegree[nodes],\n",
      "+ secondOrderDegree_sidechain[nodes],\n",
      "+ node_edge_betweenness_stride[nodes],\n",
      "+ node_edge_betweenness_sidechain_stride[nodes],\n",
      "+ node_intermodular_degree_stride[nodes],\n",
      "+ node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ node_edge_betweenness_wt[nodes],\n",
      "+ node_edge_betweenness_sidechain_wt[nodes],\n",
      "+ node_intermodular_degree_wt[nodes],\n",
      "+ node_intermodular_degree_sidechain_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_wt[nodes]\n",
      "+ )\n",
      "> rownames(out) = nodes\n",
      "> colnames(out) = c(\"Degree\",\"Degree_sidechain\",\"SecondOrderDegree\",\"SecondOrderDegree_sidechain\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_sidechain\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_sidechain\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_sidechain\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_sidechain\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_sidechain\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_sidechain\")\n",
      "> \n",
      "> #add back in nodes that were not networked as zeros\n",
      "> zeroMat = matrix(0,nrow=length(which(originalNodes%in%nodes==FALSE)),ncol=ncol(out));rownames(zeroMat)=originalNodes[which(originalNodes%in%nodes==FALSE)]\n",
      "> out = rbind(out,zeroMat)\n",
      "> \n",
      "> #Add 1 to everything to avoid zeros, except ligand\n",
      "> out[is.na(out)]=0\n",
      "> out = out+1\n",
      "> \n",
      "> #Write out\n",
      "> write.table(out,file=paste(keyword,\"_scoresEnergetics\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "> #Standard normalization\n",
      "> outZ = apply(out,2,function(x){return(scale(x))})\n",
      "> rownames(outZ) = rownames(out)\n",
      "> write.table(outZ,file=paste(keyword,\"_scoresEnergeticsZ\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "Running centroid R script\n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(igraph)\n",
      "> library(stats)\n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> ##read in data\n",
      "> ## arg1: col1 and col2 are edges, col3 is weight of edge\n",
      "> ## arg2: weighted, unweighted <-- no longer using\n",
      "> ## arg3: forced module definition <-- no longer using\n",
      "> \n",
      "> # Set desired parameters\n",
      "> removeMCMC = FALSE\n",
      "> linearcutoff = 1\n",
      "> directed = FALSE\n",
      "> secondaryStructure2 = FALSE\n",
      "> sidechainMode = TRUE ##build network of just SCSC, SCMC and MCSC. Still use unique atoms for the _uniqueAtoms directed plots\n",
      "> uniqueAtomsMode = FALSE ##build network of just unique atoms (each edge has to have at least 1). Still use unique atoms for the _uniqueAtoms directed plots\n",
      "> useDNA=TRUE\n",
      "> uniqueAtomsUnbiased=FALSE\n",
      "> uniqueAtomsGaurav=FALSE\n",
      "> uniqueAtomsOLD=FALSE\n",
      "> atomCorrection = FALSE\n",
      "> uniqueAtomsGauravPP = FALSE\n",
      "> weighted = TRUE\n",
      "> removeWaters=TRUE\n",
      "> if (tail(unlist(strsplit(getwd(),\"/\")),n=1)==\"Centroid\") {\n",
      "+     ligandCentroidMode = TRUE\n",
      "+ } else {\n",
      "+     ligandCentroidMode = FALSE\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> # Read in ligand file\n",
      "> ligandfile = scan(paste(args[1],\"_ligand\",sep=\"\"),what=\"character\")\n",
      "> if (length(ligandfile)==0){\n",
      "+     ligandmode = FALSE\n",
      "+ } else {\n",
      "+     ligandmode = TRUE\n",
      "+ }\n",
      "> \n",
      "> ligandCentroidFile = scan(paste(args[1],\"NetLigand\",sep=\"\"),what=\"character\")\n",
      "> \n",
      "> # keyword is the base filename\n",
      "> # method is \"frag\" or \"module\" or \"both\"\n",
      "> keyword = args[1]\n",
      "> args=c(paste(args[1],\"_net\",sep=\"\"),\"weighted\",dir(\"./\",\"_secondaryStructure$\"))\n",
      "> \n",
      "> ##Load data\n",
      "> #Unique atoms file\n",
      "> if(uniqueAtomsUnbiased) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"/Users/vjpatel/Dropbox/Gaurav/uniqueAtomsUnbiased\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else if (uniqueAtomsGaurav) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"/Users/vjpatel/Dropbox/Gaurav/uniqueAtomsGaurav\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else if (uniqueAtomsOLD) {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"../../../terminalAtoms\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+     }\n",
      "+ } else {\n",
      "+     terminalAtomsFile = as.matrix(read.table(\"../../../uniqueAtoms\",sep=\"\\t\"))\n",
      "+     terminalAtoms = list()\n",
      "+     otherTA = list()\n",
      "+     for (i in 1:nrow(terminalAtomsFile)) {\n",
      "+         terminalAtoms[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,2],split=\",\"))\n",
      "+         #otherTA[[terminalAtomsFile[i,1]]] = unlist(strsplit(terminalAtomsFile[i,4],split=\",\"))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> #Secondary structure file\n",
      "> #Ligands are all one secondary structure\n",
      "> secStructure = as.matrix(read.table(args[3]))\n",
      "> phiAngleDB = {}\n",
      "> for (i in 1:nrow(secStructure)) {\n",
      "+     phiAngleDB[[secStructure[i,1]]] = as.numeric(secStructure[i,3])\n",
      "+ }\n",
      "> \n",
      "> if(ligandmode) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",\"_ligand\"),sep=\"\\t\"))\n",
      "+     liganddat = unique(liganddat[,2])\n",
      "+     for (ligand in liganddat) {\n",
      "+         secStructure = rbind(secStructure,c(sub(\"-\",\"\",ligand),\"ligand\",360))\n",
      "+         phiAngleDB[[sub(\"-\",\"\",ligand)]]=0\n",
      "+     }\n",
      "+ } else {\n",
      "+     liganddat = c()\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> ##Define functions\n",
      "> \n",
      "> ##Collapse matrix functions\n",
      "> ##Will return redundant matrix\n",
      "> ##Input needs to be 7 columns\n",
      "> ##Second input is different if the main input is a subset of the larger data\n",
      "> addTerminalDetails = function(data,largerdata) {\n",
      "+     #List edges bi-directionally\n",
      "+     x=unlist(lapply(data[,3],function(x){if(x==\"MCSC\"){return(\"SCMC\")}else if(x==\"SCMC\"){return(\"MCSC\")} else {return(x)}}))\n",
      "+     data = rbind(data,cbind(data[,2],data[,1],x,data[,4],data[,5],data[,7],data[,6]))\n",
      "+     data = unique(data)\n",
      "+     out = c()\n",
      "+     out_noPP = c()\n",
      "+     out_details = c()\n",
      "+     for (i in 1:nrow(data)) {\n",
      "+         acid1 = paste(unlist(strsplit(data[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         acid2 = paste(unlist(strsplit(data[i,2],split=\"\"))[1:3],collapse=\"\")\n",
      "+         subdata = matrix(data[data[,1]==data[i,1]&data[,2]==data[i,2],],ncol=7)\n",
      "+         allsubdata = matrix(largerdata[largerdata[,1]==data[i,1],],ncol=7)\n",
      "+         code1=0;code2=0\n",
      "+         if (data[i,6] %in% terminalAtoms[[acid1]]) {\n",
      "+             code1 = 1\n",
      "+         }\n",
      "+         if (data[i,5] == \"DNA\" & useDNA==TRUE) {\n",
      "+             code1 = 1\n",
      "+         }\n",
      "+         allsubdata = matrix(largerdata[largerdata[,2]==data[i,2],],ncol=7)\n",
      "+         if (data[i,7] %in% terminalAtoms[[acid2]]) {\n",
      "+             code2 = 1\n",
      "+         }\n",
      "+         if (data[i,5] == \"DNA\" & useDNA==TRUE) {\n",
      "+             code2 = 1\n",
      "+         }\n",
      "+         if (data[i,5]==\"PICATION\"|data[i,5]==\"PIPI\") {\n",
      "+             code1=1\n",
      "+             code2=1\n",
      "+         }\n",
      "+         if (data[i,5]==\"PP\") {\n",
      "+             if (uniqueAtomsGaurav) {\n",
      "+                 if((acid1) == \"GLY\") {\n",
      "+                     code1 = 1\n",
      "+                     data[i,3]==\"SCSC\"\n",
      "+                 }\n",
      "+                 if ((acid2) == \"GLY\") {\n",
      "+                     code2 = 1\n",
      "+                     data[i,3] == \"SCSC\"\n",
      "+                 }\n",
      "+             }\n",
      "+         }\n",
      "+         out_details = rbind(out_details, c(data[i,],code1,code2))\n",
      "+     }\n",
      "+     return(out_details)\n",
      "+ }\n",
      "> \n",
      "> ##Will return non-redundant matrix\n",
      "> collapse_agnostic = function(data) {\n",
      "+     if(nrow(data)==0) {\n",
      "+         return(matrix(ncol=4,nrow=0))\n",
      "+     } else if (ncol(data)==4 & data[1,3]==\"mixed\") {\n",
      "+         return(data)\n",
      "+     } else {\n",
      "+         pairs = unique(t(apply(matrix(data[,1:2],ncol=2),1,function(x){return(sort(x))})))\n",
      "+         out = c()\n",
      "+         for (i in 1:nrow(pairs)) {\n",
      "+             subdat = matrix(data[(data[,1]==pairs[i,1]&data[,2]==pairs[i,2]) | (data[,2]==pairs[i,1]&data[,1]==pairs[i,2]),1:4],ncol=4)\n",
      "+             subdat = unique(t(apply(subdat,1,function(x){a=which(x==sort(x[1:2])[1]);if(a==1){return(x)}else{return(c(x[2],x[1],paste(unlist(strsplit(x[3],\"\"))[c(3,4,1,2)],collapse=\"\"),x[4]))}})))\n",
      "+             if (weighted==FALSE) {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",1))\n",
      "+             } else {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+             }\n",
      "+         }\n",
      "+     }\n",
      "+     return(out)\n",
      "+ }\n",
      "> \n",
      "> ##Will return directed collapsed matrix\n",
      "> collapse_directed = function(data) {\n",
      "+     if (length(data)==0) {\n",
      "+         return(matrix(nrow=0,ncol=4))\n",
      "+     } else {\n",
      "+         pairs = matrix(unique(data[,1:2]),ncol=2)\n",
      "+         out = c()\n",
      "+         for (i in 1:nrow(pairs)) {\n",
      "+             subdat = matrix(data[data[,1]==pairs[i,1]&data[,2]==pairs[i,2],1:4],ncol=4)\n",
      "+             if (weighted==FALSE) {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",1))\n",
      "+             } else {\n",
      "+                 out = rbind(out,c(subdat[1,1],subdat[1,2],\"mixed\",sum(as.numeric(subdat[,4]))))\n",
      "+             }\n",
      "+         }\n",
      "+         return(matrix(out,ncol=4))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> rsa = as.matrix(read.table(dir(\"./\",\".rsa$\")))\n",
      "> rownames(rsa) = rsa[,1]\n",
      "> rsa[,1] = rsa[,2]\n",
      "> \n",
      "> #add ligand as -1\n",
      "> for (ligand in liganddat) {\n",
      "+     rsa = rbind(rsa,c(-1,-1))\n",
      "+     rownames(rsa)[nrow(rsa)] = sub(\"-\",\"\",ligand)\n",
      "+ }\n",
      "> \n",
      "> #Function to get amino acid\n",
      "> getAcid = function(residue) {\n",
      "+     if(all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"A\")) == TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"G\"))==TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"C\"))==TRUE) | all((unlist(strsplit(residue,\"\"))[1:2] == c(\"D\",\"T\"))==TRUE)) {return(paste(unlist(strsplit(residue,\"\"))[1:2],collapse=\"\"))\n",
      "+     } else {\n",
      "+     return(paste(unlist(strsplit(residue,\"\"))[1:3],collapse=\"\"))\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> ##################\n",
      "> ## read in data ##\n",
      "> ##################\n",
      "> data_all = as.matrix(read.table(args[1]))[,1:7]\n",
      "> if(removeWaters) {\n",
      "+     y=c(grep(\"HOH\",data_all[,1]),grep(\"HOH\",data_all[,2]))\n",
      "+     if(length(y)>0) {\n",
      "+         data_all = data_all[-y,]\n",
      "+     }\n",
      "+ }\n",
      "> nodes = unique(c(data_all[,1:2])); sets = nodes\n",
      "> originalNodes = nodes\n",
      "> \n",
      "> data_all_original = rbind(data_all,data_all[,c(2,1,3,4,5,7,6)])\n",
      "> data_all = data_all[as.numeric(data_all[,4])>0,]\n",
      "> \n",
      "> basedata = addTerminalDetails(data_all,data_all_original)\n",
      "> basedata_noPP = collapse_agnostic(basedata[basedata[,5]!=\"PP\",])\n",
      "> sidechaindata = basedata[basedata[,3]!=\"MCMC\",1:4]\n",
      "> sidechaindata_detailed = basedata[basedata[,3]!=\"MCMC\",]\n",
      "> \n",
      "> influencedata_detailed = basedata\n",
      "> if (uniqueAtomsMode) {\n",
      "+     influencedata_detailed = basedata[basedata[,8]==1 | basedata[,9]==1,]\n",
      "+ }\n",
      "> influencedata = collapse_agnostic(influencedata_detailed)\n",
      "> influencedataGLY = rbind(influencedata[grep(\"GLY\",influencedata[,1]),],influencedata[grep(\"GLY\",influencedata[,1]),])\n",
      "> influencedata = influencedata[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,2])))>linearcutoff,]\n",
      "> \n",
      "> \n",
      "> influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "> influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "> \n",
      "> influencedata_directed = collapse_directed(matrix(influencedata_detailed[influencedata_detailed[,8]==1,1:4],ncol=4))\n",
      "> influencedata_directedGLY = rbind(influencedata_directed[grep(\"GLY\",influencedata_directed[,1]),],influencedata_directed[grep(\"GLY\",influencedata_directed[,2]),])\n",
      "> influencedata_directed = influencedata_directed[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,2])))>linearcutoff,]\n",
      "> \n",
      "> influencedata_directed = matrix(influencedata_directed,ncol=4)\n",
      "> influencenet_directed = graph.edgelist(matrix(influencedata_directed[,c(2,1)],ncol=2),directed=TRUE)\n",
      "> influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "> \n",
      "> \n",
      "> if(sidechainMode) {\n",
      "+     influencedata_detailed = sidechaindata_detailed\n",
      "+     influencedata = collapse_agnostic(influencedata_detailed)\n",
      "+     influencedataGLY = rbind(influencedata[grep(\"GLY\",influencedata[,1]),],influencedata[grep(\"GLY\",influencedata[,2]),])\n",
      "+     influencedata = influencedata[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata[,2])))>linearcutoff,]\n",
      "+     #Add back for GLY if uniqueAtomsGaurav\n",
      "+     if(uniqueAtomsGauravPP) {\n",
      "+         influencedata = rbind(influencedata,influencedataGLY)\n",
      "+     }\n",
      "+     influencedata = influencedata[as.numeric(influencedata[,4])>0,]\n",
      "+     \n",
      "+     influencenet = graph.edgelist(influencedata[,1:2],directed=FALSE)\n",
      "+     influencenet$weight = 1/as.numeric(influencedata[,4])\n",
      "+     \n",
      "+     influencedata_directed = collapse_directed(matrix(influencedata_detailed[influencedata_detailed[,8]==1,1:4],ncol=4))\n",
      "+     influencedata_directedGLY = rbind(influencedata_directed[grep(\"GLY\",influencedata_directed[,1]),],influencedata_directed[grep(\"GLY\",influencedata_directed[,2]),])\n",
      "+ \n",
      "+     influencedata_directed = influencedata_directed[abs(as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,1]))-as.numeric(gsub(\"[A-Z]\",\"\",influencedata_directed[,2])))>linearcutoff,]\n",
      "+     #Add back for GLY if uniqueAtomsGaurav\n",
      "+     if(uniqueAtomsGauravPP) {\n",
      "+         influencedata_directed = rbind(influencedata_directed,influencedata_directedGLY)\n",
      "+     }\n",
      "+     influencedata_directed = matrix(influencedata_directed,ncol=4)\n",
      "+     influencenet_directed = graph.edgelist(matrix(influencedata_directed[,c(2,1)],ncol=2),directed=TRUE)\n",
      "+     influencenet_directed$weight = 1/as.numeric(influencedata_directed[,4])\n",
      "+     \n",
      "+ }\n",
      "> \n",
      "> #Replace missing nodes in the igraph net\n",
      "> nodesMissing = nodes[nodes %in% influencedata[,1]==FALSE & nodes %in% influencedata[,2]==FALSE]\n",
      "> \n",
      "> ##B-factor\n",
      "> bfactor = as.matrix(read.table(dir(\"./\",\"_Bfactor\"),row.names=1))\n",
      "> \n",
      "> #Replace missing nodes in secStructure matrix\n",
      "> if (any(nodes %in% secStructure[,1]==FALSE)) {\n",
      "+     for (node in nodes) {\n",
      "+         if (node %in% secStructure[,1]==FALSE) {\n",
      "+             secStructure = rbind(secStructure,c(node,\"xxx\",360))\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> for (node in nodes) {\n",
      "+     if (node %in% rownames(rsa) == FALSE) {\n",
      "+         rsa = rbind(rsa,c(0,0))\n",
      "+         rownames(rsa)[nrow(rsa)] = node\n",
      "+     }\n",
      "+     if (node %in% rownames(bfactor) == FALSE) {\n",
      "+         bfactor = rbind(bfactor,1)\n",
      "+         rownames(bfactor)[nrow(bfactor)] = node\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> \n",
      "> \n",
      "> #Give DNA attributes\n",
      "> if (\"DNA\" %in% data_all[,5]) {\n",
      "+     for (i in 1:nrow(data_all)) {\n",
      "+         if (data_all[i,5]==\"DNA\") {\n",
      "+             phiAngleDB[[data_all[i,1]]] = 90\n",
      "+             if (data_all[i,1] %in% rownames(rsa) == FALSE) {\n",
      "+                 rsa = rbind(rsa,c(0,0))\n",
      "+                 rownames(rsa)[nrow(rsa)] = data_all[i,1]\n",
      "+             }\n",
      "+             if (data_all[i,1] %in% rownames(secStructure) == FALSE) {\n",
      "+                 secStructure = rbind(secStructure,c(data_all[i,1],\"DNA\",360))\n",
      "+             }\n",
      "+             if (data_all[i,1] %in% rownames(bfactor) == FALSE) {\n",
      "+                 bfactor = rbind(bfactor,1)\n",
      "+                 rownames(bfactor)[nrow(bfactor)] = data_all[i,1]\n",
      "+             }\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #Betweenness\n",
      "> node_betweenness = betweenness(influencenet,weights=influencenet$weight,directed=directed)\n",
      "> node_betweenness_unique = betweenness(influencenet_directed,weights=influencenet_directed$weight,directed=directed)\n",
      "> \n",
      "> #Modular calculations\n",
      "> \n",
      "> ##WALKTRAP\n",
      "> net = graph.edgelist(basedata_noPP[,1:2],directed=FALSE)\n",
      "> net$weight = 1/as.numeric(basedata_noPP[,4])\n",
      "> net_community=walktrap.community(net,weights=net$weight); net_community_vec = net_community$membership; names(net_community_vec)=net_community$names\n",
      "> net_community_vec_wt = net_community_vec\n",
      "> #net_community_vec[buriednodes] = max(net_community_vec)+1\n",
      "> nodes = V(influencenet)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> nodes.color = c()\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(influencedata)) {\n",
      "+     j = which((basedata_noPP[,1]==influencedata[i,1]&basedata_noPP[,2]==influencedata[i,2])|(basedata_noPP[,1]==influencedata[i,2]&basedata_noPP[,2]==influencedata[i,1]))\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) { edgecolors[j]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP\n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WALKTRAP - WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         if(weight1==0 & weight2==0) {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]] + 0\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]] + 0\n",
      "+         } else {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Degree and second order degree for WALKTRAP\n",
      "> firstOrderDegree = degree(influencenet)\n",
      "> firstOrderDegree_sidechain = degree(influencenet_directed,mode=c(\"in\"))\n",
      "> \n",
      "> secondOrderDegree = c(); secondOrderDegree_sidechain=c()\n",
      "> nodes = V(influencenet)$name\n",
      "> for (node in nodes) {\n",
      "+     firstorder = neighbors(influencenet,node)\n",
      "+     secondorder = c(); for (neighbor in firstorder){secondorder = c(secondorder,names(neighbors(influencenet,neighbor)))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree = c(secondOrderDegree,length(secondorder))\n",
      "+     if(node %in% V(influencenet_directed)$name) {\n",
      "+         firstorder_sidechain = neighbors(influencenet_directed,node,mode=c(\"in\"))\n",
      "+     } else {\n",
      "+         firstorder_sidechain = c()\n",
      "+     }\n",
      "+     secondorder = c(); for (neighbor in firstorder_sidechain){secondorder = c(secondorder,names(neighbors(influencenet_directed,neighbor,mode=c(\"in\"))))}; secondorder = unique(secondorder); secondorder=secondorder[secondorder!=node]\n",
      "+     secondOrderDegree_sidechain = c(secondOrderDegree_sidechain,length(secondorder))\n",
      "+ }\n",
      "> names(secondOrderDegree) = nodes; names(secondOrderDegree_sidechain) = nodes\n",
      "> \n",
      "> \n",
      "> #second order intermodular degree\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])]))\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])]))\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> #SAVE ALL AS WALKTRAP\n",
      "> node_edge_betweenness_wt = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_wt = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_wt = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_wt = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_wt = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_wt = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> \n",
      "> ##2ARY STRUCTURE\n",
      "> secStructure = secStructure[,1:2]\n",
      "> rownames(secStructure) = secStructure[,1]\n",
      "> net_community_vec = as.numeric(as.factor(secStructure[,2]))\n",
      "> names(net_community_vec) = secStructure[,1]\n",
      "> #net_community_vec[buriednodes] = max(net_community_vec)+1\n",
      "> nodes = V(influencenet)$name\n",
      "> edgecolors = rep(\"grey90\",nrow(basedata_noPP))\n",
      "> colorPalette=rainbow(max(net_community_vec),s=.5)\n",
      "> nodes.color = c()\n",
      "> for (i in 1:length(nodes)) {\n",
      "+     nodes.color = c(nodes.color,colorPalette[net_community_vec[nodes[i]]])\n",
      "+ }\n",
      "> for (i in 1:nrow(basedata_noPP)) {\n",
      "+     if (net_community_vec[basedata_noPP[i,1]]!=net_community_vec[basedata_noPP[i,2]]) { edgecolors[i]=\"grey40\" }\n",
      "+ }\n",
      "> \n",
      "> edge_betweenness = edge.betweenness(influencenet,weights=influencenet$weight)\n",
      "> node_edge_betweenness = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         node_edge_betweenness[influencedata[i,1]] = node_edge_betweenness[influencedata[i,1]]+edge_betweenness[i]\n",
      "+         node_edge_betweenness[influencedata[i,2]] = node_edge_betweenness[influencedata[i,2]]+edge_betweenness[i]\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree) = nodes\n",
      "> node_modules = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata = matrix(influencedata[influencedata[,1]==node | influencedata[,2]==node,],ncol=4)\n",
      "+     if (nrow(subdata)==0) {node_intermodular_degree[node] = 0; next}\n",
      "+     bound = unique(c(subdata[,1:2]))\n",
      "+     bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]\n",
      "+     bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules[[node]] = bound_modules\n",
      "+     node_intermodular_degree[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> ##Weighted edge betweenness WEIGHT BY SIDE CHAIN\n",
      "> node_edge_betweenness_sidechain = rep(0,length(nodes))\n",
      "> names(node_edge_betweenness_sidechain) = nodes\n",
      "> for(i in 1:length(edge_betweenness)) {\n",
      "+     if (net_community_vec[influencedata[i,1]]!=net_community_vec[influencedata[i,2]]) {\n",
      "+         weight1 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,1]&influencedata_directed[,2]==influencedata[i,2],4]);if(length(weight1)==0){weight1=0}\n",
      "+         weight2 = as.numeric(influencedata_directed[influencedata_directed[,1]==influencedata[i,2]&influencedata_directed[,2]==influencedata[i,1],4]);if(length(weight2)==0){weight2=0}\n",
      "+         if(weight1==0 & weight2==0) {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]] + 0\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]] + 0\n",
      "+         } else {\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,1]] = node_edge_betweenness_sidechain[influencedata[i,1]]+(weight1/(weight1+weight2))*edge_betweenness[i]\n",
      "+             node_edge_betweenness_sidechain[influencedata[i,2]] = node_edge_betweenness_sidechain[influencedata[i,2]]+(weight2/(weight1+weight2))*edge_betweenness[i]\n",
      "+         }\n",
      "+     }\n",
      "+ }\n",
      "> node_intermodular_degree_sidechain = rep(0,length(nodes))\n",
      "> names(node_intermodular_degree_sidechain) = nodes\n",
      "> node_modules_sidechain = list()\n",
      "> for (node in nodes) {\n",
      "+     subdata_sidechain = matrix(influencedata_directed[influencedata_directed[,1]==node,],ncol=4)\n",
      "+     if (nrow(subdata_sidechain)==0) {node_intermodular_degree_sidechain[node] = 0; next}\n",
      "+     bound = unique(c(subdata_sidechain[,1:2])); bound = bound[bound!=node]\n",
      "+     bound_modules = net_community_vec[bound]; bound_modules = bound_modules[bound_modules!=net_community_vec[node]]\n",
      "+     node_modules_sidechain[[node]] = bound_modules\n",
      "+     node_intermodular_degree_sidechain[node] = length(bound_modules)\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> ##Degree and second order degree for 2ARY STRUCTURE\n",
      "> \n",
      "> #second order intermodular degree\n",
      "> secondOrder_node_intermodular_degree = rep(0,length(node_intermodular_degree)); names(secondOrder_node_intermodular_degree) = names(node_intermodular_degree)\n",
      "> secondOrder_node_intermodular_degree_sidechain = rep(0,length(node_intermodular_degree_sidechain)); names(secondOrder_node_intermodular_degree_sidechain) = names(node_intermodular_degree_sidechain)\n",
      "> for (node in names(node_intermodular_degree)) {\n",
      "+     if (node_intermodular_degree[node]==0) {next}\n",
      "+     #secondOrder_node_intermodular_degree[node] = length(unique(unlist(node_modules[names(node_modules[[node]])])))\n",
      "+     secondOrder_node_intermodular_degree[node] = length(unlist(node_modules[names(node_modules[[node]])]))\n",
      "+     secondOrder_node_intermodular_degree_sidechain[node] = length(unlist(node_modules_sidechain[names(node_modules_sidechain[[node]])]))\n",
      "+ }\n",
      "> \n",
      "> #SAVE ALL AS STRIDE\n",
      "> node_edge_betweenness_stride = node_edge_betweenness\n",
      "> node_edge_betweenness_sidechain_stride = node_edge_betweenness_sidechain\n",
      "> node_intermodular_degree_stride = node_intermodular_degree\n",
      "> node_intermodular_degree_sidechain_stride = node_intermodular_degree_sidechain\n",
      "> secondOrder_node_intermodular_degree_stride = secondOrder_node_intermodular_degree\n",
      "> secondOrder_node_intermodular_degree_sidechain_stride = secondOrder_node_intermodular_degree_sidechain\n",
      "> \n",
      "> ### Create final dataset\n",
      "> nodes = nodes[nodes!=\"DNA1000A\"]\n",
      "> nodes = nodes[nodes %in% c(\"DA\",\"DG\",\"DC\",\"DT\") == FALSE]\n",
      "> \n",
      "> ### Remove ligand amino acids from final dataset\n",
      "> ligands = sub(\"-\",\"\",liganddat)\n",
      "> nodes = nodes[nodes %in% ligands == FALSE]\n",
      "> \n",
      "> nodes[which(nodes %in% rownames(bfactor)==FALSE)]\n",
      "character(0)\n",
      "> \n",
      "> out = cbind(\n",
      "+ rsa[nodes,1],\n",
      "+ firstOrderDegree[nodes],\n",
      "+ firstOrderDegree_sidechain[nodes],\n",
      "+ secondOrderDegree[nodes],\n",
      "+ secondOrderDegree_sidechain[nodes],\n",
      "+ node_edge_betweenness_stride[nodes],\n",
      "+ node_edge_betweenness_sidechain_stride[nodes],\n",
      "+ node_intermodular_degree_stride[nodes],\n",
      "+ node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_stride[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_stride[nodes],\n",
      "+ node_edge_betweenness_wt[nodes],\n",
      "+ node_edge_betweenness_sidechain_wt[nodes],\n",
      "+ node_intermodular_degree_wt[nodes],\n",
      "+ node_intermodular_degree_sidechain_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_wt[nodes],\n",
      "+ secondOrder_node_intermodular_degree_sidechain_wt[nodes]\n",
      "+ )\n",
      "> \n",
      "> #write column names\n",
      "> colnames(out) = c(\"RSA\",\"Degree\",\"Degree_uniqueAtoms\",\"SecondOrderDegree\",\"SecondOrderDegree_uniqueAtoms\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_unqiueAtoms\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_uniqueAtoms\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_uniqueAtoms\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_uniqueAtoms\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_uniqueAtoms\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_uniqueAtoms\")\n",
      "> \n",
      "> \n",
      "> ## LIGANDS\n",
      "> if(ligandmode==TRUE & ligandCentroidMode==FALSE) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",\"_ligand\"),sep=\"\\t\"))\n",
      "+     \n",
      "+     tmp = c()\n",
      "+     i=1\n",
      "+     for (i in 1:nrow(liganddat)) {\n",
      "+         acid=paste(unlist(strsplit(liganddat[i,1],split=\"\"))[1:3],collapse=\"\")\n",
      "+         if (unlist(strsplit(liganddat[i,1],\"-\"))[2] %in% terminalAtoms[[acid]]) {\n",
      "+             tmp = rbind(tmp,c(unlist(strsplit(liganddat[i,1],\"-\"))[1],liganddat[i,2],liganddat[i,3]))\n",
      "+         }\n",
      "+     }\n",
      "+     \n",
      "+     #ligandvec = tapply(liganddat[,3],as.factor(liganddat[,1]),min)\n",
      "+     ligandvec = tapply(as.numeric(tmp[,3]),as.factor(tmp[,1]),min)\n",
      "+ }\n",
      "> \n",
      "> if(ligandmode==TRUE & ligandCentroidMode) {\n",
      "+     liganddat = as.matrix(read.table(dir(\"./\",paste(\"_centroidNetLigand$\",sep=\"\")),sep=\"\\t\"))\n",
      "+     \n",
      "+     tmp = c()\n",
      "+     i=1\n",
      "+     for (i in 1:nrow(liganddat)) {\n",
      "+         tmp = rbind(tmp,c(liganddat[i,1],liganddat[i,2],liganddat[i,3]))\n",
      "+     }\n",
      "+     \n",
      "+     ligandvec = tapply(tmp[,3],as.factor(tmp[,1]),min)\n",
      "+ }\n",
      "> \n",
      "> if (ligandmode==FALSE) {\n",
      "+     ligandvec = rep(0,length(nodes))\n",
      "+ }\n",
      "> \n",
      "> out = cbind(out,ligandvec[rownames(out)])\n",
      "> colnames(out) = c(\"RSA\",\"Degree\",\"Degree_uniqueAtoms\",\"SecondOrderDegree\",\"SecondOrderDegree_uniqueAtoms\",\"NodeEdgeBetweennessSTRIDE\",\"NodeEdgeBetweennessSTRIDE_unqiueAtoms\",\"IntermodularDegreeSTRIDE\",\"IntermodularDegreeSTRIDE_uniqueAtoms\",\"SecondOrderIntermodularDegreeSTRIDE\",\"SecondOrderIntermodularDegreeSTRIDE_uniqueAtoms\",\"NodeEdgeBetweennessWALKTRAP\",\"NodeEdgeBetweennessWALKTRAP_uniqueAtoms\",\"IntermodularDegreeWALKTRAP\",\"IntermodularDegreeWALKTRAP_uniqueAtoms\",\"SecondOrderIntermodularDegreeWALKTRAP\",\"SecondOrderIntermodularDegreeWALKTRAP_uniqueAtoms\",\"Ligand\")\n",
      "> \n",
      "> #add back in nodes that were not networked as zeros\n",
      "> zeroMat = matrix(0,nrow=length(which(originalNodes%in%nodes==FALSE)),ncol=ncol(out));rownames(zeroMat)=originalNodes[which(originalNodes%in%nodes==FALSE)]\n",
      "> out = rbind(out,zeroMat)\n",
      "> \n",
      "> #fix RSA and Bfactor and ligand\n",
      "> out[,\"RSA\"] = rsa[rownames(out),1]\n",
      "> out[,\"Ligand\"] = ligandvec[rownames(out)]\n",
      "> \n",
      "> #Add 1 to everything to avoid zeros, except ligand\n",
      "> out[is.na(out)]=0\n",
      "> out[out[,\"Ligand\"]==0,\"Ligand\"] = 150\n",
      "> out[1:nrow(out),2:ncol(out)] = as.numeric(out[1:nrow(out),2:ncol(out)])+1\n",
      "> \n",
      "> ##create Z score file\n",
      "> outZ = apply(out,2,function(x){return(scale(as.numeric(x)))})\n",
      "> rownames(outZ) = rownames(out)\n",
      "> colnames(outZ) = colnames(out)\n",
      "> outZ[outZ==NA]=0\n",
      "> outZ[outZ==\"NaN\"]=0\n",
      "> \n",
      "> #Write out scores and z-normed scores files\n",
      "> write.table(out,file=paste(keyword,\"_scoresCentroid\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> write.table(outZ,file=paste(keyword,\"_scoresCentroidZ\",sep=\"\"),sep=\"\\t\",quote=FALSE,col.names=TRUE,row.names=TRUE)\n",
      "> \n",
      "> \n",
      "\n",
      "R version 4.0.5 (2021-03-31) -- \"Shake and Throw\"\n",
      "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-pc-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> ## run from within the monomer folder\n",
      "> ## there can only be one *_scoresEnergetics and one Centroid/*_scoresCentroid in this folder\n",
      "> ## args[1] is the folder for the multimer\n",
      "> ## NOTE: results may differ slightly from publication because of missing data in the mutagenesis experiments used in teh paper\n",
      "> \n",
      "> args <- commandArgs(trailingOnly = TRUE)\n",
      "> \n",
      "> #Load monomer\n",
      "> data1 = read.table(dir(\"./\",\"_scoresEnergetics$\"),header=T,row.names=1)\n",
      "> data1Z = as.data.frame(apply(data1,2,function(x){return(scale(as.numeric(x)))}))\n",
      "> rownames(data1Z) = rownames(data1)\n",
      "> data1Z[is.na(data1Z)] = 0\n",
      "> data2 = read.table(dir(\"./Centroid\",\"_scoresCentroid$\", full.names=TRUE),header=T,row.names=1)\n",
      "> data2Z = as.data.frame(apply(data2,2,function(x){return(scale(as.numeric(x)))}))\n",
      "> rownames(data2Z) = rownames(data2)\n",
      "> data2Z[is.na(data2Z)] = 0\n",
      "> colnames(data2Z) = paste(colnames(data2Z),\"CENTROIDSC\",sep=\"\")\n",
      "> data2Z = data2Z[rownames(data1Z),]\n",
      "> data2 = data2[rownames(data1),]\n",
      "> \n",
      "> #Load multimer\n",
      "> data1Multimer = read.table(dir(args[1],\"_scoresEnergetics$\",full.names=TRUE),header=T,row.names=1)\n",
      "> data2Multimer = read.table(dir(paste(args[1],\"Centroid\",sep=\"/\"),\"_scoresCentroid$\",full.names=TRUE),header=T,row.names=1)\n",
      "> data2Multimer = data2Multimer[rownames(data1Multimer),]\n",
      "> data1MultimerZ = as.data.frame(apply(data1Multimer,2,function(x){return(scale(as.numeric(x)))}))\n",
      "> rownames(data1MultimerZ) = rownames(data1Multimer)\n",
      "> data2MultimerZ = as.data.frame(apply(data2Multimer,2,function(x){return(scale(as.numeric(x)))}))\n",
      "> rownames(data2MultimerZ) = rownames(data2Multimer)\n",
      "> colnames(data1MultimerZ) = paste(colnames(data1MultimerZ),\"MULTIMER\",sep=\"\")\n",
      "> colnames(data2MultimerZ) = paste(colnames(data2MultimerZ),\"MULTIMERCENTROIDSC\",sep=\"\")\n",
      "> data2MultimerZ = data2MultimerZ[rownames(data1MultimerZ),]\n",
      "> \n",
      "> #Add RSA\n",
      "> data1 = cbind(data2[,\"RSA\"],data1)\n",
      "> colnames(data1)[1]=\"RSA\"\n",
      "> dataZ = cbind(data1Z,data2Z)\n",
      "> \n",
      "> #Identify max NodeEdgeBetweennessSTRIDE_sidechain\n",
      "> acidsMonomer = c()\n",
      "> for (i in 1:length(rownames(dataZ))) {\n",
      "+     node = rownames(dataZ)[i]\n",
      "+     node = unlist(strsplit(node,split=\"\"))\n",
      "+     node = paste(node[1:(length(node)-1)],collapse=\"\")\n",
      "+     acidsMonomer = c(acidsMonomer,node)\n",
      "+ }\n",
      "> \n",
      "> acidsMultimer = c()\n",
      "> for (i in 1:length(rownames(data1MultimerZ))) {\n",
      "+     node = rownames(data1MultimerZ)[i]\n",
      "+     node = unlist(strsplit(node,split=\"\"))\n",
      "+     node = paste(node[1:(length(node)-1)],collapse=\"\")\n",
      "+     acidsMultimer = c(acidsMultimer,node)\n",
      "+ }\n",
      "> \n",
      "> allAcids = unique(c(acidsMonomer,acidsMultimer))\n",
      "> \n",
      "> #SecondOrderIntermodularDegree is the average of the multimer only\n",
      "> SecondOrderIntermodularDegree_ALL = (data1MultimerZ[,\"SecondOrderIntermodularDegreeSTRIDE_sidechainMULTIMER\"]+data1MultimerZ[,\"SecondOrderIntermodularDegreeSTRIDEMULTIMER\"]+data2MultimerZ[,\"SecondOrderIntermodularDegreeSTRIDEMULTIMERCENTROIDSC\"]+data2MultimerZ[,\"SecondOrderIntermodularDegreeWALKTRAPMULTIMERCENTROIDSC\"])/4\n",
      "> SecondOrderIntermodularDegree_AVERAGE = tapply(SecondOrderIntermodularDegree_ALL,as.factor(acidsMultimer),mean)\n",
      "> SecondOrderIntermodularDegree_AVERAGE = SecondOrderIntermodularDegree_AVERAGE[allAcids]\n",
      "> missing = which(allAcids%in%acidsMultimer == FALSE)\n",
      "> SecondOrderIntermodularDegree_AVERAGE[missing] = (data1Z[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"SecondOrderIntermodularDegreeSTRIDE_sidechain\"]+data1Z[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"SecondOrderIntermodularDegreeSTRIDE\"]+data2Z[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"SecondOrderIntermodularDegreeSTRIDECENTROIDSC\"]+data2Z[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"SecondOrderIntermodularDegreeWALKTRAPCENTROIDSC\"])/4\n",
      "> names(SecondOrderIntermodularDegree_AVERAGE)[missing] = allAcids[which(allAcids%in%acidsMultimer == FALSE)]\n",
      "> \n",
      "> #NodeEdgeBetweenness is the max of the monomer/multimer\n",
      "> NodeEdgeBetweennessSTRIDE_AVERAGE_MONOMER = (dataZ[,\"NodeEdgeBetweennessSTRIDE\"]+dataZ[,\"NodeEdgeBetweennessSTRIDE_sidechain\"])/2\n",
      "> NodeEdgeBetweennessSTRIDE_AVERAGE_MULTIMER = (data1MultimerZ[,\"NodeEdgeBetweennessSTRIDEMULTIMER\"]+data1MultimerZ[,\"NodeEdgeBetweennessSTRIDE_sidechainMULTIMER\"])/2\n",
      "> NodeEdgeBetweennessSTRIDE_sidechain_ALL = c(NodeEdgeBetweennessSTRIDE_AVERAGE_MONOMER,NodeEdgeBetweennessSTRIDE_AVERAGE_MULTIMER)\n",
      "> NodeEdgeBetweennessSTRIDE_sidechain_MAX = tapply(NodeEdgeBetweennessSTRIDE_sidechain_ALL,as.factor(c(acidsMonomer,acidsMultimer)),max,na.rm=TRUE)\n",
      "> NodeEdgeBetweennessSTRIDE_sidechain_MAX = NodeEdgeBetweennessSTRIDE_sidechain_MAX[allAcids]\n",
      "> \n",
      "> #Ligand is the min of the multimer only\n",
      "> LigandMULTIMERCENTROIDSC_ALL = data2MultimerZ[,\"LigandMULTIMERCENTROIDSC\"]\n",
      "> LigandMULTIMERCENTROIDSC_MIN = tapply(LigandMULTIMERCENTROIDSC_ALL,as.factor(acidsMultimer),min)\n",
      "> LigandMULTIMERCENTROIDSC_MIN = LigandMULTIMERCENTROIDSC_MIN[allAcids]\n",
      "> missing = which(allAcids%in%acidsMultimer == FALSE)\n",
      "> LigandMULTIMERCENTROIDSC_MIN[missing] = dataZ[allAcids[which(allAcids%in%acidsMultimer == FALSE)],\"LigandCENTROIDSC\"]\n",
      "> names(LigandMULTIMERCENTROIDSC_MIN)[missing] = allAcids[which(allAcids%in%acidsMultimer == FALSE)]\n",
      "> LigandMULTIMERCENTROIDSC_MIN[is.na(LigandMULTIMERCENTROIDSC_MIN)] = 0\n",
      "> \n",
      "> #RSA\n",
      "> RSA_ALL = c(data1[,\"RSA\"],data2Multimer[,\"RSA\"])\n",
      "> RSA_MIN = tapply(RSA_ALL,as.factor(c(acidsMonomer,acidsMultimer)),min,na.rm=TRUE)\n",
      "> RSA_MIN = RSA_MIN[allAcids]\n",
      "> \n",
      "> y = SecondOrderIntermodularDegree_AVERAGE+NodeEdgeBetweennessSTRIDE_sidechain_MAX-LigandMULTIMERCENTROIDSC_MIN\n",
      "> \n",
      "> out=cbind(names(y),y)\n",
      "> acidNumbers = sapply(out[,1],substring,4)\n",
      "> out = out[order(as.numeric(acidNumbers)),]\n",
      "> write.table(out,file=\"FinalSum\",sep=\"\\t\",quote=FALSE,row.names=FALSE,col.names=FALSE)\n",
      "> \n",
      "> # also write the 3 variables to a file\n",
      "> # Concatenating all three variables into a single data frame\n",
      "> result <- data.frame(\n",
      "+   Acid = names(SecondOrderIntermodularDegree_AVERAGE),\n",
      "+   SecondOrderIntermodularDegree_AVERAGE,\n",
      "+   NodeEdgeBetweennessSTRIDE_sidechain_MAX,\n",
      "+   LigandMULTIMERCENTROIDSC_MIN\n",
      "+ )\n",
      "> \n",
      "> # Writing the data frame to a file\n",
      "> write.table(result, file = \"FinalSum_Decomp\", sep = \"\\t\", quote = FALSE, row.names = FALSE, col.names = TRUE)\n",
      "> \n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "Read 1512 items\n",
      "Read 1512 items\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "Read 7713 items\n",
      "Read 7713 items\n",
      "\n",
      "SBNA results for 7DHY succesfully downloaded\n"
     ]
    }
   ],
   "source": [
    "host = \"m3-dtn.massive.org.au\"\n",
    "username = \"yliy0004\"\n",
    "with open(\"../pw.txt\", \"r\") as f:\n",
    "    password = f.read().strip()\n",
    "\n",
    "client = paramiko.client.SSHClient()\n",
    "client.set_missing_host_key_policy(paramiko.AutoAddPolicy())\n",
    "client.connect(host, username=username, password=password)\n",
    "\n",
    "# check if R module is loaded (uses .bashrc - not the best solution because STRUDEL cannot be used, make sure to correct this if using STRUDEL)\n",
    "stdin,stdout,stderr=client.exec_command('R --version') \n",
    "print(stdout.read().decode())\n",
    "print(stderr.read().decode())\n",
    "\n",
    "# test the connection\n",
    "# stdin,stdout,stderr=client.exec_command('ls')\n",
    "# print(stdout.read().decode())\n",
    "# print(stderr.read().decode())\n",
    "\n",
    "sftp_client=client.open_sftp()\n",
    "\n",
    "\n",
    "# ==================== SBNA ====================\n",
    "# for _, row in tmp.iterrows():\n",
    "#     pdb_id = row['PDB']\n",
    "#     chains = row['Chain']\n",
    "# for pdb_id in pdbs:\n",
    "#     try:\n",
    "#         calculate_sbna_and_download(client, sftp_client, pdb_id) #chains=chains)\n",
    "#     except Exception as e:\n",
    "#         print(f\"Failed to process {pdb_id} with error:\")\n",
    "#         print(traceback.print_exc())\n",
    "#         continue\n",
    "\n",
    "# for pdb, chains in pdb_chain_dict.items():\n",
    "# for pdb in pdbs:\n",
    "for pdb in ['7DHY']:\n",
    "    try:\n",
    "        calculate_sbna_and_download(client, sftp_client, pdb), #chains=chains)\n",
    "    except Exception as e:\n",
    "        print(f\"Failed to process {pdb} with error:\")\n",
    "        print(traceback.print_exc())\n",
    "        continue\n",
    "\n",
    "# close the connection\n",
    "client.close()\n",
    "sftp_client.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>resolution</th>\n",
       "      <th>chain</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>assoc_gene</th>\n",
       "      <th>length</th>\n",
       "      <th>mutant</th>\n",
       "      <th>mutations</th>\n",
       "      <th>pair</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>7SI1</td>\n",
       "      <td>1.60</td>\n",
       "      <td>A</td>\n",
       "      <td>695</td>\n",
       "      <td>1022</td>\n",
       "      <td>P00533</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>328</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>7UKV</td>\n",
       "      <td>2.40</td>\n",
       "      <td>A</td>\n",
       "      <td>695</td>\n",
       "      <td>1022</td>\n",
       "      <td>P00533</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>328</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>3POZ</td>\n",
       "      <td>1.50</td>\n",
       "      <td>A</td>\n",
       "      <td>696</td>\n",
       "      <td>1022</td>\n",
       "      <td>P00533</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>327</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>8F1X</td>\n",
       "      <td>2.30</td>\n",
       "      <td>A</td>\n",
       "      <td>695</td>\n",
       "      <td>1022</td>\n",
       "      <td>P00533</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>328</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>4LQM</td>\n",
       "      <td>2.50</td>\n",
       "      <td>A</td>\n",
       "      <td>694</td>\n",
       "      <td>1022</td>\n",
       "      <td>P00533</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>329</td>\n",
       "      <td>Y</td>\n",
       "      <td>L858R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>ALK</td>\n",
       "      <td>4CTB</td>\n",
       "      <td>1.79</td>\n",
       "      <td>A</td>\n",
       "      <td>1093</td>\n",
       "      <td>1411</td>\n",
       "      <td>Q9UM73</td>\n",
       "      <td>ALK</td>\n",
       "      <td>319</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>ALK</td>\n",
       "      <td>2YJR</td>\n",
       "      <td>1.90</td>\n",
       "      <td>A</td>\n",
       "      <td>1093</td>\n",
       "      <td>1411</td>\n",
       "      <td>Q9UM73</td>\n",
       "      <td>ALK</td>\n",
       "      <td>319</td>\n",
       "      <td>Y</td>\n",
       "      <td>F1174L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>ALK</td>\n",
       "      <td>7R7R</td>\n",
       "      <td>1.94</td>\n",
       "      <td>A</td>\n",
       "      <td>1093</td>\n",
       "      <td>1411</td>\n",
       "      <td>Q9UM73</td>\n",
       "      <td>ALK</td>\n",
       "      <td>319</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>ALK</td>\n",
       "      <td>5IMX</td>\n",
       "      <td>2.12</td>\n",
       "      <td>A</td>\n",
       "      <td>1093</td>\n",
       "      <td>1411</td>\n",
       "      <td>Q9UM73</td>\n",
       "      <td>ALK</td>\n",
       "      <td>319</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>ALK</td>\n",
       "      <td>6CDT</td>\n",
       "      <td>1.80</td>\n",
       "      <td>A</td>\n",
       "      <td>1093</td>\n",
       "      <td>1411</td>\n",
       "      <td>Q9UM73</td>\n",
       "      <td>ALK</td>\n",
       "      <td>319</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>259 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene pdb_id  resolution chain  seq_start  seq_end uniprot_id assoc_gene  \\\n",
       "0    EGFR   7SI1        1.60     A        695     1022     P00533       EGFR   \n",
       "1    EGFR   7UKV        2.40     A        695     1022     P00533       EGFR   \n",
       "2    EGFR   3POZ        1.50     A        696     1022     P00533       EGFR   \n",
       "3    EGFR   8F1X        2.30     A        695     1022     P00533       EGFR   \n",
       "4    EGFR   4LQM        2.50     A        694     1022     P00533       EGFR   \n",
       "..    ...    ...         ...   ...        ...      ...        ...        ...   \n",
       "254   ALK   4CTB        1.79     A       1093     1411     Q9UM73        ALK   \n",
       "255   ALK   2YJR        1.90     A       1093     1411     Q9UM73        ALK   \n",
       "256   ALK   7R7R        1.94     A       1093     1411     Q9UM73        ALK   \n",
       "257   ALK   5IMX        2.12     A       1093     1411     Q9UM73        ALK   \n",
       "258   ALK   6CDT        1.80     A       1093     1411     Q9UM73        ALK   \n",
       "\n",
       "     length mutant mutations pair notes  \n",
       "0       328      N       NaN  NaN   NaN  \n",
       "1       328      N       NaN  NaN   NaN  \n",
       "2       327      N       NaN  NaN   NaN  \n",
       "3       328      N       NaN  NaN   NaN  \n",
       "4       329      Y     L858R  NaN   NaN  \n",
       "..      ...    ...       ...  ...   ...  \n",
       "254     319      N       NaN  NaN   NaN  \n",
       "255     319      Y    F1174L  NaN   NaN  \n",
       "256     319      N       NaN  NaN   NaN  \n",
       "257     319      N       NaN  NaN   NaN  \n",
       "258     319      N       NaN  NaN   NaN  \n",
       "\n",
       "[259 rows x 13 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read lung_pdb file\n",
    "lung_pdbs = pd.read_csv(\"lung genes.csv\")\n",
    "lung_pdbs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to calculate SBNA for:\n",
      "EGFR 4LQM A\n",
      "ERBB4 2AHX A\n",
      "ERBB4 2AHX B\n",
      "ERBB4 2R4B A\n",
      "ERBB4 2R4B B\n",
      "DDR2 6FER A\n",
      "DDR2 6FER B\n",
      "DDR2 6FER C\n",
      "DDR2 6FER D\n",
      "DDR2 6FER E\n",
      "DDR2 6FER F\n",
      "DDR2 6FER G\n",
      "DDR2 6FER H\n",
      "DDR2 6FER I\n",
      "DDR2 6FER J\n",
      "DDR2 6FER K\n",
      "DDR2 6FER L\n",
      "PTPRT 2OOQ A\n",
      "PTPRT 2OOQ B\n"
     ]
    }
   ],
   "source": [
    "# # check if sbna results were downloaded\n",
    "print(\"Failed to calculate SBNA for:\")\n",
    "for _, row in lung_pdbs.iterrows():\n",
    "    pdb_id = row['pdb_id']\n",
    "    chains = row['chain']\n",
    "    for chain in chains:\n",
    "    # check if the file already exists\n",
    "        if not os.path.exists(f'../sbna_results/{pdb_id}/{chain}'):\n",
    "            # if not dwl_pdb_file(pdb_id): # returns True or False if the file was downloaded\n",
    "            print(row['gene'], pdb_id, chain)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### aligning for one pdb only (is the next section but singular)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>resolution</th>\n",
       "      <th>chain</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>assoc_gene</th>\n",
       "      <th>length</th>\n",
       "      <th>mutant</th>\n",
       "      <th>mutations</th>\n",
       "      <th>pair</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>7SI1</td>\n",
       "      <td>1.6</td>\n",
       "      <td>A</td>\n",
       "      <td>695</td>\n",
       "      <td>1022</td>\n",
       "      <td>P00533</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>328</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>7UKV</td>\n",
       "      <td>2.4</td>\n",
       "      <td>A</td>\n",
       "      <td>695</td>\n",
       "      <td>1022</td>\n",
       "      <td>P00533</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>328</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene pdb_id  resolution chain  seq_start  seq_end uniprot_id assoc_gene  \\\n",
       "0  EGFR   7SI1         1.6     A        695     1022     P00533       EGFR   \n",
       "1  EGFR   7UKV         2.4     A        695     1022     P00533       EGFR   \n",
       "\n",
       "   length mutant mutations pair notes  \n",
       "0     328      N       NaN  NaN   NaN  \n",
       "1     328      N       NaN  NaN   NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read lung_pdb file\n",
    "lung_pdbs = pd.read_csv(\"lung genes.csv\")\n",
    "lung_pdbs.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A\n",
      "   res res_code  num  network_score\n",
      "0  SER        S   96      -1.126763\n",
      "1  VAL        V   97       0.411958\n",
      "2  PRO        P   98       2.463157\n",
      "     Acid  SecondOrderIntermodularDegree_AVERAGE  \\\n",
      "0  CYS229                               0.161653   \n",
      "1  TRP146                              -0.083144   \n",
      "2  ARG273                               0.125345   \n",
      "\n",
      "   NodeEdgeBetweennessSTRIDE_sidechain_MAX  LigandMULTIMERCENTROIDSC_MIN  num  \\\n",
      "0                                 1.143199                     -1.015263  229   \n",
      "1                                 0.727006                     -1.779925  146   \n",
      "2                                 0.339070                     -0.482529  273   \n",
      "\n",
      "   res res_code  \n",
      "0  CYS        C  \n",
      "1  TRP        W  \n",
      "2  ARG        R  \n",
      "B\n",
      "   res res_code  num  network_score\n",
      "0  SER        S   96      -1.126763\n",
      "1  VAL        V   97       0.206682\n",
      "2  PRO        P   98       2.042096\n",
      "     Acid  SecondOrderIntermodularDegree_AVERAGE  \\\n",
      "0  GLU271                               0.295379   \n",
      "1  LEU252                               0.408331   \n",
      "2  ARG249                               0.177501   \n",
      "\n",
      "   NodeEdgeBetweennessSTRIDE_sidechain_MAX  LigandMULTIMERCENTROIDSC_MIN  num  \\\n",
      "0                                 1.144187                     -0.296656  271   \n",
      "1                                 0.819215                      0.091743  252   \n",
      "2                                 2.260617                      0.557834  249   \n",
      "\n",
      "   res res_code  \n",
      "0  GLU        E  \n",
      "1  LEU        L  \n",
      "2  ARG        R  \n",
      "C\n",
      "   res res_code  num  network_score\n",
      "0  SER        S   96      -1.126763\n",
      "1  VAL        V   97      -0.246530\n",
      "2  PRO        P   98       2.666510\n",
      "     Acid  SecondOrderIntermodularDegree_AVERAGE  \\\n",
      "0  PRO190                              -0.314470   \n",
      "1  TYR205                               1.165403   \n",
      "2  SER149                              -0.325369   \n",
      "\n",
      "   NodeEdgeBetweennessSTRIDE_sidechain_MAX  LigandMULTIMERCENTROIDSC_MIN  num  \\\n",
      "0                                 1.711636                     -0.822114  190   \n",
      "1                                 1.370607                     -0.251305  205   \n",
      "2                                -0.417052                     -0.014318  149   \n",
      "\n",
      "   res res_code  \n",
      "0  PRO        P  \n",
      "1  TYR        Y  \n",
      "2  SER        S  \n",
      "D\n",
      "   res res_code  num  network_score\n",
      "0  SER        S   96      -1.126763\n",
      "1  VAL        V   97      -0.242941\n",
      "2  PRO        P   98       1.269103\n",
      "     Acid  SecondOrderIntermodularDegree_AVERAGE  \\\n",
      "0  ILE255                               0.799098   \n",
      "1  VAL157                               1.723400   \n",
      "2  LEU257                               1.284836   \n",
      "\n",
      "   NodeEdgeBetweennessSTRIDE_sidechain_MAX  LigandMULTIMERCENTROIDSC_MIN  num  \\\n",
      "0                                 0.144977                     -0.289410  255   \n",
      "1                                 2.991488                     -0.749197  157   \n",
      "2                                 1.392082                     -0.285457  257   \n",
      "\n",
      "   res res_code  \n",
      "0  ILE        I  \n",
      "1  VAL        V  \n",
      "2  LEU        L  \n"
     ]
    }
   ],
   "source": [
    "pdb_id = '7DHY'\n",
    "\n",
    "for chain in [x['chain'] for x in get_dbref_data(pdb_id)]:\n",
    "    print(chain)\n",
    "    print(read_finalsum(pdb_id, chain).head(3))\n",
    "    print(read_finalsum_decomp(pdb_id, chain).head(3))\n",
    "# read_finalsum(pdb_id, chain).merge(read_finalsum_decomp(pdb_id, chain), on=['num', 'res'], how='inner').drop('res_code_x', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "784\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>chain</th>\n",
       "      <th>sbna_aa3</th>\n",
       "      <th>sbna_num</th>\n",
       "      <th>sbna_aa1</th>\n",
       "      <th>uniprot_num</th>\n",
       "      <th>uniprot_aa</th>\n",
       "      <th>network_score</th>\n",
       "      <th>SecondOrderIntermodularDegree_AVERAGE</th>\n",
       "      <th>NodeEdgeBetweennessSTRIDE_sidechain_MAX</th>\n",
       "      <th>LigandMULTIMERCENTROIDSC_MIN</th>\n",
       "      <th>mutant</th>\n",
       "      <th>mutations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7DHY</td>\n",
       "      <td>A</td>\n",
       "      <td>SER</td>\n",
       "      <td>96</td>\n",
       "      <td>S</td>\n",
       "      <td>94</td>\n",
       "      <td>S</td>\n",
       "      <td>-1.126763</td>\n",
       "      <td>-0.768789</td>\n",
       "      <td>-0.417127</td>\n",
       "      <td>-0.059153</td>\n",
       "      <td>Y</td>\n",
       "      <td>G245S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7DHY</td>\n",
       "      <td>A</td>\n",
       "      <td>VAL</td>\n",
       "      <td>97</td>\n",
       "      <td>V</td>\n",
       "      <td>97</td>\n",
       "      <td>V</td>\n",
       "      <td>0.411958</td>\n",
       "      <td>-0.573712</td>\n",
       "      <td>0.448125</td>\n",
       "      <td>-0.537545</td>\n",
       "      <td>Y</td>\n",
       "      <td>G245S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7DHY</td>\n",
       "      <td>A</td>\n",
       "      <td>PRO</td>\n",
       "      <td>98</td>\n",
       "      <td>P</td>\n",
       "      <td>98</td>\n",
       "      <td>P</td>\n",
       "      <td>2.463157</td>\n",
       "      <td>-0.170044</td>\n",
       "      <td>1.897883</td>\n",
       "      <td>-0.735318</td>\n",
       "      <td>Y</td>\n",
       "      <td>G245S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7DHY</td>\n",
       "      <td>A</td>\n",
       "      <td>SER</td>\n",
       "      <td>99</td>\n",
       "      <td>S</td>\n",
       "      <td>99</td>\n",
       "      <td>S</td>\n",
       "      <td>0.848752</td>\n",
       "      <td>-0.720570</td>\n",
       "      <td>-0.347014</td>\n",
       "      <td>-1.916336</td>\n",
       "      <td>Y</td>\n",
       "      <td>G245S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7DHY</td>\n",
       "      <td>A</td>\n",
       "      <td>GLN</td>\n",
       "      <td>100</td>\n",
       "      <td>Q</td>\n",
       "      <td>100</td>\n",
       "      <td>Q</td>\n",
       "      <td>0.547100</td>\n",
       "      <td>-0.353719</td>\n",
       "      <td>0.184136</td>\n",
       "      <td>-0.716683</td>\n",
       "      <td>Y</td>\n",
       "      <td>G245S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene pdb_id chain sbna_aa3  sbna_num sbna_aa1  uniprot_num uniprot_aa  \\\n",
       "0  TP53   7DHY     A      SER        96        S           94          S   \n",
       "1  TP53   7DHY     A      VAL        97        V           97          V   \n",
       "2  TP53   7DHY     A      PRO        98        P           98          P   \n",
       "3  TP53   7DHY     A      SER        99        S           99          S   \n",
       "4  TP53   7DHY     A      GLN       100        Q          100          Q   \n",
       "\n",
       "   network_score  SecondOrderIntermodularDegree_AVERAGE  \\\n",
       "0      -1.126763                              -0.768789   \n",
       "1       0.411958                              -0.573712   \n",
       "2       2.463157                              -0.170044   \n",
       "3       0.848752                              -0.720570   \n",
       "4       0.547100                              -0.353719   \n",
       "\n",
       "   NodeEdgeBetweennessSTRIDE_sidechain_MAX  LigandMULTIMERCENTROIDSC_MIN  \\\n",
       "0                                -0.417127                     -0.059153   \n",
       "1                                 0.448125                     -0.537545   \n",
       "2                                 1.897883                     -0.735318   \n",
       "3                                -0.347014                     -1.916336   \n",
       "4                                 0.184136                     -0.716683   \n",
       "\n",
       "  mutant mutations  \n",
       "0      Y     G245S  \n",
       "1      Y     G245S  \n",
       "2      Y     G245S  \n",
       "3      Y     G245S  \n",
       "4      Y     G245S  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ns_data = pd.DataFrame()\n",
    "\n",
    "for _, row in lung_pdbs[lung_pdbs['pdb_id']==pdb_id].iterrows():\n",
    "    pdb_id = row['pdb_id']\n",
    "    chain = row['chain']\n",
    "    # print(pdb_id, chain)\n",
    "    try:\n",
    "        tmp = read_finalsum(pdb_id, chain).merge(read_finalsum_decomp(pdb_id, chain), on=['num', 'res'], how='inner').drop('res_code_x', axis=1)\n",
    "    except:\n",
    "        continue\n",
    "    # rename columns\n",
    "    tmp = tmp.rename(columns={\"res_code_y\": \"res_code\"})\n",
    "    tmp = align_finalsum_with_uniprot(tmp, pdb_id, chain)\n",
    "    # tmp = tmp[tmp['num']==12].reset_index(drop=True)\n",
    "    tmp['pdb_id'] = pdb_id\n",
    "    tmp['chain'] = chain\n",
    "    tmp['gene'] = row['gene']\n",
    "    tmp['mutant'] = row['mutant']\n",
    "    tmp['mutations'] = row['mutations']\n",
    "    ns_data = pd.concat([ns_data, tmp])\n",
    "\n",
    "# rename columns\n",
    "ns_data = ns_data.rename(columns={\"res\": \"sbna_aa3\", \"res_code\": \"sbna_aa1\", \"num\": \"sbna_num\", \"uniprot_res\": \"uniprot_aa\"})\n",
    "ns_data = ns_data[[\"gene\", \"pdb_id\", \"chain\", \"sbna_aa3\", \"sbna_num\", \"sbna_aa1\", \"uniprot_num\", \"uniprot_aa\", \"network_score\", \"SecondOrderIntermodularDegree_AVERAGE\", \"NodeEdgeBetweennessSTRIDE_sidechain_MAX\", \"LigandMULTIMERCENTROIDSC_MIN\", 'mutant', 'mutations']]\n",
    "\n",
    "print(len(ns_data))\n",
    "ns_data.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'I': 'Hydrophobic',\n",
       " 'A': 'Hydrophobic',\n",
       " 'F': 'Hydrophobic',\n",
       " 'L': 'Hydrophobic',\n",
       " 'M': 'Hydrophobic',\n",
       " 'P': 'Special cases',\n",
       " 'V': 'Hydrophobic',\n",
       " 'W': 'Hydrophobic',\n",
       " 'C': 'Special cases',\n",
       " 'G': 'Special cases',\n",
       " 'T': 'Polar uncharged',\n",
       " 'S': 'Polar uncharged',\n",
       " 'Y': 'Hydrophobic',\n",
       " 'Q': 'Polar uncharged',\n",
       " 'N': 'Polar uncharged',\n",
       " 'H': 'Polar charged positive',\n",
       " 'K': 'Polar charged positive',\n",
       " 'E': 'Polar charged negative',\n",
       " 'R': 'Polar charged positive',\n",
       " 'D': 'Polar charged negative'}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_properties = pd.read_csv(\"../data_prep/aa_properties.csv\")\n",
    "# make dictionary of aa1:Acid Type\n",
    "aa_properties = dict(zip(aa_properties['aa1'], aa_properties['Acid Type']))\n",
    "aa_properties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>chain</th>\n",
       "      <th>sbna_aa3</th>\n",
       "      <th>sbna_num</th>\n",
       "      <th>sbna_aa1</th>\n",
       "      <th>uniprot_num</th>\n",
       "      <th>uniprot_aa</th>\n",
       "      <th>network_score</th>\n",
       "      <th>SecondOrderIntermodularDegree_AVERAGE</th>\n",
       "      <th>NodeEdgeBetweennessSTRIDE_sidechain_MAX</th>\n",
       "      <th>LigandMULTIMERCENTROIDSC_MIN</th>\n",
       "      <th>mutant</th>\n",
       "      <th>mutations</th>\n",
       "      <th>cgc_tier</th>\n",
       "      <th>cancer_role</th>\n",
       "      <th>uniprot_aa_type</th>\n",
       "      <th>sbna_aa_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7DHY</td>\n",
       "      <td>A</td>\n",
       "      <td>SER</td>\n",
       "      <td>96</td>\n",
       "      <td>S</td>\n",
       "      <td>94</td>\n",
       "      <td>S</td>\n",
       "      <td>-1.126763</td>\n",
       "      <td>-0.768789</td>\n",
       "      <td>-0.417127</td>\n",
       "      <td>-0.059153</td>\n",
       "      <td>Y</td>\n",
       "      <td>G245S</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>Polar uncharged</td>\n",
       "      <td>Polar uncharged</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7DHY</td>\n",
       "      <td>A</td>\n",
       "      <td>VAL</td>\n",
       "      <td>97</td>\n",
       "      <td>V</td>\n",
       "      <td>97</td>\n",
       "      <td>V</td>\n",
       "      <td>0.411958</td>\n",
       "      <td>-0.573712</td>\n",
       "      <td>0.448125</td>\n",
       "      <td>-0.537545</td>\n",
       "      <td>Y</td>\n",
       "      <td>G245S</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>Hydrophobic</td>\n",
       "      <td>Hydrophobic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7DHY</td>\n",
       "      <td>A</td>\n",
       "      <td>PRO</td>\n",
       "      <td>98</td>\n",
       "      <td>P</td>\n",
       "      <td>98</td>\n",
       "      <td>P</td>\n",
       "      <td>2.463157</td>\n",
       "      <td>-0.170044</td>\n",
       "      <td>1.897883</td>\n",
       "      <td>-0.735318</td>\n",
       "      <td>Y</td>\n",
       "      <td>G245S</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>Special cases</td>\n",
       "      <td>Special cases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7DHY</td>\n",
       "      <td>A</td>\n",
       "      <td>SER</td>\n",
       "      <td>99</td>\n",
       "      <td>S</td>\n",
       "      <td>99</td>\n",
       "      <td>S</td>\n",
       "      <td>0.848752</td>\n",
       "      <td>-0.720570</td>\n",
       "      <td>-0.347014</td>\n",
       "      <td>-1.916336</td>\n",
       "      <td>Y</td>\n",
       "      <td>G245S</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>Polar uncharged</td>\n",
       "      <td>Polar uncharged</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7DHY</td>\n",
       "      <td>A</td>\n",
       "      <td>GLN</td>\n",
       "      <td>100</td>\n",
       "      <td>Q</td>\n",
       "      <td>100</td>\n",
       "      <td>Q</td>\n",
       "      <td>0.547100</td>\n",
       "      <td>-0.353719</td>\n",
       "      <td>0.184136</td>\n",
       "      <td>-0.716683</td>\n",
       "      <td>Y</td>\n",
       "      <td>G245S</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>Polar uncharged</td>\n",
       "      <td>Polar uncharged</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7DHY</td>\n",
       "      <td>D</td>\n",
       "      <td>GLU</td>\n",
       "      <td>287</td>\n",
       "      <td>E</td>\n",
       "      <td>287</td>\n",
       "      <td>E</td>\n",
       "      <td>-0.052733</td>\n",
       "      <td>-1.013234</td>\n",
       "      <td>-0.466253</td>\n",
       "      <td>-1.426754</td>\n",
       "      <td>Y</td>\n",
       "      <td>G245S</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>Polar charged negative</td>\n",
       "      <td>Polar charged negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7DHY</td>\n",
       "      <td>D</td>\n",
       "      <td>ASN</td>\n",
       "      <td>288</td>\n",
       "      <td>N</td>\n",
       "      <td>288</td>\n",
       "      <td>N</td>\n",
       "      <td>-1.089911</td>\n",
       "      <td>-1.169358</td>\n",
       "      <td>-0.441454</td>\n",
       "      <td>-0.520901</td>\n",
       "      <td>Y</td>\n",
       "      <td>G245S</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>Polar uncharged</td>\n",
       "      <td>Polar uncharged</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7DHY</td>\n",
       "      <td>D</td>\n",
       "      <td>LEU</td>\n",
       "      <td>289</td>\n",
       "      <td>L</td>\n",
       "      <td>289</td>\n",
       "      <td>L</td>\n",
       "      <td>-0.687241</td>\n",
       "      <td>-0.748215</td>\n",
       "      <td>-0.118174</td>\n",
       "      <td>-0.179147</td>\n",
       "      <td>Y</td>\n",
       "      <td>G245S</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>Hydrophobic</td>\n",
       "      <td>Hydrophobic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7DHY</td>\n",
       "      <td>D</td>\n",
       "      <td>ARG</td>\n",
       "      <td>290</td>\n",
       "      <td>R</td>\n",
       "      <td>290</td>\n",
       "      <td>R</td>\n",
       "      <td>-4.354151</td>\n",
       "      <td>-1.169271</td>\n",
       "      <td>-0.466253</td>\n",
       "      <td>2.718628</td>\n",
       "      <td>Y</td>\n",
       "      <td>G245S</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>Polar charged positive</td>\n",
       "      <td>Polar charged positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>TP53</td>\n",
       "      <td>7DHY</td>\n",
       "      <td>D</td>\n",
       "      <td>LYS</td>\n",
       "      <td>291</td>\n",
       "      <td>K</td>\n",
       "      <td>291</td>\n",
       "      <td>K</td>\n",
       "      <td>-4.405226</td>\n",
       "      <td>-1.169271</td>\n",
       "      <td>-0.466253</td>\n",
       "      <td>2.769703</td>\n",
       "      <td>Y</td>\n",
       "      <td>G245S</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>Polar charged positive</td>\n",
       "      <td>Polar charged positive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>784 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene pdb_id chain sbna_aa3  sbna_num sbna_aa1  uniprot_num uniprot_aa  \\\n",
       "0    TP53   7DHY     A      SER        96        S           94          S   \n",
       "1    TP53   7DHY     A      VAL        97        V           97          V   \n",
       "2    TP53   7DHY     A      PRO        98        P           98          P   \n",
       "3    TP53   7DHY     A      SER        99        S           99          S   \n",
       "4    TP53   7DHY     A      GLN       100        Q          100          Q   \n",
       "..    ...    ...   ...      ...       ...      ...          ...        ...   \n",
       "191  TP53   7DHY     D      GLU       287        E          287          E   \n",
       "192  TP53   7DHY     D      ASN       288        N          288          N   \n",
       "193  TP53   7DHY     D      LEU       289        L          289          L   \n",
       "194  TP53   7DHY     D      ARG       290        R          290          R   \n",
       "195  TP53   7DHY     D      LYS       291        K          291          K   \n",
       "\n",
       "     network_score  SecondOrderIntermodularDegree_AVERAGE  \\\n",
       "0        -1.126763                              -0.768789   \n",
       "1         0.411958                              -0.573712   \n",
       "2         2.463157                              -0.170044   \n",
       "3         0.848752                              -0.720570   \n",
       "4         0.547100                              -0.353719   \n",
       "..             ...                                    ...   \n",
       "191      -0.052733                              -1.013234   \n",
       "192      -1.089911                              -1.169358   \n",
       "193      -0.687241                              -0.748215   \n",
       "194      -4.354151                              -1.169271   \n",
       "195      -4.405226                              -1.169271   \n",
       "\n",
       "     NodeEdgeBetweennessSTRIDE_sidechain_MAX  LigandMULTIMERCENTROIDSC_MIN  \\\n",
       "0                                  -0.417127                     -0.059153   \n",
       "1                                   0.448125                     -0.537545   \n",
       "2                                   1.897883                     -0.735318   \n",
       "3                                  -0.347014                     -1.916336   \n",
       "4                                   0.184136                     -0.716683   \n",
       "..                                       ...                           ...   \n",
       "191                                -0.466253                     -1.426754   \n",
       "192                                -0.441454                     -0.520901   \n",
       "193                                -0.118174                     -0.179147   \n",
       "194                                -0.466253                      2.718628   \n",
       "195                                -0.466253                      2.769703   \n",
       "\n",
       "    mutant mutations  cgc_tier cancer_role         uniprot_aa_type  \\\n",
       "0        Y     G245S         1         TSG         Polar uncharged   \n",
       "1        Y     G245S         1         TSG             Hydrophobic   \n",
       "2        Y     G245S         1         TSG           Special cases   \n",
       "3        Y     G245S         1         TSG         Polar uncharged   \n",
       "4        Y     G245S         1         TSG         Polar uncharged   \n",
       "..     ...       ...       ...         ...                     ...   \n",
       "191      Y     G245S         1         TSG  Polar charged negative   \n",
       "192      Y     G245S         1         TSG         Polar uncharged   \n",
       "193      Y     G245S         1         TSG             Hydrophobic   \n",
       "194      Y     G245S         1         TSG  Polar charged positive   \n",
       "195      Y     G245S         1         TSG  Polar charged positive   \n",
       "\n",
       "               sbna_aa_type  \n",
       "0           Polar uncharged  \n",
       "1               Hydrophobic  \n",
       "2             Special cases  \n",
       "3           Polar uncharged  \n",
       "4           Polar uncharged  \n",
       "..                      ...  \n",
       "191  Polar charged negative  \n",
       "192         Polar uncharged  \n",
       "193             Hydrophobic  \n",
       "194  Polar charged positive  \n",
       "195  Polar charged positive  \n",
       "\n",
       "[784 rows x 18 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ns_data['cgc_tier'] = 1\n",
    "ns_data['cancer_role'] = \"TSG\"\n",
    "ns_data[\"uniprot_aa_type\"] = ns_data['uniprot_aa'].map(aa_properties)\n",
    "ns_data[\"sbna_aa_type\"] = ns_data['sbna_aa1'].map(aa_properties)\n",
    "ns_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "ns_data.to_csv(\"tmp.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# merge final_sum with cosmic data (adding cgc_tier and cancer_role)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>resolution</th>\n",
       "      <th>chain</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>seq_end</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>assoc_gene</th>\n",
       "      <th>length</th>\n",
       "      <th>mutant</th>\n",
       "      <th>mutations</th>\n",
       "      <th>pair</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>7SI1</td>\n",
       "      <td>1.6</td>\n",
       "      <td>A</td>\n",
       "      <td>695</td>\n",
       "      <td>1022</td>\n",
       "      <td>P00533</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>328</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>7UKV</td>\n",
       "      <td>2.4</td>\n",
       "      <td>A</td>\n",
       "      <td>695</td>\n",
       "      <td>1022</td>\n",
       "      <td>P00533</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>328</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene pdb_id  resolution chain  seq_start  seq_end uniprot_id assoc_gene  \\\n",
       "0  EGFR   7SI1         1.6     A        695     1022     P00533       EGFR   \n",
       "1  EGFR   7UKV         2.4     A        695     1022     P00533       EGFR   \n",
       "\n",
       "   length mutant mutations pair notes  \n",
       "0     328      N       NaN  NaN   NaN  \n",
       "1     328      N       NaN  NaN   NaN  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read lung_pdb file\n",
    "lung_pdbs = pd.read_csv(\"lung genes.csv\")\n",
    "lung_pdbs.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Tier</th>\n",
       "      <th>Role in Cancer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>2</td>\n",
       "      <td>oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG, fusion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene Symbol  Tier Role in Cancer\n",
       "0        A1CF     2       oncogene\n",
       "1        ABI1     1    TSG, fusion"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# first add  cgc_tier and cancer_role\n",
    "cgc_data = pd.read_csv(\"../data_prep/Census_allTue Apr  9 04_58_38 2024.csv\", usecols=['Gene Symbol', 'Tier', 'Role in Cancer'])\n",
    "cgc_data.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FinalSum file not found for 4LQM\n",
      "FinalSum file not found for 2AHX\n",
      "FinalSum file not found for 2AHX\n",
      "FinalSum file not found for 2R4B\n",
      "FinalSum file not found for 2R4B\n",
      "FinalSum file not found for 6FER\n",
      "FinalSum file not found for 6FER\n",
      "FinalSum file not found for 6FER\n",
      "FinalSum file not found for 6FER\n",
      "FinalSum file not found for 6FER\n",
      "FinalSum file not found for 6FER\n",
      "FinalSum file not found for 6FER\n",
      "FinalSum file not found for 6FER\n",
      "FinalSum file not found for 6FER\n",
      "FinalSum file not found for 6FER\n",
      "FinalSum file not found for 6FER\n",
      "FinalSum file not found for 6FER\n",
      "FinalSum file not found for 2OOQ\n",
      "FinalSum file not found for 2OOQ\n",
      "60305\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>chain</th>\n",
       "      <th>sbna_aa3</th>\n",
       "      <th>sbna_num</th>\n",
       "      <th>sbna_aa1</th>\n",
       "      <th>uniprot_num</th>\n",
       "      <th>uniprot_aa</th>\n",
       "      <th>network_score</th>\n",
       "      <th>SecondOrderIntermodularDegree_AVERAGE</th>\n",
       "      <th>NodeEdgeBetweennessSTRIDE_sidechain_MAX</th>\n",
       "      <th>LigandMULTIMERCENTROIDSC_MIN</th>\n",
       "      <th>mutant</th>\n",
       "      <th>mutations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>7SI1</td>\n",
       "      <td>A</td>\n",
       "      <td>SER</td>\n",
       "      <td>695</td>\n",
       "      <td>S</td>\n",
       "      <td>695</td>\n",
       "      <td>S</td>\n",
       "      <td>-1.946806</td>\n",
       "      <td>-0.964584</td>\n",
       "      <td>-0.368011</td>\n",
       "      <td>0.614212</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>7SI1</td>\n",
       "      <td>A</td>\n",
       "      <td>GLY</td>\n",
       "      <td>696</td>\n",
       "      <td>G</td>\n",
       "      <td>696</td>\n",
       "      <td>G</td>\n",
       "      <td>-1.901057</td>\n",
       "      <td>-0.900271</td>\n",
       "      <td>-0.530286</td>\n",
       "      <td>0.470500</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>7SI1</td>\n",
       "      <td>A</td>\n",
       "      <td>GLU</td>\n",
       "      <td>697</td>\n",
       "      <td>E</td>\n",
       "      <td>697</td>\n",
       "      <td>E</td>\n",
       "      <td>-0.544335</td>\n",
       "      <td>-0.288050</td>\n",
       "      <td>-0.341759</td>\n",
       "      <td>-0.085474</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>7SI1</td>\n",
       "      <td>A</td>\n",
       "      <td>ALA</td>\n",
       "      <td>698</td>\n",
       "      <td>A</td>\n",
       "      <td>698</td>\n",
       "      <td>A</td>\n",
       "      <td>-2.229062</td>\n",
       "      <td>-1.095856</td>\n",
       "      <td>-0.596766</td>\n",
       "      <td>0.536440</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>7SI1</td>\n",
       "      <td>A</td>\n",
       "      <td>PRO</td>\n",
       "      <td>699</td>\n",
       "      <td>P</td>\n",
       "      <td>699</td>\n",
       "      <td>P</td>\n",
       "      <td>-2.403619</td>\n",
       "      <td>-1.158033</td>\n",
       "      <td>-0.596766</td>\n",
       "      <td>0.648820</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene pdb_id chain sbna_aa3  sbna_num sbna_aa1 uniprot_num uniprot_aa  \\\n",
       "0  EGFR   7SI1     A      SER       695        S         695          S   \n",
       "1  EGFR   7SI1     A      GLY       696        G         696          G   \n",
       "2  EGFR   7SI1     A      GLU       697        E         697          E   \n",
       "3  EGFR   7SI1     A      ALA       698        A         698          A   \n",
       "4  EGFR   7SI1     A      PRO       699        P         699          P   \n",
       "\n",
       "   network_score  SecondOrderIntermodularDegree_AVERAGE  \\\n",
       "0      -1.946806                              -0.964584   \n",
       "1      -1.901057                              -0.900271   \n",
       "2      -0.544335                              -0.288050   \n",
       "3      -2.229062                              -1.095856   \n",
       "4      -2.403619                              -1.158033   \n",
       "\n",
       "   NodeEdgeBetweennessSTRIDE_sidechain_MAX  LigandMULTIMERCENTROIDSC_MIN  \\\n",
       "0                                -0.368011                      0.614212   \n",
       "1                                -0.530286                      0.470500   \n",
       "2                                -0.341759                     -0.085474   \n",
       "3                                -0.596766                      0.536440   \n",
       "4                                -0.596766                      0.648820   \n",
       "\n",
       "  mutant mutations  \n",
       "0      N       NaN  \n",
       "1      N       NaN  \n",
       "2      N       NaN  \n",
       "3      N       NaN  \n",
       "4      N       NaN  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ns_data = pd.DataFrame()\n",
    "\n",
    "for _, row in lung_pdbs.iterrows():\n",
    "    pdb_id = row['pdb_id']\n",
    "    chain = row['chain']\n",
    "    # print(pdb_id, chain)\n",
    "    try:\n",
    "        tmp = read_finalsum(pdb_id, chain).merge(read_finalsum_decomp(pdb_id, chain), on=['num', 'res'], how='inner').drop('res_code_x', axis=1)\n",
    "    except:\n",
    "        continue\n",
    "    # rename columns\n",
    "    tmp = tmp.rename(columns={\"res_code_y\": \"res_code\"})\n",
    "    tmp = align_finalsum_with_uniprot(tmp, pdb_id, chain)\n",
    "    # tmp = tmp[tmp['num']==12].reset_index(drop=True)\n",
    "    tmp['pdb_id'] = pdb_id\n",
    "    tmp['chain'] = chain\n",
    "    tmp['gene'] = row['gene']\n",
    "    tmp['mutant'] = row['mutant']\n",
    "    tmp['mutations'] = row['mutations']\n",
    "    ns_data = pd.concat([ns_data, tmp])\n",
    "\n",
    "# rename columns\n",
    "ns_data = ns_data.rename(columns={\"res\": \"sbna_aa3\", \"res_code\": \"sbna_aa1\", \"num\": \"sbna_num\", \"uniprot_res\": \"uniprot_aa\"})\n",
    "ns_data = ns_data[[\"gene\", \"pdb_id\", \"chain\", \"sbna_aa3\", \"sbna_num\", \"sbna_aa1\", \"uniprot_num\", \"uniprot_aa\", \"network_score\", \"SecondOrderIntermodularDegree_AVERAGE\", \"NodeEdgeBetweennessSTRIDE_sidechain_MAX\", \"LigandMULTIMERCENTROIDSC_MIN\", 'mutant', 'mutations']]\n",
    "\n",
    "print(len(ns_data))\n",
    "ns_data.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>chain</th>\n",
       "      <th>sbna_aa3</th>\n",
       "      <th>sbna_num</th>\n",
       "      <th>sbna_aa1</th>\n",
       "      <th>uniprot_num</th>\n",
       "      <th>uniprot_aa</th>\n",
       "      <th>network_score</th>\n",
       "      <th>SecondOrderIntermodularDegree_AVERAGE</th>\n",
       "      <th>NodeEdgeBetweennessSTRIDE_sidechain_MAX</th>\n",
       "      <th>LigandMULTIMERCENTROIDSC_MIN</th>\n",
       "      <th>mutant</th>\n",
       "      <th>mutations</th>\n",
       "      <th>cgc_tier</th>\n",
       "      <th>cancer_role</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>7SI1</td>\n",
       "      <td>A</td>\n",
       "      <td>SER</td>\n",
       "      <td>695</td>\n",
       "      <td>S</td>\n",
       "      <td>695</td>\n",
       "      <td>S</td>\n",
       "      <td>-1.946806</td>\n",
       "      <td>-0.964584</td>\n",
       "      <td>-0.368011</td>\n",
       "      <td>0.614212</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>7SI1</td>\n",
       "      <td>A</td>\n",
       "      <td>GLY</td>\n",
       "      <td>696</td>\n",
       "      <td>G</td>\n",
       "      <td>696</td>\n",
       "      <td>G</td>\n",
       "      <td>-1.901057</td>\n",
       "      <td>-0.900271</td>\n",
       "      <td>-0.530286</td>\n",
       "      <td>0.470500</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>7SI1</td>\n",
       "      <td>A</td>\n",
       "      <td>GLU</td>\n",
       "      <td>697</td>\n",
       "      <td>E</td>\n",
       "      <td>697</td>\n",
       "      <td>E</td>\n",
       "      <td>-0.544335</td>\n",
       "      <td>-0.288050</td>\n",
       "      <td>-0.341759</td>\n",
       "      <td>-0.085474</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>7SI1</td>\n",
       "      <td>A</td>\n",
       "      <td>ALA</td>\n",
       "      <td>698</td>\n",
       "      <td>A</td>\n",
       "      <td>698</td>\n",
       "      <td>A</td>\n",
       "      <td>-2.229062</td>\n",
       "      <td>-1.095856</td>\n",
       "      <td>-0.596766</td>\n",
       "      <td>0.536440</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>7SI1</td>\n",
       "      <td>A</td>\n",
       "      <td>PRO</td>\n",
       "      <td>699</td>\n",
       "      <td>P</td>\n",
       "      <td>699</td>\n",
       "      <td>P</td>\n",
       "      <td>-2.403619</td>\n",
       "      <td>-1.158033</td>\n",
       "      <td>-0.596766</td>\n",
       "      <td>0.648820</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene pdb_id chain sbna_aa3  sbna_num sbna_aa1 uniprot_num uniprot_aa  \\\n",
       "0  EGFR   7SI1     A      SER       695        S         695          S   \n",
       "1  EGFR   7SI1     A      GLY       696        G         696          G   \n",
       "2  EGFR   7SI1     A      GLU       697        E         697          E   \n",
       "3  EGFR   7SI1     A      ALA       698        A         698          A   \n",
       "4  EGFR   7SI1     A      PRO       699        P         699          P   \n",
       "\n",
       "   network_score  SecondOrderIntermodularDegree_AVERAGE  \\\n",
       "0      -1.946806                              -0.964584   \n",
       "1      -1.901057                              -0.900271   \n",
       "2      -0.544335                              -0.288050   \n",
       "3      -2.229062                              -1.095856   \n",
       "4      -2.403619                              -1.158033   \n",
       "\n",
       "   NodeEdgeBetweennessSTRIDE_sidechain_MAX  LigandMULTIMERCENTROIDSC_MIN  \\\n",
       "0                                -0.368011                      0.614212   \n",
       "1                                -0.530286                      0.470500   \n",
       "2                                -0.341759                     -0.085474   \n",
       "3                                -0.596766                      0.536440   \n",
       "4                                -0.596766                      0.648820   \n",
       "\n",
       "  mutant mutations  cgc_tier cancer_role  \n",
       "0      N       NaN         1    oncogene  \n",
       "1      N       NaN         1    oncogene  \n",
       "2      N       NaN         1    oncogene  \n",
       "3      N       NaN         1    oncogene  \n",
       "4      N       NaN         1    oncogene  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add cgc_tier and cancer_role from cgc_data\n",
    "cgc_data = pd.read_csv(\"../data_prep/Census_all_with_pdb.csv\", usecols=['Gene Symbol', 'Tier', 'Role in Cancer'])\n",
    "\n",
    "# rename columns\n",
    "cgc_data = cgc_data.rename(columns={\"Gene Symbol\": \"gene\", \"Tier\": \"cgc_tier\", \"Role in Cancer\": \"cancer_role\"})\n",
    "\n",
    "# merge with tmp2\n",
    "ns_data = ns_data.merge(cgc_data, left_on='gene', right_on='gene', how='left')\n",
    "ns_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "56754\n"
     ]
    }
   ],
   "source": [
    "# ns_data.to_csv(\"lung_genes_sbna.csv\", index=False)\n",
    "\n",
    "ns_data = pd.read_csv(\"lung_genes_sbna.csv\")\n",
    "print(len(ns_data))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # change cancer_role for \n",
    "# # TP53 to \"TSG\", NOTCH1 to \"oncogene\", ERBB4 to \"oncogene\"\n",
    "# ns_data.loc[ns_data['gene']==\"TP53\", \"cancer_role\"] = \"TSG\"\n",
    "# ns_data.loc[ns_data['gene']==\"NOTCH1\", \"cancer_role\"] = \"oncogene\"\n",
    "# ns_data.loc[ns_data['gene']==\"ERBB4\", \"cancer_role\"] = \"oncogene\"\n",
    "\n",
    "# # change \"oncogene, fusion\" to \"oncogene\"\n",
    "# ns_data.loc[ns_data['cancer_role']==\"oncogene, fusion\", \"cancer_role\"] = \"oncogene\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Aggreate data by position. (only for wild types)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>uniprot_num</th>\n",
       "      <th>uniprot_aa</th>\n",
       "      <th>sbna_aa1</th>\n",
       "      <th>network_score</th>\n",
       "      <th>SecondOrderIntermodularDegree_AVERAGE</th>\n",
       "      <th>NodeEdgeBetweennessSTRIDE_sidechain_MAX</th>\n",
       "      <th>LigandMULTIMERCENTROIDSC_MIN</th>\n",
       "      <th>cgc_tier</th>\n",
       "      <th>cancer_role</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>-1.442056</td>\n",
       "      <td>-0.831746</td>\n",
       "      <td>-0.478642</td>\n",
       "      <td>0.131669</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>2</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>-1.028050</td>\n",
       "      <td>-0.712966</td>\n",
       "      <td>-0.406600</td>\n",
       "      <td>-0.091516</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>3</td>\n",
       "      <td>D</td>\n",
       "      <td>D</td>\n",
       "      <td>-0.337630</td>\n",
       "      <td>-0.602712</td>\n",
       "      <td>-0.611381</td>\n",
       "      <td>-0.876463</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>4</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "      <td>0.369376</td>\n",
       "      <td>-0.335788</td>\n",
       "      <td>-0.067975</td>\n",
       "      <td>-0.773139</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>5</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.258724</td>\n",
       "      <td>-0.365234</td>\n",
       "      <td>-0.080011</td>\n",
       "      <td>-0.703969</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene  uniprot_num uniprot_aa sbna_aa1  network_score  \\\n",
       "0  AKT1            1          M        M      -1.442056   \n",
       "1  AKT1            2          S        S      -1.028050   \n",
       "2  AKT1            3          D        D      -0.337630   \n",
       "3  AKT1            4          V        V       0.369376   \n",
       "4  AKT1            5          A        A       0.258724   \n",
       "\n",
       "   SecondOrderIntermodularDegree_AVERAGE  \\\n",
       "0                              -0.831746   \n",
       "1                              -0.712966   \n",
       "2                              -0.602712   \n",
       "3                              -0.335788   \n",
       "4                              -0.365234   \n",
       "\n",
       "   NodeEdgeBetweennessSTRIDE_sidechain_MAX  LigandMULTIMERCENTROIDSC_MIN  \\\n",
       "0                                -0.478642                      0.131669   \n",
       "1                                -0.406600                     -0.091516   \n",
       "2                                -0.611381                     -0.876463   \n",
       "3                                -0.067975                     -0.773139   \n",
       "4                                -0.080011                     -0.703969   \n",
       "\n",
       "   cgc_tier cancer_role  \n",
       "0         1    oncogene  \n",
       "1         1    oncogene  \n",
       "2         1    oncogene  \n",
       "3         1    oncogene  \n",
       "4         1    oncogene  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter for wild types\n",
    "ns_data = ns_data[ns_data['mutant']==\"N\"]\n",
    "\n",
    "# aggregate data by gene, uniprot_num, uniprot_aa, sbna_aa1\n",
    "ns_data = ns_data[ns_data['uniprot_num']!=\"?\"]\n",
    "ns_data['uniprot_num'] = ns_data['uniprot_num'].astype(int)\n",
    "ns_data = ns_data.groupby(['gene', 'uniprot_num', 'uniprot_aa', 'sbna_aa1']).agg(\\\n",
    "    {'network_score': 'mean', 'SecondOrderIntermodularDegree_AVERAGE': 'mean', 'NodeEdgeBetweennessSTRIDE_sidechain_MAX': 'mean', \\\n",
    "     'LigandMULTIMERCENTROIDSC_MIN': 'mean', 'cgc_tier': 'first', 'cancer_role': 'first'}).reset_index()\n",
    "ns_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ns_data.to_csv(\"wt_ns_aggregated.csv\", index=False)\n",
    "ns_data.to_csv(\"ns_aggregated_all.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "72\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>uniprot_num</th>\n",
       "      <th>uniprot_aa</th>\n",
       "      <th>sbna_aa1</th>\n",
       "      <th>network_score</th>\n",
       "      <th>SecondOrderIntermodularDegree_AVERAGE</th>\n",
       "      <th>NodeEdgeBetweennessSTRIDE_sidechain_MAX</th>\n",
       "      <th>LigandMULTIMERCENTROIDSC_MIN</th>\n",
       "      <th>cgc_tier</th>\n",
       "      <th>cancer_role</th>\n",
       "      <th>mutant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>17</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "      <td>0.217918</td>\n",
       "      <td>-0.232960</td>\n",
       "      <td>-0.532687</td>\n",
       "      <td>-0.983564</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>17</td>\n",
       "      <td>E</td>\n",
       "      <td>K</td>\n",
       "      <td>0.178303</td>\n",
       "      <td>-0.516073</td>\n",
       "      <td>-0.260502</td>\n",
       "      <td>-0.954878</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ALK</td>\n",
       "      <td>1097</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>-2.128042</td>\n",
       "      <td>-0.866995</td>\n",
       "      <td>-0.384675</td>\n",
       "      <td>0.876372</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ALK</td>\n",
       "      <td>1097</td>\n",
       "      <td>C</td>\n",
       "      <td>S</td>\n",
       "      <td>-2.087775</td>\n",
       "      <td>-0.949926</td>\n",
       "      <td>-0.454795</td>\n",
       "      <td>0.683053</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ALK</td>\n",
       "      <td>1174</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "      <td>2.649683</td>\n",
       "      <td>0.714982</td>\n",
       "      <td>1.521564</td>\n",
       "      <td>-0.413137</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene  uniprot_num uniprot_aa sbna_aa1  network_score  \\\n",
       "0  AKT1           17          E        E       0.217918   \n",
       "1  AKT1           17          E        K       0.178303   \n",
       "2   ALK         1097          C        C      -2.128042   \n",
       "3   ALK         1097          C        S      -2.087775   \n",
       "4   ALK         1174          F        F       2.649683   \n",
       "\n",
       "   SecondOrderIntermodularDegree_AVERAGE  \\\n",
       "0                              -0.232960   \n",
       "1                              -0.516073   \n",
       "2                              -0.866995   \n",
       "3                              -0.949926   \n",
       "4                               0.714982   \n",
       "\n",
       "   NodeEdgeBetweennessSTRIDE_sidechain_MAX  LigandMULTIMERCENTROIDSC_MIN  \\\n",
       "0                                -0.532687                     -0.983564   \n",
       "1                                -0.260502                     -0.954878   \n",
       "2                                -0.384675                      0.876372   \n",
       "3                                -0.454795                      0.683053   \n",
       "4                                 1.521564                     -0.413137   \n",
       "\n",
       "   cgc_tier cancer_role  mutant  \n",
       "0         1    oncogene   False  \n",
       "1         1    oncogene    True  \n",
       "2         1    oncogene   False  \n",
       "3         1    oncogene    True  \n",
       "4         1    oncogene   False  "
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ns_data = pd.read_csv(\"ns_aggregated_all.csv\")\n",
    "\n",
    "# find rows where there are more than one row for each uniprot_aa\n",
    "diff_aas = ns_data[ns_data.duplicated(subset=['gene', 'uniprot_num'], keep=False)].sort_values(['gene', 'uniprot_num'], ascending=True).reset_index(drop=True)\n",
    "diff_aas['mutant'] = diff_aas['uniprot_aa']!=diff_aas['sbna_aa1']\n",
    "diff_aas.to_csv(\"diff_aas.csv\", index=False)\n",
    "print(len(diff_aas))\n",
    "diff_aas.head()\n",
    "\n",
    "# merged_data[merged_data['sbna_aa1'] != merged_data['uniprot_aa']].sort_values(['gene', 'uniprot_num'], ascending=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### try to calculate network scores again for all comparison of WT vs mutant \n",
    "* (i did a mistake before of taking NS of all structures even like mutant vs mutant)\n",
    "* note there is no distinction between chains here for comparisons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>mutations</th>\n",
       "      <th>uniprot_num</th>\n",
       "      <th>uniprot_aa</th>\n",
       "      <th>sbna_aa1</th>\n",
       "      <th>network_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>E17K</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>-0.160771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>E17K</td>\n",
       "      <td>2</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>0.357676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>E17K</td>\n",
       "      <td>3</td>\n",
       "      <td>D</td>\n",
       "      <td>D</td>\n",
       "      <td>-0.140439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>E17K</td>\n",
       "      <td>4</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "      <td>0.953059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>E17K</td>\n",
       "      <td>5</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.181981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6298</th>\n",
       "      <td>TP53</td>\n",
       "      <td>V272M</td>\n",
       "      <td>285</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "      <td>1.349385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6299</th>\n",
       "      <td>TP53</td>\n",
       "      <td>V272M</td>\n",
       "      <td>286</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "      <td>-0.337362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6300</th>\n",
       "      <td>TP53</td>\n",
       "      <td>V272M</td>\n",
       "      <td>287</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "      <td>-2.176509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6301</th>\n",
       "      <td>TP53</td>\n",
       "      <td>V272M</td>\n",
       "      <td>288</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>-2.209905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6302</th>\n",
       "      <td>TP53</td>\n",
       "      <td>V272M</td>\n",
       "      <td>289</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>-1.470848</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6303 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene mutations  uniprot_num uniprot_aa sbna_aa1  network_score\n",
       "0     AKT1      E17K            1          M        M      -0.160771\n",
       "1     AKT1      E17K            2          S        S       0.357676\n",
       "2     AKT1      E17K            3          D        D      -0.140439\n",
       "3     AKT1      E17K            4          V        V       0.953059\n",
       "4     AKT1      E17K            5          A        A       0.181981\n",
       "...    ...       ...          ...        ...      ...            ...\n",
       "6298  TP53     V272M          285          E        E       1.349385\n",
       "6299  TP53     V272M          286          E        E      -0.337362\n",
       "6300  TP53     V272M          287          E        E      -2.176509\n",
       "6301  TP53     V272M          288          N        N      -2.209905\n",
       "6302  TP53     V272M          289          L        L      -1.470848\n",
       "\n",
       "[6303 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_data = pd.read_csv(\"lung_genes_sbna.csv\")\n",
    "mut_data = mut_data[mut_data['uniprot_num'] != \"?\"]\n",
    "mut_data['uniprot_num'] = mut_data['uniprot_num'].astype(int)\n",
    "mut_data = mut_data.groupby(['gene', 'mutations', 'uniprot_num', 'uniprot_aa', 'sbna_aa1'])['network_score'].mean().reset_index()\n",
    "mut_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>uniprot_num</th>\n",
       "      <th>uniprot_aa</th>\n",
       "      <th>sbna_aa1</th>\n",
       "      <th>network_score</th>\n",
       "      <th>SecondOrderIntermodularDegree_AVERAGE</th>\n",
       "      <th>NodeEdgeBetweennessSTRIDE_sidechain_MAX</th>\n",
       "      <th>LigandMULTIMERCENTROIDSC_MIN</th>\n",
       "      <th>cgc_tier</th>\n",
       "      <th>cancer_role</th>\n",
       "      <th>cbioportal_freq</th>\n",
       "      <th>uniprot_aa_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>-1.442056</td>\n",
       "      <td>-0.831746</td>\n",
       "      <td>-0.478642</td>\n",
       "      <td>0.131669</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonpolar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>63</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>-0.798672</td>\n",
       "      <td>-0.378048</td>\n",
       "      <td>-0.040428</td>\n",
       "      <td>0.380196</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonpolar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>118</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>-1.451314</td>\n",
       "      <td>-0.850082</td>\n",
       "      <td>-0.685275</td>\n",
       "      <td>-0.084043</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonpolar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ALK</td>\n",
       "      <td>1089</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>-0.358116</td>\n",
       "      <td>1.118639</td>\n",
       "      <td>-0.459551</td>\n",
       "      <td>1.017204</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonpolar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ALK</td>\n",
       "      <td>1166</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>-0.979742</td>\n",
       "      <td>-0.512846</td>\n",
       "      <td>-0.413243</td>\n",
       "      <td>0.053652</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonpolar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10749</th>\n",
       "      <td>TP53</td>\n",
       "      <td>238</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>3.678561</td>\n",
       "      <td>1.766149</td>\n",
       "      <td>1.912412</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>57</td>\n",
       "      <td>Polar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10750</th>\n",
       "      <td>TP53</td>\n",
       "      <td>242</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.418538</td>\n",
       "      <td>0.852515</td>\n",
       "      <td>-0.433977</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>77</td>\n",
       "      <td>Polar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10751</th>\n",
       "      <td>TP53</td>\n",
       "      <td>275</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>2.052943</td>\n",
       "      <td>1.229223</td>\n",
       "      <td>0.823720</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>66</td>\n",
       "      <td>Polar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10752</th>\n",
       "      <td>TP53</td>\n",
       "      <td>277</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.779711</td>\n",
       "      <td>-0.483881</td>\n",
       "      <td>-0.295830</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>45</td>\n",
       "      <td>Polar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10753</th>\n",
       "      <td>TPR</td>\n",
       "      <td>75</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.497148</td>\n",
       "      <td>0.394188</td>\n",
       "      <td>0.260483</td>\n",
       "      <td>0.157523</td>\n",
       "      <td>1</td>\n",
       "      <td>fusion</td>\n",
       "      <td>0</td>\n",
       "      <td>Polar</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10754 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene  uniprot_num uniprot_aa sbna_aa1  network_score  \\\n",
       "0      AKT1            1          M        M      -1.442056   \n",
       "1      AKT1           63          M        M      -0.798672   \n",
       "2      AKT1          118          M        M      -1.451314   \n",
       "3       ALK         1089          M        M      -0.358116   \n",
       "4       ALK         1166          M        M      -0.979742   \n",
       "...     ...          ...        ...      ...            ...   \n",
       "10749  TP53          238          C        C       3.678561   \n",
       "10750  TP53          242          C        C       0.418538   \n",
       "10751  TP53          275          C        C       2.052943   \n",
       "10752  TP53          277          C        C      -0.779711   \n",
       "10753   TPR           75          C        C       0.497148   \n",
       "\n",
       "       SecondOrderIntermodularDegree_AVERAGE  \\\n",
       "0                                  -0.831746   \n",
       "1                                  -0.378048   \n",
       "2                                  -0.850082   \n",
       "3                                   1.118639   \n",
       "4                                  -0.512846   \n",
       "...                                      ...   \n",
       "10749                               1.766149   \n",
       "10750                               0.852515   \n",
       "10751                               1.229223   \n",
       "10752                              -0.483881   \n",
       "10753                               0.394188   \n",
       "\n",
       "       NodeEdgeBetweennessSTRIDE_sidechain_MAX  LigandMULTIMERCENTROIDSC_MIN  \\\n",
       "0                                    -0.478642                      0.131669   \n",
       "1                                    -0.040428                      0.380196   \n",
       "2                                    -0.685275                     -0.084043   \n",
       "3                                    -0.459551                      1.017204   \n",
       "4                                    -0.413243                      0.053652   \n",
       "...                                        ...                           ...   \n",
       "10749                                 1.912412                      0.000000   \n",
       "10750                                -0.433977                      0.000000   \n",
       "10751                                 0.823720                      0.000000   \n",
       "10752                                -0.295830                      0.000000   \n",
       "10753                                 0.260483                      0.157523   \n",
       "\n",
       "       cgc_tier cancer_role  cbioportal_freq uniprot_aa_type  \n",
       "0             1    oncogene                0        Nonpolar  \n",
       "1             1    oncogene                0        Nonpolar  \n",
       "2             1    oncogene                0        Nonpolar  \n",
       "3             1    oncogene                0        Nonpolar  \n",
       "4             1    oncogene                0        Nonpolar  \n",
       "...         ...         ...              ...             ...  \n",
       "10749         1         TSG               57           Polar  \n",
       "10750         1         TSG               77           Polar  \n",
       "10751         1         TSG               66           Polar  \n",
       "10752         1         TSG               45           Polar  \n",
       "10753         1      fusion                0           Polar  \n",
       "\n",
       "[10754 rows x 12 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read wt_ns_aggregated.csv\n",
    "wt_data = pd.read_csv(\"wt_ns_aggregated.csv\")\n",
    "wt_data = wt_data[wt_data['uniprot_num'] != \"?\"]\n",
    "wt_data['uniprot_num'] = wt_data['uniprot_num'].astype(int)\n",
    "wt_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>uniprot_num</th>\n",
       "      <th>uniprot_aa</th>\n",
       "      <th>sbna_aa1</th>\n",
       "      <th>network_score_wt</th>\n",
       "      <th>SecondOrderIntermodularDegree_AVERAGE</th>\n",
       "      <th>NodeEdgeBetweennessSTRIDE_sidechain_MAX</th>\n",
       "      <th>LigandMULTIMERCENTROIDSC_MIN</th>\n",
       "      <th>cgc_tier</th>\n",
       "      <th>cancer_role</th>\n",
       "      <th>cbioportal_freq</th>\n",
       "      <th>uniprot_aa_type</th>\n",
       "      <th>mutations</th>\n",
       "      <th>network_score_mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>-1.442056</td>\n",
       "      <td>-0.831746</td>\n",
       "      <td>-0.478642</td>\n",
       "      <td>0.131669</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonpolar</td>\n",
       "      <td>E17K</td>\n",
       "      <td>-0.160771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>2</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>-1.028050</td>\n",
       "      <td>-0.712966</td>\n",
       "      <td>-0.406600</td>\n",
       "      <td>-0.091516</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>0</td>\n",
       "      <td>Polar</td>\n",
       "      <td>E17K</td>\n",
       "      <td>0.357676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>3</td>\n",
       "      <td>D</td>\n",
       "      <td>D</td>\n",
       "      <td>-0.337630</td>\n",
       "      <td>-0.602712</td>\n",
       "      <td>-0.611381</td>\n",
       "      <td>-0.876463</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>1</td>\n",
       "      <td>Charged</td>\n",
       "      <td>E17K</td>\n",
       "      <td>-0.140439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>4</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "      <td>0.369376</td>\n",
       "      <td>-0.335788</td>\n",
       "      <td>-0.067975</td>\n",
       "      <td>-0.773139</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonpolar</td>\n",
       "      <td>E17K</td>\n",
       "      <td>0.953059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1436</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>5</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.258724</td>\n",
       "      <td>-0.365234</td>\n",
       "      <td>-0.080011</td>\n",
       "      <td>-0.703969</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>0</td>\n",
       "      <td>Nonpolar</td>\n",
       "      <td>E17K</td>\n",
       "      <td>0.181981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3813</th>\n",
       "      <td>TP53</td>\n",
       "      <td>289</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>-1.299322</td>\n",
       "      <td>-0.861989</td>\n",
       "      <td>-0.437332</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>1</td>\n",
       "      <td>Nonpolar</td>\n",
       "      <td>N239Y</td>\n",
       "      <td>-1.167156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3814</th>\n",
       "      <td>TP53</td>\n",
       "      <td>289</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>-1.299322</td>\n",
       "      <td>-0.861989</td>\n",
       "      <td>-0.437332</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>1</td>\n",
       "      <td>Nonpolar</td>\n",
       "      <td>R273C, T284R</td>\n",
       "      <td>-3.219438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3815</th>\n",
       "      <td>TP53</td>\n",
       "      <td>289</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>-1.299322</td>\n",
       "      <td>-0.861989</td>\n",
       "      <td>-0.437332</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>1</td>\n",
       "      <td>Nonpolar</td>\n",
       "      <td>R273H</td>\n",
       "      <td>0.671923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3816</th>\n",
       "      <td>TP53</td>\n",
       "      <td>289</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>-1.299322</td>\n",
       "      <td>-0.861989</td>\n",
       "      <td>-0.437332</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>1</td>\n",
       "      <td>Nonpolar</td>\n",
       "      <td>R280K</td>\n",
       "      <td>-0.849960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3817</th>\n",
       "      <td>TP53</td>\n",
       "      <td>289</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>-1.299322</td>\n",
       "      <td>-0.861989</td>\n",
       "      <td>-0.437332</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>TSG</td>\n",
       "      <td>1</td>\n",
       "      <td>Nonpolar</td>\n",
       "      <td>V272M</td>\n",
       "      <td>-1.470848</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6158 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene  uniprot_num uniprot_aa sbna_aa1  network_score_wt  \\\n",
       "0     AKT1            1          M        M         -1.442056   \n",
       "208   AKT1            2          S        S         -1.028050   \n",
       "649   AKT1            3          D        D         -0.337630   \n",
       "990   AKT1            4          V        V          0.369376   \n",
       "1436  AKT1            5          A        A          0.258724   \n",
       "...    ...          ...        ...      ...               ...   \n",
       "3813  TP53          289          L        L         -1.299322   \n",
       "3814  TP53          289          L        L         -1.299322   \n",
       "3815  TP53          289          L        L         -1.299322   \n",
       "3816  TP53          289          L        L         -1.299322   \n",
       "3817  TP53          289          L        L         -1.299322   \n",
       "\n",
       "      SecondOrderIntermodularDegree_AVERAGE  \\\n",
       "0                                 -0.831746   \n",
       "208                               -0.712966   \n",
       "649                               -0.602712   \n",
       "990                               -0.335788   \n",
       "1436                              -0.365234   \n",
       "...                                     ...   \n",
       "3813                              -0.861989   \n",
       "3814                              -0.861989   \n",
       "3815                              -0.861989   \n",
       "3816                              -0.861989   \n",
       "3817                              -0.861989   \n",
       "\n",
       "      NodeEdgeBetweennessSTRIDE_sidechain_MAX  LigandMULTIMERCENTROIDSC_MIN  \\\n",
       "0                                   -0.478642                      0.131669   \n",
       "208                                 -0.406600                     -0.091516   \n",
       "649                                 -0.611381                     -0.876463   \n",
       "990                                 -0.067975                     -0.773139   \n",
       "1436                                -0.080011                     -0.703969   \n",
       "...                                       ...                           ...   \n",
       "3813                                -0.437332                      0.000000   \n",
       "3814                                -0.437332                      0.000000   \n",
       "3815                                -0.437332                      0.000000   \n",
       "3816                                -0.437332                      0.000000   \n",
       "3817                                -0.437332                      0.000000   \n",
       "\n",
       "      cgc_tier cancer_role  cbioportal_freq uniprot_aa_type     mutations  \\\n",
       "0            1    oncogene                0        Nonpolar          E17K   \n",
       "208          1    oncogene                0           Polar          E17K   \n",
       "649          1    oncogene                1         Charged          E17K   \n",
       "990          1    oncogene                0        Nonpolar          E17K   \n",
       "1436         1    oncogene                0        Nonpolar          E17K   \n",
       "...        ...         ...              ...             ...           ...   \n",
       "3813         1         TSG                1        Nonpolar         N239Y   \n",
       "3814         1         TSG                1        Nonpolar  R273C, T284R   \n",
       "3815         1         TSG                1        Nonpolar         R273H   \n",
       "3816         1         TSG                1        Nonpolar         R280K   \n",
       "3817         1         TSG                1        Nonpolar         V272M   \n",
       "\n",
       "      network_score_mut  \n",
       "0             -0.160771  \n",
       "208            0.357676  \n",
       "649           -0.140439  \n",
       "990            0.953059  \n",
       "1436           0.181981  \n",
       "...                 ...  \n",
       "3813          -1.167156  \n",
       "3814          -3.219438  \n",
       "3815           0.671923  \n",
       "3816          -0.849960  \n",
       "3817          -1.470848  \n",
       "\n",
       "[6158 rows x 14 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge with mut_data\n",
    "merged_data = wt_data.merge(mut_data, on=['gene', 'uniprot_num', 'uniprot_aa', 'sbna_aa1'], how='inner', suffixes=('_wt', '_mut'))\n",
    "# fill nan in cancer_role with unknown\n",
    "merged_data['cancer_role'] = merged_data['cancer_role'].fillna(\"unknown\")\n",
    "merged_data = merged_data.sort_values(['gene', 'uniprot_num'])\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data['diff_ns'] = merged_data['network_score_mut'] - merged_data['network_score_wt']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* note: cbioportal_freq here means the total number of mutations at that residue. not how many times this particular mutation has happened"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data.to_csv(\"sbna_wt_vs_mut.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# adding TCGA data\n",
    "* Using wild type structures only "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>uniprot_num</th>\n",
       "      <th>uniprot_aa</th>\n",
       "      <th>sbna_aa1</th>\n",
       "      <th>network_score</th>\n",
       "      <th>SecondOrderIntermodularDegree_AVERAGE</th>\n",
       "      <th>NodeEdgeBetweennessSTRIDE_sidechain_MAX</th>\n",
       "      <th>LigandMULTIMERCENTROIDSC_MIN</th>\n",
       "      <th>cgc_tier</th>\n",
       "      <th>cancer_role</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>-1.442056</td>\n",
       "      <td>-0.831746</td>\n",
       "      <td>-0.478642</td>\n",
       "      <td>0.131669</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>2</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>-1.028050</td>\n",
       "      <td>-0.712966</td>\n",
       "      <td>-0.406600</td>\n",
       "      <td>-0.091516</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>3</td>\n",
       "      <td>D</td>\n",
       "      <td>D</td>\n",
       "      <td>-0.337630</td>\n",
       "      <td>-0.602712</td>\n",
       "      <td>-0.611381</td>\n",
       "      <td>-0.876463</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>4</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "      <td>0.369376</td>\n",
       "      <td>-0.335788</td>\n",
       "      <td>-0.067975</td>\n",
       "      <td>-0.773139</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>5</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.258724</td>\n",
       "      <td>-0.365234</td>\n",
       "      <td>-0.080011</td>\n",
       "      <td>-0.703969</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10829</th>\n",
       "      <td>TPR</td>\n",
       "      <td>138</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>1.147110</td>\n",
       "      <td>0.583953</td>\n",
       "      <td>0.932649</td>\n",
       "      <td>0.369492</td>\n",
       "      <td>1</td>\n",
       "      <td>fusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10830</th>\n",
       "      <td>TPR</td>\n",
       "      <td>139</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "      <td>-1.110253</td>\n",
       "      <td>-0.500517</td>\n",
       "      <td>-0.218686</td>\n",
       "      <td>0.391051</td>\n",
       "      <td>1</td>\n",
       "      <td>fusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10831</th>\n",
       "      <td>TPR</td>\n",
       "      <td>140</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>-1.315210</td>\n",
       "      <td>-0.560337</td>\n",
       "      <td>-0.245537</td>\n",
       "      <td>0.509335</td>\n",
       "      <td>1</td>\n",
       "      <td>fusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10832</th>\n",
       "      <td>TPR</td>\n",
       "      <td>141</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>-0.579938</td>\n",
       "      <td>0.123206</td>\n",
       "      <td>-0.154097</td>\n",
       "      <td>0.549048</td>\n",
       "      <td>1</td>\n",
       "      <td>fusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10833</th>\n",
       "      <td>TPR</td>\n",
       "      <td>142</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.024671</td>\n",
       "      <td>0.155236</td>\n",
       "      <td>0.391037</td>\n",
       "      <td>0.521603</td>\n",
       "      <td>1</td>\n",
       "      <td>fusion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10834 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene  uniprot_num uniprot_aa sbna_aa1  network_score  \\\n",
       "0      AKT1            1          M        M      -1.442056   \n",
       "1      AKT1            2          S        S      -1.028050   \n",
       "2      AKT1            3          D        D      -0.337630   \n",
       "3      AKT1            4          V        V       0.369376   \n",
       "4      AKT1            5          A        A       0.258724   \n",
       "...     ...          ...        ...      ...            ...   \n",
       "10829   TPR          138          L        L       1.147110   \n",
       "10830   TPR          139          E        E      -1.110253   \n",
       "10831   TPR          140          Y        Y      -1.315210   \n",
       "10832   TPR          141          L        L      -0.579938   \n",
       "10833   TPR          142          T        T       0.024671   \n",
       "\n",
       "       SecondOrderIntermodularDegree_AVERAGE  \\\n",
       "0                                  -0.831746   \n",
       "1                                  -0.712966   \n",
       "2                                  -0.602712   \n",
       "3                                  -0.335788   \n",
       "4                                  -0.365234   \n",
       "...                                      ...   \n",
       "10829                               0.583953   \n",
       "10830                              -0.500517   \n",
       "10831                              -0.560337   \n",
       "10832                               0.123206   \n",
       "10833                               0.155236   \n",
       "\n",
       "       NodeEdgeBetweennessSTRIDE_sidechain_MAX  LigandMULTIMERCENTROIDSC_MIN  \\\n",
       "0                                    -0.478642                      0.131669   \n",
       "1                                    -0.406600                     -0.091516   \n",
       "2                                    -0.611381                     -0.876463   \n",
       "3                                    -0.067975                     -0.773139   \n",
       "4                                    -0.080011                     -0.703969   \n",
       "...                                        ...                           ...   \n",
       "10829                                 0.932649                      0.369492   \n",
       "10830                                -0.218686                      0.391051   \n",
       "10831                                -0.245537                      0.509335   \n",
       "10832                                -0.154097                      0.549048   \n",
       "10833                                 0.391037                      0.521603   \n",
       "\n",
       "       cgc_tier cancer_role  \n",
       "0             1    oncogene  \n",
       "1             1    oncogene  \n",
       "2             1    oncogene  \n",
       "3             1    oncogene  \n",
       "4             1    oncogene  \n",
       "...         ...         ...  \n",
       "10829         1      fusion  \n",
       "10830         1      fusion  \n",
       "10831         1      fusion  \n",
       "10832         1      fusion  \n",
       "10833         1      fusion  \n",
       "\n",
       "[10834 rows x 10 columns]"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ns_data = pd.read_csv(\"wt_ns_aggregated.csv\")\n",
    "\n",
    "ns_data = ns_data[ns_data['uniprot_num']!=\"?\"].reset_index(drop=True)\n",
    "ns_data['uniprot_num'] = ns_data['uniprot_num'].astype(int)\n",
    "ns_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Match mutations in residue data only!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_position(mutation_position):\n",
    "    try:\n",
    "        chromosome, position_range = mutation_position.split(':')\n",
    "        start_position, end_position = position_range.split('-')\n",
    "        # Verify that the start and end positions are the same\n",
    "        if start_position == end_position:\n",
    "            return int(start_position)\n",
    "        else:\n",
    "            return None\n",
    "    except:\n",
    "        # if mutation_position is float then return it\n",
    "        if type(mutation_position) == float:\n",
    "            return mutation_position\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21583\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th>Mutation AA</th>\n",
       "      <th>AA_MUT_START</th>\n",
       "      <th>AA_MUT_STOP</th>\n",
       "      <th>GENOMIC_WT_ALLELE_SEQ</th>\n",
       "      <th>GENOMIC_MUT_ALLELE_SEQ</th>\n",
       "      <th>AA_WT_ALLELE_SEQ</th>\n",
       "      <th>AA_MUT_ALLELE_SEQ</th>\n",
       "      <th>Mutation Description AA</th>\n",
       "      <th>Mutation genome position GRCh37</th>\n",
       "      <th>Mutation genome position GRCh38</th>\n",
       "      <th>MUTATION_SIGNIFICANCE_TIER</th>\n",
       "      <th>GRCh38 position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>68794</th>\n",
       "      <td>ERBB4</td>\n",
       "      <td>p.R842L</td>\n",
       "      <td>842</td>\n",
       "      <td>842</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>R</td>\n",
       "      <td>L</td>\n",
       "      <td>Substitution - Missense</td>\n",
       "      <td>2:212295788-212295788</td>\n",
       "      <td>2:211431063-211431063</td>\n",
       "      <td>Other</td>\n",
       "      <td>211431063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68795</th>\n",
       "      <td>ERBB4</td>\n",
       "      <td>p.E810Q</td>\n",
       "      <td>810</td>\n",
       "      <td>810</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>E</td>\n",
       "      <td>Q</td>\n",
       "      <td>Substitution - Missense</td>\n",
       "      <td>2:212426687-212426687</td>\n",
       "      <td>2:211561962-211561962</td>\n",
       "      <td>Other</td>\n",
       "      <td>211561962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68796</th>\n",
       "      <td>ERBB4</td>\n",
       "      <td>p.L567F</td>\n",
       "      <td>567</td>\n",
       "      <td>567</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>L</td>\n",
       "      <td>F</td>\n",
       "      <td>Substitution - Missense</td>\n",
       "      <td>2:212537906-212537906</td>\n",
       "      <td>2:211673181-211673181</td>\n",
       "      <td>Other</td>\n",
       "      <td>211673181</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      GENE_NAME Mutation AA  AA_MUT_START  AA_MUT_STOP GENOMIC_WT_ALLELE_SEQ  \\\n",
       "68794     ERBB4     p.R842L           842          842                     C   \n",
       "68795     ERBB4     p.E810Q           810          810                     C   \n",
       "68796     ERBB4     p.L567F           567          567                     G   \n",
       "\n",
       "      GENOMIC_MUT_ALLELE_SEQ AA_WT_ALLELE_SEQ AA_MUT_ALLELE_SEQ  \\\n",
       "68794                      A                R                 L   \n",
       "68795                      G                E                 Q   \n",
       "68796                      A                L                 F   \n",
       "\n",
       "       Mutation Description AA Mutation genome position GRCh37  \\\n",
       "68794  Substitution - Missense           2:212295788-212295788   \n",
       "68795  Substitution - Missense           2:212426687-212426687   \n",
       "68796  Substitution - Missense           2:212537906-212537906   \n",
       "\n",
       "      Mutation genome position GRCh38 MUTATION_SIGNIFICANCE_TIER  \\\n",
       "68794           2:211431063-211431063                      Other   \n",
       "68795           2:211561962-211561962                      Other   \n",
       "68796           2:211673181-211673181                      Other   \n",
       "\n",
       "       GRCh38 position  \n",
       "68794        211431063  \n",
       "68795        211561962  \n",
       "68796        211673181  "
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read cmc data\n",
    "cmc_data = pd.read_csv('../CMC/cmc_export.tsv', sep='\\t', usecols=[\n",
    "    'GENE_NAME', 'AA_MUT_START', 'AA_MUT_STOP', 'AA_WT_ALLELE_SEQ', 'AA_MUT_ALLELE_SEQ', 'Mutation AA', 'Mutation Description AA', \\\n",
    "        'MUTATION_SIGNIFICANCE_TIER', 'GENOMIC_WT_ALLELE_SEQ', 'GENOMIC_MUT_ALLELE_SEQ', \"Mutation genome position GRCh37\", \"Mutation genome position GRCh38\"])\n",
    "\n",
    "cmc_data = cmc_data[cmc_data['AA_MUT_START'] == cmc_data['AA_MUT_STOP']]\n",
    "\n",
    "# mutation not nan in AA_WT_ALLELE_SEQ\tMutation AA\tMutation Description AA\n",
    "cmc_data = cmc_data.dropna(subset=['AA_WT_ALLELE_SEQ', 'Mutation AA'])\n",
    "\n",
    "# only use missense mutations\n",
    "cmc_data = cmc_data[cmc_data['Mutation Description AA'] == 'Substitution - Missense']\n",
    "\n",
    "# remove rows with null AA_WT_ALLELE_SEQ\n",
    "cmc_data = cmc_data[cmc_data['AA_WT_ALLELE_SEQ'].notnull()] \n",
    "    \n",
    "cmc_data['GRCh38 position'] = cmc_data['Mutation genome position GRCh38'].apply(extract_position)\n",
    "cmc_data = cmc_data[cmc_data['GRCh38 position'].notnull()]\n",
    "cmc_data['GRCh38 position'] = cmc_data['GRCh38 position'].astype(int)\n",
    "\n",
    "# filter genes that are in lung_genes\n",
    "cmc_data = cmc_data[cmc_data['GENE_NAME'].isin(ns_data['gene'])]\n",
    "\n",
    "print(len(cmc_data))\n",
    "cmc_data.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find rows where chromosome position match to more than one uniprot position\n",
    "is_one_to_one = cmc_data.groupby(['GENE_NAME', 'GRCh38 position'])['AA_MUT_START'].nunique() == 1\n",
    "is_one_to_one = is_one_to_one.reset_index()\n",
    "tmp = is_one_to_one[is_one_to_one[\"AA_MUT_START\"]==False]['GENE_NAME'].unique()\n",
    "\n",
    "# assert none of them are in lung genes (otherwise would need to do some manual mapping)\n",
    "assert len(set(tmp).intersection(set(ns_data['gene'].unique()))) == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th>AA_MUT_START</th>\n",
       "      <th>GRCh38 position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>3</td>\n",
       "      <td>{104792637}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>4</td>\n",
       "      <td>{104792633, 104792634}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>11</td>\n",
       "      <td>{104792611}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>14</td>\n",
       "      <td>{104792602, 104792603}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>15</td>\n",
       "      <td>{104792600}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  GENE_NAME  AA_MUT_START         GRCh38 position\n",
       "0      AKT1             3             {104792637}\n",
       "1      AKT1             4  {104792633, 104792634}\n",
       "2      AKT1            11             {104792611}\n",
       "3      AKT1            14  {104792602, 104792603}\n",
       "4      AKT1            15             {104792600}"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now match chromosome position to uniprot num\n",
    "# for each gene_name, aa_mut_start create set of chromosome positions\n",
    "chromosome_positions = cmc_data.groupby(['GENE_NAME', 'AA_MUT_START'])['GRCh38 position'].apply(set).reset_index()\n",
    "chromosome_positions.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: this assumes that CMC data is as large as TCGA...as in every mutation in TCGA is also in CMC...will need to check and manually add after if no match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>uniprot_num</th>\n",
       "      <th>uniprot_aa</th>\n",
       "      <th>sbna_aa1</th>\n",
       "      <th>network_score</th>\n",
       "      <th>SecondOrderIntermodularDegree_AVERAGE</th>\n",
       "      <th>NodeEdgeBetweennessSTRIDE_sidechain_MAX</th>\n",
       "      <th>LigandMULTIMERCENTROIDSC_MIN</th>\n",
       "      <th>cgc_tier</th>\n",
       "      <th>cancer_role</th>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th>AA_MUT_START</th>\n",
       "      <th>GRCh38 position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>-1.442056</td>\n",
       "      <td>-0.831746</td>\n",
       "      <td>-0.478642</td>\n",
       "      <td>0.131669</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>2</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>-1.028050</td>\n",
       "      <td>-0.712966</td>\n",
       "      <td>-0.406600</td>\n",
       "      <td>-0.091516</td>\n",
       "      <td>1</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene  uniprot_num uniprot_aa sbna_aa1  network_score  \\\n",
       "0  AKT1            1          M        M      -1.442056   \n",
       "1  AKT1            2          S        S      -1.028050   \n",
       "\n",
       "   SecondOrderIntermodularDegree_AVERAGE  \\\n",
       "0                              -0.831746   \n",
       "1                              -0.712966   \n",
       "\n",
       "   NodeEdgeBetweennessSTRIDE_sidechain_MAX  LigandMULTIMERCENTROIDSC_MIN  \\\n",
       "0                                -0.478642                      0.131669   \n",
       "1                                -0.406600                     -0.091516   \n",
       "\n",
       "   cgc_tier cancer_role GENE_NAME  AA_MUT_START GRCh38 position  \n",
       "0         1    oncogene       NaN           NaN             NaN  \n",
       "1         1    oncogene       NaN           NaN             NaN  "
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ns_data = ns_data.merge(chromosome_positions, left_on=['gene', 'uniprot_num'], right_on=['GENE_NAME', 'AA_MUT_START'], how='left')\n",
    "ns_data.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ssm_id</th>\n",
       "      <th>dna_change</th>\n",
       "      <th>protein_change</th>\n",
       "      <th>type</th>\n",
       "      <th>consequence</th>\n",
       "      <th>num_ssm_affected_cases</th>\n",
       "      <th>num_cases_in_gene</th>\n",
       "      <th>ssm_affected_cases_percentage</th>\n",
       "      <th>num_gdc_ssm_affected_cases</th>\n",
       "      <th>num_gdc_ssm_cases</th>\n",
       "      <th>gdc_ssm_affected_cases_percentage</th>\n",
       "      <th>vep_impact</th>\n",
       "      <th>sift_impact</th>\n",
       "      <th>sift_score</th>\n",
       "      <th>polyphen_impact</th>\n",
       "      <th>polyphen_score</th>\n",
       "      <th>gene</th>\n",
       "      <th>GRCh38_pos</th>\n",
       "      <th>wt_aa</th>\n",
       "      <th>res_pos</th>\n",
       "      <th>mut_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ab96fb54-dfc0-58d9-b727-20a857d58dad</td>\n",
       "      <td>chr14:g.104780214C&gt;T</td>\n",
       "      <td>AKT1 E17K</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>Missense</td>\n",
       "      <td>69</td>\n",
       "      <td>4230</td>\n",
       "      <td>1.63</td>\n",
       "      <td>69</td>\n",
       "      <td>16477</td>\n",
       "      <td>0.42</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.00</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>1.000</td>\n",
       "      <td>AKT1</td>\n",
       "      <td>104780214</td>\n",
       "      <td>E</td>\n",
       "      <td>17</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5bedd0f4-b0e1-503a-b857-a2f54f3bdc6c</td>\n",
       "      <td>chr14:g.104780145C&gt;T</td>\n",
       "      <td>AKT1 E40K</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>Missense</td>\n",
       "      <td>5</td>\n",
       "      <td>4230</td>\n",
       "      <td>0.12</td>\n",
       "      <td>5</td>\n",
       "      <td>16477</td>\n",
       "      <td>0.03</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.16</td>\n",
       "      <td>possibly_damaging</td>\n",
       "      <td>0.515</td>\n",
       "      <td>AKT1</td>\n",
       "      <td>104780145</td>\n",
       "      <td>E</td>\n",
       "      <td>40</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>85931b10-2dfc-5711-ad10-52e8df9d4116</td>\n",
       "      <td>chr14:g.104776708A&gt;G</td>\n",
       "      <td>AKT1 W80R</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>Missense</td>\n",
       "      <td>4</td>\n",
       "      <td>4230</td>\n",
       "      <td>0.09</td>\n",
       "      <td>4</td>\n",
       "      <td>16477</td>\n",
       "      <td>0.02</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.01</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>0.930</td>\n",
       "      <td>AKT1</td>\n",
       "      <td>104776708</td>\n",
       "      <td>W</td>\n",
       "      <td>80</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>95594822-4344-5194-912d-99baa0a765b1</td>\n",
       "      <td>chr14:g.104780108A&gt;C</td>\n",
       "      <td>AKT1 L52R</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>Missense</td>\n",
       "      <td>4</td>\n",
       "      <td>4230</td>\n",
       "      <td>0.09</td>\n",
       "      <td>4</td>\n",
       "      <td>16477</td>\n",
       "      <td>0.02</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.00</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>0.993</td>\n",
       "      <td>AKT1</td>\n",
       "      <td>104780108</td>\n",
       "      <td>L</td>\n",
       "      <td>52</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>79df6966-11ae-5d3b-ab71-36da48e404f3</td>\n",
       "      <td>chr14:g.104776711G&gt;T</td>\n",
       "      <td>AKT1 Q79K</td>\n",
       "      <td>Substitution</td>\n",
       "      <td>Missense</td>\n",
       "      <td>2</td>\n",
       "      <td>4230</td>\n",
       "      <td>0.05</td>\n",
       "      <td>2</td>\n",
       "      <td>16477</td>\n",
       "      <td>0.01</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.02</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>0.967</td>\n",
       "      <td>AKT1</td>\n",
       "      <td>104776711</td>\n",
       "      <td>Q</td>\n",
       "      <td>79</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 ssm_id            dna_change protein_change  \\\n",
       "0  ab96fb54-dfc0-58d9-b727-20a857d58dad  chr14:g.104780214C>T      AKT1 E17K   \n",
       "1  5bedd0f4-b0e1-503a-b857-a2f54f3bdc6c  chr14:g.104780145C>T      AKT1 E40K   \n",
       "2  85931b10-2dfc-5711-ad10-52e8df9d4116  chr14:g.104776708A>G      AKT1 W80R   \n",
       "3  95594822-4344-5194-912d-99baa0a765b1  chr14:g.104780108A>C      AKT1 L52R   \n",
       "4  79df6966-11ae-5d3b-ab71-36da48e404f3  chr14:g.104776711G>T      AKT1 Q79K   \n",
       "\n",
       "           type consequence  num_ssm_affected_cases  num_cases_in_gene  \\\n",
       "0  Substitution    Missense                      69               4230   \n",
       "1  Substitution    Missense                       5               4230   \n",
       "2  Substitution    Missense                       4               4230   \n",
       "3  Substitution    Missense                       4               4230   \n",
       "4  Substitution    Missense                       2               4230   \n",
       "\n",
       "   ssm_affected_cases_percentage  num_gdc_ssm_affected_cases  \\\n",
       "0                           1.63                          69   \n",
       "1                           0.12                           5   \n",
       "2                           0.09                           4   \n",
       "3                           0.09                           4   \n",
       "4                           0.05                           2   \n",
       "\n",
       "   num_gdc_ssm_cases  gdc_ssm_affected_cases_percentage vep_impact  \\\n",
       "0              16477                               0.42   MODERATE   \n",
       "1              16477                               0.03   MODERATE   \n",
       "2              16477                               0.02   MODERATE   \n",
       "3              16477                               0.02   MODERATE   \n",
       "4              16477                               0.01   MODERATE   \n",
       "\n",
       "   sift_impact  sift_score    polyphen_impact  polyphen_score  gene  \\\n",
       "0  deleterious        0.00  probably_damaging           1.000  AKT1   \n",
       "1    tolerated        0.16  possibly_damaging           0.515  AKT1   \n",
       "2  deleterious        0.01  probably_damaging           0.930  AKT1   \n",
       "3  deleterious        0.00  probably_damaging           0.993  AKT1   \n",
       "4  deleterious        0.02  probably_damaging           0.967  AKT1   \n",
       "\n",
       "   GRCh38_pos wt_aa res_pos mut_aa  \n",
       "0   104780214     E      17      K  \n",
       "1   104780145     E      40      K  \n",
       "2   104776708     W      80      R  \n",
       "3   104780108     L      52      R  \n",
       "4   104776711     Q      79      K  "
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def tcga_get_chr_position(dna_change):\n",
    "    # example \"chr12:g.25245350C>A\" to 25245350\n",
    "    match = re.search(r'chr\\d+:g\\.(\\d+)', dna_change)\n",
    "    if match:\n",
    "        return match.group(1)\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "# read all TCGA files ending with .tsv\n",
    "tcga_data = pd.DataFrame()\n",
    "for file in os.listdir(\"../TCGA\"):\n",
    "    if file.endswith(\".tsv\"):\n",
    "        tmp = pd.read_csv(f\"../TCGA/{file}\", sep='\\t')\n",
    "        gene = file.split('_')[0] # assume the file name is written properly...\n",
    "        tmp['gene'] = gene\n",
    "        tmp = tmp.rename(columns={f'num_{gene}_cases': 'num_cases_in_gene'}) # rename num_<GENE>_cases to num_cases\n",
    "        tcga_data = pd.concat([tcga_data, tmp])\n",
    "\n",
    "# missense only\n",
    "tcga_data = tcga_data[(tcga_data['type']==\"Substitution\") & (tcga_data['consequence']==\"Missense\")] \n",
    "\n",
    "# extract position\n",
    "tcga_data['GRCh38_pos'] = tcga_data['dna_change'].apply(tcga_get_chr_position).astype(int)\n",
    "\n",
    "tcga_data = tcga_data.reset_index(drop=True)\n",
    "\n",
    "tcga_data[['gene', 'wt_aa', 'res_pos', 'mut_aa']] = tcga_data['protein_change'].str.extract(r'(\\w+) (\\w)(\\d+)(\\w)')\n",
    "tcga_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "# assert that there are unique chromosome positions per gene\n",
    "assert (tcga_data.groupby(['gene', 'GRCh38_pos'])['res_pos'].nunique() == 1).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save tcga_data\n",
    "tcga_data.to_csv(\"../TCGA/lung_genes_all_muts_tcga.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = []\n",
    "for _, row in ns_data.iterrows():\n",
    "    gene = row['gene']\n",
    "    # find corresponding row in tcga_data\n",
    "    \n",
    "    # position is nan\n",
    "    if pd.isnull(row['GRCh38 position']):\n",
    "        counts.append(0)\n",
    "        continue\n",
    "\n",
    "    # find corresponding rows in tcga_data\n",
    "    tmp = tcga_data[(tcga_data['gene']==gene) & (tcga_data['GRCh38_pos'].isin(row['GRCh38 position'])) & (tcga_data['wt_aa'] == row['uniprot_aa'])]\n",
    "    \n",
    "    # SUM num_gdc_ssm_affected_cases BY POSITION\n",
    "    counts.append(tmp['num_gdc_ssm_affected_cases'].sum()) \n",
    "\n",
    "ns_data['tcga_num_cases'] = counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "# also fill missing values in cancer_role\n",
    "ns_data['cancer_role'] = ns_data['cancer_role'].fillna('Unknown')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "ns_data.to_csv(\"wt_ns_aggregated.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sbna",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
